--- before/net/minecraft/entity/projectile/EntityLlamaSpit.java
+++ after/net/minecraft/entity/projectile/EntityLlamaSpit.java
@@ -20,7 +20,7 @@
 
 public class EntityLlamaSpit extends Entity implements IProjectile
 {
-    public EntityLlama owner;
+    public net.minecraft.entity.EntityLivingBase owner; // CraftBukkit - type EntityLlama -> EntityLivingBase
     private NBTTagCompound ownerNbt;
 
     public EntityLlamaSpit(World worldIn)
@@ -87,7 +87,7 @@
             raytraceresult = new RayTraceResult(entity);
         }
 
-        if (raytraceresult != null)
+        if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
         {
             this.onHit(raytraceresult);
         }
@@ -222,6 +222,7 @@
 
     public void onHit(RayTraceResult p_190536_1_)
     {
+        org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, p_190536_1_);
         if (p_190536_1_.entityHit != null && this.owner != null)
         {
             p_190536_1_.entityHit.attackEntityFrom(DamageSource.causeIndirectDamage(this, this.owner).setProjectile(), 1.0F);
