--- before/net/minecraft/entity/projectile/EntityPotion.java
+++ after/net/minecraft/entity/projectile/EntityPotion.java
@@ -1,12 +1,18 @@
 package net.minecraft.entity.projectile;
 
 import com.google.common.base.Predicate;
+
+import java.util.HashMap;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import javax.annotation.Nullable;
 import net.minecraft.entity.EntityAreaEffectCloud;
+import net.minecraft.entity.EntityLiving;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.monster.EntityBlaze;
 import net.minecraft.entity.monster.EntityEnderman;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Items;
 import net.minecraft.init.PotionTypes;
@@ -30,6 +36,8 @@
 import net.minecraft.world.World;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_12_R1.entity.CraftLivingEntity;
+import org.bukkit.entity.LivingEntity;
 
 public class EntityPotion extends EntityThrowable
 {
@@ -126,7 +134,7 @@
             {
                 this.applyWater();
             }
-            else if (!list.isEmpty())
+            else if (true || !list.isEmpty()) // CraftBukkit - Call event even if no effects to apply
             {
                 if (this.isLingering())
                 {
@@ -167,6 +175,7 @@
     {
         AxisAlignedBB axisalignedbb = this.getEntityBoundingBox().grow(4.0, 2.0, 4.0);
         List<EntityLivingBase> list = this.world.getEntitiesWithinAABB(EntityLivingBase.class, axisalignedbb);
+        Map<LivingEntity, Double> affected = new HashMap<LivingEntity, Double>();
 
         if (!list.isEmpty())
         {
@@ -184,26 +193,46 @@
                         {
                             d1 = 1.0;
                         }
-
-                        for (PotionEffect potioneffect : p_190543_2_)
-                        {
-                            Potion potion = potioneffect.getPotion();
-
-                            if (potion.isInstant())
-                            {
-                                potion.affectEntity(this, this.getThrower(), entitylivingbase, potioneffect.getAmplifier(), d1);
-                            }
-                            else
-                            {
-                                int i = (int)(d1 * potioneffect.getDuration() + 0.5);
-
-                                if (i > 20)
-                                {
-                                    entitylivingbase.addPotionEffect(
-                                        new PotionEffect(potion, i, potioneffect.getAmplifier(), potioneffect.getIsAmbient(), potioneffect.doesShowParticles())
-                                    );
-                                }
-                            }
+                        affected.put((LivingEntity)entitylivingbase.getBukkitEntity(), d1);
+                    }
+                }
+            }
+        }
+
+        org.bukkit.event.entity.PotionSplashEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPotionSplashEvent(this, affected);
+        if (!event.isCancelled() && p_190543_2_ != null && !p_190543_2_.isEmpty())
+        {
+            for (LivingEntity victim : event.getAffectedEntities())
+            {
+                if (!(victim instanceof CraftLivingEntity))
+                {
+                    continue;
+                }
+                EntityLivingBase entityliving = ((CraftLivingEntity)victim).getHandle();
+                double d1 = event.getIntensity(victim);
+
+                for (PotionEffect mobeffect : p_190543_2_)
+                {
+                    Potion mobeffectlist = mobeffect.getPotion();
+                    if (!this.world.pvpMode && this.getThrower() instanceof EntityPlayer && entityliving instanceof EntityPlayer && entityliving != this.getThrower())
+                    {
+                        int i = Potion.getIdFromPotion(mobeffectlist);
+                        if (i == 2 || i == 4 || i == 7 || i == 15 || i == 17 || i == 18 || i == 19)
+                        {
+                            continue;
+                        }
+                    }
+
+                    if (mobeffectlist.isInstant())
+                    {
+                        mobeffectlist.affectEntity(this, this.getThrower(), entityliving, mobeffect.getAmplifier(), d1);
+                    }
+                    else
+                    {
+                        int i = (int)(d1 * mobeffect.getDuration() + 0.5D);
+                        if (i > 20)
+                        {
+                            entityliving.addPotionEffect(new PotionEffect(mobeffectlist, i, mobeffect.getAmplifier(), mobeffect.getIsAmbient(), mobeffect.doesShowParticles()));
                         }
                     }
                 }
@@ -233,10 +262,16 @@
             entityareaeffectcloud.setColor(nbttagcompound.getInteger("CustomPotionColor"));
         }
 
-        this.world.spawnEntity(entityareaeffectcloud);
+        // this.world.spawnEntity(entityareaeffectcloud);
+        org.bukkit.event.entity.LingeringPotionSplashEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callLingeringPotionSplashEvent(this, entityareaeffectcloud);
+        if (!(event.isCancelled() || entityareaeffectcloud.isDead)) {
+            this.world.spawnEntity(entityareaeffectcloud);
+        } else {
+            entityareaeffectcloud.isDead = true;
+        }
     }
 
-    private boolean isLingering()
+    public boolean isLingering()
     {
         return this.getPotion().getItem() == Items.LINGERING_POTION;
     }
