--- before/net/minecraft/entity/projectile/EntityFishHook.java
+++ after/net/minecraft/entity/projectile/EntityFishHook.java
@@ -38,7 +38,7 @@
     private static final DataParameter<Integer> DATA_HOOKED_ENTITY = EntityDataManager.createKey(EntityFishHook.class, DataSerializers.VARINT);
     private boolean inGround;
     private int ticksInGround;
-    private EntityPlayer angler;
+    public EntityPlayer angler;
     private int ticksInAir;
     private int ticksCatchable;
     private int ticksCaughtDelay;
@@ -280,8 +280,8 @@
     {
         ItemStack itemstack = this.angler.getHeldItemMainhand();
         ItemStack itemstack1 = this.angler.getHeldItemOffhand();
-        boolean flag = itemstack.getItem() == Items.FISHING_ROD;
-        boolean flag1 = itemstack1.getItem() == Items.FISHING_ROD;
+        boolean flag = itemstack.getItem() instanceof net.minecraft.item.ItemFishingRod;
+        boolean flag1 = itemstack1.getItem() instanceof net.minecraft.item.ItemFishingRod;
 
         if (!this.angler.isDead && this.angler.isEntityAlive() && (flag || flag1) && !(this.getDistanceSq(this.angler) > 1024.0))
         {
@@ -369,6 +369,7 @@
 
         if (raytraceresult != null && raytraceresult.typeOfHit != RayTraceResult.Type.MISS)
         {
+            org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
             if (raytraceresult.typeOfHit == RayTraceResult.Type.ENTITY)
             {
                 this.caughtEntity = raytraceresult.entityHit;
@@ -410,6 +411,8 @@
             {
                 this.ticksCaughtDelay = 0;
                 this.ticksCatchableDelay = 0;
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), null, (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.FAILED_ATTEMPT);
+                this.world.getServer().getPluginManager().callEvent(playerFishEvent);
             }
             else
             {
@@ -429,9 +432,9 @@
                 double d0 = this.posX + f1 * this.ticksCatchableDelay * 0.1F;
                 double d1 = MathHelper.floor(this.getEntityBoundingBox().minY) + 1.0F;
                 double d2 = this.posZ + f2 * this.ticksCatchableDelay * 0.1F;
-                Block block = worldserver.getBlockState(new BlockPos(d0, d1 - 1.0, d2)).getBlock();
+                IBlockState state = worldserver.getBlockState(new BlockPos(d0, d1 - 1.0, d2));
 
-                if (block == Blocks.WATER || block == Blocks.FLOWING_WATER)
+                if (state.getMaterial() == Material.WATER)
                 {
                     if (this.rand.nextFloat() < 0.15F)
                     {
@@ -446,6 +449,11 @@
             }
             else
             {
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), null, (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.BITE);
+                this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+                if (playerFishEvent.isCancelled()) {
+                    return;
+                }
                 this.motionY = -0.4F * MathHelper.nextFloat(this.rand, 0.6F, 1.0F);
                 this.playSound(SoundEvents.ENTITY_BOBBER_SPLASH, 0.25F, 1.0F + (this.rand.nextFloat() - this.rand.nextFloat()) * 0.4F);
                 double d3 = this.getEntityBoundingBox().minY + 0.5;
@@ -499,9 +507,9 @@
                 double d4 = this.posX + MathHelper.sin(f6) * f7 * 0.1F;
                 double d5 = MathHelper.floor(this.getEntityBoundingBox().minY) + 1.0F;
                 double d6 = this.posZ + MathHelper.cos(f6) * f7 * 0.1F;
-                Block block1 = worldserver.getBlockState(new BlockPos((int)d4, (int)d5 - 1, (int)d6)).getBlock();
+                IBlockState state = worldserver.getBlockState(new BlockPos((int) d4, (int) d5 - 1, (int) d6));
 
-                if (block1 == Blocks.WATER || block1 == Blocks.FLOWING_WATER)
+                if (state.getMaterial() == Material.WATER)
                 {
                     worldserver.spawnParticle(EnumParticleTypes.WATER_SPLASH, d4, d5, d6, 2 + this.rand.nextInt(2), 0.1F, 0.0, 0.1F, 0.0);
                 }
@@ -541,8 +549,14 @@
         {
             int i = 0;
 
+            net.minecraftforge.event.entity.player.ItemFishedEvent event = null;
             if (this.caughtEntity != null)
             {
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), this.caughtEntity.getBukkitEntity(), (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_ENTITY);
+                this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+                if (playerFishEvent.isCancelled()) {
+                    return 0;
+                }
                 this.bringInHookedEntity();
                 this.world.setEntityState(this, (byte)31);
                 i = this.caughtEntity instanceof EntityItem ? 3 : 5;
@@ -550,14 +564,26 @@
             else if (this.ticksCatchable > 0)
             {
                 LootContext.Builder lootcontext$builder = new LootContext.Builder((WorldServer)this.world);
-                lootcontext$builder.withLuck(this.luck + this.angler.getLuck());
+                lootcontext$builder.withLuck((float)this.luck + this.angler.getLuck()).withPlayer(this.angler).withLootedEntity(this); // Forge: add player & looted entity to LootContext
+                List<ItemStack> result = this.world.getLootTableManager().getLootTableFromLocation(LootTableList.GAMEPLAY_FISHING).generateLootForPools(this.rand, lootcontext$builder.build());
+                event = new net.minecraftforge.event.entity.player.ItemFishedEvent(result, this.inGround ? 2 : 1, this);
+                net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+                if (event.isCanceled())
+                {
+                    this.setDead();
+                    return event.getRodDamage();
+                }
 
-                for (ItemStack itemstack : this.world
-                        .getLootTableManager()
-                        .getLootTableFromLocation(LootTableList.GAMEPLAY_FISHING)
-                        .generateLootForPools(this.rand, lootcontext$builder.build()))
+                for (ItemStack itemstack : result)
                 {
                     EntityItem entityitem = new EntityItem(this.world, this.posX, this.posY, this.posZ, itemstack);
+                    org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), entityitem.getBukkitEntity(), (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_FISH);
+                    playerFishEvent.setExpToDrop(this.rand.nextInt(6) + 1);
+                    this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+                    if (playerFishEvent.isCancelled()) {
+                        return 0;
+                    }
                     double d0 = this.angler.posX - this.posX;
                     double d1 = this.angler.posY - this.posY;
                     double d2 = this.angler.posZ - this.posZ;
@@ -567,17 +593,12 @@
                     entityitem.motionY = d1 * 0.1 + MathHelper.sqrt(d3) * 0.08;
                     entityitem.motionZ = d2 * 0.1;
                     this.world.spawnEntity(entityitem);
-                    this.angler
-                    .world
-                    .spawnEntity(
-                        new EntityXPOrb(
-                            this.angler.world,
-                            this.angler.posX,
-                            this.angler.posY + 0.5,
-                            this.angler.posZ + 0.5,
-                            this.rand.nextInt(6) + 1
-                        )
-                    );
+                    // this.angler.world.spawnEntity(new EntityXPOrb(this.angler.world, this.angler.posX, this.angler.posY + 0.5D, this.angler.posZ + 0.5D, this.rand.nextInt(6) + 1));
+                    // CraftBukkit start - this.random.nextInt(6) + 1 -> playerFishEvent.getExpToDrop()
+                    if (playerFishEvent.getExpToDrop() > 0) {
+                        this.angler.world.spawnEntity(new EntityXPOrb(this.angler.world, this.angler.posX, this.angler.posY + 0.5D, this.angler.posZ + 0.5D, playerFishEvent.getExpToDrop()));
+                    }
+                    // CraftBukkit end
                     Item item = itemstack.getItem();
 
                     if (item == Items.FISH || item == Items.COOKED_FISH)
@@ -591,11 +612,24 @@
 
             if (this.inGround)
             {
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), null, (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.IN_GROUND);
+                this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+
+                if (playerFishEvent.isCancelled()) {
+                    return 0;
+                }
                 i = 2;
             }
+            if (i == 0) {
+                org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) this.angler.getBukkitEntity(), null, (org.bukkit.entity.Fish) this.getBukkitEntity(), org.bukkit.event.player.PlayerFishEvent.State.FAILED_ATTEMPT);
+                this.world.getServer().getPluginManager().callEvent(playerFishEvent);
+                if (playerFishEvent.isCancelled()) {
+                    return 0;
+                }
+            }
 
             this.setDead();
-            return i;
+            return event == null ? i : event.getRodDamage();
         }
         else
         {
