--- before/net/minecraft/entity/projectile/EntityThrowable.java
+++ after/net/minecraft/entity/projectile/EntityThrowable.java
@@ -31,8 +31,8 @@
     private Block inTile;
     protected boolean inGround;
     public int throwableShake;
-    protected EntityLivingBase thrower;
-    private String throwerName;
+    public EntityLivingBase thrower;
+    public String throwerName;
     private int ticksInGround;
     private int ticksInAir;
     public Entity ignoreEntity;
@@ -54,6 +54,7 @@
     {
         this(worldIn, throwerIn.posX, throwerIn.posY + throwerIn.getEyeHeight() - 0.1F, throwerIn.posZ);
         this.thrower = throwerIn;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) throwerIn.getBukkitEntity();
     }
 
     @Override
@@ -200,7 +201,7 @@
                 {
                     flag = true;
                 }
-                else if (this.thrower != null && this.ticksExisted < 2 && this.ignoreEntity == null)
+                else if (this.thrower != null && this.ticksExisted < 2 && this.ignoreEntity == null && this.thrower == entity1) // CraftBukkit - MC-88491
                 {
                     this.ignoreEntity = entity1;
                     flag = true;
@@ -249,9 +250,12 @@
             {
                 this.setPortal(raytraceresult.getBlockPos());
             }
-            else
+            else if (!net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onImpact(raytraceresult);
+                if (this.isDead) {
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
+                }
             }
         }
 
