--- before/net/minecraft/entity/projectile/EntityFireball.java
+++ after/net/minecraft/entity/projectile/EntityFireball.java
@@ -24,6 +24,9 @@
     public double accelerationY;
     public double accelerationZ;
 
+    public float bukkitYield = 1;
+    public boolean isIncendiary = true;
+
     public EntityFireball(World worldIn)
     {
         super(worldIn);
@@ -66,12 +69,18 @@
     {
         super(worldIn);
         this.shootingEntity = shooter;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) shooter.getBukkitEntity();
         this.setSize(1.0F, 1.0F);
         this.setLocationAndAngles(shooter.posX, shooter.posY, shooter.posZ, shooter.rotationYaw, shooter.rotationPitch);
         this.setPosition(this.posX, this.posY, this.posZ);
         this.motionX = 0.0;
         this.motionY = 0.0;
         this.motionZ = 0.0;
+        // CraftBukkit start - Added setDirection method
+        this.setDirection(accelX, accelY, accelZ);
+    }
+
+    public void setDirection(double accelX, double accelY, double accelZ) {
         accelX += this.rand.nextGaussian() * 0.4;
         accelY += this.rand.nextGaussian() * 0.4;
         accelZ += this.rand.nextGaussian() * 0.4;
@@ -97,9 +106,12 @@
             this.ticksInAir++;
             RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, this.ticksInAir >= 25, this.shootingEntity);
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onImpact(raytraceresult);
+                if (this.isDead) {
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, raytraceresult);
+                }
             }
 
             this.posX = this.posX + this.motionX;
@@ -227,6 +239,9 @@
 
             if (source.getTrueSource() != null)
             {
+                if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
+                    return false;
+                }
                 Vec3d vec3d = source.getTrueSource().getLookVec();
 
                 if (vec3d != null)
@@ -242,6 +257,7 @@
                 if (source.getTrueSource() instanceof EntityLivingBase)
                 {
                     this.shootingEntity = (EntityLivingBase)source.getTrueSource();
+                    this.projectileSource = (org.bukkit.projectiles.ProjectileSource) this.shootingEntity.getBukkitEntity();
                 }
 
                 return true;
