--- before/net/minecraft/entity/projectile/EntityWitherSkull.java
+++ after/net/minecraft/entity/projectile/EntityWitherSkull.java
@@ -67,7 +67,7 @@
         float f = super.getExplosionResistance(explosionIn, worldIn, pos, blockStateIn);
         Block block = blockStateIn.getBlock();
 
-        if (this.isInvulnerable() && EntityWither.canDestroyBlock(block))
+        if (this.isInvulnerable() && block.canEntityDestroy(blockStateIn, worldIn, pos, this) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this.shootingEntity, pos, blockStateIn))
         {
             f = Math.min(0.8F, f);
         }
@@ -84,7 +84,7 @@
             {
                 if (this.shootingEntity != null)
                 {
-                    if (result.entityHit.attackEntityFrom(DamageSource.causeMobDamage(this.shootingEntity), 8.0F))
+                    if (result.entityHit.attackEntityFrom(DamageSource.causeThrownDamage(this, this.shootingEntity), 8.0F))
                     {
                         if (result.entityHit.isEntityAlive())
                         {
@@ -92,7 +92,7 @@
                         }
                         else
                         {
-                            this.shootingEntity.heal(5.0F);
+                            this.shootingEntity.heal(5.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.WITHER);
                         }
                     }
                 }
@@ -121,10 +121,13 @@
                 }
             }
 
-            this.world
-            .newExplosion(
-                this, this.posX, this.posY, this.posZ, 1.0F, false, this.world.getGameRules().getBoolean("mobGriefing")
-            );
+            // this.world.newExplosion(this, this.posX, this.posY, this.posZ, 1.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity));
+            org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 1.0F, false);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                this.world.newExplosion(this, this.posX, this.posY, this.posZ, event.getRadius(), event.getFire(), net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this.shootingEntity));
+            }
             this.setDead();
         }
     }
