--- before/net/minecraft/entity/projectile/EntityShulkerBullet.java
+++ after/net/minecraft/entity/projectile/EntityShulkerBullet.java
@@ -80,7 +80,28 @@
         this.target = targetIn;
         this.direction = EnumFacing.UP;
         this.selectNextMoveDirection(p_i46772_4_);
-    }
+        projectileSource = (org.bukkit.entity.LivingEntity) ownerIn.getBukkitEntity();
+    }
+
+    // CraftBukkit start
+    public EntityLivingBase getShooter() {
+        return this.owner;
+    }
+
+    public void setShooter(EntityLivingBase e) {
+        this.owner = e;
+    }
+
+    public Entity getTarget() {
+        return this.target;
+    }
+
+    public void setTarget(Entity e) {
+        this.target = e;
+        this.direction = EnumFacing.UP;
+        this.selectNextMoveDirection(EnumFacing.Axis.X);
+    }
+    // CraftBukkit end
 
     @Override
     protected void writeEntityToNBT(NBTTagCompound compound)
@@ -325,7 +346,7 @@
 
                 RayTraceResult raytraceresult = ProjectileHelper.forwardsRaycast(this, true, false, this.owner);
 
-                if (raytraceresult != null)
+                if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
                 {
                     this.bulletHit(raytraceresult);
                 }
@@ -412,6 +433,7 @@
 
     protected void bulletHit(RayTraceResult result)
     {
+        org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, result);
         if (result.entityHit == null)
         {
             ((WorldServer)this.world)
