--- before/net/minecraft/entity/projectile/EntityArrow.java
+++ after/net/minecraft/entity/projectile/EntityArrow.java
@@ -49,12 +49,12 @@
     }
             );
     private static final DataParameter<Byte> CRITICAL = EntityDataManager.createKey(EntityArrow.class, DataSerializers.BYTE);
-    private int xTile = -1;
-    private int yTile = -1;
-    private int zTile = -1;
+    public int xTile = -1; // PAIL: private->public
+    public int yTile = -1; // PAIL: private->public
+    public int zTile = -1; // PAIL: private->public
     private Block inTile;
     private int inData;
-    protected boolean inGround;
+    public boolean inGround; // Spigot: protected->public
     protected int timeInGround;
     public EntityArrow.PickupStatus pickupStatus = EntityArrow.PickupStatus.DISALLOWED;
     public int arrowShake;
@@ -62,7 +62,19 @@
     private int ticksInGround;
     private int ticksInAir;
     private double damage = 2.0;
-    private int knockbackStrength;
+    public int knockbackStrength;
+
+    // Spigot Start
+    @Override
+    public void inactiveTick()
+    {
+        if ( this.inGround )
+        {
+            this.ticksInGround += 1; // Despawn counter. First int after shooter
+        }
+        super.inactiveTick();
+    }
+    // Spigot End
 
     public EntityArrow(World worldIn)
     {
@@ -80,6 +92,7 @@
     {
         this(worldIn, shooter.posX, shooter.posY + shooter.getEyeHeight() - 0.1F, shooter.posZ);
         this.shootingEntity = shooter;
+        this.projectileSource = (org.bukkit.entity.LivingEntity) shooter.getBukkitEntity();
 
         if (shooter instanceof EntityPlayer)
         {
@@ -228,7 +241,7 @@
             {
                 this.ticksInGround++;
 
-                if (this.ticksInGround >= 1200)
+                if (this.ticksInGround >= world.spigotConfig.arrowDespawnRate) // Spigot - First int after shooter
                 {
                     this.setDead();
                 }
@@ -270,7 +283,7 @@
                 }
             }
 
-            if (raytraceresult != null)
+            if (raytraceresult != null && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult))
             {
                 this.onHit(raytraceresult);
             }
@@ -366,6 +379,7 @@
     protected void onHit(RayTraceResult raytraceResultIn)
     {
         Entity entity = raytraceResultIn.entityHit;
+        org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileHitEvent(this, raytraceResultIn);
 
         if (entity != null)
         {
@@ -392,7 +406,12 @@
 
             if (this.isBurning() && !(entity instanceof EntityEnderman))
             {
-                entity.setFire(5);
+//                entity.setFire(5);
+                org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+                org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+                if (!combustEvent.isCancelled()) {
+                    entity.setFire(combustEvent.getDuration());
+                }
             }
 
             if (entity.attackEntityFrom(damagesource, i))
@@ -612,10 +631,20 @@
     {
         if (!this.world.isRemote && this.inGround && this.arrowShake <= 0)
         {
-            boolean flag = this.pickupStatus == EntityArrow.PickupStatus.ALLOWED
-                           || this.pickupStatus == EntityArrow.PickupStatus.CREATIVE_ONLY && entityIn.capabilities.isCreativeMode;
-
-            if (this.pickupStatus == EntityArrow.PickupStatus.ALLOWED && !entityIn.inventory.addItemStackToInventory(this.getArrowStack()))
+            ItemStack itemstack = this.getArrowStack();
+            net.minecraft.entity.item.EntityItem item = new net.minecraft.entity.item.EntityItem(this.world, this.posX, this.posY, this.posZ, itemstack);
+            if (this.pickupStatus == PickupStatus.ALLOWED && entityIn.inventory.canHold(itemstack) > 0) {
+                org.bukkit.event.player.PlayerPickupArrowEvent event = new org.bukkit.event.player.PlayerPickupArrowEvent((org.bukkit.entity.Player) entityIn.getBukkitEntity(), new org.bukkit.craftbukkit.v1_12_R1.entity.CraftItem(this.world.getServer(), this, item), (org.bukkit.entity.Arrow) this.getBukkitEntity());
+                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+            }
+            boolean flag = this.pickupStatus == PickupStatus.ALLOWED || this.pickupStatus == PickupStatus.CREATIVE_ONLY && entityIn.capabilities.isCreativeMode;
+
+            if (this.pickupStatus == PickupStatus.ALLOWED && !entityIn.inventory.addItemStackToInventory(item.getItem()))
             {
                 flag = false;
             }
@@ -701,7 +730,13 @@
 
         if (EnchantmentHelper.getMaxEnchantmentLevel(Enchantments.FLAME, p_190547_1_) > 0)
         {
-            this.setFire(100);
+//            this.setFire(100);
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustEvent(this.getBukkitEntity(), 100);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                this.setFire(event.getDuration());
+            }
         }
     }
 
