--- before/net/minecraft/entity/projectile/EntityEvokerFangs.java
+++ after/net/minecraft/entity/projectile/EntityEvokerFangs.java
@@ -93,11 +93,11 @@
         {
             if (this.clientSideAttackStarted)
             {
-                this.lifeTicks--;
+                --this.lifeTicks;
 
                 if (this.lifeTicks == 14)
                 {
-                    for (int i = 0; i < 12; i++)
+                    for (int i = 0; i < 12; ++i)
                     {
                         double d0 = this.posX + (this.rand.nextDouble() * 2.0 - 1.0) * this.width * 0.5;
                         double d1 = this.posY + 0.05 + this.rand.nextDouble() * 1.0;
@@ -141,7 +141,9 @@
         {
             if (entitylivingbase == null)
             {
+                org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = this;
                 p_190551_1_.attackEntityFrom(DamageSource.MAGIC, 6.0F);
+                org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = null;
             }
             else
             {
