--- before/net/minecraft/item/crafting/RecipeRepairItem.java
+++ after/net/minecraft/item/crafting/RecipeRepairItem.java
@@ -7,11 +7,15 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
 import net.minecraft.world.World;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
-public class RecipeRepairItem implements IRecipe
+public class RecipeRepairItem extends ShapelessRecipes implements IRecipe
 {
+    // CraftBukkit start - Delegate to new parent class
+    public RecipeRepairItem() {
+        super("", new ItemStack(net.minecraft.init.Items.LEATHER_HELMET), NonNullList.from(Ingredient.EMPTY, Ingredient.fromItem(net.minecraft.init.Items.LEATHER_HELMET)));
+    }
+    // CraftBukkit end
+
     @Override
     public boolean matches(InventoryCrafting inv, World worldIn)
     {
@@ -32,7 +36,7 @@
                     if (itemstack.getItem() != itemstack1.getItem()
                             || itemstack1.getCount() != 1
                             || itemstack.getCount() != 1
-                            || !itemstack1.getItem().isDamageable())
+                            || !itemstack1.getItem().isRepairable())
                     {
                         return false;
                     }
@@ -63,7 +67,7 @@
                     if (itemstack.getItem() != itemstack1.getItem()
                             || itemstack1.getCount() != 1
                             || itemstack.getCount() != 1
-                            || !itemstack1.getItem().isDamageable())
+                            || !itemstack1.getItem().isRepairable())
                     {
                         return ItemStack.EMPTY;
                     }
@@ -79,20 +83,32 @@
             if (itemstack2.getItem() == itemstack3.getItem()
                     && itemstack2.getCount() == 1
                     && itemstack3.getCount() == 1
-                    && itemstack2.getItem().isDamageable())
+                    && itemstack2.getItem().isRepairable())
             {
-                Item item = itemstack2.getItem();
-                int j = item.getMaxDamage() - itemstack2.getItemDamage();
-                int k = item.getMaxDamage() - itemstack3.getItemDamage();
-                int l = j + k + item.getMaxDamage() * 5 / 100;
-                int i1 = item.getMaxDamage() - l;
+                // FORGE: Make itemstack sensitive // Item item = itemstack2.getItem();
+                int j = itemstack2.getMaxDamage() - itemstack2.getItemDamage();
+                int k = itemstack2.getMaxDamage() - itemstack3.getItemDamage();
+                int l = j + k + itemstack2.getMaxDamage() * 5 / 100;
+                int i1 = itemstack2.getMaxDamage() - l;
 
                 if (i1 < 0)
                 {
                     i1 = 0;
                 }
 
-                return new ItemStack(itemstack2.getItem(), 1, i1);
+                // CraftBukkit start - Construct a dummy repair recipe
+                ItemStack result = new ItemStack(itemstack3.getItem(), 1, i1);
+                NonNullList<Ingredient> ingredients = NonNullList.create();
+                ingredients.add(Ingredient.fromStacks(itemstack2.copy()));
+                ingredients.add(Ingredient.fromStacks(itemstack3.copy()));
+                ShapelessRecipes recipe = new ShapelessRecipes("", result.copy(), ingredients);
+                recipe.key = new net.minecraft.util.ResourceLocation("repairitem");
+                inv.currentRecipe = recipe;
+                if (inv.resultInventory != null && inv.eventHandler.getBukkitView() != null) // CatServer - mods bypass
+                    result = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPreCraftEvent(inv, result, inv.eventHandler.getBukkitView(), true);
+                return result;
+                // return new ItemStack(itemstack2.getItem(), 1, i1);
+                // CraftBukkit end
             }
         }
 
@@ -113,11 +129,7 @@
         for (int i = 0; i < nonnulllist.size(); i++)
         {
             ItemStack itemstack = inv.getStackInSlot(i);
-
-            if (itemstack.getItem().hasContainerItem())
-            {
-                nonnulllist.set(i, new ItemStack(itemstack.getItem().getContainerItem()));
-            }
+            nonnulllist.set(i, net.minecraftforge.common.ForgeHooks.getContainerItem(itemstack));
         }
 
         return nonnulllist;
@@ -129,7 +141,6 @@
         return true;
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public boolean canFit(int width, int height)
     {
