--- before/net/minecraft/util/ClassInheritanceMultiMap.java
+++ after/net/minecraft/util/ClassInheritanceMultiMap.java
@@ -13,7 +13,8 @@
 
 public class ClassInheritanceMultiMap<T> extends AbstractSet<T>
 {
-    private static final Set < Class<? >> ALL_KNOWN = Sets.newHashSet();
+    // Forge: Use concurrent collection to allow creating chunks from multiple threads safely
+    private static final Set < Class<? >> ALL_KNOWN = java.util.Collections.newSetFromMap(new java.util.concurrent.ConcurrentHashMap<>());
     private final Map < Class<?>, List<T >> map = Maps.newHashMap();
     private final Set < Class<? >> knownKeys = Sets.newIdentityHashSet();
     private final Class<T> baseClass;
@@ -126,17 +127,30 @@
             @Override
             public Iterator<S> iterator()
             {
-                List<T> list = ClassInheritanceMultiMap.this.map.get(ClassInheritanceMultiMap.this.initializeClassLookup(clazz));
-
+                // CatServer start - Optimization from Foamfix
+                Class c = ClassInheritanceMultiMap.this.initializeClassLookup(clazz);
+                List<T> list = ClassInheritanceMultiMap.this.map.get(c);
                 if (list == null)
                 {
                     return Collections.emptyIterator();
+                } else {
+                    Iterator<T> iterator = list.iterator();
+                    return clazz == c ? (Iterator<S>) iterator : Iterators.filter(iterator, clazz);
+                }
+                // CatServer end
+                /*
+                List<T> list = (List)ClassInheritanceMultiMap.this.map.get(ClassInheritanceMultiMap.this.initializeClassLookup(clazz));
+
+                if (list == null)
+                {
+                    return Collections.<S>emptyIterator();
                 }
                 else
                 {
                     Iterator<T> iterator = list.iterator();
                     return Iterators.filter(iterator, clazz);
                 }
+                */
             }
         };
     }
@@ -144,7 +158,8 @@
     @Override
     public Iterator<T> iterator()
     {
-        return (Iterator<T>)(this.values.isEmpty() ? Collections.emptyIterator() : Iterators.unmodifiableIterator(this.values.iterator()));
+        // return (Iterator<T>)(this.values.isEmpty() ? Collections.emptyIterator() : Iterators.unmodifiableIterator(this.values.iterator()));
+        return this.values.isEmpty() ? Collections.emptyIterator() : this.values.iterator(); // CREF: world edit fix
     }
 
     @Override
