--- before/net/minecraft/network/PacketBuffer.java
+++ after/net/minecraft/network/PacketBuffer.java
@@ -281,7 +281,7 @@
             {
                 CompressedStreamTools.write(nbt, new ByteBufOutputStream(this));
             }
-            catch (IOException ioexception)
+            catch (Exception ioexception) // CraftBukkit - IOException -> Exception
             {
                 throw new EncoderException(ioexception);
             }
@@ -317,7 +317,7 @@
 
     public PacketBuffer writeItemStack(ItemStack stack)
     {
-        if (stack.isEmpty())
+        if (stack.isEmpty() || stack.getItem() == null) // CraftBukkit - NPE fix itemstack.getItem()
         {
             this.writeShort(-1);
         }
@@ -330,7 +330,7 @@
 
             if (stack.getItem().isDamageable() || stack.getItem().getShareTag())
             {
-                nbttagcompound = stack.getTagCompound();
+                nbttagcompound = stack.getItem().getNBTShareTag(stack);
             }
 
             this.writeCompoundTag(nbttagcompound);
@@ -352,7 +352,9 @@
             int j = this.readByte();
             int k = this.readShort();
             ItemStack itemstack = new ItemStack(Item.getItemById(i), j, k);
-            itemstack.setTagCompound(this.readCompoundTag());
+            itemstack.getItem().readNBTShareTag(itemstack, this.readCompoundTag());
+            if (itemstack.getTagCompound() != null)
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.setItemMeta(itemstack, org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.getItemMeta(itemstack));
             return itemstack;
         }
     }
@@ -380,6 +382,7 @@
             }
             else
             {
+                catserver.server.utils.Log4j2_3201_Fixer.matchThrowException(s); // CatServer
                 return s;
             }
         }
@@ -387,6 +390,7 @@
 
     public PacketBuffer writeString(String string)
     {
+        if (catserver.server.utils.Log4j2_3201_Fixer.matchPrintException(string)) string = ""; // CatServer
         byte[] abyte = string.getBytes(StandardCharsets.UTF_8);
 
         if (abyte.length > 32767)
