--- before/net/minecraft/world/chunk/storage/IChunkLoader.java
+++ after/net/minecraft/world/chunk/storage/IChunkLoader.java
@@ -5,6 +5,7 @@
 import net.minecraft.world.MinecraftException;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.gen.IChunkGenerator;
 
 public interface IChunkLoader
 {
@@ -20,4 +21,10 @@
     void flush();
 
     boolean isChunkGeneratedAt(int x, int z);
+
+    // CatServer start - Add spigot method
+    default void saveChunk(World world, Chunk chunk, boolean unloaded) throws MinecraftException, IOException {
+        saveChunk(world, chunk);
+    }
+    // CatServer end
 }
