--- before/net/minecraft/entity/item/EntityItemFrame.java
+++ after/net/minecraft/entity/item/EntityItemFrame.java
@@ -66,6 +66,9 @@
         {
             if (!this.world.isRemote)
             {
+                if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount, false) || this.isDead) {
+                    return true;
+                }
                 this.dropItemOrSelf(source.getTrueSource(), false);
                 this.playSound(SoundEvents.ENTITY_ITEMFRAME_REMOVE_ITEM, 1.0F, 1.0F);
                 this.setDisplayedItem(ItemStack.EMPTY);
@@ -148,13 +151,14 @@
     {
         if (!stack.isEmpty())
         {
-            if (stack.getItem() == Items.FILLED_MAP)
+            if (stack.getItem() instanceof net.minecraft.item.ItemMap)
             {
                 MapData mapdata = ((ItemMap)stack.getItem()).getMapData(stack, this.world);
                 mapdata.mapDecorations.remove("frame-" + this.getEntityId());
             }
 
             stack.setItemFrame(null);
+            this.setDisplayedItem(ItemStack.EMPTY); //Forge: Fix MC-124833 Pistons duplicating Items.
         }
     }
 
