--- before/net/minecraft/entity/item/EntityFallingBlock.java
+++ after/net/minecraft/entity/item/EntityFallingBlock.java
@@ -38,7 +38,7 @@
     public int fallTime;
     public boolean shouldDropItem = true;
     private boolean dontSetBlock;
-    private boolean hurtEntities;
+    public boolean hurtEntities;
     private int fallHurtMax = 40;
     private float fallHurtAmount = 2.0F;
     public NBTTagCompound tileEntityData;
@@ -119,7 +119,7 @@
             {
                 BlockPos blockpos = new BlockPos(this);
 
-                if (this.world.getBlockState(blockpos).getBlock() == block)
+                if (this.world.getBlockState(blockpos).getBlock() == block && !org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR, 0).isCancelled())
                 {
                     this.world.setBlockToAir(blockpos);
                 }
@@ -177,13 +177,14 @@
                 {
                     IBlockState iblockstate = this.world.getBlockState(blockpos1);
 
+                    if (this.world.isAirBlock(new BlockPos(this.posX, this.posY - 0.01F, this.posZ))) //Forge: Don't indent below.
                     if (!flag1
                             && BlockFalling.canFallThrough(
                                 this.world.getBlockState(new BlockPos(this.posX, this.posY - 0.01F, this.posZ))
                             ))
                     {
                         this.onGround = false;
-                        return;
+                        // return; // CraftBukkit
                     }
 
                     this.motionX *= 0.7F;
@@ -196,16 +197,20 @@
 
                         if (!this.dontSetBlock)
                         {
-                            if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, null)
+                            if (this.world.mayPlace(block, blockpos1, true, EnumFacing.UP, this)
                                     && (flag1 || !BlockFalling.canFallThrough(this.world.getBlockState(blockpos1.down())))
-                                    && this.world.setBlockState(blockpos1, this.fallTile, 3))
+                                    /* && this.world.setBlockState(blockpos1, this.fallTile, 3)*/)
                             {
+                                if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos1, this.fallTile.getBlock(), this.fallTile.getBlock().getMetaFromState(this.fallTile)).isCancelled()) {
+                                    return;
+                                }
+                                this.world.setBlockState(blockpos1, this.fallTile, 3);
                                 if (block instanceof BlockFalling)
                                 {
                                     ((BlockFalling)block).onEndFalling(this.world, blockpos1, this.fallTile, iblockstate);
                                 }
 
-                                if (this.tileEntityData != null && block instanceof ITileEntityProvider)
+                                if (this.tileEntityData != null && block.hasTileEntity(this.fallTile))
                                 {
                                     TileEntity tileentity = this.world.getTileEntity(blockpos1);
 
@@ -264,7 +269,9 @@
 
                 for (Entity entity : list)
                 {
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = this;
                     entity.attackEntityFrom(damagesource, Math.min(MathHelper.floor(i * this.fallHurtAmount), this.fallHurtMax));
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = null;
                 }
 
                 if (flag && this.rand.nextFloat() < 0.05F + i * 0.05)
