--- before/net/minecraft/entity/item/EntityBoat.java
+++ after/net/minecraft/entity/item/EntityBoat.java
@@ -68,6 +68,13 @@
     private EntityBoat.Status previousStatus;
     private double lastYd;
 
+    // Some of these haven't worked since a few updates, and since 1.9 they are less and less applicable.
+    public double maxSpeed = 0.4D;
+    public double occupiedDeceleration = 0.2D;
+    public double unoccupiedDeceleration = -1;
+    public boolean landBoats = false;
+    private org.bukkit.Location lastLocation;
+
     public EntityBoat(World worldIn)
     {
         super(worldIn);
@@ -150,6 +157,16 @@
         }
         else
         {
+            org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+            org.bukkit.entity.Entity attacker = (source.getTrueSource() == null) ? null : source.getTrueSource().getBukkitEntity();
+            
+            org.bukkit.event.vehicle.VehicleDamageEvent event = new org.bukkit.event.vehicle.VehicleDamageEvent(vehicle, attacker, amount);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return true;
+            }
+            // amount = event.getDamage(); // TODO Why don't we do this?
             this.setForwardDirection(-this.getForwardDirection());
             this.setTimeSinceHit(10);
             this.setDamageTaken(this.getDamageTaken() + amount * 10.0F);
@@ -158,6 +175,13 @@
 
             if (flag || this.getDamageTaken() > 40.0F)
             {
+                org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, attacker);
+                this.world.getServer().getPluginManager().callEvent(destroyEvent);
+
+                if (destroyEvent.isCancelled()) {
+                    this.setDamageTaken(40F); // Maximize damage so this doesn't get triggered again right away
+                    return true;
+                }
                 if (!flag && this.world.getGameRules().getBoolean("doEntityDrops"))
                 {
                     this.dropItemWithOffset(this.getItemBoat(), 1, 0.0F);
@@ -177,11 +201,23 @@
         {
             if (entityIn.getEntityBoundingBox().minY < this.getEntityBoundingBox().maxY)
             {
+                org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entityIn.getBukkitEntity());
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
                 super.applyEntityCollision(entityIn);
             }
         }
         else if (entityIn.getEntityBoundingBox().minY <= this.getEntityBoundingBox().minY)
         {
+            org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent((org.bukkit.entity.Vehicle) this.getBukkitEntity(), entityIn.getBukkitEntity());
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
             super.applyEntityCollision(entityIn);
         }
     }
@@ -301,6 +337,20 @@
             this.motionZ = 0.0;
         }
 
+        org.bukkit.Server server = this.world.getServer();
+        org.bukkit.World bworld = this.world.getWorld();
+        
+        org.bukkit.Location to = new org.bukkit.Location(bworld, this.posX, this.posY, this.posZ, this.rotationYaw, this.rotationPitch);
+        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+
+        server.getPluginManager().callEvent(new org.bukkit.event.vehicle.VehicleUpdateEvent(vehicle));
+
+        if (lastLocation != null && !lastLocation.equals(to)) {
+            org.bukkit.event.vehicle.VehicleMoveEvent event = new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, lastLocation, to);
+            server.getPluginManager().callEvent(event);
+        }
+        lastLocation = vehicle.getLocation();
+
         for (int i = 0; i <= 1; i++)
         {
             if (this.getPaddleState(i))
@@ -475,6 +525,12 @@
                             blockpos$pooledmutableblockpos.setPos(l1, k1, i2);
                             IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
 
+                            Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, new AxisAlignedBB(blockpos$pooledmutableblockpos), Material.WATER);
+                            if (result != null) {
+                                if (!result) continue;
+                                f = Math.max(f, iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) + blockpos$pooledmutableblockpos.getY());
+                            }
+                            
                             if (iblockstate.getMaterial() == Material.WATER)
                             {
                                 f = Math.max(f, BlockLiquid.getBlockLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos));
@@ -550,7 +606,7 @@
 
                                 if (!list.isEmpty())
                                 {
-                                    f += iblockstate.getBlock().slipperiness;
+                                    f += iblockstate.getBlock().getSlipperiness(iblockstate, this.world, blockpos$pooledmutableblockpos, this);
                                     k1++;
                                 }
 
@@ -592,7 +648,16 @@
                     {
                         blockpos$pooledmutableblockpos.setPos(k1, l1, i2);
                         IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
+                        
+                        Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, axisalignedbb, Material.WATER);
+                        if (result != null) {
+                            if (!result) continue;
 
+                            float f = iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) + blockpos$pooledmutableblockpos.getY();
+                            this.waterLevel = Math.max(f, this.waterLevel);
+                            flag |= axisalignedbb.minY < (double)f;
+                        }
+                        
                         if (iblockstate.getMaterial() == Material.WATER)
                         {
                             float f = BlockLiquid.getLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos);
@@ -636,6 +701,18 @@
                         blockpos$pooledmutableblockpos.setPos(k1, l1, i2);
                         IBlockState iblockstate = this.world.getBlockState(blockpos$pooledmutableblockpos);
 
+                        Boolean result = iblockstate.getBlock().isAABBInsideMaterial(world, blockpos$pooledmutableblockpos, axisalignedbb, Material.WATER);
+                        if (result != null) {
+                            if (!result) continue;
+
+                            if(iblockstate.getBlock().getBlockLiquidHeight(world, blockpos$pooledmutableblockpos, iblockstate, Material.WATER) > 0)
+                            {
+                                blockpos$pooledmutableblockpos.release();
+                                return EntityBoat.Status.UNDER_FLOWING_WATER;
+                            } else
+                                continue;
+                        }
+
                         if (iblockstate.getMaterial() == Material.WATER
                                 && d0 < BlockLiquid.getLiquidHeight(iblockstate, this.world, blockpos$pooledmutableblockpos))
                         {
@@ -875,6 +952,10 @@
 
                     if (!this.world.isRemote && !this.isDead)
                     {
+                        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+                        org.bukkit.event.vehicle.VehicleDestroyEvent destroyEvent = new org.bukkit.event.vehicle.VehicleDestroyEvent(vehicle, null);
+                        this.world.getServer().getPluginManager().callEvent(destroyEvent);
+                        if (destroyEvent.isCancelled()) {this.fallDistance = 0.0F; return;}
                         this.setDead();
 
                         if (this.world.getGameRules().getBoolean("doEntityDrops"))
@@ -967,6 +1048,22 @@
         this.rightInputDown = p_184442_2_;
         this.forwardInputDown = p_184442_3_;
         this.backInputDown = p_184442_4_;
+    }
+
+    // Forge: Fix MC-119811 by instantly completing lerp on board
+    @Override
+    protected void addPassenger(Entity passenger)
+    {
+        super.addPassenger(passenger);
+        if(this.canPassengerSteer() && this.lerpSteps > 0)
+        {
+            this.lerpSteps = 0;
+            this.posX = this.lerpX;
+            this.posY = this.lerpY;
+            this.posZ = this.lerpZ;
+            this.rotationYaw = (float)this.lerpYaw;
+            this.rotationPitch = (float)this.lerpPitch;
+        }
     }
 
     public static enum Status
