--- before/net/minecraft/entity/item/EntityArmorStand.java
+++ after/net/minecraft/entity/item/EntityArmorStand.java
@@ -58,7 +58,7 @@
     {
         public boolean apply(@Nullable Entity p_apply_1_)
         {
-            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).getType() == EntityMinecart.Type.RIDEABLE;
+            return p_apply_1_ instanceof EntityMinecart && ((EntityMinecart)p_apply_1_).canBeRidden();
         }
     };
     private final NonNullList<ItemStack> handItems = NonNullList.withSize(2, ItemStack.EMPTY);
@@ -67,12 +67,14 @@
     public long punchCooldown;
     private int disabledSlots;
     private boolean wasMarker;
-    private Rotations headRotation = DEFAULT_HEAD_ROTATION;
-    private Rotations bodyRotation = DEFAULT_BODY_ROTATION;
-    private Rotations leftArmRotation = DEFAULT_LEFTARM_ROTATION;
-    private Rotations rightArmRotation = DEFAULT_RIGHTARM_ROTATION;
-    private Rotations leftLegRotation = DEFAULT_LEFTLEG_ROTATION;
-    private Rotations rightLegRotation = DEFAULT_RIGHTLEG_ROTATION;
+    public Rotations headRotation = DEFAULT_HEAD_ROTATION;
+    public Rotations bodyRotation = DEFAULT_BODY_ROTATION;
+    public Rotations leftArmRotation = DEFAULT_LEFTARM_ROTATION;
+    public Rotations rightArmRotation = DEFAULT_RIGHTARM_ROTATION;
+    public Rotations leftLegRotation = DEFAULT_LEFTLEG_ROTATION;
+    public Rotations rightLegRotation = DEFAULT_RIGHTLEG_ROTATION;
+
+    private java.util.ArrayList<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<>(); // CatServer
 
     public EntityArmorStand(World worldIn)
     {
@@ -87,7 +89,13 @@
         this.setPosition(posX, posY, posZ);
     }
 
+    // CraftBukkit start - SPIGOT-3607, SPIGOT-3637
     @Override
+    public float getBukkitYaw() {
+        return this.rotationYaw;
+    }
+    // CraftBukkit end
+
     protected final void setSize(float width, float height)
     {
         double d0 = this.posX;
@@ -460,6 +468,19 @@
         {
             if (!itemstack.isEmpty() || (this.disabledSlots & 1 << p_184795_2_.getSlotIndex() + 16) == 0)
             {
+                org.bukkit.inventory.ItemStack armorStandItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+                org.bukkit.inventory.ItemStack playerHeldItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(p_184795_3_);
+                
+                org.bukkit.entity.Player bukkitPlayer = (org.bukkit.entity.Player) player.getBukkitEntity();
+                org.bukkit.entity.ArmorStand self = (org.bukkit.entity.ArmorStand) this.getBukkitEntity();
+                
+                org.bukkit.inventory.EquipmentSlot slot = org.bukkit.craftbukkit.v1_12_R1.CraftEquipmentSlot.getSlot(p_184795_2_);
+                org.bukkit.event.player.PlayerArmorStandManipulateEvent armorStandManipulateEvent = new org.bukkit.event.player.PlayerArmorStandManipulateEvent(bukkitPlayer, self, playerHeldItem, armorStandItem, slot);
+                this.world.getServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+                if (armorStandManipulateEvent.isCancelled()) {
+                    return;
+                }
                 if (player.capabilities.isCreativeMode && itemstack.isEmpty() && !p_184795_3_.isEmpty())
                 {
                     ItemStack itemstack2 = p_184795_3_.copy();
@@ -485,13 +506,16 @@
     @Override
     public boolean attackEntityFrom(DamageSource source, float amount)
     {
+        if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, amount)) {
+            return false;
+        }
         if (this.world.isRemote || this.isDead)
         {
             return false;
         }
         else if (DamageSource.OUT_OF_WORLD.equals(source))
         {
-            this.setDead();
+            this.onKillCommand();
             return false;
         }
         else if (this.isEntityInvulnerable(source) || this.canInteract || this.hasMarker())
@@ -501,7 +525,7 @@
         else if (source.isExplosion())
         {
             this.dropContents();
-            this.setDead();
+            this.onKillCommand();
             return false;
         }
         else if (DamageSource.IN_FIRE.equals(source))
@@ -546,7 +570,7 @@
                 {
                     this.playBrokenSound();
                     this.playParticles();
-                    this.setDead();
+                    this.onKillCommand();
                     return false;
                 }
                 else
@@ -562,7 +586,7 @@
                     {
                         this.dropBlock();
                         this.playParticles();
-                        this.setDead();
+                        this.onKillCommand();
                     }
 
                     return false;
@@ -633,7 +657,8 @@
         if (f <= 0.5F)
         {
             this.dropContents();
-            this.setDead();
+//            this.setDead();
+            this.onKillCommand();
         }
         else
         {
@@ -643,7 +668,8 @@
 
     private void dropBlock()
     {
-        Block.spawnAsEntity(this.world, new BlockPos(this), new ItemStack(Items.ARMOR_STAND));
+//        Block.spawnAsEntity(this.world, new BlockPos(this), new ItemStack(Items.ARMOR_STAND));
+        drops.add(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asBukkitCopy(new ItemStack(Items.ARMOR_STAND))); // CraftBukkit - add to drops
         this.dropContents();
     }
 
@@ -657,18 +683,20 @@
 
             if (!itemstack.isEmpty())
             {
-                Block.spawnAsEntity(this.world, new BlockPos(this).up(), itemstack);
+//                Block.spawnAsEntity(this.world, (new BlockPos(this)).up(), itemstack);
+                drops.add(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asBukkitCopy(itemstack)); // CraftBukkit - add to drops
                 this.handItems.set(i, ItemStack.EMPTY);
             }
         }
 
-        for (int j = 0; j < this.armorItems.size(); j++)
+        for (int j = 0; j < this.armorItems.size(); ++j)
         {
             ItemStack itemstack1 = this.armorItems.get(j);
 
             if (!itemstack1.isEmpty())
             {
-                Block.spawnAsEntity(this.world, new BlockPos(this).up(), itemstack1);
+//                Block.spawnAsEntity(this.world, (new BlockPos(this)).up(), itemstack1);
+                drops.add(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asBukkitCopy(itemstack1)); // CraftBukkit - add to drops
                 this.armorItems.set(j, ItemStack.EMPTY);
             }
         }
@@ -813,7 +841,14 @@
     @Override
     public void onKillCommand()
     {
+        org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityDeathEvent(this, drops); // CraftBukkit - call event
         this.setDead();
+        // CatServer start
+        for (EntityItem item : capturedDrops)
+        {
+            world.spawnEntity(item);
+        }
+        // CatServer end
     }
 
     @Override
@@ -828,7 +863,7 @@
         return this.hasMarker() ? EnumPushReaction.IGNORE : super.getPushReaction();
     }
 
-    private void setSmall(boolean small)
+    public void setSmall(boolean small)
     {
         this.dataManager.set(STATUS, this.setBit(this.dataManager.get(STATUS), 1, small));
         this.setSize(0.5F, 1.975F);
@@ -839,7 +874,7 @@
         return (this.dataManager.get(STATUS) & 1) != 0;
     }
 
-    private void setShowArms(boolean showArms)
+    public void setShowArms(boolean showArms)
     {
         this.dataManager.set(STATUS, this.setBit(this.dataManager.get(STATUS), 4, showArms));
     }
@@ -849,7 +884,7 @@
         return (this.dataManager.get(STATUS) & 4) != 0;
     }
 
-    private void setNoBasePlate(boolean noBasePlate)
+    public void setNoBasePlate(boolean noBasePlate)
     {
         this.dataManager.set(STATUS, this.setBit(this.dataManager.get(STATUS), 8, noBasePlate));
     }
@@ -859,7 +894,7 @@
         return (this.dataManager.get(STATUS) & 8) != 0;
     }
 
-    private void setMarker(boolean marker)
+    public void setMarker(boolean marker)
     {
         this.dataManager.set(STATUS, this.setBit(this.dataManager.get(STATUS), 16, marker));
         this.setSize(0.5F, 1.975F);
