--- before/net/minecraft/entity/item/EntityFireworkRocket.java
+++ after/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -3,6 +3,7 @@
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityLivingBase;
 import net.minecraft.entity.MoverType;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.item.ItemStack;
 import net.minecraft.nbt.NBTTagCompound;
@@ -22,13 +23,14 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<ItemStack> FIREWORK_ITEM = EntityDataManager.createKey(EntityFireworkRocket.class, DataSerializers.ITEM_STACK);
+    public static final DataParameter<ItemStack> FIREWORK_ITEM = EntityDataManager.createKey(EntityFireworkRocket.class, DataSerializers.ITEM_STACK);
     private static final DataParameter<Integer> BOOSTED_ENTITY_ID = EntityDataManager.createKey(EntityFireworkRocket.class, DataSerializers.VARINT);
     private int fireworkAge;
-    private int lifetime;
+    public int lifetime;
     private EntityLivingBase boostedEntity;
 
     public EntityFireworkRocket(World worldIn)
@@ -37,7 +39,14 @@
         this.setSize(0.25F, 0.25F);
     }
 
+    // Spigot Start
     @Override
+    public void inactiveTick() {
+        this.fireworkAge += 1;
+        super.inactiveTick();
+    }
+    // Spigot End
+
     protected void entityInit()
     {
         this.dataManager.register(FIREWORK_ITEM, ItemStack.EMPTY);
@@ -187,7 +196,7 @@
             .playSound(null, this.posX, this.posY, this.posZ, SoundEvents.ENTITY_FIREWORK_LAUNCH, SoundCategory.AMBIENT, 3.0F, 1.0F);
         }
 
-        this.fireworkAge++;
+        ++this.fireworkAge;
 
         if (this.world.isRemote && this.fireworkAge % 2 < 2)
         {
@@ -205,8 +214,10 @@
 
         if (!this.world.isRemote && this.fireworkAge > this.lifetime)
         {
-            this.world.setEntityState(this, (byte)17);
-            this.dealExplosionDamage();
+            if (!org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.world.setEntityState(this, (byte) 17);
+                this.dealExplosionDamage();
+            }
             this.setDead();
         }
     }
@@ -227,7 +238,9 @@
         {
             if (this.boostedEntity != null)
             {
+                CraftEventFactory.entityDamage = this;
                 this.boostedEntity.attackEntityFrom(DamageSource.FIREWORKS, 5 + nbttaglist.tagCount() * 2);
+                CraftEventFactory.entityDamage = null;
             }
 
             double d0 = 5.0;
@@ -239,7 +252,7 @@
                 {
                     boolean flag = false;
 
-                    for (int i = 0; i < 2; i++)
+                    for (int i = 0; i < 2; ++i)
                     {
                         RayTraceResult raytraceresult = this.world
                                                         .rayTraceBlocks(
@@ -264,7 +277,9 @@
                     if (flag)
                     {
                         float f1 = f * (float)Math.sqrt((5.0 - this.getDistance(entitylivingbase)) / 5.0);
+                        CraftEventFactory.entityDamage = this;
                         entitylivingbase.attackEntityFrom(DamageSource.FIREWORKS, f1);
+                        CraftEventFactory.entityDamage = null;
                     }
                 }
             }
