--- before/net/minecraft/entity/item/EntityEnderPearl.java
+++ after/net/minecraft/entity/item/EntityEnderPearl.java
@@ -110,6 +110,19 @@
                         && entityplayermp.world == this.world
                         && !entityplayermp.isPlayerSleeping())
                 {
+                    net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(entityplayermp, this.posX, this.posY, this.posZ, 5.0F);
+                    if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event))
+                    { // Don't indent to lower patch size
+                        // CraftBukkit start - Fire PlayerTeleportEvent
+                        org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer player = entityplayermp.getBukkitEntity();
+                        org.bukkit.Location location = getBukkitEntity().getLocation();
+                        location.setPitch(player.getLocation().getPitch());
+                        location.setYaw(player.getLocation().getYaw());
+                        
+                        org.bukkit.event.player.PlayerTeleportEvent teleEvent = new org.bukkit.event.player.PlayerTeleportEvent(player, player.getLocation(), location, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.ENDER_PEARL);
+                        org.bukkit.Bukkit.getPluginManager().callEvent(teleEvent);
+
+                        if (!teleEvent.isCancelled() && !entityplayermp.connection.isDisconnected()) {
                     if (this.rand.nextFloat() < 0.05F && this.world.getGameRules().getBoolean("doMobSpawning"))
                     {
                         EntityEndermite entityendermite = new EntityEndermite(this.world);
@@ -121,7 +134,7 @@
                             entitylivingbase.rotationYaw,
                             entitylivingbase.rotationPitch
                         );
-                        this.world.spawnEntity(entityendermite);
+                        this.world.addEntity(entityendermite, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.ENDER_PEARL);
                     }
 
                     if (entitylivingbase.isRiding())
@@ -129,9 +142,16 @@
                         entitylivingbase.dismountRidingEntity();
                     }
 
-                    entitylivingbase.setPositionAndUpdate(this.posX, this.posY, this.posZ);
-                    entitylivingbase.fallDistance = 0.0F;
-                    entitylivingbase.attackEntityFrom(DamageSource.FALL, 5.0F);
+                            entityplayermp.connection.teleport(teleEvent.getTo());
+                            entitylivingbase.fallDistance = 0.0F;
+                            org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = this;
+                            entitylivingbase.attackEntityFrom(DamageSource.FALL, 5.0F);
+                            org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.entityDamage = null;
+                        }
+//                            entitylivingbase.setPositionAndUpdate(event.getTargetX(), event.getTargetY(), event.getTargetZ());
+//                            entitylivingbase.fallDistance = 0.0F;
+//                            entitylivingbase.attackEntityFrom(DamageSource.FALL, event.getAttackDamage());
+                    }
                 }
             }
             else if (entitylivingbase != null)
@@ -161,13 +181,13 @@
 
     @Nullable
     @Override
-    public Entity changeDimension(int dimensionIn)
+    public Entity changeDimension(int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
     {
         if (this.thrower.dimension != dimensionIn)
         {
             this.thrower = null;
         }
 
-        return super.changeDimension(dimensionIn);
+        return super.changeDimension(dimensionIn, teleporter);
     }
 }
