--- before/net/minecraft/entity/boss/EntityWither.java
+++ after/net/minecraft/entity/boss/EntityWither.java
@@ -295,24 +295,39 @@
 
             if (j1 <= 0)
             {
-                this.world
-                .newExplosion(
-                    this,
-                    this.posX,
-                    this.posY + this.getEyeHeight(),
-                    this.posZ,
-                    7.0F,
-                    false,
-                    this.world.getGameRules().getBoolean("mobGriefing")
-                );
-                this.world.playBroadcastSound(1023, new BlockPos(this), 0);
+                // this.world.newExplosion(this, this.posX, this.posY + (double)this.getEyeHeight(), this.posZ, 7.0F, false, net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this));
+                org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 7.0F, false);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    this.world.newExplosion(this, this.posX, this.posY + this.getEyeHeight(), this.posZ, event.getRadius(), event.getFire(), net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this));
+                }
+
+                // CraftBukkit start - Use relative location for far away sounds
+                int viewDistance = this.world.getServer().getViewDistance() * 16;
+                for (EntityPlayerMP player : world.getMinecraftServer().getServer().getPlayerList().playerEntityList) {
+                    if (catserver.server.CatServer.getConfig().fixPlayBossSoundToOtherWorld && player.world != this.world) continue; // CatServer
+                    double deltaX = this.posX - player.posX;
+                    double deltaZ = this.posZ - player.posZ;
+                    double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                    if (world.spigotConfig.witherSpawnSoundRadius > 0 && distanceSquared > world.spigotConfig.witherSpawnSoundRadius * world.spigotConfig.witherSpawnSoundRadius) continue; // Spigot
+                    if (distanceSquared > viewDistance * viewDistance) {
+                        double deltaLength = Math.sqrt(distanceSquared);
+                        double relativeX = player.posX + (deltaX / deltaLength) * viewDistance;
+                        double relativeZ = player.posZ + (deltaZ / deltaLength) * viewDistance;
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1013, new BlockPos(relativeX, this.posY, relativeZ), 0, true));
+                    } else {
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1013, new BlockPos(this.posX, this.posY, this.posZ), 0, true));
+                    }
+                }
+                // CraftBukkit end
             }
 
             this.setInvulTime(j1);
 
             if (this.ticksExisted % 10 == 0)
             {
-                this.heal(10.0F);
+                this.heal(10.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.WITHER_SPAWN);
             }
         }
         else
@@ -407,7 +422,7 @@
             {
                 this.blockBreakCounter--;
 
-                if (this.blockBreakCounter == 0 && this.world.getGameRules().getBoolean("mobGriefing"))
+                if (this.blockBreakCounter == 0 && net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this))
                 {
                     int i1 = MathHelper.floor(this.posY);
                     int l1 = MathHelper.floor(this.posX);
@@ -427,8 +442,13 @@
                                 IBlockState iblockstate = this.world.getBlockState(blockpos);
                                 Block block = iblockstate.getBlock();
 
-                                if (iblockstate.getMaterial() != Material.AIR && canDestroyBlock(block))
+                                if (!block.isAir(iblockstate, this.world, blockpos)
+                                       && block.canEntityDestroy(iblockstate, world, blockpos, this)
+                                       && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                                 {
+                                    if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR, 0).isCancelled()) {
+                                        continue;
+                                    }
                                     flag = this.world.destroyBlock(blockpos, true) || flag;
                                 }
                             }
@@ -444,7 +464,7 @@
 
             if (this.ticksExisted % 20 == 0)
             {
-                this.heal(1.0F);
+                this.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN);
             }
 
             this.bossInfo.setPercent(this.getHealth() / this.getMaxHealth());
