--- before/net/minecraft/entity/boss/EntityDragon.java
+++ after/net/minecraft/entity/boss/EntityDragon.java
@@ -78,6 +78,8 @@
     private final int[] neighbors = new int[24];
     private final PathHeap pathFindQueue = new PathHeap();
 
+    private net.minecraft.world.Explosion explosionSource = new net.minecraft.world.Explosion(null, this, Double.NaN, Double.NaN, Double.NaN, Float.NaN, true, true); // CraftBukkit - reusable source for CraftTNTPrimed.getSource()
+
     public EntityDragon(World worldIn)
     {
         super(worldIn);
@@ -283,7 +285,7 @@
 
                     Vec3d vec3d = iphase.getTargetLocation();
 
-                    if (vec3d != null)
+                    if (vec3d != null && iphase.getType() != PhaseList.HOVER) // CraftBukkit - Don't move when hovering
                     {
                         double d6 = vec3d.x - this.posX;
                         double d7 = vec3d.y - this.posY;
@@ -477,7 +479,13 @@
             }
             else if (this.ticksExisted % 10 == 0 && this.getHealth() < this.getMaxHealth())
             {
-                this.setHealth(this.getHealth() + 1.0F);
+//                this.setHealth(this.getHealth() + 1.0F);
+                org.bukkit.event.entity.EntityRegainHealthEvent event = new org.bukkit.event.entity.EntityRegainHealthEvent(this.getBukkitEntity(), 1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.ENDER_CRYSTAL);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    this.setHealth((float) (this.getHealth() + event.getAmount()));
+                }
             }
         }
 
@@ -554,7 +562,10 @@
         int j1 = MathHelper.floor(p_70972_1_.maxZ);
         boolean flag = false;
         boolean flag1 = false;
-
+        // CraftBukkit start - Create a list to hold all the destroyed blocks
+        List<org.bukkit.block.Block> destroyedBlocks = new java.util.ArrayList<>();
+        org.bukkit.craftbukkit.v1_12_R1.CraftWorld craftWorld = this.world.getWorld();
+        // CraftBukkit end
         for (int k1 = i; k1 <= l; k1++)
         {
             for (int l1 = j; l1 <= i1; l1++)
@@ -565,18 +576,14 @@
                     IBlockState iblockstate = this.world.getBlockState(blockpos);
                     Block block = iblockstate.getBlock();
 
-                    if (iblockstate.getMaterial() != Material.AIR && iblockstate.getMaterial() != Material.FIRE)
+                    if (!block.isAir(iblockstate, this.world, blockpos) && iblockstate.getMaterial() != Material.FIRE)
                     {
-                        if (!this.world.getGameRules().getBoolean("mobGriefing"))
+                        if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this))
                         {
                             flag = true;
                         }
-                        else if (block == Blocks.BARRIER
-                                 || block == Blocks.OBSIDIAN
-                                 || block == Blocks.END_STONE
-                                 || block == Blocks.BEDROCK
-                                 || block == Blocks.END_PORTAL
-                                 || block == Blocks.END_PORTAL_FRAME)
+                        else if (!block.canEntityDestroy(iblockstate, this.world, blockpos, this)
+                            || !net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this, blockpos, iblockstate))
                         {
                             flag = true;
                         }
@@ -586,7 +593,11 @@
                                  && block != Blocks.IRON_BARS
                                  && block != Blocks.END_GATEWAY)
                         {
-                            flag1 = this.world.setBlockToAir(blockpos) || flag1;
+                            // CraftBukkit start - Add blocks to list rather than destroying them
+                            // flag1 = this.world.setBlockToAir(blockpos) || flag1;
+                            flag1 = true;
+                            destroyedBlocks.add(craftWorld.getBlockAt(k1, l1, i2));
+                            // CraftBukkit end
                         }
                         else
                         {
@@ -596,6 +607,41 @@
                 }
             }
         }
+
+        // CraftBukkit start - Set off an EntityExplodeEvent for the dragon exploding all these blocks
+        org.bukkit.entity.Entity bukkitEntity = this.getBukkitEntity();
+        org.bukkit.event.entity.EntityExplodeEvent event = new org.bukkit.event.entity.EntityExplodeEvent(bukkitEntity, bukkitEntity.getLocation(), destroyedBlocks, 0F);
+        bukkitEntity.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            // This flag literally means 'Dragon hit something hard' (Obsidian, White Stone or Bedrock) and will cause the dragon to slow down.
+            // We should consider adding an event extension for it, or perhaps returning true if the event is cancelled.
+            return flag;
+        } else if (event.getYield() == 0F) {
+            // Yield zero ==> no drops
+            for (org.bukkit.block.Block block : event.blockList()) {
+                this.world.setBlockToAir(new BlockPos(block.getX(), block.getY(), block.getZ()));
+            }
+        } else {
+            for (org.bukkit.block.Block block : event.blockList()) {
+                org.bukkit.Material blockId = block.getType();
+                if (blockId == org.bukkit.Material.AIR) {
+                    continue;
+                }
+
+                int blockX = block.getX();
+                int blockY = block.getY();
+                int blockZ = block.getZ();
+
+                Block nmsBlock = org.bukkit.craftbukkit.v1_12_R1.util.CraftMagicNumbers.getBlock(blockId);
+                if (nmsBlock.canDropFromExplosion(explosionSource)) {
+                    nmsBlock.dropBlockAsItemWithChance(this.world, new BlockPos(blockX, blockY, blockZ), nmsBlock.getDefaultState(), event.getYield(), 0);
+                }
+                nmsBlock.onExplosionDestroy(world, new BlockPos(blockX, blockY, blockZ), explosionSource);
+
+                this.world.setBlockToAir(new BlockPos(blockX, blockY, blockZ));
+            }
+        }
+        // CraftBukkit end
 
         if (flag1)
         {
@@ -715,7 +761,24 @@
 
             if (this.deathTicks == 1)
             {
-                this.world.playBroadcastSound(1028, new BlockPos(this), 0);
+                // CraftBukkit start - Use relative location for far away sounds
+                int viewDistance = this.world.getServer().getViewDistance() * 16;
+                for (net.minecraft.entity.player.EntityPlayerMP player : this.world.getMinecraftServer().getPlayerList().playerEntityList) {
+                    if (catserver.server.CatServer.getConfig().fixPlayBossSoundToOtherWorld && player.world != this.world) continue; // CatServer
+                    double deltaX = this.posX - player.posX;
+                    double deltaZ = this.posZ - player.posZ;
+                    double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                    if (world.spigotConfig.dragonDeathSoundRadius > 0 && distanceSquared > world.spigotConfig.dragonDeathSoundRadius * world.spigotConfig.dragonDeathSoundRadius) continue; // Spigot
+                    if (distanceSquared > viewDistance * viewDistance) {
+                        double deltaLength = Math.sqrt(distanceSquared);
+                        double relativeX = player.posX + (deltaX / deltaLength) * viewDistance;
+                        double relativeZ = player.posZ + (deltaZ / deltaLength) * viewDistance;
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1028, new BlockPos((int) relativeX, (int) this.posY, (int) relativeZ), 0, true));
+                    } else {
+                        player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEffect(1028, new BlockPos((int) this.posX, (int) this.posY, (int) this.posZ), 0, true));
+                    }
+                }
+                // CraftBukkit end
             }
         }
 
@@ -741,6 +804,8 @@
 
     private void dropExperience(int p_184668_1_)
     {
+        p_184668_1_ = net.minecraftforge.event.ForgeEventFactory.getExperienceDrop(this, this.attackingPlayer, p_184668_1_);
+         
         while (p_184668_1_ > 0)
         {
             int i = EntityXPOrb.getXPSplit(p_184668_1_);
