--- before/net/minecraft/item/ItemRedstone.java
+++ after/net/minecraft/item/ItemRedstone.java
@@ -34,8 +34,8 @@
         BlockPos blockpos = flag ? pos : pos.offset(facing);
         ItemStack itemstack = player.getHeldItem(hand);
 
-        if (player.canPlayerEdit(blockpos, facing, itemstack)
-                && worldIn.mayPlace(worldIn.getBlockState(blockpos).getBlock(), blockpos, false, facing, null)
+        if (!itemstack.isEmpty() && player.canPlayerEdit(blockpos, facing, itemstack)
+                && worldIn.mayPlace(worldIn.getBlockState(blockpos).getBlock(), blockpos, false, facing, player)
                 && Blocks.REDSTONE_WIRE.canPlaceBlockAt(worldIn, blockpos))
         {
             worldIn.setBlockState(blockpos, Blocks.REDSTONE_WIRE.getDefaultState());
