--- before/net/minecraft/item/ItemBow.java
+++ after/net/minecraft/item/ItemBow.java
@@ -40,7 +40,7 @@
                 }
                 else
                 {
-                    return entityIn.getActiveItemStack().getItem() != Items.BOW
+                    return !(entityIn.getActiveItemStack().getItem() instanceof ItemBow)
                            ? 0.0F
                            : (stack.getMaxItemUseDuration() - entityIn.getItemInUseCount()) / 20.0F;
                 }
@@ -98,6 +98,10 @@
             boolean flag = entityplayer.capabilities.isCreativeMode || EnchantmentHelper.getEnchantmentLevel(Enchantments.INFINITY, stack) > 0;
             ItemStack itemstack = this.findAmmo(entityplayer);
 
+            int i = this.getMaxItemUseDuration(stack) - timeLeft;
+            i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(stack, worldIn, entityplayer, i, !itemstack.isEmpty() || flag);
+            if (i < 0) return;
+
             if (!itemstack.isEmpty() || flag)
             {
                 if (itemstack.isEmpty())
@@ -105,17 +109,17 @@
                     itemstack = new ItemStack(Items.ARROW);
                 }
 
-                int i = this.getMaxItemUseDuration(stack) - timeLeft;
                 float f = getArrowVelocity(i);
 
                 if (!(f < 0.1))
                 {
-                    boolean flag1 = flag && itemstack.getItem() == Items.ARROW;
+                    boolean flag1 = entityplayer.capabilities.isCreativeMode || (itemstack.getItem() instanceof ItemArrow && ((ItemArrow) itemstack.getItem()).isInfinite(itemstack, stack, entityplayer));
 
                     if (!worldIn.isRemote)
                     {
                         ItemArrow itemarrow = (ItemArrow)(itemstack.getItem() instanceof ItemArrow ? itemstack.getItem() : Items.ARROW);
                         EntityArrow entityarrow = itemarrow.createArrow(worldIn, itemstack, entityplayer);
+                        entityarrow = this.customizeArrow(entityarrow);
                         entityarrow.shoot(entityplayer, entityplayer.rotationPitch, entityplayer.rotationYaw, 0.0F, f * 3.0F, 1.0F);
 
                         if (f == 1.0F)
@@ -139,7 +143,18 @@
 
                         if (EnchantmentHelper.getEnchantmentLevel(Enchantments.FLAME, stack) > 0)
                         {
-                            entityarrow.setFire(100);
+                            // entityarrow.setFire(100);
+                            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustEvent(entityarrow.getBukkitEntity(), 100);
+                            entityarrow.world.getServer().getPluginManager().callEvent(event);
+
+                            if (!event.isCancelled()) {
+                                entityarrow.setFire(event.getDuration());
+                            }
+                        }
+                        org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityShootBowEvent(entityplayer, itemstack, entityarrow, f);
+                        if (event.isCancelled()) {
+                            event.getProjectile().remove();
+                            return;
                         }
 
                         stack.damageItem(1, entityplayer);
@@ -151,7 +166,15 @@
                             entityarrow.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY;
                         }
 
-                        worldIn.spawnEntity(entityarrow);
+                        // worldIn.spawnEntity(entityarrow);
+                        if (event.getProjectile() == entityarrow.getBukkitEntity()) {
+                            if (!worldIn.spawnEntity(entityarrow)) {
+                                if (entityplayer instanceof net.minecraft.entity.player.EntityPlayerMP) {
+                                    ((net.minecraft.entity.player.EntityPlayerMP) entityplayer).getBukkitEntity().updateInventory();
+                                }
+                                return;
+                            }
+                        }
                     }
 
                     worldIn.playSound(
@@ -212,6 +235,9 @@
         ItemStack itemstack = playerIn.getHeldItem(handIn);
         boolean flag = !this.findAmmo(playerIn).isEmpty();
 
+        ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onArrowNock(itemstack, worldIn, playerIn, handIn, flag);
+        if (ret != null) return ret;
+
         if (playerIn.capabilities.isCreativeMode || flag)
         {
             playerIn.setActiveHand(handIn);
@@ -227,5 +253,10 @@
     public int getItemEnchantability()
     {
         return 1;
+    }
+
+    public EntityArrow customizeArrow(EntityArrow arrow)
+    {
+        return arrow;
     }
 }
