--- before/net/minecraft/item/ItemBucket.java
+++ after/net/minecraft/item/ItemBucket.java
@@ -41,6 +41,8 @@
         boolean flag = this.containedBlock == Blocks.AIR;
         ItemStack itemstack = playerIn.getHeldItem(handIn);
         RayTraceResult raytraceresult = this.rayTrace(worldIn, playerIn, flag);
+        ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onBucketUse(playerIn, worldIn, itemstack, raytraceresult);
+        if (ret != null) return ret;
 
         if (raytraceresult == null)
         {
@@ -71,17 +73,27 @@
 
                     if (material == Material.WATER && iblockstate.getValue(BlockLiquid.LEVEL) == 0)
                     {
+                        org.bukkit.event.player.PlayerBucketFillEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerBucketFillEvent(playerIn, blockpos.getX(), blockpos.getY(), blockpos.getZ(), null, itemstack, Items.WATER_BUCKET);
+
+                        if (event.isCancelled()) {
+                            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                        }
                         worldIn.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 11);
                         playerIn.addStat(StatList.getObjectUseStats(this));
                         playerIn.playSound(SoundEvents.ITEM_BUCKET_FILL, 1.0F, 1.0F);
-                        return new ActionResult<>(EnumActionResult.SUCCESS, this.fillBucket(itemstack, playerIn, Items.WATER_BUCKET));
+                        return new ActionResult<>(EnumActionResult.SUCCESS, this.fillBucket(itemstack, playerIn, Items.WATER_BUCKET, event.getItemStack()));
                     }
                     else if (material == Material.LAVA && iblockstate.getValue(BlockLiquid.LEVEL) == 0)
                     {
+                        org.bukkit.event.player.PlayerBucketFillEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerBucketFillEvent(playerIn, blockpos.getX(), blockpos.getY(), blockpos.getZ(), null, itemstack, Items.LAVA_BUCKET);
+
+                        if (event.isCancelled()) {
+                            return new ActionResult<>(EnumActionResult.FAIL, itemstack);
+                        }
                         playerIn.playSound(SoundEvents.ITEM_BUCKET_FILL_LAVA, 1.0F, 1.0F);
                         worldIn.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 11);
                         playerIn.addStat(StatList.getObjectUseStats(this));
-                        return new ActionResult<>(EnumActionResult.SUCCESS, this.fillBucket(itemstack, playerIn, Items.LAVA_BUCKET));
+                        return new ActionResult<>(EnumActionResult.SUCCESS, this.fillBucket(itemstack, playerIn, Items.LAVA_BUCKET, event.getItemStack()));
                     }
                     else
                     {
@@ -98,7 +110,7 @@
                 {
                     return new ActionResult<>(EnumActionResult.FAIL, itemstack);
                 }
-                else if (this.tryPlaceContainedLiquid(playerIn, worldIn, blockpos1))
+                else if (this.tryPlaceContainedLiquid(playerIn, worldIn, blockpos1, raytraceresult.sideHit, blockpos, itemstack))
                 {
                     if (playerIn instanceof EntityPlayerMP)
                     {
@@ -144,8 +156,40 @@
         }
     }
 
+    // CraftBukkit - added ob.ItemStack result - TODO: Is this... the right way to handle this?
+    private ItemStack fillBucket(ItemStack emptyBuckets, EntityPlayer player, Item fullBucket, org.bukkit.inventory.ItemStack result)
+    {
+        if (player.capabilities.isCreativeMode)
+        {
+            return emptyBuckets;
+        }
+        else
+        {
+            emptyBuckets.shrink(1);
+
+            if (emptyBuckets.isEmpty())
+            {
+                // return new ItemStack(fullBucket);
+                return org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(result);
+            }
+            else
+            {
+                if (!player.inventory.addItemStackToInventory(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(result)))
+                {
+                    player.dropItem(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(result), false);
+                }
+
+                return emptyBuckets;
+            }
+        }
+    }
+
     public boolean tryPlaceContainedLiquid(@Nullable EntityPlayer player, World worldIn, BlockPos posIn)
     {
+        return tryPlaceContainedLiquid(player, worldIn, posIn, EnumFacing.UP, posIn.down(), ItemStack.EMPTY); // CatServer - fake clicked position to call event
+    }
+
+    public boolean tryPlaceContainedLiquid(@Nullable EntityPlayer player, World worldIn, BlockPos posIn, EnumFacing enumdirection, BlockPos clicked, ItemStack itemstack) {
         if (this.containedBlock == Blocks.AIR)
         {
             return false;
@@ -163,6 +207,14 @@
             }
             else
             {
+                if (player != null) {
+                    org.bukkit.event.player.PlayerBucketEmptyEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerBucketEmptyEvent(player, clicked.getX(), clicked.getY(), clicked.getZ(), enumdirection, itemstack);
+                    if (event.isCancelled()) {
+                        // TODO: inventory not updated
+                        return false;
+                    }
+                }
+
                 if (worldIn.provider.doesWaterVaporize() && this.containedBlock == Blocks.FLOWING_WATER)
                 {
                     int l = posIn.getX();
@@ -196,6 +248,18 @@
 
                 return true;
             }
+        }
+    }
+
+    @Override
+    public net.minecraftforge.common.capabilities.ICapabilityProvider initCapabilities(ItemStack stack, @Nullable net.minecraft.nbt.NBTTagCompound nbt) {
+        if (this.getClass() == ItemBucket.class)
+        {
+            return new net.minecraftforge.fluids.capability.wrappers.FluidBucketWrapper(stack);
+        }
+        else
+        {
+            return super.initCapabilities(stack, nbt);
         }
     }
 }
