--- before/net/minecraft/world/gen/layer/GenLayerVoronoiZoom.java
+++ after/net/minecraft/world/gen/layer/GenLayerVoronoiZoom.java
@@ -11,25 +11,28 @@
     @Override
     public int[] getInts(int areaX, int areaY, int areaWidth, int areaHeight)
     {
-        areaX -= 2;
-        areaY -= 2;
+        // CREF start - FIXED: in some cases, arraycopy throws an ArrayIndexOutOfBoundsException
+        //areaX -= 2;
+        //areaY -= 2;
         int i = areaX >> 2;
         int j = areaY >> 2;
-        int k = (areaWidth >> 2) + 2;
-        int l = (areaHeight >> 2) + 2;
+        //int k = (areaWidth >> 2) + 2;
+        //int l = (areaHeight >> 2) + 2;
+        int k = (areaWidth >> 2) + 3;
+        int l = (areaHeight >> 2) + 3;
+        // CREF end
         int[] aint = this.parent.getInts(i, j, k, l);
         int i1 = k - 1 << 2;
         int j1 = l - 1 << 2;
         int[] aint1 = IntCache.getIntCache(i1 * j1);
 
-        for (int k1 = 0; k1 < l - 1; k1++)
+        for (int k1 = 0; k1 < l - 1; ++k1)
         {
             int l1 = 0;
-            int i2 = aint[l1 + 0 + (k1 + 0) * k];
+            int i2 = aint[l1 + k1 * k];
 
-            for (int j2 = aint[l1 + 0 + (k1 + 1) * k]; l1 < k - 1; l1++)
+            for (int j2 = aint[l1 + (k1 + 1) * k]; l1 < k - 1; ++l1)
             {
-                double d0 = 3.6;
                 this.initChunkSeed(l1 + i << 2, k1 + j << 2);
                 double d1 = (this.nextInt(1024) / 1024.0 - 0.5) * 3.6;
                 double d2 = (this.nextInt(1024) / 1024.0 - 0.5) * 3.6;
@@ -42,14 +45,14 @@
                 this.initChunkSeed(l1 + i + 1 << 2, k1 + j + 1 << 2);
                 double d7 = (this.nextInt(1024) / 1024.0 - 0.5) * 3.6 + 4.0;
                 double d8 = (this.nextInt(1024) / 1024.0 - 0.5) * 3.6 + 4.0;
-                int k2 = aint[l1 + 1 + (k1 + 0) * k] & 0xFF;
+                int k2 = aint[l1 + 1 + k1 * k] & 0xFF;
                 int l2 = aint[l1 + 1 + (k1 + 1) * k] & 0xFF;
 
-                for (int i3 = 0; i3 < 4; i3++)
+                for (int i3 = 0; i3 < 4; ++i3)
                 {
                     int j3 = ((k1 << 2) + i3) * i1 + (l1 << 2);
 
-                    for (int k3 = 0; k3 < 4; k3++)
+                    for (int k3 = 0; k3 < 4; ++k3)
                     {
                         double d9 = (i3 - d2) * (i3 - d2) + (k3 - d1) * (k3 - d1);
                         double d10 = (i3 - d4) * (i3 - d4) + (k3 - d3) * (k3 - d3);
@@ -82,7 +85,7 @@
 
         int[] aint2 = IntCache.getIntCache(areaWidth * areaHeight);
 
-        for (int l3 = 0; l3 < areaHeight; l3++)
+        for (int l3 = 0; l3 < areaHeight; ++l3)
         {
             System.arraycopy(aint1, (l3 + (areaY & 3)) * i1 + (areaX & 3), aint2, l3 * areaWidth, areaWidth);
         }
