--- before/net/minecraft/world/gen/structure/ComponentScatteredFeaturePieces.java
+++ after/net/minecraft/world/gen/structure/ComponentScatteredFeaturePieces.java
@@ -93,7 +93,7 @@
                     false
                 );
 
-                for (int i = 1; i <= 9; i++)
+                for (int i = 1; i <= 9; ++i)
                 {
                     this.fillWithBlocks(
                         worldIn,
@@ -123,9 +123,9 @@
                     );
                 }
 
-                for (int i2 = 0; i2 < this.width; i2++)
+                for (int i2 = 0; i2 < this.width; ++i2)
                 {
-                    for (int j = 0; j < this.depth; j++)
+                    for (int j = 0; j < this.depth; ++j)
                     {
                         int k = -5;
                         this.replaceAirAndLiquidDownwards(worldIn, Blocks.SANDSTONE.getDefaultState(), i2, -5, j, structureBoundingBoxIn);
@@ -615,16 +615,16 @@
                     int j = 0;
                     BlockPos.MutableBlockPos blockpos$mutableblockpos = new BlockPos.MutableBlockPos();
 
-                    for (int k = this.boundingBox.minZ; k <= this.boundingBox.maxZ; k++)
+                    for (int k = this.boundingBox.minZ; k <= this.boundingBox.maxZ; ++k)
                     {
-                        for (int l = this.boundingBox.minX; l <= this.boundingBox.maxX; l++)
+                        for (int l = this.boundingBox.minX; l <= this.boundingBox.maxX; ++l)
                         {
                             blockpos$mutableblockpos.setPos(l, 64, k);
 
                             if (structurebb.isVecInside(blockpos$mutableblockpos))
                             {
                                 i += Math.max(worldIn.getTopSolidOrLiquidBlock(blockpos$mutableblockpos).getY(), worldIn.provider.getAverageGroundLevel());
-                                j++;
+                                ++j;
                             }
                         }
                     }
@@ -685,7 +685,7 @@
                         Template template2 = templatemanager.getTemplate(minecraftserver, IGLOO_BOTTOM_ID);
                         int i = randomIn.nextInt(8) + 4;
 
-                        for (int j = 0; j < i; j++)
+                        for (int j = 0; j < i; ++j)
                         {
                             BlockPos blockpos1 = template.calculateConnectedPos(placementsettings, new BlockPos(3, -1 - j * 3, 5), placementsettings, new BlockPos(1, 2, 1));
                             template1.addBlocksToWorldChunk(worldIn, blockpos.add(blockpos1), placementsettings);
@@ -851,7 +851,7 @@
                     this.setBlockState(worldIn, iblockstate2, 4, 4, 5, structureBoundingBoxIn);
                     this.setBlockState(worldIn, iblockstate3, 7, 4, 5, structureBoundingBoxIn);
 
-                    for (int k = 0; k < 4; k++)
+                    for (int k = 0; k < 4; ++k)
                     {
                         this.setBlockState(worldIn, iblockstate, 5, 0 - k, 6 + k, structureBoundingBoxIn);
                         this.setBlockState(worldIn, iblockstate, 6, 0 - k, 6 + k, structureBoundingBoxIn);
@@ -1205,7 +1205,7 @@
                             entitywitch.enablePersistence();
                             entitywitch.setLocationAndAngles(l + 0.5, i1, k + 0.5, 0.0F, 0.0F);
                             entitywitch.onInitialSpawn(worldIn.getDifficultyForLocation(new BlockPos(l, i1, k)), null);
-                            worldIn.spawnEntity(entitywitch);
+                            worldIn.addEntity(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN);
                         }
                     }
 
