--- before/net/minecraft/init/Bootstrap.java
+++ after/net/minecraft/init/Bootstrap.java
@@ -232,15 +232,41 @@
                 double d0 = source.getX() + enumfacing.getXOffset();
                 double d1 = source.getBlockPos().getY() + enumfacing.getYOffset() + 0.2F;
                 double d2 = source.getZ() + enumfacing.getZOffset();
-                Entity entity = ItemMonsterPlacer.spawnCreature(source.getWorld(), ItemMonsterPlacer.getNamedIdFrom(stack), d0, d1, d2);
-
+//                Entity entity = ItemMonsterPlacer.spawnCreature(source.getWorld(), ItemMonsterPlacer.getNamedIdFrom(stack), d0, d1, d2);
+
+                World world = source.getWorld();
+                ItemStack itemstack1 = stack.splitStack(1);
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2));
+                world.getServer().getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    stack.grow(1);
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    stack.grow(1);
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
+                itemstack1 = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+
+                Entity entity = ItemMonsterPlacer.spawnCreature(source.getWorld(), ItemMonsterPlacer.getNamedIdFrom(stack), event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DISPENSE_EGG);
                 if (entity instanceof EntityLivingBase && stack.hasDisplayName())
                 {
                     entity.setCustomNameTag(stack.getDisplayName());
                 }
 
                 ItemMonsterPlacer.applyItemEntityDataToEntity(source.getWorld(), null, stack, entity);
-                stack.shrink(1);
+//                stack.shrink(1); // Handled during event processing
                 return stack;
             }
         });
@@ -253,7 +279,33 @@
                 double d0 = source.getX() + enumfacing.getXOffset();
                 double d1 = source.getBlockPos().getY() + 0.2F;
                 double d2 = source.getZ() + enumfacing.getZOffset();
-                EntityFireworkRocket entityfireworkrocket = new EntityFireworkRocket(source.getWorld(), d0, d1, d2, stack);
+//                EntityFireworkRocket entityfireworkrocket = new EntityFireworkRocket(source.getWorld(), d0, d1, d2, stack);
+                World world = source.getWorld();
+                ItemStack itemstack1 = stack.splitStack(1);
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1, d2));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    stack.grow(1);
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    stack.grow(1);
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
+                itemstack1 = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                EntityFireworkRocket entityfireworkrocket = new EntityFireworkRocket(source.getWorld(), event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), itemstack1);
                 source.getWorld().spawnEntity(entityfireworkrocket);
                 stack.shrink(1);
                 return stack;
@@ -279,8 +331,34 @@
                 double d3 = random.nextGaussian() * 0.05 + enumfacing.getXOffset();
                 double d4 = random.nextGaussian() * 0.05 + enumfacing.getYOffset();
                 double d5 = random.nextGaussian() * 0.05 + enumfacing.getZOffset();
-                world.spawnEntity(new EntitySmallFireball(world, d0, d1, d2, d3, d4, d5));
-                stack.shrink(1);
+//                world.spawnEntity(new EntitySmallFireball(world, d0, d1, d2, d3, d4, d5));
+                ItemStack itemstack1 = stack.splitStack(1);
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d3, d4, d5));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    stack.grow(1);
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    stack.grow(1);
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
+                EntitySmallFireball fireball = new EntitySmallFireball(world, d0, d1, d2, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
+                fireball.projectileSource = new org.bukkit.craftbukkit.v1_12_R1.projectiles.CraftBlockProjectileSource((TileEntityDispenser) source.getBlockTileEntity());
+
+//                stack.shrink(1);
                 return stack;
             }
             @Override
@@ -303,13 +381,57 @@
             {
                 ItemBucket itembucket = (ItemBucket)stack.getItem();
                 BlockPos blockpos = source.getBlockPos().offset(source.getBlockState().getValue(BlockDispenser.FACING));
-                return itembucket.tryPlaceContainedLiquid(null, source.getWorld(), blockpos)
-                       ? new ItemStack(Items.BUCKET)
-                       : this.dispenseBehavior.dispense(source, stack);
+//                return itembucket.tryPlaceContainedLiquid((EntityPlayer)null, source.getWorld(), blockpos) ? new ItemStack(Items.BUCKET) : this.dispenseBehavior.dispense(source, stack);
+                World world = source.getWorld();
+                int x = blockpos.getX();
+                int y = blockpos.getY();
+                int z = blockpos.getZ();
+                if (world.isAirBlock(blockpos) || !world.getBlockState(blockpos).getMaterial().isSolid()) {
+                    org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                    org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                    org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(x, y, z));
+                    world.getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return stack;
+                    }
+
+                    if (!event.getItem().equals(craftItem)) {
+                        // Chain to handler for new item
+                        ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                        IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                        if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                            idispensebehavior.dispense(source, eventStack);
+                            return stack;
+                        }
+                    }
+
+                    itembucket = (ItemBucket) org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem()).getItem();
+                }
+
+                if (itembucket.tryPlaceContainedLiquid(null, source.getWorld(), blockpos)) {
+                    // CraftBukkit start - Handle stacked buckets
+                    Item item = Items.BUCKET;
+                    stack.shrink(1);
+                    if (stack.isEmpty()) {
+                        stack.setItem(Items.BUCKET);
+                        stack.setCount(1);
+                    } else if (((TileEntityDispenser) source.getBlockTileEntity()).addItemStack(new ItemStack(item)) < 0) {
+                        this.dispenseBehavior.dispense(source, new ItemStack(item));
+                    }
+                    // CraftBukkit end
+                    return stack;
+                } else {
+                    return this.dispenseBehavior.dispense(source, stack);
+                }
             }
         };
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.LAVA_BUCKET, ibehaviordispenseitem);
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.WATER_BUCKET, ibehaviordispenseitem);
+        BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.MILK_BUCKET, net.minecraftforge.fluids.DispenseFluidContainer.getInstance());
+        BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.BUCKET, net.minecraftforge.fluids.DispenseFluidContainer.getInstance());
+        if (false)
         BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.putObject(Items.BUCKET, new BehaviorDefaultDispenseItem()
         {
             private final BehaviorDefaultDispenseItem dispenseBehavior = new BehaviorDefaultDispenseItem();
@@ -337,6 +459,26 @@
                     item = Items.LAVA_BUCKET;
                 }
 
+                org.bukkit.block.Block bukkitBlock = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
                 world.setBlockToAir(blockpos);
                 stack.shrink(1);
 
@@ -361,17 +503,40 @@
             protected ItemStack dispenseStack(IBlockSource source, ItemStack stack)
             {
                 World world = source.getWorld();
+
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
                 this.successful = true;
                 BlockPos blockpos = source.getBlockPos().offset(source.getBlockState().getValue(BlockDispenser.FACING));
 
                 if (world.isAirBlock(blockpos))
                 {
+                    if (!org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos.getX(), blockpos.getY(), blockpos.getZ(), source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ()).isCancelled()) {
                     world.setBlockState(blockpos, Blocks.FIRE.getDefaultState());
 
                     if (stack.attemptDamageItem(1, world.rand, null))
                     {
                         stack.setCount(0);
                     }
+                    }
                 }
                 else if (world.getBlockState(blockpos).getBlock() == Blocks.TNT)
                 {
@@ -397,6 +562,27 @@
                 {
                     World world = source.getWorld();
                     BlockPos blockpos = source.getBlockPos().offset(source.getBlockState().getValue(BlockDispenser.FACING));
+                    org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                    org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                    org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(0, 0, 0));
+                    world.getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return stack;
+                    }
+
+                    if (!event.getItem().equals(craftItem)) {
+                        // Chain to handler for new item
+                        ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                        IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                        if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                            idispensebehavior.dispense(source, eventStack);
+                            return stack;
+                        }
+                    }
+
+                    world.captureTreeGeneration = true;
 
                     if (ItemDye.applyBonemeal(stack, world, blockpos))
                     {
@@ -410,6 +596,27 @@
                         this.successful = false;
                     }
 
+                    world.captureTreeGeneration = false;
+                    if (world.capturedBlockSnapshots.size() > 0) {
+                        org.bukkit.TreeType treeType = net.minecraft.block.BlockSapling.treeType;
+                        net.minecraft.block.BlockSapling.treeType = null;
+                        org.bukkit.Location location = new org.bukkit.Location(world.getWorld(), blockpos.getX(), blockpos.getY(), blockpos.getZ());
+                        java.util.List<org.bukkit.block.BlockState> blockstates = new java.util.ArrayList();
+                        for (net.minecraftforge.common.util.BlockSnapshot snapshot : world.capturedBlockSnapshots) {
+                            blockstates.add(new org.bukkit.craftbukkit.v1_12_R1.block.CraftBlockState(snapshot));
+                        }
+                        world.capturedBlockSnapshots.clear();
+                        org.bukkit.event.world.StructureGrowEvent structureEvent = null;
+                        if (treeType != null) {
+                            structureEvent = new org.bukkit.event.world.StructureGrowEvent(location, treeType, false, null, blockstates);
+                            org.bukkit.Bukkit.getPluginManager().callEvent(structureEvent);
+                        }
+                        if (structureEvent == null || !structureEvent.isCancelled()) {
+                            for (org.bukkit.block.BlockState blockstate : blockstates) {
+                                blockstate.update(true);
+                            }
+                        }
+                    }
                     return stack;
                 }
                 else
@@ -428,9 +635,31 @@
             {
                 World world = source.getWorld();
                 BlockPos blockpos = source.getBlockPos().offset(source.getBlockState().getValue(BlockDispenser.FACING));
-                EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(
-                    world, blockpos.getX() + 0.5, blockpos.getY(), blockpos.getZ() + 0.5, null
-                );
+//                EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, (double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D, (EntityLivingBase)null);
+                ItemStack itemstack1 = stack.splitStack(1);
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector((double) blockpos.getX() + 0.5D, (double) blockpos.getY(), (double) blockpos.getZ() + 0.5D));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    stack.grow(1);
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    stack.grow(1);
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
+                EntityTNTPrimed entitytntprimed = new EntityTNTPrimed(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ(), null);
                 world.spawnEntity(entitytntprimed);
                 world.playSound(
                     null,
@@ -442,7 +671,7 @@
                     1.0F,
                     1.0F
                 );
-                stack.shrink(1);
+//                stack.shrink(1); // CraftBukkit - handled above
                 return stack;
             }
         }
@@ -456,6 +685,27 @@
                 EnumFacing enumfacing = source.getBlockState().getValue(BlockDispenser.FACING);
                 BlockPos blockpos = source.getBlockPos().offset(enumfacing);
                 BlockSkull blockskull = Blocks.SKULL;
+
+                org.bukkit.block.Block bukkitBlock = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
                 this.successful = true;
 
                 if (world.isAirBlock(blockpos) && blockskull.canDispenserPlace(world, blockpos, stack))
@@ -520,6 +770,27 @@
                 World world = source.getWorld();
                 BlockPos blockpos = source.getBlockPos().offset(source.getBlockState().getValue(BlockDispenser.FACING));
                 BlockPumpkin blockpumpkin = (BlockPumpkin)Blocks.PUMPKIN;
+
+                org.bukkit.block.Block bukkitBlock = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
                 this.successful = true;
 
                 if (world.isAirBlock(blockpos) && blockpumpkin.canDispenserPlace(world, blockpos))
@@ -557,6 +828,7 @@
         if (!alreadyRegistered)
         {
             alreadyRegistered = true;
+            if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
             redirectOutputToLog();
             SoundEvent.registerSounds();
             Block.registerBlocks();
@@ -592,6 +864,7 @@
                     LOGGER.error("Errors with built-in loot tables");
                 }
             }
+            net.minecraftforge.registries.GameData.vanillaSnapshot();
         }
     }
 
@@ -651,11 +924,37 @@
                     d3 = 0.0;
                 }
 
-                EntityBoat entityboat = new EntityBoat(world, d0, d1 + d3, d2);
+//                EntityBoat entityboat = new EntityBoat(world, d0, d1 + d3, d2);
+                ItemStack itemstack1 = stack.splitStack(1);
+                org.bukkit.block.Block block = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack1);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(block, craftItem.clone(), new org.bukkit.util.Vector(d0, d1 + d3, d2));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    stack.grow(1);
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    stack.grow(1);
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
+                EntityBoat entityboat = new EntityBoat(world, event.getVelocity().getX(), event.getVelocity().getY(), event.getVelocity().getZ());
                 entityboat.setBoatType(this.boatType);
                 entityboat.rotationYaw = enumfacing.getHorizontalAngle();
-                world.spawnEntity(entityboat);
-                stack.shrink(1);
+                if (!world.spawnEntity(entityboat))
+                    stack.grow(1);
+//                world.spawnEntity(entityboat);
+//                stack.shrink(1); // CraftBukkit - handled during event processing
                 return stack;
             }
 
@@ -690,6 +989,27 @@
                 World world = source.getWorld();
                 EnumFacing enumfacing = source.getBlockState().getValue(BlockDispenser.FACING);
                 BlockPos blockpos = source.getBlockPos().offset(enumfacing);
+
+                org.bukkit.block.Block bukkitBlock = world.getWorld().getBlockAt(source.getBlockPos().getX(), source.getBlockPos().getY(), source.getBlockPos().getZ());
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(stack);
+
+                org.bukkit.event.block.BlockDispenseEvent event = new org.bukkit.event.block.BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return stack;
+                }
+
+                if (!event.getItem().equals(craftItem)) {
+                    // Chain to handler for new item
+                    ItemStack eventStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getItem());
+                    IBehaviorDispenseItem idispensebehavior = BlockDispenser.DISPENSE_BEHAVIOR_REGISTRY.getObject(eventStack.getItem());
+                    if (idispensebehavior != IBehaviorDispenseItem.DEFAULT_BEHAVIOR && idispensebehavior != this) {
+                        idispensebehavior.dispense(source, eventStack);
+                        return stack;
+                    }
+                }
+
                 this.successful = world.mayPlace(block, blockpos, false, EnumFacing.DOWN, null);
 
                 if (this.successful)
@@ -699,6 +1019,8 @@
                     world.setBlockState(blockpos, iblockstate);
                     TileEntity tileentity = world.getTileEntity(blockpos);
                     ItemStack itemstack = stack.splitStack(1);
+
+                    if (tileentity == null) return stack; // CatServer - fix mojang bug
 
                     if (itemstack.hasTagCompound())
                     {
