--- before/net/minecraft/pathfinding/PathNavigate.java
+++ after/net/minecraft/pathfinding/PathNavigate.java
@@ -8,6 +8,7 @@
 import net.minecraft.entity.SharedMonsterAttributes;
 import net.minecraft.entity.ai.attributes.IAttributeInstance;
 import net.minecraft.init.Blocks;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
@@ -22,11 +23,11 @@
     @Nullable
     protected Path currentPath;
     protected double speed;
-    private final IAttributeInstance pathSearchRange;
+    public final IAttributeInstance pathSearchRange;
     protected int totalTicks;
     private int ticksAtLastPos;
-    private Vec3d lastPosCheck = Vec3d.ZERO;
-    private Vec3d timeoutCachedNode = Vec3d.ZERO;
+    public Vec3d lastPosCheck = Vec3d.ZERO; // CatServer - private -> public
+    public Vec3d timeoutCachedNode = Vec3d.ZERO; // CatServer - private -> public
     private long timeoutTimer;
     private long lastTimeoutCheck;
     private double timeoutLimit;
@@ -148,9 +149,27 @@
 
     public boolean tryMoveToEntityLiving(Entity entityIn, double speedIn)
     {
+        // Paper start - Pathfinding optimizations
+        if (this.pathfindFailures > 10 && this.currentPath == null && MinecraftServer.currentTick < this.lastFailure + 40) {
+            return false;
+        }
+
         Path path = this.getPathToEntityLiving(entityIn);
-        return path != null && this.setPath(path, speedIn);
+
+        if (path != null && this.setPath(path, speedIn)) {
+            this.lastFailure = 0;
+            this.pathfindFailures = 0;
+            return true;
+        } else {
+            this.pathfindFailures++;
+            this.lastFailure = MinecraftServer.currentTick;
+            return false;
+        }
+
     }
+    private int lastFailure = 0;
+    private int pathfindFailures = 0;
+    // Paper end
 
     public boolean setPath(@Nullable Path pathentityIn, double speedIn)
     {
@@ -191,7 +210,7 @@
 
     public void onUpdateNavigation()
     {
-        this.totalTicks++;
+        ++this.totalTicks;
 
         if (this.tryUpdatePath)
         {
@@ -240,7 +259,7 @@
         Vec3d vec3d = this.getEntityPosition();
         int i = this.currentPath.getCurrentPathLength();
 
-        for (int j = this.currentPath.getCurrentPathIndex(); j < this.currentPath.getCurrentPathLength(); j++)
+        for (int j = this.currentPath.getCurrentPathIndex(); j < this.currentPath.getCurrentPathLength(); ++j)
         {
             if (this.currentPath.getPathPointFromIndex(j).y != Math.floor(vec3d.y))
             {
@@ -265,7 +284,7 @@
         int l = MathHelper.ceil(this.entity.height);
         int i1 = k;
 
-        for (int j1 = i - 1; j1 >= this.currentPath.getCurrentPathIndex(); j1--)
+        for (int j1 = i - 1; j1 >= this.currentPath.getCurrentPathIndex(); --j1)
         {
             if (this.isDirectPathBetweenPoints(vec3d, this.currentPath.getVectorFromIndex(this.entity, j1), k, l, i1))
             {
@@ -296,7 +315,7 @@
 
             if (vec3d.equals(this.timeoutCachedNode))
             {
-                this.timeoutTimer = this.timeoutTimer + (System.currentTimeMillis() - this.lastTimeoutCheck);
+                this.timeoutTimer += System.currentTimeMillis() - this.lastTimeoutCheck;
             }
             else
             {
@@ -324,6 +343,7 @@
 
     public void clearPath()
     {
+        this.pathfindFailures = 0; this.lastFailure = 0; // Paper - Pathfinding optimizations
         this.currentPath = null;
     }
 
@@ -340,7 +360,7 @@
     {
         if (this.currentPath != null)
         {
-            for (int i = 0; i < this.currentPath.getCurrentPathLength(); i++)
+            for (int i = 0; i < this.currentPath.getCurrentPathLength(); ++i)
             {
                 PathPoint pathpoint = this.currentPath.getPathPointFromIndex(i);
                 PathPoint pathpoint1 = i + 1 < this.currentPath.getCurrentPathLength() ? this.currentPath.getPathPointFromIndex(i + 1) : null;
