--- before/net/minecraft/stats/StatisticsManagerServer.java
+++ after/net/minecraft/stats/StatisticsManagerServer.java
@@ -34,6 +34,14 @@
     {
         this.server = serverIn;
         this.statsFile = statsFileIn;
+        // Spigot start
+        for (String name : org.spigotmc.SpigotConfig.forcedStats.keySet())
+        {
+            TupleIntJsonSerializable wrapper = new TupleIntJsonSerializable();
+            wrapper.setIntegerValue(org.spigotmc.SpigotConfig.forcedStats.get(name));
+            statsData.put(StatList.getOneShotStat(name), wrapper);
+        }
+        // Spigot end
     }
 
     public void readStatFile()
@@ -58,6 +66,7 @@
 
     public void saveStatFile()
     {
+        if (org.spigotmc.SpigotConfig.disableStatSaving) return; // Spigot
         try
         {
             FileUtils.writeStringToFile(this.statsFile, dumpJson(this.statsData));
@@ -84,7 +93,7 @@
 
     public Map<StatBase, TupleIntJsonSerializable> parseJson(String p_150881_1_)
     {
-        JsonElement jsonelement = new JsonParser().parse(p_150881_1_);
+        JsonElement jsonelement = JsonParser.parseString(p_150881_1_);
 
         if (!jsonelement.isJsonObject())
         {
