--- before/net/minecraft/world/BossInfoServer.java
+++ after/net/minecraft/world/BossInfoServer.java
@@ -9,14 +9,12 @@
 import net.minecraft.network.play.server.SPacketUpdateBossInfo;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.text.ITextComponent;
-import net.minecraftforge.fml.relauncher.Side;
-import net.minecraftforge.fml.relauncher.SideOnly;
 
 public class BossInfoServer extends BossInfo
 {
     private final Set<EntityPlayerMP> players = Sets.newHashSet();
     private final Set<EntityPlayerMP> readOnlyPlayers = Collections.unmodifiableSet(this.players);
-    private boolean visible = true;
+    public boolean visible = true;
 
     public BossInfoServer(ITextComponent nameIn, BossInfo.Color colorIn, BossInfo.Overlay overlayIn)
     {
@@ -33,7 +31,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public void setColor(BossInfo.Color colorIn)
     {
@@ -44,7 +41,6 @@
         }
     }
 
-    @SideOnly(Side.CLIENT)
     @Override
     public void setOverlay(BossInfo.Overlay overlayIn)
     {
@@ -101,7 +97,7 @@
         }
     }
 
-    private void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
+    public void sendUpdate(SPacketUpdateBossInfo.Operation operationIn)
     {
         if (this.visible)
         {
