--- before/net/minecraft/world/ServerWorldEventHandler.java
+++ after/net/minecraft/world/ServerWorldEventHandler.java
@@ -98,7 +98,7 @@
             y,
             z,
             volume > 1.0F ? 16.0F * volume : 16.0,
-            this.world.provider.getDimensionType().getId(),
+            this.world.provider.getDimension(),
             new SPacketSoundEffect(soundIn, category, x, y, z, volume, pitch)
         );
     }
@@ -135,7 +135,7 @@
             blockPosIn.getY(),
             blockPosIn.getZ(),
             64.0,
-            this.world.provider.getDimensionType().getId(),
+            this.world.provider.getDimension(),
             new SPacketEffect(type, blockPosIn, data, false)
         );
     }
@@ -149,6 +149,9 @@
     @Override
     public void sendBlockBreakProgress(int breakerId, BlockPos pos, int progress)
     {
+        EntityPlayer entityhuman = null;
+        Entity entity = world.getEntityByID(breakerId);
+        if (entity instanceof EntityPlayer) entityhuman = (EntityPlayer) entity;
         for (EntityPlayerMP entityplayermp : this.server.getPlayerList().getPlayers())
         {
             if (entityplayermp != null && entityplayermp.world == this.world && entityplayermp.getEntityId() != breakerId)
@@ -156,6 +159,10 @@
                 double d0 = pos.getX() - entityplayermp.posX;
                 double d1 = pos.getY() - entityplayermp.posY;
                 double d2 = pos.getZ() - entityplayermp.posZ;
+
+                if (entityhuman != null && entityhuman instanceof EntityPlayerMP && !entityplayermp.getBukkitEntity().canSee(((EntityPlayerMP) entityhuman).getBukkitEntity())) {
+                    continue;
+                }
 
                 if (d0 * d0 + d1 * d1 + d2 * d2 < 1024.0)
                 {
