--- before/net/minecraft/world/WorldProviderHell.java
+++ after/net/minecraft/world/WorldProviderHell.java
@@ -31,7 +31,7 @@
     {
         float f = 0.1F;
 
-        for (int i = 0; i <= 15; i++)
+        for (int i = 0; i <= 15; ++i)
         {
             float f1 = 1.0F - i / 15.0F;
             this.lightBrightnessTable[i] = (1.0F - f1) / (f1 * 3.0F + 1.0F) * 0.9F + 0.1F;
@@ -83,12 +83,14 @@
             @Override
             public double getCenterX()
             {
-                return super.getCenterX() / 8.0;
+                // return super.getCenterX() / 8.0D;
+                return super.getCenterX(); // CraftBukkit
             }
             @Override
             public double getCenterZ()
             {
-                return super.getCenterZ() / 8.0;
+                // return super.getCenterZ() / 8.0D;
+                return super.getCenterZ(); // CraftBukkit
             }
         };
     }
