--- before/net/minecraft/world/World.java
+++ after/net/minecraft/world/World.java
@@ -62,8 +62,16 @@
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
 
-public abstract class World implements IBlockAccess
+public abstract class World implements IBlockAccess, net.minecraftforge.common.capabilities.ICapabilityProvider
 {
+    /**
+     * Used in the getEntitiesWithinAABB functions to expand the search area for entities.
+     * Modders should change this variable to a higher value if it is less then the radius
+     * of one of there entities.
+     */
+    public static double MAX_ENTITY_RADIUS = 2.0D;
+
+    private int tilesThisCycle = 0; // CatRoom - Move to member variable and keep LVT clean
     private int seaLevel = 63;
     protected boolean scheduledUpdatesAreImmediate;
     public final List<Entity> loadedEntityList = Lists.newArrayList();
@@ -85,36 +93,149 @@
     public float thunderingStrength;
     private int lastLightningBolt;
     public final Random rand = new Random();
-    public final WorldProvider provider;
+    public WorldProvider provider;
     protected PathWorldListener pathListener = new PathWorldListener();
     protected List<IWorldEventListener> eventListeners = Lists.newArrayList(this.pathListener);
     protected IChunkProvider chunkProvider;
     protected final ISaveHandler saveHandler;
-    protected WorldInfo worldInfo;
+    public WorldInfo worldInfo;
     protected boolean findingSpawnPoint;
-    protected MapStorage mapStorage;
+    public MapStorage mapStorage;
     public VillageCollection villageCollection;
     protected LootTableManager lootTable;
     protected AdvancementManager advancementManager;
     protected FunctionManager functionManager;
     public final Profiler profiler;
     private final Calendar calendar = Calendar.getInstance();
-    protected Scoreboard worldScoreboard = new Scoreboard();
+    public Scoreboard worldScoreboard = new Scoreboard();
     public final boolean isRemote;
-    protected boolean spawnHostileMobs = true;
-    protected boolean spawnPeacefulMobs = true;
+    public boolean spawnHostileMobs = true;
+    public boolean spawnPeacefulMobs = true;
     private boolean processingLoadedTiles;
     private final WorldBorder worldBorder;
     int[] lightUpdateBlockList = new int[32768];
 
+    public boolean restoringBlockSnapshots = false;
+    public boolean captureBlockSnapshots = false;
+    public it.unimi.dsi.fastutil.objects.ObjectArrayList<net.minecraftforge.common.util.BlockSnapshot> capturedBlockSnapshots = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    private net.minecraftforge.common.capabilities.CapabilityDispatcher capabilities;
+    private net.minecraftforge.common.util.WorldCapabilityData capabilityData;
+
+    private final org.bukkit.craftbukkit.v1_12_R1.CraftWorld world;
+    public boolean pvpMode;
+    public boolean keepSpawnInMemory = true;
+    public org.bukkit.generator.ChunkGenerator generator;
+
+    public boolean captureTreeGeneration = false;
+
+    public List<net.minecraft.entity.item.EntityItem> captureDrops;
+    public long ticksPerAnimalSpawns;
+    public long ticksPerMonsterSpawns;
+    public boolean populating;
+    private int tickPosition;
+    public final org.spigotmc.SpigotWorldConfig spigotConfig; // Spigot
+
+    public static boolean haveWeSilencedAPhysicsCrash;
+    public static String blockLocation;
+    private org.spigotmc.TickLimiter entityLimiter;
+    private org.spigotmc.TickLimiter tileLimiter;
+    private int tileTickPosition;
+    public final it.unimi.dsi.fastutil.objects.Object2FloatOpenHashMap<Explosion.CacheKey> explosionDensityCache = new it.unimi.dsi.fastutil.objects.Object2FloatOpenHashMap<>(); // Paper - Optimize explosions // CREF - HashMap -> Object2FloatOpenHashMap
+
+    public org.bukkit.craftbukkit.v1_12_R1.CraftWorld getWorld() {
+        return this.world;
+    }
+
+    public org.bukkit.craftbukkit.v1_12_R1.CraftServer getServer() {
+        return (org.bukkit.craftbukkit.v1_12_R1.CraftServer) org.bukkit.Bukkit.getServer();
+    }
+
+    public Chunk getChunkIfLoaded(int x, int z) {
+        if (this.chunkProvider instanceof net.minecraft.world.gen.ChunkProviderServer providerServer) {
+            return providerServer.getChunkIfLoaded(x, z);
+        } else {
+            return this.isChunkLoaded(x, z, true) ? this.chunkProvider.getLoadedChunk(x, z) : null;
+        }
+    }
+
+    protected World(ISaveHandler saveHandlerIn, WorldInfo info, WorldProvider providerIn, Profiler profilerIn, boolean client, org.bukkit.generator.ChunkGenerator gen, org.bukkit.World.Environment env) {
+        this.spigotConfig = new org.spigotmc.SpigotWorldConfig( info.getWorldName() ); // Spigot
+        this.generator = gen;
+        this.world = new org.bukkit.craftbukkit.v1_12_R1.CraftWorld((WorldServer) this, gen, env);
+        this.ticksPerAnimalSpawns = this.getServer().getTicksPerAnimalSpawns(); // CraftBukkit
+        this.ticksPerMonsterSpawns = this.getServer().getTicksPerMonsterSpawns(); // CraftBukkit
+        // CraftBukkit end
+        this.saveHandler = saveHandlerIn;
+        this.profiler = profilerIn;
+        this.worldInfo = info;
+        this.provider = providerIn;
+        this.isRemote = client;
+        this.worldBorder = providerIn.createWorldBorder();
+        perWorldStorage = new MapStorage(null);
+        // CatServer start
+        if(this.worldInfo != null) // Use saved dimension from level.dat. Fixes issues with MultiVerse
+        {
+            if (this.worldInfo.getDimension() != 0)
+                this.provider.setDimension(this.worldInfo.getDimension());
+            else
+            {
+                this.worldInfo.setDimension(this.provider.getDimension());
+            }
+        }
+
+        if (this.worldInfo.getDimension() == 0) // Set generator for overworld
+        {
+            generator = this.getServer().getGenerator(this.worldInfo.getWorldName());
+            getWorld().generator = generator;
+        }
+        // CatServer end
+        // CraftBukkit start
+        getWorldBorder().world = (WorldServer) this;
+        // From PlayerList.setPlayerFileData
+        getWorldBorder().addListener(new net.minecraft.world.border.IBorderListener() {
+            public void onSizeChanged(WorldBorder worldborder, double d0) {
+                getServer().getHandle().sendAll(new net.minecraft.network.play.server.SPacketWorldBorder(worldborder, net.minecraft.network.play.server.SPacketWorldBorder.Action.SET_SIZE), worldborder.world);
+            }
+
+            public void onTransitionStarted(WorldBorder worldborder, double d0, double d1, long i) {
+                getServer().getHandle().sendAll(new net.minecraft.network.play.server.SPacketWorldBorder(worldborder, net.minecraft.network.play.server.SPacketWorldBorder.Action.LERP_SIZE), worldborder.world);
+            }
+
+            public void onCenterChanged(WorldBorder worldborder, double d0, double d1) {
+                getServer().getHandle().sendAll(new net.minecraft.network.play.server.SPacketWorldBorder(worldborder, net.minecraft.network.play.server.SPacketWorldBorder.Action.SET_CENTER), worldborder.world);
+            }
+
+            public void onWarningTimeChanged(WorldBorder worldborder, int i) {
+                getServer().getHandle().sendAll(new net.minecraft.network.play.server.SPacketWorldBorder(worldborder, net.minecraft.network.play.server.SPacketWorldBorder.Action.SET_WARNING_TIME), worldborder.world);
+            }
+
+            public void onWarningDistanceChanged(WorldBorder worldborder, int i) {
+                getServer().getHandle().sendAll(new net.minecraft.network.play.server.SPacketWorldBorder(worldborder, net.minecraft.network.play.server.SPacketWorldBorder.Action.SET_WARNING_BLOCKS), worldborder.world);
+            }
+
+            public void onDamageAmountChanged(WorldBorder worldborder, double d0) {}
+
+            public void onDamageBufferChanged(WorldBorder worldborder, double d0) {}
+        });
+        this.getServer().addWorld(this.world);
+        // CraftBukkit end
+        this.entityLimiter = new org.spigotmc.TickLimiter(spigotConfig.entityMaxTickTime);
+        this.tileLimiter = new org.spigotmc.TickLimiter(spigotConfig.tileMaxTickTime);
+    }
+
     protected World(ISaveHandler saveHandlerIn, WorldInfo info, WorldProvider providerIn, Profiler profilerIn, boolean client)
     {
+        this.spigotConfig = new org.spigotmc.SpigotWorldConfig(info.getWorldName()); // Spigot
+        this.world = net.minecraftforge.common.DimensionManager.getWorld(0) != null ? net.minecraftforge.common.DimensionManager.getWorld(0).getWorld() : null;
         this.saveHandler = saveHandlerIn;
         this.profiler = profilerIn;
         this.worldInfo = info;
         this.provider = providerIn;
         this.isRemote = client;
         this.worldBorder = providerIn.createWorldBorder();
+        perWorldStorage = new MapStorage(null);
+        this.entityLimiter = new org.spigotmc.TickLimiter(spigotConfig.entityMaxTickTime);
+        this.tileLimiter = new org.spigotmc.TickLimiter(spigotConfig.tileMaxTickTime);
     }
 
     public World init()
@@ -125,6 +246,11 @@
     @Override
     public Biome getBiome(final BlockPos pos)
     {
+        return this.provider.getBiomeForCoords(pos);
+    }
+
+    public Biome getBiomeForCoordsBody(final BlockPos pos)
+    {
         if (this.isBlockLoaded(pos))
         {
             Chunk chunk = this.getChunk(pos);
@@ -206,7 +332,7 @@
     @Override
     public boolean isAirBlock(BlockPos pos)
     {
-        return this.getBlockState(pos).getMaterial() == Material.AIR;
+        return this.getBlockState(pos).getBlock().isAir(this.getBlockState(pos), this, pos);
     }
 
     public boolean isBlockLoaded(BlockPos pos)
@@ -301,7 +427,7 @@
         }
     }
 
-    protected abstract boolean isChunkLoaded(int x, int z, boolean allowEmpty);
+    public abstract boolean isChunkLoaded(int x, int z, boolean allowEmpty);
 
     public Chunk getChunk(BlockPos pos)
     {
@@ -320,6 +446,26 @@
 
     public boolean setBlockState(BlockPos pos, IBlockState newState, int flags)
     {
+        // CatServer start - tree generation
+        if (this.captureTreeGeneration) {
+            net.minecraftforge.common.util.BlockSnapshot blocksnapshot = null;
+
+            for (net.minecraftforge.common.util.BlockSnapshot previous : this.capturedBlockSnapshots)
+            {
+                if (previous.getPos().equals(pos))
+                {
+                    blocksnapshot = previous;
+                    break;
+                }
+            }
+            if (blocksnapshot != null)
+            {
+                this.capturedBlockSnapshots.remove(blocksnapshot);
+            }
+            this.capturedBlockSnapshots.add(new net.minecraftforge.common.util.BlockSnapshot(this, pos, newState, flags));
+            return true;
+        }
+        // CatServer end
         if (this.isOutsideBuildHeight(pos))
         {
             return false;
@@ -331,23 +477,50 @@
         else
         {
             Chunk chunk = this.getChunk(pos);
-            Block block = newState.getBlock();
+
+            pos = pos.toImmutable(); // Forge - prevent mutable BlockPos leaks
+            net.minecraftforge.common.util.BlockSnapshot blockSnapshot = null;
+            if (this.captureBlockSnapshots && !this.isRemote)
+            {
+                blockSnapshot = net.minecraftforge.common.util.BlockSnapshot.getBlockSnapshot(this, pos, flags);
+                this.capturedBlockSnapshots.add(blockSnapshot);
+            }
+            IBlockState oldState = getBlockState(pos);
+            int oldLight = oldState.getLightValue(this, pos);
+            int oldOpacity = oldState.getLightOpacity(this, pos);
+
             IBlockState iblockstate = chunk.setBlockState(pos, newState);
 
             if (iblockstate == null)
             {
+                if (blockSnapshot != null) this.capturedBlockSnapshots.remove(blockSnapshot);
                 return false;
             }
             else
             {
-                if (newState.getLightOpacity() != iblockstate.getLightOpacity() || newState.getLightValue() != iblockstate.getLightValue())
+                if (newState.getLightOpacity(this, pos) != oldOpacity || newState.getLightValue(this, pos) != oldLight)
                 {
                     this.profiler.startSection("checkLight");
                     this.checkLight(pos);
                     this.profiler.endSection();
                 }
 
-                if ((flags & 2) != 0 && (!this.isRemote || (flags & 4) == 0) && chunk.isPopulated())
+                if (blockSnapshot == null) // Don't notify clients or update physics while capturing blockstates
+                {
+                    this.markAndNotifyBlock(pos, chunk, iblockstate, newState, flags);
+                }
+                return true;
+            }
+        }
+    }
+
+    // Split off from original setBlockState(BlockPos, IBlockState, int) method in order to directly send client and physic updates
+    public void markAndNotifyBlock(BlockPos pos, @Nullable Chunk chunk, IBlockState iblockstate, IBlockState newState, int flags)
+    {
+        Block block = newState.getBlock();
+        {
+            {
+                if ((flags & 2) != 0 && (!this.isRemote || (flags & 4) == 0) && (chunk == null || chunk.isPopulated()))
                 {
                     this.notifyBlockUpdate(pos, iblockstate, newState, flags);
                 }
@@ -365,8 +538,6 @@
                 {
                     this.updateObservingBlocksAt(pos, block);
                 }
-
-                return true;
             }
         }
     }
@@ -381,7 +552,7 @@
         IBlockState iblockstate = this.getBlockState(pos);
         Block block = iblockstate.getBlock();
 
-        if (iblockstate.getMaterial() == Material.AIR)
+        if (block.isAir(iblockstate, this, pos))
         {
             return false;
         }
@@ -415,6 +586,9 @@
     {
         if (this.worldInfo.getTerrainType() != WorldType.DEBUG_ALL_BLOCK_STATES)
         {
+            if (populating) {
+                return;
+            }
             this.notifyNeighborsOfStateChange(pos, blockType, updateObservers);
         }
     }
@@ -471,6 +645,9 @@
 
     public void notifyNeighborsOfStateChange(BlockPos pos, Block blockType, boolean updateObservers)
     {
+        if(net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(this, pos, this.getBlockState(pos), java.util.EnumSet.allOf(EnumFacing.class), updateObservers).isCanceled())
+            return;
+
         this.neighborChanged(pos.west(), blockType, pos);
         this.neighborChanged(pos.east(), blockType, pos);
         this.neighborChanged(pos.down(), blockType, pos);
@@ -486,6 +663,11 @@
 
     public void notifyNeighborsOfStateExcept(BlockPos pos, Block blockType, EnumFacing skipSide)
     {
+        java.util.EnumSet<EnumFacing> directions = java.util.EnumSet.allOf(EnumFacing.class);
+        directions.remove(skipSide);
+        if(net.minecraftforge.event.ForgeEventFactory.onNeighborNotify(this, pos, this.getBlockState(pos), directions, false).isCanceled())
+            return;
+
         if (skipSide != EnumFacing.WEST)
         {
             this.neighborChanged(pos.west(), blockType, pos);
@@ -525,6 +707,15 @@
 
             try
             {
+                org.bukkit.craftbukkit.v1_12_R1.CraftWorld world = this.getWorld();
+                if (world != null) {
+                    org.bukkit.event.block.BlockPhysicsEvent event = new org.bukkit.event.block.BlockPhysicsEvent(world.getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.craftbukkit.v1_12_R1.util.CraftMagicNumbers.getId(blockIn));
+                    this.getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return;
+                    }
+                }
                 iblockstate.neighborChanged(this, pos, blockIn, fromPos);
             }
             catch (Throwable throwable)
@@ -540,7 +731,7 @@
                         try
                         {
                             return String.format(
-                                       "ID #%d (%s // %s)", Block.getIdFromBlock(blockIn), blockIn.getTranslationKey(), blockIn.getClass().getCanonicalName()
+                                       "ID #%d (%s // %s // %s)", Block.getIdFromBlock(blockIn), blockIn.getTranslationKey(), blockIn.getClass().getName(), blockIn.getRegistryName()
                                    );
                         }
                         catch (Throwable throwable1)
@@ -562,11 +753,16 @@
         {
             IBlockState iblockstate = this.getBlockState(pos);
 
-            if (iblockstate.getBlock() == Blocks.OBSERVER)
+            if (true)
             {
                 try
                 {
-                    ((BlockObserver)iblockstate.getBlock()).observedNeighborChanged(iblockstate, this, pos, changedBlock, changedBlockPos);
+                    iblockstate.getBlock().observedNeighborChange(iblockstate, this, pos, changedBlock, changedBlockPos);
+                }
+                catch (StackOverflowError stackoverflowerror) { // Spigot Start
+                    haveWeSilencedAPhysicsCrash = true;
+                    blockLocation = pos.getX() + ", " + pos.getY() + ", " + pos.getZ();
+                    // Spigot End
                 }
                 catch (Throwable throwable)
                 {
@@ -581,7 +777,7 @@
                             try
                             {
                                 return String.format(
-                                           "ID #%d (%s // %s)", Block.getIdFromBlock(changedBlock), changedBlock.getTranslationKey(), changedBlock.getClass().getCanonicalName()
+                                           "ID #%d (%s // %s // %s)", Block.getIdFromBlock(changedBlock), changedBlock.getTranslationKey(), changedBlock.getClass().getName(), changedBlock.getRegistryName()
                                        );
                             }
                             catch (Throwable throwable1)
@@ -628,7 +824,7 @@
                 {
                     IBlockState iblockstate = this.getBlockState(blockpos1);
 
-                    if (iblockstate.getLightOpacity() > 0 && !iblockstate.getMaterial().isLiquid())
+                    if (iblockstate.getBlock().getLightOpacity(iblockstate, this, blockpos) > 0 && !iblockstate.getMaterial().isLiquid())
                     {
                         return false;
                     }
@@ -893,12 +1089,24 @@
     @Override
     public IBlockState getBlockState(BlockPos pos)
     {
+        // CatServer start - tree generation
+        if (captureTreeGeneration)
+        {
+            for (net.minecraftforge.common.util.BlockSnapshot blocksnapshot : this.capturedBlockSnapshots)
+            {
+                if (blocksnapshot.getPos().equals(pos)) { 
+                    return blocksnapshot.getReplacedBlock();
+                }
+            }
+        }
+        // CraftBukkit end
         if (this.isOutsideBuildHeight(pos))
         {
             return Blocks.AIR.getDefaultState();
         }
         else
         {
+            if (catserver.server.CatServer.getConfig().preventBlockLoadChunk && !isBlockLoaded(pos)) return Blocks.AIR.getDefaultState();
             Chunk chunk = this.getChunk(pos);
             return chunk.getBlockState(pos);
         }
@@ -906,7 +1114,7 @@
 
     public boolean isDaytime()
     {
-        return this.skylightSubtracted < 4;
+        return this.provider.isDaytime();
     }
 
     @Nullable
@@ -1126,6 +1334,13 @@
         float pitch
     )
     {
+        net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(player, soundIn, category, volume, pitch);
+        if (event.isCanceled() || event.getSound() == null) return;
+        soundIn = event.getSound();
+        category = event.getCategory();
+        volume = event.getVolume();
+        pitch = event.getPitch();
+
         for (int i = 0; i < this.eventListeners.size(); i++)
         {
             this.eventListeners.get(i).playSoundToAllNearExcept(player, soundIn, category, x, y, z, volume, pitch);
@@ -1236,12 +1451,69 @@
 
     public boolean addWeatherEffect(Entity entityIn)
     {
+        if(net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(entityIn, this)))
+            return false;
+
         this.weatherEffects.add(entityIn);
         return true;
     }
 
     public boolean spawnEntity(Entity entityIn)
     {
+        // CraftBukkit start - Used for entities other than creatures
+        return addEntity(entityIn, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public boolean addEntity(Entity entityIn, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+        if (catserver.server.CatAsyncCatcher.checkAsync("add entity")) return catserver.server.CatAsyncCatcher.ensureExecuteOnPrimaryThread(() -> addEntity(entityIn, spawnReason)); // CatServer
+        if (entityIn == null) return false;
+
+        org.bukkit.event.Cancellable event = null;
+        if (entityIn instanceof net.minecraft.entity.EntityLivingBase && !(entityIn instanceof EntityPlayerMP)) {
+            boolean isAnimal = entityIn instanceof net.minecraft.entity.passive.EntityAnimal || entityIn instanceof net.minecraft.entity.passive.EntityWaterMob || entityIn instanceof net.minecraft.entity.monster.EntityGolem;
+            boolean isMonster = entityIn instanceof net.minecraft.entity.monster.EntityMob || entityIn instanceof net.minecraft.entity.monster.EntityGhast || entityIn instanceof net.minecraft.entity.monster.EntitySlime;
+            boolean isNpc = entityIn instanceof org.bukkit.entity.NPC;
+
+            if (spawnReason != org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CUSTOM) {
+                if (isAnimal && !spawnPeacefulMobs || isMonster && !spawnHostileMobs || isNpc && !getServer().getServer().getCanSpawnNPCs()) {
+                    entityIn.isDead = true;
+                    return false;
+                }
+            }
+
+            event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callCreatureSpawnEvent((net.minecraft.entity.EntityLivingBase) entityIn, spawnReason);
+        } else if (entityIn instanceof net.minecraft.entity.item.EntityItem) {
+            event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callItemSpawnEvent((net.minecraft.entity.item.EntityItem) entityIn);
+        } else if (entityIn.getBukkitEntity() instanceof org.bukkit.entity.Projectile) {
+            // Not all projectiles extend EntityProjectile, so check for Bukkit interface instead
+            event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callProjectileLaunchEvent(entityIn);
+        } else if (entityIn.getBukkitEntity() instanceof org.bukkit.entity.Vehicle){
+            event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callVehicleCreateEvent(entityIn);
+        }
+        // Spigot start
+        else if (entityIn instanceof net.minecraft.entity.item.EntityXPOrb xp) {
+            double radius = spigotConfig.expMerge;
+            if (radius > 0) {
+                List<Entity> entities = this.getEntitiesWithinAABBExcludingEntity(entityIn, entityIn.getEntityBoundingBox().grow(radius, radius, radius));
+                for (Entity e : entities) {
+                    if (e instanceof net.minecraft.entity.item.EntityXPOrb loopItem) {
+                        if (!loopItem.isDead) {
+                            xp.xpValue += loopItem.xpValue;
+                            loopItem.setDead();
+                        }
+                    }
+                }
+            }
+        } // Spigot end
+
+        if (event != null && (event.isCancelled() || entityIn.isDead)) {
+            entityIn.isDead = true;
+            return false;
+        }
+        // CraftBukkit end
+        // do not drop any items while restoring blocksnapshots. Prevents dupes
+        if (!this.isRemote && (entityIn == null || (entityIn instanceof net.minecraft.entity.item.EntityItem && this.restoringBlockSnapshots))) return false;
+
         int i = MathHelper.floor(entityIn.posX / 16.0);
         int j = MathHelper.floor(entityIn.posZ / 16.0);
         boolean flag = entityIn.forceSpawn;
@@ -1264,6 +1536,8 @@
                 this.updateAllPlayersSleepingFlag();
             }
 
+            if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(entityIn, this)) && !flag) return false;
+
             this.getChunk(i, j).addEntity(entityIn);
             this.loadedEntityList.add(entityIn);
             this.onEntityAdded(entityIn);
@@ -1277,6 +1551,8 @@
         {
             this.eventListeners.get(i).onEntityAdded(entityIn);
         }
+        entityIn.onAddedToWorld();
+        entityIn.valid = true;
     }
 
     public void onEntityRemoved(Entity entityIn)
@@ -1285,10 +1561,13 @@
         {
             this.eventListeners.get(i).onEntityRemoved(entityIn);
         }
+        entityIn.onRemovedFromWorld();
+        entityIn.valid = false;
     }
 
     public void removeEntity(Entity entityIn)
     {
+        if (catserver.server.CatAsyncCatcher.checkAsync("kill entity")) { catserver.server.CatAsyncCatcher.ensureExecuteOnPrimaryThread(() -> removeEntity(entityIn)); return; } // CatServer
         if (entityIn.isBeingRidden())
         {
             entityIn.removePassengers();
@@ -1311,6 +1590,7 @@
 
     public void removeEntityDangerously(Entity entityIn)
     {
+        if (catserver.server.CatAsyncCatcher.checkAsync("remove entity")) { catserver.server.CatAsyncCatcher.ensureExecuteOnPrimaryThread(() -> removeEntityDangerously(entityIn)); return; } // CatServer
         entityIn.setDropItemsWhenDead(false);
         entityIn.setDead();
 
@@ -1328,7 +1608,16 @@
             this.getChunk(i, j).removeEntity(entityIn);
         }
 
-        this.loadedEntityList.remove(entityIn);
+        // this.loadedEntityList.remove(entityIn);
+        // CraftBukkit start - Decrement loop variable field if we've already ticked this entity
+        int index = this.loadedEntityList.indexOf(entityIn);
+        if (index != -1) {
+            if (index <= this.tickPosition) {
+                this.tickPosition--;
+            }
+            this.loadedEntityList.remove(index);
+        }
+        // CraftBukkit end
         this.onEntityRemoved(entityIn);
     }
 
@@ -1337,7 +1626,6 @@
         this.eventListeners.add(listener);
     }
 
-    @SideOnly(Side.CLIENT)
     public void removeEventListener(IWorldEventListener listener)
     {
         this.eventListeners.remove(listener);
@@ -1357,6 +1645,7 @@
         IBlockState iblockstate = Blocks.STONE.getDefaultState();
         BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.retain();
 
+        if (p_191504_3_ && !net.minecraftforge.event.ForgeEventFactory.gatherCollisionBoxes(this, entityIn, aabb, outList)) return true;
         try
         {
             for (int k1 = i; k1 < j; k1++)
@@ -1398,7 +1687,7 @@
 
                                 iblockstate1.addCollisionBoxToList(this, blockpos$pooledmutableblockpos, aabb, outList, entityIn, false);
 
-                                if (p_191504_3_ && !outList.isEmpty())
+                                if (p_191504_3_ && !net.minecraftforge.event.ForgeEventFactory.gatherCollisionBoxes(this, entityIn, aabb, outList))
                                 {
                                     return true;
                                 }
@@ -1447,7 +1736,7 @@
                 }
             }
         }
-
+        net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.GetCollisionBoxesEvent(this, entityIn, aabb, list));
         return list;
     }
 
@@ -1483,19 +1772,38 @@
 
     public int calculateSkylightSubtracted(float partialTicks)
     {
+        float f = provider.getSunBrightnessFactor(partialTicks);
+        f = 1 - f;
+        return (int)(f * 11);
+    }
+
+    /**
+     * The current sun brightness factor for this dimension.
+     * 0.0f means no light at all, and 1.0f means maximum sunlight.
+     * Highly recommended for sunlight detection like solar panel.
+     *
+     * @return The current brightness factor
+     * */
+    public float getSunBrightnessFactor(float partialTicks)
+    {
         float f = this.getCelestialAngle(partialTicks);
         float f1 = 1.0F - (MathHelper.cos(f * (float)(Math.PI * 2)) * 2.0F + 0.5F);
         f1 = MathHelper.clamp(f1, 0.0F, 1.0F);
         f1 = 1.0F - f1;
         f1 = (float)(f1 * (1.0 - this.getRainStrength(partialTicks) * 5.0F / 16.0));
         f1 = (float)(f1 * (1.0 - this.getThunderStrength(partialTicks) * 5.0F / 16.0));
-        f1 = 1.0F - f1;
-        return (int)(f1 * 11.0F);
+        return f1;
     }
 
     @SideOnly(Side.CLIENT)
     public float getSunBrightness(float partialTicks)
     {
+        return this.provider.getSunBrightness(partialTicks);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public float getSunBrightnessBody(float partialTicks)
+    {
         float f = this.getCelestialAngle(partialTicks);
         float f1 = 1.0F - (MathHelper.cos(f * (float)(Math.PI * 2)) * 2.0F + 0.2F);
         f1 = MathHelper.clamp(f1, 0.0F, 1.0F);
@@ -1508,6 +1816,12 @@
     @SideOnly(Side.CLIENT)
     public Vec3d getSkyColor(Entity entityIn, float partialTicks)
     {
+        return this.provider.getSkyColor(entityIn, partialTicks);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public Vec3d getSkyColorBody(Entity entityIn, float partialTicks)
+    {
         float f = this.getCelestialAngle(partialTicks);
         float f1 = MathHelper.cos(f * (float)(Math.PI * 2)) * 2.0F + 0.5F;
         f1 = MathHelper.clamp(f1, 0.0F, 1.0F);
@@ -1515,9 +1829,7 @@
         int j = MathHelper.floor(entityIn.posY);
         int k = MathHelper.floor(entityIn.posZ);
         BlockPos blockpos = new BlockPos(i, j, k);
-        Biome biome = this.getBiome(blockpos);
-        float f2 = biome.getTemperature(blockpos);
-        int l = biome.getSkyColorByTemp(f2);
+        int l = net.minecraftforge.client.ForgeHooksClient.getSkyBlendColour(this, blockpos);
         float f3 = (l >> 16 & 0xFF) / 255.0F;
         float f4 = (l >> 8 & 0xFF) / 255.0F;
         float f5 = (l & 0xFF) / 255.0F;
@@ -1566,18 +1878,23 @@
 
     public float getCelestialAngle(float partialTicks)
     {
-        return this.provider.calculateCelestialAngle(this.worldInfo.getWorldTime(), partialTicks);
+        return this.provider.calculateCelestialAngle(this.getWorldTime(), partialTicks);
     }
 
     @SideOnly(Side.CLIENT)
     public int getMoonPhase()
     {
-        return this.provider.getMoonPhase(this.worldInfo.getWorldTime());
+        return this.provider.getMoonPhase(this.getWorldTime());
     }
 
     public float getCurrentMoonPhaseFactor()
     {
-        return WorldProvider.MOON_PHASE_FACTORS[this.provider.getMoonPhase(this.worldInfo.getWorldTime())];
+        return provider.getCurrentMoonPhaseFactor();
+    }
+
+    public float getCurrentMoonPhaseFactorBody()
+    {
+        return WorldProvider.MOON_PHASE_FACTORS[this.provider.getMoonPhase(this.getWorldTime())];
     }
 
     public float getCelestialAngleRadians(float partialTicks)
@@ -1589,6 +1906,12 @@
     @SideOnly(Side.CLIENT)
     public Vec3d getCloudColour(float partialTicks)
     {
+        return this.provider.getCloudColor(partialTicks);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public Vec3d getCloudColorBody(float partialTicks)
+    {
         float f = this.getCelestialAngle(partialTicks);
         float f1 = MathHelper.cos(f * (float)(Math.PI * 2)) * 2.0F + 0.5F;
         f1 = MathHelper.clamp(f1, 0.0F, 1.0F);
@@ -1643,9 +1966,9 @@
         while (blockpos.getY() >= 0)
         {
             BlockPos blockpos1 = blockpos.down();
-            Material material = chunk.getBlockState(blockpos1).getMaterial();
+            IBlockState state = chunk.getBlockState(blockpos1);
 
-            if (material.blocksMovement() && material != Material.LEAVES)
+            if (state.getMaterial().blocksMovement() && !state.getBlock().isLeaves(state, this, blockpos1) && !state.getBlock().isFoliage(this, blockpos1))
             {
                 break;
             }
@@ -1659,6 +1982,12 @@
     @SideOnly(Side.CLIENT)
     public float getStarBrightness(float partialTicks)
     {
+        return this.provider.getStarBrightness(partialTicks);
+    }
+
+    @SideOnly(Side.CLIENT)
+    public float getStarBrightnessBody(float partialTicks)
+    {
         float f = this.getCelestialAngle(partialTicks);
         float f1 = 1.0F - (MathHelper.cos(f * (float)(Math.PI * 2)) * 2.0F + 0.25F);
         f1 = MathHelper.clamp(f1, 0.0F, 1.0F);
@@ -1690,9 +2019,15 @@
         for (int i = 0; i < this.weatherEffects.size(); i++)
         {
             Entity entity = this.weatherEffects.get(i);
+            // CraftBukkit start - Fixed an NPE
+            if (entity == null) {
+                continue;
+            }
+            // CraftBukkit end
 
             try
             {
+                if(entity.updateBlocked) continue;
                 entity.ticksExisted++;
                 entity.onUpdate();
             }
@@ -1710,6 +2045,12 @@
                     entity.addEntityCrashInfo(crashreportcategory);
                 }
 
+                if (net.minecraftforge.common.ForgeModContainer.removeErroringEntities)
+                {
+                    net.minecraftforge.fml.common.FMLLog.log.fatal("{}", crashreport.getCompleteReport());
+                    removeEntity(entity);
+                }
+                else
                 throw new ReportedException(crashreport);
             }
 
@@ -1743,9 +2084,19 @@
         this.tickPlayers();
         this.profiler.endStartSection("regular");
 
-        for (int i1 = 0; i1 < this.loadedEntityList.size(); i1++)
-        {
-            Entity entity2 = this.loadedEntityList.get(i1);
+        org.spigotmc.ActivationRange.activateEntities(this); // Spigot
+        { // CatServer - Dont break asm
+        // CraftBukkit start - Use field for loop variable
+        int entitiesThisCycle = 0;
+        if (tickPosition < 0) tickPosition = 0;
+        for ( entityLimiter.initTick();
+                entitiesThisCycle < loadedEntityList.size() && (entitiesThisCycle % 10 != 0 || entityLimiter.shouldContinue());
+                tickPosition++, entitiesThisCycle++) {
+            if (catserver.server.WorldTickSkipManager.shouldSkipWorldEntityTick && playerEntities.size() == 0) continue; // CatServer - Don't update Entity for no player world
+            tickPosition = (tickPosition < loadedEntityList.size()) ? tickPosition : 0;
+            Entity entity2 = this.loadedEntityList.get(this.tickPosition);
+            // CraftBukkit end
+            if (catserver.server.WorldTickSkipManager.shouldSkipEntity && entity2.skipTick) continue; // CatServer
             Entity entity3 = entity2.getRidingEntity();
 
             if (entity3 != null)
@@ -1764,13 +2115,23 @@
             {
                 try
                 {
+                    net.minecraftforge.server.timings.TimeTracker.ENTITY_UPDATE.trackStart(entity2);
+                    catserver.server.command.internal.CommandChunkStats.onTickStart();
                     this.updateEntity(entity2);
+                    catserver.server.command.internal.CommandChunkStats.onTickEnd(this, entity2.getPosition());
+                    net.minecraftforge.server.timings.TimeTracker.ENTITY_UPDATE.trackEnd(entity2);
                 }
                 catch (Throwable throwable1)
                 {
                     CrashReport crashreport1 = CrashReport.makeCrashReport(throwable1, "Ticking entity");
                     CrashReportCategory crashreportcategory1 = crashreport1.makeCategory("Entity being ticked");
                     entity2.addEntityCrashInfo(crashreportcategory1);
+                    if (net.minecraftforge.common.ForgeModContainer.removeErroringEntities)
+                    {
+                        net.minecraftforge.fml.common.FMLLog.log.fatal("{}", crashreport1.getCompleteReport());
+                        removeEntity(entity2);
+                    }
+                    else
                     throw new ReportedException(crashreport1);
                 }
             }
@@ -1788,39 +2149,65 @@
                     this.getChunk(l1, i2).removeEntity(entity2);
                 }
 
-                this.loadedEntityList.remove(i1--);
+                this.loadedEntityList.remove(this.tickPosition--); // CraftBukkit - Use field for loop variable
                 this.onEntityRemoved(entity2);
             }
 
             this.profiler.endSection();
         }
+        } // CatServer - Dont break asm
 
         this.profiler.endStartSection("blockEntities");
 
+        this.processingLoadedTiles = true; //FML Move above remove to prevent CMEs
         if (!this.tileEntitiesToBeRemoved.isEmpty())
         {
-            this.tickableTileEntities.removeAll(this.tileEntitiesToBeRemoved);
-            this.loadedTileEntityList.removeAll(this.tileEntitiesToBeRemoved);
+            for (Object tile : tileEntitiesToBeRemoved)
+            {
+               ((TileEntity)tile).onChunkUnload();
+            }
+
+            // forge: faster "contains" makes this removal much more efficient
+            java.util.Set<TileEntity> remove = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
+            remove.addAll(tileEntitiesToBeRemoved);
+            this.tickableTileEntities.removeAll(remove);
+            this.loadedTileEntityList.removeAll(remove);
             this.tileEntitiesToBeRemoved.clear();
         }
 
-        this.processingLoadedTiles = true;
-        Iterator<TileEntity> iterator = this.tickableTileEntities.iterator();
-
-        while (iterator.hasNext())
-        {
-            TileEntity tileentity = iterator.next();
-
+        // Iterator<TileEntity> iterator = this.tickableTileEntities.iterator();
+        Iterator<TileEntity> iterator = null; // The code is no used, but do not remove, if remove will asm boom
+
+        { // CatServer - Dont break asm
+        this.tilesThisCycle = 0; // CatRoom - Reset value
+        for (tileLimiter.initTick();
+             tilesThisCycle < this.tickableTileEntities.size() && (tilesThisCycle % 10 != 0 || tileLimiter.shouldContinue());
+             tileTickPosition++, tilesThisCycle++) {
+            tileTickPosition = (tileTickPosition < tickableTileEntities.size()) ? tileTickPosition : 0;
+            TileEntity tileentity = this.tickableTileEntities.get(tileTickPosition);
+            // Spigot start
+            if (tileentity == null) {
+                getServer().getLogger().severe("Spigot has detected a null entity and has removed it, preventing a crash");
+                tilesThisCycle--;
+                this.tickableTileEntities.remove(tileTickPosition--);
+                continue;
+            }
+            // Spigot end
             if (!tileentity.isInvalid() && tileentity.hasWorld())
             {
+                if (catserver.server.WorldTickSkipManager.shouldSkipTileEntity && tileentity.skipTick) continue; // CatServer
                 BlockPos blockpos = tileentity.getPos();
 
-                if (this.isBlockLoaded(blockpos) && this.worldBorder.contains(blockpos))
+                if (this.isBlockLoaded(blockpos, false) && this.worldBorder.contains(blockpos)) //Forge: Fix TE's getting an extra tick on the client side....
                 {
                     try
                     {
-                        this.profiler.func_194340_a(() -> String.valueOf(TileEntity.getKey((Class <? extends TileEntity >)tileentity.getClass())));
+                        this.profiler.func_194340_a(() -> String.valueOf(TileEntity.getKey(tileentity.getClass())));
+                        net.minecraftforge.server.timings.TimeTracker.TILE_ENTITY_UPDATE.trackStart(tileentity);
+                        catserver.server.command.internal.CommandChunkStats.onTickStart();
                         ((ITickable)tileentity).update();
+                        catserver.server.command.internal.CommandChunkStats.onTickEnd(this, tileentity.getPos());
+                        net.minecraftforge.server.timings.TimeTracker.TILE_ENTITY_UPDATE.trackEnd(tileentity);
                         this.profiler.endSection();
                     }
                     catch (Throwable throwable)
@@ -1828,6 +2215,13 @@
                         CrashReport crashreport2 = CrashReport.makeCrashReport(throwable, "Ticking block entity");
                         CrashReportCategory crashreportcategory2 = crashreport2.makeCategory("Block entity being ticked");
                         tileentity.addInfoToCrashReport(crashreportcategory2);
+                        if (net.minecraftforge.common.ForgeModContainer.removeErroringTileEntities)
+                        {
+                            net.minecraftforge.fml.common.FMLLog.log.fatal("{}", crashreport2.getCompleteReport());
+                            tileentity.invalidate();
+                            this.removeTileEntity(tileentity.getPos());
+                        }
+                        else
                         throw new ReportedException(crashreport2);
                     }
                 }
@@ -1835,15 +2229,20 @@
 
             if (tileentity.isInvalid())
             {
-                iterator.remove();
+                tilesThisCycle--;
+                this.tickableTileEntities.remove(tileTickPosition--);
                 this.loadedTileEntityList.remove(tileentity);
 
                 if (this.isBlockLoaded(tileentity.getPos()))
                 {
-                    this.getChunk(tileentity.getPos()).removeTileEntity(tileentity.getPos());
+                    //Forge: Bugfix: If we set the tile entity it immediately sets it in the chunk, so we could be desyned
+                    Chunk chunk = this.getChunk(tileentity.getPos());
+                    if (chunk.getTileEntity(tileentity.getPos(), net.minecraft.world.chunk.Chunk.EnumCreateEntityType.CHECK) == tileentity)
+                        chunk.removeTileEntity(tileentity.getPos());
                 }
             }
         }
+        } // CatServer - Dont break asm
 
         this.processingLoadedTiles = false;
         this.profiler.endStartSection("pendingBlockEntities");
@@ -1856,10 +2255,12 @@
 
                 if (!tileentity1.isInvalid())
                 {
+                    /* CraftBukkit start - Order matters, moved down
                     if (!this.loadedTileEntityList.contains(tileentity1))
                     {
                         this.addTileEntity(tileentity1);
                     }
+                    // CraftBukkit end */
 
                     if (this.isBlockLoaded(tileentity1.getPos()))
                     {
@@ -1867,6 +2268,12 @@
                         IBlockState iblockstate = chunk.getBlockState(tileentity1.getPos());
                         chunk.addTileEntity(tileentity1.getPos(), tileentity1);
                         this.notifyBlockUpdate(tileentity1.getPos(), iblockstate, iblockstate, 3);
+                        // CraftBukkit start
+                        // From above, don't screw this up - SPIGOT-1746
+                        if (!this.loadedTileEntityList.contains(tileentity1)) {
+                            this.addTileEntity(tileentity1);
+                        }
+                        // CraftBukkit end
                     }
                 }
             }
@@ -1884,12 +2291,18 @@
 
     public boolean addTileEntity(TileEntity tile)
     {
+        // Forge - set the world early as vanilla doesn't set it until next tick
+        if (tile.getWorld() != this) tile.setWorld(this);
+        // Forge: wait to add new TE if we're currently processing existing ones
+        if (processingLoadedTiles) return addedTileEntityList.add(tile);
+
         boolean flag = this.loadedTileEntityList.add(tile);
 
         if (flag && tile instanceof ITickable)
         {
             this.tickableTileEntities.add(tile);
         }
+        tile.onLoad();
 
         if (this.isRemote)
         {
@@ -1905,6 +2318,11 @@
     {
         if (this.processingLoadedTiles)
         {
+            for (TileEntity te : tileEntityCollection)
+            {
+                if (te.getWorld() != this) // Forge - set the world early as vanilla doesn't set it until next tick
+                    te.setWorld(this);
+            }
             this.addedTileEntityList.addAll(tileEntityCollection);
         }
         else
@@ -1927,13 +2345,25 @@
         {
             int i = MathHelper.floor(entityIn.posX);
             int j = MathHelper.floor(entityIn.posZ);
-            int k = 32;
-
-            if (forceUpdate && !this.isAreaLoaded(i - 32, 0, j - 32, i + 32, 0, j + 32, true))
+
+            boolean isForced = !this.isRemote && getPersistentChunks().containsKey(new net.minecraft.util.math.ChunkPos(i >> 4, j >> 4));
+            int range = isForced ? 0 : 32;
+            boolean canUpdate = !forceUpdate || this.isAreaLoaded(i - range, 0, j - range, i + range, 0, j + range, true);
+            if (!canUpdate) canUpdate = net.minecraftforge.event.ForgeEventFactory.canEntityUpdate(entityIn);
+
+            if (!canUpdate)
             {
                 return;
             }
         }
+
+        // Spigot start
+        if (forceUpdate && !org.spigotmc.ActivationRange.checkIfActive(entityIn)) {
+            entityIn.ticksExisted++;
+            entityIn.inactiveTick();
+            return;
+        }
+       // Spigot end
 
         entityIn.lastTickPosX = entityIn.posX;
         entityIn.lastTickPosY = entityIn.posY;
@@ -1951,7 +2381,9 @@
             }
             else
             {
+                if(!entityIn.updateBlocked)
                 entityIn.onUpdate();
+                entityIn.postTick();
             }
         }
 
@@ -2034,7 +2466,7 @@
         {
             Entity entity = list.get(i);
 
-            if (!entity.isDead && entity.preventEntitySpawning && entity != entityIn && (entityIn == null || entity.isRidingSameEntity(entityIn)))
+            if (!entity.isDead && entity.preventEntitySpawning && entity != entityIn && (entityIn == null || !entity.isRidingSameEntity(entityIn))) // Forge: fix MC-103516
             {
                 return false;
             }
@@ -2092,6 +2524,12 @@
                 {
                     IBlockState iblockstate = this.getBlockState(blockpos$pooledmutableblockpos.setPos(k1, l1, i2));
 
+                    Boolean result = iblockstate.getBlock().isAABBInsideLiquid(this, blockpos$pooledmutableblockpos, bb);
+                    if (result != null) {
+                        if (!result) continue;
+                        blockpos$pooledmutableblockpos.release();
+                        return true;
+                    }
                     if (iblockstate.getMaterial().isLiquid())
                     {
                         blockpos$pooledmutableblockpos.release();
@@ -2131,6 +2569,11 @@
                             blockpos$pooledmutableblockpos.release();
                             return true;
                         }
+                        else if (block.isBurning(this, new BlockPos(k1, l1, i2)))
+                        {
+                            blockpos$pooledmutableblockpos.release();
+                            return true;
+                        }
                     }
                 }
             }
@@ -2170,6 +2613,16 @@
                         IBlockState iblockstate = this.getBlockState(blockpos$pooledmutableblockpos);
                         Block block = iblockstate.getBlock();
 
+                        Boolean result = block.isEntityInsideMaterial(this, blockpos$pooledmutableblockpos, iblockstate, entityIn, (double)l, materialIn, false);
+                        if (result != null && result == true)
+                        {
+                            // Forge: When requested call blocks modifyAcceleration method, and more importantly cause this method to return true, which results in an entity being "inWater"
+                            flag = true;
+                            vec3d = block.modifyAcceleration(this, blockpos$pooledmutableblockpos, entityIn, vec3d);
+                            continue;
+                        }
+                        else if (result != null && result == false) continue;
+
                         if (iblockstate.getMaterial() == materialIn)
                         {
                             double d0 = l1 + 1 - BlockLiquid.getLiquidHeightPercent(iblockstate.getValue(BlockLiquid.LEVEL));
@@ -2215,7 +2668,14 @@
             {
                 for (int i2 = i1; i2 < j1; i2++)
                 {
-                    if (this.getBlockState(blockpos$pooledmutableblockpos.setPos(k1, l1, i2)).getMaterial() == materialIn)
+                    IBlockState iblockstate1 = this.getBlockState(blockpos$pooledmutableblockpos.setPos(k1, l1, i2));
+                    Boolean result = iblockstate1.getBlock().isAABBInsideMaterial(this, blockpos$pooledmutableblockpos, bb, materialIn);
+                    if (result != null) {
+                        if (!result) continue;
+                        blockpos$pooledmutableblockpos.release();
+                        return true;
+                    }
+                    if (iblockstate1.getMaterial() == materialIn)
                     {
                         blockpos$pooledmutableblockpos.release();
                         return true;
@@ -2238,6 +2698,7 @@
     )
     {
         Explosion explosion = new Explosion(this, entityIn, x, y, z, strength, causesFire, damagesTerrain);
+        if (net.minecraftforge.event.ForgeEventFactory.onExplosionStart(this, explosion)) return explosion;
         explosion.doExplosionA();
         explosion.doExplosionB(true);
         return explosion;
@@ -2329,7 +2790,7 @@
                 tileentity = this.getPendingTileEntityAt(pos);
             }
 
-            if (tileentity == null)
+            if (tileentity == null && (!catserver.server.CatServer.getConfig().preventBlockLoadChunk || isBlockLoaded(pos)))
             {
                 tileentity = this.getChunk(pos).getTileEntity(pos, Chunk.EnumCreateEntityType.IMMEDIATE);
             }
@@ -2361,6 +2822,7 @@
 
     public void setTileEntity(BlockPos pos, @Nullable TileEntity tileEntityIn)
     {
+        pos = pos.toImmutable(); // Forge - prevent mutable BlockPos leaks
         if (!this.isOutsideBuildHeight(pos))
         {
             if (tileEntityIn != null && !tileEntityIn.isInvalid())
@@ -2368,7 +2830,10 @@
                 if (this.processingLoadedTiles)
                 {
                     tileEntityIn.setPos(pos);
+                    if (tileEntityIn.getWorld() != this)
+                        tileEntityIn.setWorld(this); // Forge - set the world early as vanilla doesn't set it until next tick
                     Iterator<TileEntity> iterator = this.addedTileEntityList.iterator();
+                    List<TileEntity> toInvalidate = Lists.newArrayList();
 
                     while (iterator.hasNext())
                     {
@@ -2376,16 +2841,19 @@
 
                         if (tileentity.getPos().equals(pos))
                         {
-                            tileentity.invalidate();
+                            toInvalidate.add(tileentity); // Forge - don't call invalidate while iterating
                             iterator.remove();
                         }
                     }
 
+                    tileEntityIn.setWorld(this); // Spigot - No null worlds
+                    toInvalidate.forEach(TileEntity::invalidate);
                     this.addedTileEntityList.add(tileEntityIn);
                 }
                 else
                 {
-                    this.getChunk(pos).addTileEntity(pos, tileEntityIn);
+                    Chunk chunk = this.getChunk(pos);
+                    if (chunk != null) chunk.addTileEntity(pos, tileEntityIn);
                     this.addTileEntity(tileEntityIn);
                 }
             }
@@ -2400,6 +2868,8 @@
         {
             tileentity.invalidate();
             this.addedTileEntityList.remove(tileentity);
+            if (!(tileentity instanceof ITickable)) //Forge: If they are not tickable they wont be removed in the update loop.
+                this.loadedTileEntityList.remove(tileentity);
         }
         else
         {
@@ -2412,6 +2882,7 @@
 
             this.getChunk(pos).removeTileEntity(pos);
         }
+        this.updateComparatorOutputLevel(pos, getBlockState(pos).getBlock()); //Notify neighbors of changes
     }
 
     public void markTileEntityForRemoval(TileEntity tileEntityIn)
@@ -2438,7 +2909,7 @@
             if (chunk != null && !chunk.isEmpty())
             {
                 IBlockState iblockstate = this.getBlockState(pos);
-                return iblockstate.getMaterial().isOpaque() && iblockstate.isFullCube();
+                return iblockstate.getBlock().isNormalCube(iblockstate, this, pos);
             }
             else
             {
@@ -2461,6 +2932,7 @@
     {
         this.spawnHostileMobs = hostile;
         this.spawnPeacefulMobs = peaceful;
+        this.provider.setAllowedSpawnTypes(hostile, peaceful);
     }
 
     public void tick()
@@ -2470,6 +2942,11 @@
 
     protected void calculateInitialWeather()
     {
+        this.provider.calculateInitialWeather();
+    }
+
+    public void calculateInitialWeatherBody()
+    {
         if (this.worldInfo.isRaining())
         {
             this.rainingStrength = 1.0F;
@@ -2483,6 +2960,11 @@
 
     protected void updateWeather()
     {
+        this.provider.updateWeather();
+    }
+
+    public void updateWeatherBody()
+    {
         if (this.provider.hasSkyLight())
         {
             if (!this.isRemote)
@@ -2571,6 +3053,11 @@
                 }
 
                 this.rainingStrength = MathHelper.clamp(this.rainingStrength, 0.0F, 1.0F);
+                for (int idx = 0; idx < this.playerEntities.size(); ++idx) {
+                    if (this.playerEntities.get(idx).world == this) {
+                        ((EntityPlayerMP) this.playerEntities.get(idx)).tickWeather();
+                    }
+                }
             }
         }
     }
@@ -2604,6 +3091,11 @@
 
     public boolean canBlockFreeze(BlockPos pos, boolean noWaterAdj)
     {
+        return this.provider.canBlockFreeze(pos, noWaterAdj);
+    }
+
+    public boolean canBlockFreezeBody(BlockPos pos, boolean noWaterAdj)
+    {
         Biome biome = this.getBiome(pos);
         float f = biome.getTemperature(pos);
 
@@ -2648,6 +3140,11 @@
 
     public boolean canSnowAt(BlockPos pos, boolean checkLight)
     {
+        return this.provider.canSnowAt(pos, checkLight);
+    }
+
+    public boolean canSnowAtBody(BlockPos pos, boolean checkLight)
+    {
         Biome biome = this.getBiome(pos);
         float f = biome.getTemperature(pos);
 
@@ -2665,7 +3162,7 @@
             {
                 IBlockState iblockstate = this.getBlockState(pos);
 
-                if (iblockstate.getMaterial() == Material.AIR && Blocks.SNOW_LAYER.canPlaceBlockAt(this, pos))
+                if (iblockstate.getBlock().isAir(iblockstate, this, pos) && Blocks.SNOW_LAYER.canPlaceBlockAt(this, pos))
                 {
                     return true;
                 }
@@ -2696,10 +3193,10 @@
         else
         {
             IBlockState iblockstate = this.getBlockState(pos);
-            int i = lightType == EnumSkyBlock.SKY ? 0 : iblockstate.getLightValue();
-            int j = iblockstate.getLightOpacity();
+            int i = lightType == EnumSkyBlock.SKY ? 0 : iblockstate.getBlock().getLightValue(iblockstate, this, pos);
+            int j = iblockstate.getBlock().getLightOpacity(iblockstate, this, pos);
 
-            if (j >= 15 && iblockstate.getLightValue() > 0)
+            if (false) // Forge: fix MC-119932
             {
                 j = 1;
             }
@@ -2711,7 +3208,7 @@
 
             if (j >= 15)
             {
-                return 0;
+                return i; // Forge: fix MC-119932
             }
             else if (i >= 14)
             {
@@ -2751,12 +3248,16 @@
 
     public boolean checkLightFor(EnumSkyBlock lightType, BlockPos pos)
     {
-        if (!this.isAreaLoaded(pos, 17, false))
+        // CraftBukkit start - Use neighbor cache instead of looking up
+        Chunk chunk = this.getChunkIfLoaded(pos.getX() >> 4, pos.getZ() >> 4);
+        if (chunk == null || !chunk.areNeighborsLoaded(1) /*!this.isAreaLoaded(pos, 16, false)*/)
         {
+            // CraftBukkit end
             return false;
         }
         else
         {
+            int updateRange = this.isAreaLoaded(pos, 18, false) ? 17 : 15;
             int i = 0;
             int j = 0;
             this.profiler.startSection("getBrightness");
@@ -2794,7 +3295,7 @@
                             int k3 = MathHelper.abs(j2 - j1);
                             int l3 = MathHelper.abs(k2 - k1);
 
-                            if (j3 + k3 + l3 < 17)
+                            if (j3 + k3 + l3 < updateRange)
                             {
                                 BlockPos.PooledMutableBlockPos blockpos$pooledmutableblockpos = BlockPos.PooledMutableBlockPos.retain();
 
@@ -2804,7 +3305,8 @@
                                     int j4 = j2 + enumfacing.getYOffset();
                                     int k4 = k2 + enumfacing.getZOffset();
                                     blockpos$pooledmutableblockpos.setPos(i4, j4, k4);
-                                    int l4 = Math.max(1, this.getBlockState(blockpos$pooledmutableblockpos).getLightOpacity());
+                                    IBlockState bs = this.getBlockState(blockpos$pooledmutableblockpos);
+                                    int l4 = Math.max(1, bs.getBlock().getLightOpacity(bs, this, blockpos$pooledmutableblockpos));
                                     i3 = this.getLightFor(lightType, blockpos$pooledmutableblockpos);
 
                                     if (i3 == l2 - l4 && j < this.lightUpdateBlockList.length)
@@ -2827,55 +3329,55 @@
 
             while (i < j)
             {
-                int i5 = this.lightUpdateBlockList[i++];
-                int j5 = (i5 & 63) - 32 + i1;
-                int k5 = (i5 >> 6 & 63) - 32 + j1;
-                int l5 = (i5 >> 12 & 63) - 32 + k1;
-                BlockPos blockpos1 = new BlockPos(j5, k5, l5);
-                int i6 = this.getLightFor(lightType, blockpos1);
-                int j6 = this.getRawLight(blockpos1, lightType);
+                int j7 = this.lightUpdateBlockList[i++];
+                int k7 = (j7 & 63) - 32 + i1;
+                int l7 = (j7 >> 6 & 63) - 32 + j1;
+                int i8 = (j7 >> 12 & 63) - 32 + k1;
+                BlockPos blockpos2 = new BlockPos(k7, l7, i8);
+                int j8 = this.getLightFor(lightType, blockpos2);
+                int k8 = this.getRawLight(blockpos2, lightType);
 
-                if (j6 != i6)
+                if (k8 != j8)
                 {
-                    this.setLightFor(lightType, blockpos1, j6);
+                    this.setLightFor(lightType, blockpos2, k8);
 
-                    if (j6 > i6)
+                    if (k8 > j8)
                     {
-                        int k6 = Math.abs(j5 - i1);
-                        int l6 = Math.abs(k5 - j1);
-                        int i7 = Math.abs(l5 - k1);
+                        int l8 = Math.abs(k7 - i1);
+                        int i9 = Math.abs(l7 - j1);
+                        int j9 = Math.abs(i8 - k1);
                         boolean flag = j < this.lightUpdateBlockList.length - 6;
 
-                        if (k6 + l6 + i7 < 17 && flag)
+                        if (l8 + i9 + j9 < updateRange && flag)
                         {
-                            if (this.getLightFor(lightType, blockpos1.west()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 - 1 - i1 + 32 + (k5 - j1 + 32 << 6) + (l5 - k1 + 32 << 12);
-                            }
-
-                            if (this.getLightFor(lightType, blockpos1.east()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 + 1 - i1 + 32 + (k5 - j1 + 32 << 6) + (l5 - k1 + 32 << 12);
-                            }
-
-                            if (this.getLightFor(lightType, blockpos1.down()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 - i1 + 32 + (k5 - 1 - j1 + 32 << 6) + (l5 - k1 + 32 << 12);
-                            }
-
-                            if (this.getLightFor(lightType, blockpos1.up()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 - i1 + 32 + (k5 + 1 - j1 + 32 << 6) + (l5 - k1 + 32 << 12);
-                            }
-
-                            if (this.getLightFor(lightType, blockpos1.north()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 - i1 + 32 + (k5 - j1 + 32 << 6) + (l5 - 1 - k1 + 32 << 12);
-                            }
-
-                            if (this.getLightFor(lightType, blockpos1.south()) < j6)
-                            {
-                                this.lightUpdateBlockList[j++] = j5 - i1 + 32 + (k5 - j1 + 32 << 6) + (l5 + 1 - k1 + 32 << 12);
+                            if (this.getLightFor(lightType, blockpos2.west()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 - 1 - i1 + 32 + (l7 - j1 + 32 << 6) + (i8 - k1 + 32 << 12);
+                            }
+
+                            if (this.getLightFor(lightType, blockpos2.east()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 + 1 - i1 + 32 + (l7 - j1 + 32 << 6) + (i8 - k1 + 32 << 12);
+                            }
+
+                            if (this.getLightFor(lightType, blockpos2.down()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 - i1 + 32 + (l7 - 1 - j1 + 32 << 6) + (i8 - k1 + 32 << 12);
+                            }
+
+                            if (this.getLightFor(lightType, blockpos2.up()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 - i1 + 32 + (l7 + 1 - j1 + 32 << 6) + (i8 - k1 + 32 << 12);
+                            }
+
+                            if (this.getLightFor(lightType, blockpos2.north()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 - i1 + 32 + (l7 - j1 + 32 << 6) + (i8 - 1 - k1 + 32 << 12);
+                            }
+
+                            if (this.getLightFor(lightType, blockpos2.south()) < k8)
+                            {
+                                this.lightUpdateBlockList[j++] = k7 - i1 + 32 + (l7 - j1 + 32 << 6) + (i8 + 1 - k1 + 32 << 12);
                             }
                         }
                     }
@@ -2912,10 +3414,10 @@
     public List<Entity> getEntitiesInAABBexcluding(@Nullable Entity entityIn, AxisAlignedBB boundingBox, @Nullable Predicate <? super Entity > predicate)
     {
         List<Entity> list = Lists.newArrayList();
-        int i = MathHelper.floor((boundingBox.minX - 2.0) / 16.0);
-        int j = MathHelper.floor((boundingBox.maxX + 2.0) / 16.0);
-        int k = MathHelper.floor((boundingBox.minZ - 2.0) / 16.0);
-        int l = MathHelper.floor((boundingBox.maxZ + 2.0) / 16.0);
+        int i = MathHelper.floor((boundingBox.minX - MAX_ENTITY_RADIUS) / 16.0);
+        int j = MathHelper.floor((boundingBox.maxX + MAX_ENTITY_RADIUS) / 16.0);
+        int k = MathHelper.floor((boundingBox.minZ - MAX_ENTITY_RADIUS) / 16.0);
+        int l = MathHelper.floor((boundingBox.maxZ + MAX_ENTITY_RADIUS) / 16.0);
 
         for (int i1 = i; i1 <= j; i1++)
         {
@@ -2968,10 +3470,10 @@
 
     public <T extends Entity> List<T> getEntitiesWithinAABB(Class <? extends T > clazz, AxisAlignedBB aabb, @Nullable Predicate <? super T > filter)
     {
-        int i = MathHelper.floor((aabb.minX - 2.0) / 16.0);
-        int j = MathHelper.ceil((aabb.maxX + 2.0) / 16.0);
-        int k = MathHelper.floor((aabb.minZ - 2.0) / 16.0);
-        int l = MathHelper.ceil((aabb.maxZ + 2.0) / 16.0);
+        int i = MathHelper.floor((aabb.minX - MAX_ENTITY_RADIUS) / 16.0);
+        int j = MathHelper.ceil((aabb.maxX + MAX_ENTITY_RADIUS) / 16.0);
+        int k = MathHelper.floor((aabb.minZ - MAX_ENTITY_RADIUS) / 16.0);
+        int l = MathHelper.ceil((aabb.maxZ + MAX_ENTITY_RADIUS) / 16.0);
         List<T> list = Lists.newArrayList();
 
         for (int i1 = i; i1 < j; i1++)
@@ -3040,7 +3542,16 @@
 
         for (Entity entity : this.loadedEntityList)
         {
-            if ((!(entity instanceof EntityLiving) || !((EntityLiving)entity).isNoDespawnRequired()) && entityType.isAssignableFrom(entity.getClass()))
+            // CraftBukkit start - Split out persistent check, don't apply it to special persistent mobs
+            if (entity instanceof EntityLiving) {
+                EntityLiving entityinsentient = (EntityLiving)entity;
+                if (entityinsentient.canDespawn() && entityinsentient.isNoDespawnRequired()) {
+                    continue;
+                }
+            }
+
+            // if ((!(entity instanceof EntityLiving) || !((EntityLiving)entity).isNoDespawnRequired()) && entityType.isAssignableFrom(entity.getClass()))
+            if (entityType.isAssignableFrom(entity.getClass()))
             {
                 i++;
             }
@@ -3051,11 +3562,15 @@
 
     public void loadEntities(Collection<Entity> entityCollection)
     {
-        this.loadedEntityList.addAll(entityCollection);
-
+        if (catserver.server.CatAsyncCatcher.checkAsync("load entities")) { catserver.server.CatAsyncCatcher.ensureExecuteOnPrimaryThread(() -> loadEntities(entityCollection)); return; } // CatServer
         for (Entity entity : entityCollection)
         {
-            this.onEntityAdded(entity);
+            if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(entity, this)))
+            {
+                if (entity == null) continue;
+                loadedEntityList.add(entity);
+                this.onEntityAdded(entity);
+            }
         }
     }
 
@@ -3069,16 +3584,24 @@
         IBlockState iblockstate = this.getBlockState(pos);
         AxisAlignedBB axisalignedbb = skipCollisionCheck ? null : blockIn.getDefaultState().getCollisionBoundingBox(this, pos);
 
-        if (axisalignedbb != Block.NULL_AABB && !this.checkNoEntityCollision(axisalignedbb.offset(pos), placer))
-        {
-            return false;
+        if (!((placer instanceof EntityPlayer) || !net.minecraftforge.event.ForgeEventFactory.onBlockPlace(placer, new net.minecraftforge.common.util.BlockSnapshot(this, pos, blockIn.getDefaultState()), sidePlacedOn).isCanceled())) return false;
+        boolean defaultReturn;
+        if (axisalignedbb != Block.NULL_AABB && !this.checkNoEntityCollision(axisalignedbb.offset(pos)))  // Forge: Remove second parameter, we patch placer to be non-null, passing it here skips collision checks for the placer
+        {
+            defaultReturn = false;
+        }
+        else if (iblockstate.getMaterial() == Material.CIRCUITS && blockIn == Blocks.ANVIL)
+        {
+            defaultReturn = true;
         }
         else
         {
-            return iblockstate.getMaterial() == Material.CIRCUITS && blockIn == Blocks.ANVIL
-                   ? true
-                   : iblockstate.getMaterial().isReplaceable() && blockIn.canPlaceBlockOnSide(this, pos, sidePlacedOn);
+            defaultReturn = iblockstate.getBlock().isReplaceable(this, pos)  && blockIn.canPlaceBlockOnSide(this, pos, sidePlacedOn);
         }
+        org.bukkit.event.block.BlockCanBuildEvent event = new org.bukkit.event.block.BlockCanBuildEvent(this.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.craftbukkit.v1_12_R1.util.CraftMagicNumbers.getId(blockIn), defaultReturn);
+        this.getServer().getPluginManager().callEvent(event);
+
+        return event.isBuildable();
     }
 
     public int getSeaLevel()
@@ -3163,7 +3686,7 @@
     public int getRedstonePower(BlockPos pos, EnumFacing facing)
     {
         IBlockState iblockstate = this.getBlockState(pos);
-        return iblockstate.isNormalCube() ? this.getStrongPower(pos) : iblockstate.getWeakPower(this, pos, facing);
+        return iblockstate.getBlock().shouldCheckWeakPower(iblockstate, this, pos, facing) ? this.getStrongPower(pos) : iblockstate.getWeakPower(this, pos, facing);
     }
 
     public boolean isBlockPowered(BlockPos pos)
@@ -3243,6 +3766,12 @@
         {
             EntityPlayer entityplayer1 = this.playerEntities.get(i);
 
+            // CraftBukkit start - Fixed an NPE
+            if (entityplayer1 == null || entityplayer1.isDead) {
+                continue;
+            }
+            // CraftBukkit end
+
             if (predicate.apply(entityplayer1))
             {
                 double d1 = entityplayer1.getDistanceSq(x, y, z);
@@ -3340,6 +3869,8 @@
                     d2 *= MoreObjects.firstNonNull(playerToDouble.apply(entityplayer1), 1.0);
                 }
 
+                d2 = net.minecraftforge.common.ForgeHooks.getPlayerVisibilityDistance(entityplayer1, d2, maxYDistance);
+
                 if ((maxYDistance < 0.0 || Math.abs(entityplayer1.posY - posY) < maxYDistance * maxYDistance)
                         && (maxXZDistance < 0.0 || d1 < d2 * d2)
                         && (d0 == -1.0 || d1 < d0))
@@ -3403,7 +3934,7 @@
 
     public long getSeed()
     {
-        return this.worldInfo.getSeed();
+        return this.provider.getSeed();
     }
 
     public long getTotalWorldTime()
@@ -3413,17 +3944,17 @@
 
     public long getWorldTime()
     {
-        return this.worldInfo.getWorldTime();
+        return this.provider.getWorldTime();
     }
 
     public void setWorldTime(long time)
     {
-        this.worldInfo.setWorldTime(time);
+        this.provider.setWorldTime(time);
     }
 
     public BlockPos getSpawnPoint()
     {
-        BlockPos blockpos = new BlockPos(this.worldInfo.getSpawnX(), this.worldInfo.getSpawnY(), this.worldInfo.getSpawnZ());
+        BlockPos blockpos = this.provider.getSpawnPoint();
 
         if (!this.getWorldBorder().contains(blockpos))
         {
@@ -3435,7 +3966,7 @@
 
     public void setSpawnPoint(BlockPos pos)
     {
-        this.worldInfo.setSpawn(pos);
+        this.provider.setSpawnPoint(pos);
     }
 
     @SideOnly(Side.CLIENT)
@@ -3455,12 +3986,18 @@
 
         if (!this.loadedEntityList.contains(entityIn))
         {
+            if (!net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.entity.EntityJoinWorldEvent(entityIn, this)))
             this.loadedEntityList.add(entityIn);
         }
     }
 
     public boolean isBlockModifiable(EntityPlayer player, BlockPos pos)
     {
+        return this.provider.canMineBlock(player, pos);
+    }
+
+    public boolean canMineBlockBody(EntityPlayer player, BlockPos pos)
+    {
         return true;
     }
 
@@ -3497,6 +4034,16 @@
     {
     }
 
+    // CraftBukkit start
+    // Calls the method that checks to see if players are sleeping
+    // Called by CraftPlayer.setPermanentSleeping()
+    public void checkSleepStatus() {
+        if (!this.isRemote) {
+            this.updateAllPlayersSleepingFlag();
+        }
+    }
+    // CraftBukkit end
+
     public float getThunderStrength(float delta)
     {
         return (this.prevThunderingStrength + (this.thunderingStrength - this.prevThunderingStrength) * delta) * this.getRainStrength(delta);
@@ -3562,8 +4109,7 @@
 
     public boolean isBlockinHighHumidity(BlockPos pos)
     {
-        Biome biome = this.getBiome(pos);
-        return biome.isHighHumidity();
+        return this.provider.isBlockHighHumidity(pos);
     }
 
     @Nullable
@@ -3624,12 +4170,12 @@
 
     public int getHeight()
     {
-        return 256;
+        return this.provider.getHeight();
     }
 
     public int getActualHeight()
     {
-        return this.provider.isNether() ? 128 : 256;
+        return this.provider.getActualHeight();
     }
 
     public Random setRandomSeed(int seedX, int seedY, int seedZ)
@@ -3642,7 +4188,7 @@
     @SideOnly(Side.CLIENT)
     public double getHorizon()
     {
-        return this.worldInfo.getTerrainType() == WorldType.FLAT ? 0.0 : 63.0;
+        return provider.getHorizon();
     }
 
     public CrashReportCategory addWorldInfoToCrashReport(CrashReport report)
@@ -3709,7 +4255,7 @@
 
     public void updateComparatorOutputLevel(BlockPos pos, Block blockIn)
     {
-        for (EnumFacing enumfacing : EnumFacing.Plane.HORIZONTAL)
+        for (EnumFacing enumfacing : EnumFacing.VALUES)
         {
             BlockPos blockpos = pos.offset(enumfacing);
 
@@ -3717,18 +4263,15 @@
             {
                 IBlockState iblockstate = this.getBlockState(blockpos);
 
-                if (Blocks.UNPOWERED_COMPARATOR.isSameDiode(iblockstate))
-                {
-                    iblockstate.neighborChanged(this, blockpos, blockIn, pos);
-                }
-                else if (iblockstate.isNormalCube())
+                iblockstate.getBlock().onNeighborChange(this, blockpos, pos);
+                if (iblockstate.getBlock().isNormalCube(iblockstate, this, blockpos))
                 {
                     blockpos = blockpos.offset(enumfacing);
                     iblockstate = this.getBlockState(blockpos);
 
-                    if (Blocks.UNPOWERED_COMPARATOR.isSameDiode(iblockstate))
+                    if (iblockstate.getBlock().getWeakChanges(this, blockpos))
                     {
-                        iblockstate.neighborChanged(this, blockpos, blockIn, pos);
+                        iblockstate.getBlock().onNeighborChange(this, blockpos, pos);
                     }
                 }
             }
@@ -3791,7 +4334,125 @@
         int i = x * 16 + 8 - blockpos.getX();
         int j = z * 16 + 8 - blockpos.getZ();
         int k = 128;
-        return i >= -128 && i <= 128 && j >= -128 && j <= 128;
+        return i >= -128 && i <= 128 && j >= -128 && j <= 128 && this.keepSpawnInMemory;
+    }
+
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Determine if the given block is considered solid on the
+     * specified side.  Used by placement logic.
+     *
+     * @param pos Block Position
+     * @param side The Side in question
+     * @return True if the side is solid
+    */
+    public boolean isSideSolid(BlockPos pos, EnumFacing side)
+    {
+       return isSideSolid(pos, side, false);
+    }
+
+    /**
+     * Determine if the given block is considered solid on the
+     * specified side.  Used by placement logic.
+     *
+     * @param pos Block Position
+     * @param side The Side in question
+     * @param _default The default to return if the block doesn't exist.
+     * @return True if the side is solid
+     */
+    @Override
+    public boolean isSideSolid(BlockPos pos, EnumFacing side, boolean _default)
+    {
+        if (!this.isValid(pos)) return _default;
+
+        Chunk chunk = getChunk(pos);
+        if (chunk == null || chunk.isEmpty()) return _default;
+        return getBlockState(pos).isSideSolid(this, pos, side);
+    }
+
+    /**
+     * Get the persistent chunks for this world
+     *
+     * @return
+     */
+    public com.google.common.collect.ImmutableSetMultimap<net.minecraft.util.math.ChunkPos, net.minecraftforge.common.ForgeChunkManager.Ticket> getPersistentChunks()
+    {
+        return net.minecraftforge.common.ForgeChunkManager.getPersistentChunksFor(this);
+    }
+
+    public Iterator<Chunk> getPersistentChunkIterable(Iterator<Chunk> chunkIterator)
+    {
+        return net.minecraftforge.common.ForgeChunkManager.getPersistentChunksIterableFor(this, chunkIterator);
+    }
+    /**
+     * Readded as it was removed, very useful helper function
+     *
+     * @param pos Block position
+     * @return The blocks light opacity
+     */
+    public int getBlockLightOpacity(BlockPos pos)
+    {
+        if (!this.isValid(pos)) return 0;
+        return getChunk(pos).getBlockLightOpacity(pos);
+    }
+
+    /**
+     * Returns a count of entities that classify themselves as the specified creature type.
+     */
+    public int countEntities(net.minecraft.entity.EnumCreatureType type, boolean forSpawnCount)
+    {
+        int count = 0;
+        for (int x = 0; x < loadedEntityList.size(); x++)
+        {
+            if (((Entity)loadedEntityList.get(x)).isCreatureType(type, forSpawnCount))
+            {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    @Deprecated // remove in 1.13
+    public void markTileEntitiesInChunkForRemoval(Chunk chunk)
+    {
+        for (TileEntity tileentity : chunk.getTileEntityMap().values())
+        {
+            markTileEntityForRemoval(tileentity);
+        }
+    }
+
+    protected void initCapabilities()
+    {
+        net.minecraftforge.common.capabilities.ICapabilityProvider parent = provider.initCapabilities();
+        capabilities = net.minecraftforge.event.ForgeEventFactory.gatherCapabilities(this, parent);
+        net.minecraftforge.common.util.WorldCapabilityData data = (net.minecraftforge.common.util.WorldCapabilityData)perWorldStorage.getOrLoadData(net.minecraftforge.common.util.WorldCapabilityData.class, net.minecraftforge.common.util.WorldCapabilityData.ID);
+        if (data == null)
+        {
+            capabilityData = new net.minecraftforge.common.util.WorldCapabilityData(capabilities);
+            perWorldStorage.setData(capabilityData.mapName, capabilityData);
+        }
+        else
+        {
+            capabilityData = data;
+            capabilityData.setCapabilities(provider, capabilities);
+        }
+    }
+    @Override
+    public boolean hasCapability(net.minecraftforge.common.capabilities.Capability<?> capability, @Nullable EnumFacing facing)
+    {
+        return capabilities == null ? false : capabilities.hasCapability(capability, facing);
+    }
+    @Override
+    @Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @Nullable EnumFacing facing)
+    {
+        return capabilities == null ? null : capabilities.getCapability(capability, facing);
+    }
+
+    protected MapStorage perWorldStorage; //Moved to a getter to simulate final without being final so we can load in subclasses.
+    public MapStorage getPerWorldStorage()
+    {
+        return perWorldStorage;
     }
 
     public void sendPacketToServer(Packet<?> packetIn)
@@ -3809,4 +4470,6 @@
     {
         return null;
     }
+
+    public java.util.Map<BlockPos, TileEntity> capturedTileEntities = com.google.common.collect.Maps.newHashMap(); // CatServer - Add spigot field, no used
 }
