--- before/net/minecraft/block/BlockJukebox.java
+++ after/net/minecraft/block/BlockJukebox.java
@@ -75,7 +75,7 @@
         }
     }
 
-    private void dropRecord(World worldIn, BlockPos pos, IBlockState state)
+    public void dropRecord(World worldIn, BlockPos pos, IBlockState state)
     {
         if (!worldIn.isRemote)
         {
@@ -215,6 +215,10 @@
 
             public void setRecord(ItemStack recordStack)
             {
+                // CraftBukkit start - There can only be one
+                if (!recordStack.isEmpty()) {
+                    recordStack.setCount(1);
+                }
                 this.record = recordStack;
                 this.markDirty();
             }
