--- before/net/minecraft/block/BlockPumpkin.java
+++ after/net/minecraft/block/BlockPumpkin.java
@@ -62,19 +62,22 @@
     private void trySpawnGolem(World worldIn, BlockPos pos)
     {
         BlockPattern.PatternHelper blockpattern$patternhelper = this.getSnowmanPattern().match(worldIn, pos);
-
+        org.bukkit.craftbukkit.v1_12_R1.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_12_R1.util.BlockStateListPopulator(worldIn.getWorld());
         if (blockpattern$patternhelper != null)
         {
             for (int i = 0; i < this.getSnowmanPattern().getThumbLength(); i++)
             {
                 BlockWorldState blockworldstate = blockpattern$patternhelper.translateOffset(0, i, 0);
-                worldIn.setBlockState(blockworldstate.getPos(), Blocks.AIR.getDefaultState(), 2);
+//                worldIn.setBlockState(blockworldstate.getPos(), Blocks.AIR.getDefaultState(), 2);
+                BlockPos pos1 = blockworldstate.getPos();
+                blockList.setTypeId(pos1.getX(), pos1.getY(), pos1.getZ(), 0);
             }
 
             EntitySnowman entitysnowman = new EntitySnowman(worldIn);
             BlockPos blockpos1 = blockpattern$patternhelper.translateOffset(0, 2, 0).getPos();
-            entitysnowman.setLocationAndAngles(blockpos1.getX() + 0.5, blockpos1.getY() + 0.05, blockpos1.getZ() + 0.5, 0.0F, 0.0F);
-            worldIn.spawnEntity(entitysnowman);
+            entitysnowman.setLocationAndAngles((double)blockpos1.getX() + 0.5D, (double)blockpos1.getY() + 0.05D, (double)blockpos1.getZ() + 0.5D, 0.0F, 0.0F);
+            if (worldIn.addEntity(entitysnowman, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN)) {
+                blockList.updateList();
 
             for (EntityPlayerMP entityplayermp : worldIn.getEntitiesWithinAABB(EntityPlayerMP.class, entitysnowman.getEntityBoundingBox().grow(5.0)))
             {
@@ -99,6 +102,7 @@
                 BlockWorldState blockworldstate2 = blockpattern$patternhelper.translateOffset(0, i1, 0);
                 worldIn.notifyNeighborsRespectDebug(blockworldstate2.getPos(), Blocks.AIR, false);
             }
+            }
         }
         else
         {
@@ -110,7 +114,9 @@
                 {
                     for (int k = 0; k < this.getGolemPattern().getThumbLength(); k++)
                     {
-                        worldIn.setBlockState(blockpattern$patternhelper.translateOffset(j, k, 0).getPos(), Blocks.AIR.getDefaultState(), 2);
+//                        worldIn.setBlockState(blockpattern$patternhelper.translateOffset(j, k, 0).getPos(), Blocks.AIR.getDefaultState(), 2);
+                        BlockPos pos2 = blockpattern$patternhelper.translateOffset(j, k, 0).getPos();
+                        blockList.setTypeId(pos2.getX(), pos2.getY(), pos2.getZ(), 0);
                     }
                 }
 
@@ -118,7 +124,8 @@
                 EntityIronGolem entityirongolem = new EntityIronGolem(worldIn);
                 entityirongolem.setPlayerCreated(true);
                 entityirongolem.setLocationAndAngles(blockpos.getX() + 0.5, blockpos.getY() + 0.05, blockpos.getZ() + 0.5, 0.0F, 0.0F);
-                worldIn.spawnEntity(entityirongolem);
+                if (worldIn.addEntity(entityirongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM)) {
+                    blockList.updateList();
 
                 for (EntityPlayerMP entityplayermp1 : worldIn.getEntitiesWithinAABB(EntityPlayerMP.class, entityirongolem.getEntityBoundingBox().grow(5.0)))
                 {
@@ -145,6 +152,7 @@
                         BlockWorldState blockworldstate1 = blockpattern$patternhelper.translateOffset(k1, l1, 0);
                         worldIn.notifyNeighborsRespectDebug(blockworldstate1.getPos(), Blocks.AIR, false);
                     }
+                    }
                 }
             }
         }
@@ -153,8 +161,8 @@
     @Override
     public boolean canPlaceBlockAt(World worldIn, BlockPos pos)
     {
-        return worldIn.getBlockState(pos).getBlock().material.isReplaceable()
-               && worldIn.getBlockState(pos.down()).isTopSolid();
+        return worldIn.getBlockState(pos).getBlock().isReplaceable(worldIn, pos)
+               && worldIn.isSideSolid(pos.down(), EnumFacing.UP);
     }
 
     @Override
