--- before/net/minecraft/block/BlockRailDetector.java
+++ after/net/minecraft/block/BlockRailDetector.java
@@ -118,6 +118,13 @@
             flag1 = true;
         }
 
+        if (flag != flag1) {
+            org.bukkit.block.Block block = worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+            org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(block, flag ? 15 : 0, flag1 ? 15 : 0);
+            worldIn.getServer().getPluginManager().callEvent(eventRedstone);
+            flag1 = eventRedstone.getNewCurrent() > 0;
+        }
+
         if (flag1 && !flag)
         {
             worldIn.setBlockState(pos, state.withProperty(POWERED, true), 3);
@@ -183,6 +190,8 @@
     {
         if (blockState.getValue(POWERED))
         {
+            List<EntityMinecart> carts = this.findMinecarts(worldIn, pos, EntityMinecart.class);
+            if (!carts.isEmpty() && carts.getFirst().getComparatorLevel() > -1) return carts.getFirst().getComparatorLevel();
             List<EntityMinecartCommandBlock> list = this.findMinecarts(worldIn, pos, EntityMinecartCommandBlock.class);
 
             if (!list.isEmpty())
