--- before/net/minecraft/block/BlockChorusFlower.java
+++ after/net/minecraft/block/BlockChorusFlower.java
@@ -4,6 +4,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.material.MapColor;
 import net.minecraft.block.material.Material;
+import net.minecraft.block.properties.IProperty;
 import net.minecraft.block.properties.PropertyInteger;
 import net.minecraft.block.state.BlockFaceShape;
 import net.minecraft.block.state.BlockStateContainer;
@@ -22,6 +23,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory;
 
 public class BlockChorusFlower extends Block
 {
@@ -56,7 +58,7 @@
             {
                 int i = state.getValue(AGE);
 
-                if (i < 5 && rand.nextInt(1) == 0)
+                if (i < 5 &&  net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, blockpos, state, rand.nextInt(1) == 0))
                 {
                     boolean flag = false;
                     boolean flag1 = false;
@@ -107,8 +109,17 @@
 
                     if (flag && areAllNeighborsEmpty(worldIn, blockpos, null) && worldIn.isAirBlock(pos.up(2)))
                     {
-                        worldIn.setBlockState(pos, Blocks.CHORUS_PLANT.getDefaultState(), 2);
-                        this.placeGrownFlower(worldIn, blockpos, i);
+//                        worldIn.setBlockState(pos, Blocks.CHORUS_PLANT.getDefaultState(), 2);
+//                        this.placeGrownFlower(worldIn, blockpos, i);
+                        if (CraftEventFactory.handleBlockSpreadEvent(
+                                worldIn.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()),
+                                worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()),
+                                this,
+                                this.getMetaFromState(this.getDefaultState().withProperty(BlockChorusFlower.AGE, i))
+                        )) {
+                            worldIn.setBlockState(pos, Blocks.CHORUS_PLANT.getDefaultState(), 2);
+                            worldIn.playEvent(1033, pos, 0);
+                        }
                     }
                     else if (i < 4)
                     {
@@ -129,7 +140,15 @@
                                     && worldIn.isAirBlock(blockpos1.down())
                                     && areAllNeighborsEmpty(worldIn, blockpos1, enumfacing.getOpposite()))
                             {
-                                this.placeGrownFlower(worldIn, blockpos1, i + 1);
+//                                this.placeGrownFlower(worldIn, blockpos1, i + 1);
+	                            if (CraftEventFactory.handleBlockSpreadEvent(
+                                        worldIn.getWorld().getBlockAt(blockpos1.getX(), blockpos1.getY(), blockpos1.getZ()),
+                                        worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()),
+                                        this,
+                                        this.getMetaFromState(this.getDefaultState().withProperty(BlockChorusFlower.AGE, Integer.valueOf(i + 1)))
+                                )) {
+                                    worldIn.playEvent(1033, pos, 0);
+                                }
                                 flag2 = true;
                             }
                         }
@@ -140,13 +159,34 @@
                         }
                         else
                         {
-                            this.placeDeadFlower(worldIn, pos);
+//                            this.placeDeadFlower(worldIn, pos);
+                            if (CraftEventFactory.handleBlockGrowEvent(
+                                    worldIn,
+                                    pos.getX(),
+                                    pos.getY(),
+                                    pos.getZ(),
+                                    this,
+                                    this.getMetaFromState(state.withProperty(BlockChorusFlower.AGE, Integer.valueOf(5)))
+                            )) {
+                                worldIn.playEvent(1034, pos, 0);
+                            }
                         }
                     }
                     else if (i == 4)
                     {
-                        this.placeDeadFlower(worldIn, pos);
+//                        this.placeDeadFlower(worldIn, pos);
+                        if (CraftEventFactory.handleBlockGrowEvent(
+                                worldIn,
+                                pos.getX(),
+                                pos.getY(),
+                                pos.getZ(),
+                                this,
+                                getMetaFromState(state.withProperty(BlockChorusFlower.AGE, Integer.valueOf(5)))
+                        )) {
+                            worldIn.playEvent(1034, pos, 0);
+                        }
                     }
+                    net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
                 }
             }
         }
