--- before/net/minecraft/block/BlockRedstoneOre.java
+++ after/net/minecraft/block/BlockRedstoneOre.java
@@ -42,15 +42,29 @@
     @Override
     public void onBlockClicked(World worldIn, BlockPos pos, EntityPlayer playerIn)
     {
-        this.activate(worldIn, pos);
+        this.activate(worldIn, pos, playerIn); // CraftBukkit - add playerIn
         super.onBlockClicked(worldIn, pos, playerIn);
     }
 
     @Override
     public void onEntityWalk(World worldIn, BlockPos pos, Entity entityIn)
     {
-        this.activate(worldIn, pos);
-        super.onEntityWalk(worldIn, pos, entityIn);
+//        this.activate(worldIn, pos);
+//        super.onEntityWalk(worldIn, pos, entityIn);
+        if (entityIn instanceof EntityPlayer) {
+            org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerInteractEvent((EntityPlayer) entityIn, org.bukkit.event.block.Action.PHYSICAL, pos, null, null, null);
+            if (!event.isCancelled()) {
+                this.activate(worldIn, pos, entityIn); // add entity
+                super.onEntityWalk(worldIn, pos, entityIn);
+            }
+        } else {
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(entityIn.getBukkitEntity(), worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+            worldIn.getServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                this.activate(worldIn, pos, entityIn); // add entity
+                super.onEntityWalk(worldIn, pos, entityIn);
+            }
+        }
     }
 
     @Override
@@ -66,7 +80,7 @@
         float hitZ
     )
     {
-        this.activate(worldIn, pos);
+        this.activate(worldIn, pos, playerIn); // CraftBukkit - add playerIn
         return super.onBlockActivated(worldIn, pos, state, playerIn, hand, facing, hitX, hitY, hitZ);
     }
 
@@ -80,11 +94,27 @@
         }
     }
 
+    private void activate(World worldIn, BlockPos pos, Entity entity)
+    {
+        this.spawnParticles(worldIn, pos);
+
+        if (this == Blocks.REDSTONE_ORE)
+        {
+            if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.LIT_REDSTONE_ORE, 0).isCancelled()) {
+                return;
+            }
+            worldIn.setBlockState(pos, Blocks.LIT_REDSTONE_ORE.getDefaultState());
+        }
+    }
+
     @Override
     public void updateTick(World worldIn, BlockPos pos, IBlockState state, Random rand)
     {
         if (this == Blocks.LIT_REDSTONE_ORE)
         {
+            if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callBlockFadeEvent(worldIn.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), Blocks.REDSTONE_ORE).isCancelled()) {
+                return;
+            }
             worldIn.setBlockState(pos, Blocks.REDSTONE_ORE.getDefaultState());
         }
     }
@@ -111,12 +141,16 @@
     public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
     {
         super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
+    }
 
-        if (this.getItemDropped(state, worldIn.rand, fortune) != Item.getItemFromBlock(this))
+    @Override
+    public int getExpDrop(IBlockState state, net.minecraft.world.IBlockAccess world, BlockPos pos, int fortune)
+    {
+        if (this.getItemDropped(state, RANDOM, fortune) != Item.getItemFromBlock(this))
         {
-            int i = 1 + worldIn.rand.nextInt(5);
-            this.dropXpOnBlockBreak(worldIn, pos, i);
+            return 1 + RANDOM.nextInt(5);
         }
+        return 0;
     }
 
     @SideOnly(Side.CLIENT)
