--- before/net/minecraft/block/BlockShulkerBox.java
+++ after/net/minecraft/block/BlockShulkerBox.java
@@ -43,7 +43,7 @@
 public class BlockShulkerBox extends BlockContainer
 {
     public static final PropertyEnum<EnumFacing> FACING = PropertyDirection.create("facing");
-    private final EnumDyeColor color;
+    public final EnumDyeColor color;
 
     public BlockShulkerBox(EnumDyeColor colorIn)
     {
@@ -409,8 +409,12 @@
     {
         state = this.getActualState(state, worldIn, pos);
         EnumFacing enumfacing = state.getValue(FACING);
-        TileEntityShulkerBox.AnimationStatus tileentityshulkerbox$animationstatus = ((TileEntityShulkerBox)worldIn.getTileEntity(pos))
-                .getAnimationStatus();
+        TileEntityShulkerBox te = (TileEntityShulkerBox) worldIn.getTileEntity(pos);
+
+        // CREF: getTileEntity may return null
+        TileEntityShulkerBox.AnimationStatus tileentityshulkerbox$animationstatus =
+            te == null ? TileEntityShulkerBox.AnimationStatus.CLOSED : te.getAnimationStatus();
+
         return tileentityshulkerbox$animationstatus != TileEntityShulkerBox.AnimationStatus.CLOSED
                && (
                    tileentityshulkerbox$animationstatus != TileEntityShulkerBox.AnimationStatus.OPENED
