--- before/net/minecraft/block/BlockPortal.java
+++ after/net/minecraft/block/BlockPortal.java
@@ -65,7 +65,7 @@
     {
         super.updateTick(worldIn, pos, state, rand);
 
-        if (worldIn.provider.isSurfaceWorld()
+        if (worldIn.spigotConfig.enableZombiePigmenPortalSpawns && worldIn.provider.isSurfaceWorld() // Spigot
                 && worldIn.getGameRules().getBoolean("doMobSpawning")
                 && rand.nextInt(2000) < worldIn.getDifficulty().getId())
         {
@@ -124,19 +124,19 @@
     {
         BlockPortal.Size blockportal$size = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.X);
 
-        if (blockportal$size.isValid() && blockportal$size.portalBlockCount == 0)
+        if (blockportal$size.isValid() && blockportal$size.portalBlockCount == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, blockportal$size))
         {
-            blockportal$size.placePortalBlocks();
-            return true;
+            return blockportal$size.placePortalBlocksCB();
+//            return true;
         }
         else
         {
             BlockPortal.Size blockportal$size1 = new BlockPortal.Size(worldIn, pos, EnumFacing.Axis.Z);
 
-            if (blockportal$size1.isValid() && blockportal$size1.portalBlockCount == 0)
+            if (blockportal$size1.isValid() && blockportal$size1.portalBlockCount == 0 && !net.minecraftforge.event.ForgeEventFactory.onTrySpawnPortal(worldIn, pos, blockportal$size1))
             {
-                blockportal$size1.placePortalBlocks();
-                return true;
+                return blockportal$size1.placePortalBlocksCB();
+//                return true;
             }
             else
             {
@@ -240,6 +240,8 @@
     {
         if (!entityIn.isRiding() && !entityIn.isBeingRidden() && entityIn.isNonBoss())
         {
+            org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entityIn.getBukkitEntity(), new org.bukkit.Location(worldIn.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            worldIn.getServer().getPluginManager().callEvent(event);
             entityIn.setPortal(pos);
         }
     }
@@ -425,6 +427,7 @@
             private BlockPos bottomLeft;
             private int height;
             private int width;
+            java.util.Collection<org.bukkit.block.Block> blocks = new java.util.HashSet<>();
 
             public Size(World worldIn, BlockPos p_i45694_2_, EnumFacing.Axis p_i45694_3_)
             {
@@ -504,6 +507,8 @@
 
             protected int calculatePortalHeight()
             {
+                this.blocks.clear();
+                org.bukkit.World bworld = this.world.getWorld();
                 label56:
 
                 for (this.height = 0; this.height < 21; this.height++)
@@ -530,6 +535,9 @@
                             if (block != Blocks.OBSIDIAN)
                             {
                                 break label56;
+                            } else {
+                                BlockPos pos = blockpos.offset(this.leftDir);
+                                blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                             }
                         }
                         else if (i == this.width - 1)
@@ -539,6 +547,9 @@
                             if (block != Blocks.OBSIDIAN)
                             {
                                 break label56;
+                            } else {
+                                BlockPos pos = blockpos.offset(this.rightDir);
+                                blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                             }
                         }
                     }
@@ -551,6 +562,9 @@
                     {
                         this.height = 0;
                         break;
+                    } else {
+                        BlockPos pos = this.bottomLeft.offset(this.rightDir, j).up(this.height);
+                        blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
                     }
                 }
 
@@ -591,6 +605,39 @@
                         );
                     }
                 }
+            }
+
+            public boolean placePortalBlocksCB()
+            {
+                org.bukkit.World bworld = this.world.getWorld();
+
+                // Copy below for loop
+                for (int i = 0; i < this.width; ++i) {
+                    BlockPos blockposition = this.bottomLeft.offset(this.rightDir, i);
+
+                    for (int j = 0; j < this.height; ++j) {
+                        BlockPos pos = blockposition.up(j);
+                        blocks.add(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()));
+                    }
+                }
+
+                org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent(blocks, bworld, org.bukkit.event.world.PortalCreateEvent.CreateReason.FIRE);
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return false;
+                }
+                for (int i = 0; i < this.width; ++i)
+                {
+                    BlockPos blockpos = this.bottomLeft.offset(this.rightDir, i);
+
+                    for (int j = 0; j < this.height; ++j)
+                    {
+                        this.world.setBlockState(blockpos.up(j), Blocks.PORTAL.getDefaultState().withProperty(BlockPortal.AXIS, this.axis), 2);
+                    }
+                }
+
+                return true;
             }
         }
 }
