--- before/net/minecraft/block/BlockSapling.java
+++ after/net/minecraft/block/BlockSapling.java
@@ -29,6 +29,7 @@
     public static final PropertyEnum<BlockPlanks.EnumType> TYPE = PropertyEnum.create("type", BlockPlanks.EnumType.class);
     public static final PropertyInteger STAGE = PropertyInteger.create("stage", 0, 1);
     protected static final AxisAlignedBB SAPLING_AABB = new AxisAlignedBB(0.099999994F, 0.0, 0.099999994F, 0.9F, 0.8F, 0.9F);
+    public static org.bukkit.TreeType treeType;
 
     protected BlockSapling()
     {
@@ -55,9 +56,32 @@
         {
             super.updateTick(worldIn, pos, state, rand);
 
-            if (worldIn.getLightFromNeighbors(pos.up()) >= 9 && rand.nextInt(7) == 0)
+            if (!worldIn.isAreaLoaded(pos, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light
+            if (worldIn.getLightFromNeighbors(pos.up()) >= 9 && rand.nextInt(Math.max(2, (int) (((100.0F / worldIn.spigotConfig.saplingModifier) * 7) + 0.5F))) == 0)
             {
+                worldIn.captureTreeGeneration = true;
                 this.grow(worldIn, pos, state, rand);
+                worldIn.captureTreeGeneration = false;
+                if (!worldIn.capturedBlockSnapshots.isEmpty()) {
+                    org.bukkit.TreeType treeType = BlockSapling.treeType;
+                    BlockSapling.treeType = null;
+                    org.bukkit.Location location = new org.bukkit.Location(worldIn.getWorld(), pos.getX(), pos.getY(), pos.getZ());
+                    java.util.List<org.bukkit.block.BlockState> blockstates = new java.util.ArrayList();
+                    for (net.minecraftforge.common.util.BlockSnapshot snapshot : worldIn.capturedBlockSnapshots) {
+                        blockstates.add(new org.bukkit.craftbukkit.v1_12_R1.block.CraftBlockState(snapshot));
+                    }
+                    worldIn.capturedBlockSnapshots.clear();
+                    org.bukkit.event.world.StructureGrowEvent event = null;
+                    if (treeType != null) {
+                        event = new org.bukkit.event.world.StructureGrowEvent(location, treeType, false, null, blockstates);
+                        org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                    }
+                    if (event == null || !event.isCancelled()) {
+                        for (org.bukkit.block.BlockState blockstate : blockstates) {
+                            blockstate.update(true);
+                        }
+                    }
+                }
             }
         }
     }
@@ -76,7 +100,18 @@
 
     public void generateTree(World worldIn, BlockPos pos, IBlockState state, Random rand)
     {
-        WorldGenerator worldgenerator = (WorldGenerator)(rand.nextInt(10) == 0 ? new WorldGenBigTree(true) : new WorldGenTrees(true));
+        if (!net.minecraftforge.event.terraingen.TerrainGen.saplingGrowTree(worldIn, rand, pos)) return;
+        // CraftBukkit start - Turn ternary operator into if statement to set treeType
+        // WorldGenerator worldgenerator = (WorldGenerator)(rand.nextInt(10) == 0 ? new WorldGenBigTree(true) : new WorldGenTrees(true));
+        WorldGenerator worldgenerator;
+        if (rand.nextInt(10) == 0) {
+            treeType = org.bukkit.TreeType.BIG_TREE;
+            worldgenerator = new WorldGenBigTree(true);
+        } else {
+            treeType = org.bukkit.TreeType.TREE;
+            worldgenerator = new WorldGenTrees(true);
+        }
+        // CraftBukkit end
         int i = 0;
         int j = 0;
         boolean flag = false;
@@ -92,6 +127,7 @@
                     {
                         if (this.isTwoByTwoOfType(worldIn, pos, i, j, BlockPlanks.EnumType.SPRUCE))
                         {
+                            treeType = org.bukkit.TreeType.MEGA_REDWOOD;
                             worldgenerator = new WorldGenMegaPineTree(false, rand.nextBoolean());
                             flag = true;
                             break label68;
@@ -103,11 +139,13 @@
                 {
                     i = 0;
                     j = 0;
+                    treeType = org.bukkit.TreeType.REDWOOD;
                     worldgenerator = new WorldGenTaiga2(true);
                 }
 
                 break;
             case BIRCH:
+                treeType = org.bukkit.TreeType.BIRCH;
                 worldgenerator = new WorldGenBirchTree(true, false);
                 break;
             case JUNGLE:
@@ -124,6 +162,7 @@
                     {
                         if (this.isTwoByTwoOfType(worldIn, pos, i, j, BlockPlanks.EnumType.JUNGLE))
                         {
+                            treeType = org.bukkit.TreeType.JUNGLE;
                             worldgenerator = new WorldGenMegaJungle(true, 10, 20, iblockstate, iblockstate1);
                             flag = true;
                             break label82;
@@ -135,11 +174,13 @@
                 {
                     i = 0;
                     j = 0;
+                    treeType = org.bukkit.TreeType.SMALL_JUNGLE;
                     worldgenerator = new WorldGenTrees(true, 4 + rand.nextInt(7), iblockstate, iblockstate1, false);
                 }
 
                 break;
             case ACACIA:
+                treeType = org.bukkit.TreeType.ACACIA;
                 worldgenerator = new WorldGenSavannaTree(true);
                 break;
             case DARK_OAK:
@@ -151,6 +192,7 @@
                     {
                         if (this.isTwoByTwoOfType(worldIn, pos, i, j, BlockPlanks.EnumType.DARK_OAK))
                         {
+                            treeType = org.bukkit.TreeType.DARK_OAK;
                             worldgenerator = new WorldGenCanopyTree(true);
                             flag = true;
                             break label96;
