--- before/net/minecraft/block/BlockRailPowered.java
+++ after/net/minecraft/block/BlockRailPowered.java
@@ -30,9 +30,17 @@
             );
     public static final PropertyBool POWERED = PropertyBool.create("powered");
 
+    private final boolean isActivator;
+
     protected BlockRailPowered()
     {
+        this(false);
+    }
+
+    protected BlockRailPowered(boolean isActivator)
+    {
         super(true);
+        this.isActivator = isActivator;
         this.setDefaultState(
             this.blockState.getBaseState().withProperty(SHAPE, BlockRailBase.EnumRailDirection.NORTH_SOUTH).withProperty(POWERED, false)
         );
@@ -149,7 +157,7 @@
     {
         IBlockState iblockstate = worldIn.getBlockState(pos);
 
-        if (iblockstate.getBlock() != this)
+        if (!(iblockstate.getBlock() instanceof BlockRailPowered) || isActivator != ((BlockRailPowered)iblockstate.getBlock()).isActivator)
         {
             return false;
         }
@@ -200,6 +208,11 @@
 
         if (flag1 != flag)
         {
+            int power = state.getValue(POWERED) ? 15 : 0;
+            int newPower = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callRedstoneChange(worldIn, pos.getX(), pos.getY(), pos.getZ(), power, 15 - power).getNewCurrent();
+            if (newPower == power) {
+                return;
+            }
             worldIn.setBlockState(pos, state.withProperty(POWERED, flag1), 3);
             worldIn.notifyNeighborsOfStateChange(pos.down(), this, false);
 
