--- before/net/minecraft/block/BlockTNT.java
+++ after/net/minecraft/block/BlockTNT.java
@@ -143,6 +143,10 @@
 
             if (entityarrow.isBurning())
             {
+                if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(entityarrow, pos, Blocks.AIR, 0).isCancelled()) {
+                    return;
+                }
+
                 this.explode(
                     worldIn,
                     pos,
