--- before/net/minecraft/block/BlockTrapDoor.java
+++ after/net/minecraft/block/BlockTrapDoor.java
@@ -151,6 +151,18 @@
 
             if (flag || blockIn.getDefaultState().canProvidePower())
             {
+                org.bukkit.World bworld = worldIn.getWorld();
+                org.bukkit.block.Block bblock = bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+
+                int power = bblock.getBlockPower();
+                int oldPower = state.getValue(OPEN) ? 15 : 0;
+
+                if (oldPower == 0 ^ power == 0 || blockIn.getDefaultState().hasComparatorInputOverride()) {
+                    org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(bblock, oldPower, power);
+                    worldIn.getServer().getPluginManager().callEvent(eventRedstone);
+                    flag = eventRedstone.getNewCurrent() > 0;
+                }
+
                 boolean flag1 = state.getValue(OPEN);
 
                 if (flag1 != flag)
@@ -296,6 +308,18 @@
                && !state.getValue(OPEN)
                ? BlockFaceShape.SOLID
                : BlockFaceShape.UNDEFINED;
+    }
+
+    @Override
+    public boolean isLadder(IBlockState state, IBlockAccess world, BlockPos pos, EntityLivingBase entity)
+    {
+        if (state.getValue(OPEN))
+        {
+            IBlockState down = world.getBlockState(pos.down());
+            if (down.getBlock() == net.minecraft.init.Blocks.LADDER)
+                return down.getValue(BlockLadder.FACING) == state.getValue(FACING);
+        }
+        return false;
     }
 
     public static enum DoorHalf implements IStringSerializable
