--- before/net/minecraft/block/BlockDropper.java
+++ after/net/minecraft/block/BlockDropper.java
@@ -29,7 +29,7 @@
     }
 
     @Override
-    protected void dispense(World worldIn, BlockPos pos)
+    public void dispense(World worldIn, BlockPos pos) // Bukkit: protected -> public
     {
         BlockSourceImpl blocksourceimpl = new BlockSourceImpl(worldIn, pos);
         TileEntityDispenser tileentitydispenser = blocksourceimpl.getBlockTileEntity();
@@ -46,7 +46,7 @@
             {
                 ItemStack itemstack = tileentitydispenser.getStackInSlot(i);
 
-                if (!itemstack.isEmpty())
+                if (!itemstack.isEmpty() && net.minecraftforge.items.VanillaInventoryCodeHooks.dropperInsertHook(worldIn, pos, tileentitydispenser, i, itemstack))
                 {
                     EnumFacing enumfacing = worldIn.getBlockState(pos).getValue(FACING);
                     BlockPos blockpos = pos.offset(enumfacing);
