--- before/net/minecraft/block/BlockCauldron.java
+++ after/net/minecraft/block/BlockCauldron.java
@@ -30,6 +30,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class BlockCauldron extends Block
 {
@@ -90,8 +91,11 @@
 
         if (!worldIn.isRemote && entityIn.isBurning() && i > 0 && entityIn.getEntityBoundingBox().minY <= f)
         {
+            if (!this.changeLevel(worldIn, pos, state, i - 1, entityIn, CauldronLevelChangeEvent.ChangeReason.EXTINGUISH)) {
+                return;
+            }
             entityIn.extinguish();
-            this.setWaterLevel(worldIn, pos, state, i - 1);
+//            this.setWaterLevel(worldIn, pos, state, i - 1);
         }
     }
 
@@ -123,13 +127,16 @@
             {
                 if (i < 3 && !worldIn.isRemote)
                 {
+                    if (!this.changeLevel(worldIn, pos, state, 3, playerIn, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+                        return true;
+                    }
                     if (!playerIn.capabilities.isCreativeMode)
                     {
                         playerIn.setHeldItem(hand, new ItemStack(Items.BUCKET));
                     }
 
                     playerIn.addStat(StatList.CAULDRON_FILLED);
-                    this.setWaterLevel(worldIn, pos, state, 3);
+//                    this.setWaterLevel(worldIn, pos, state, 3);
                     worldIn.playSound(null, pos, SoundEvents.ITEM_BUCKET_EMPTY, SoundCategory.BLOCKS, 1.0F, 1.0F);
                 }
 
@@ -139,6 +146,9 @@
             {
                 if (i == 3 && !worldIn.isRemote)
                 {
+                    if (!this.changeLevel(worldIn, pos, state, 0, playerIn, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+                        return true;
+                    }
                     if (!playerIn.capabilities.isCreativeMode)
                     {
                         itemstack.shrink(1);
@@ -154,7 +164,7 @@
                     }
 
                     playerIn.addStat(StatList.CAULDRON_USED);
-                    this.setWaterLevel(worldIn, pos, state, 0);
+//                    this.setWaterLevel(worldIn, pos, state, 0);
                     worldIn.playSound(null, pos, SoundEvents.ITEM_BUCKET_FILL, SoundCategory.BLOCKS, 1.0F, 1.0F);
                 }
 
@@ -164,6 +174,9 @@
             {
                 if (i > 0 && !worldIn.isRemote)
                 {
+                    if (!this.changeLevel(worldIn, pos, state, i - 1, playerIn, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                        return true;
+                    }
                     if (!playerIn.capabilities.isCreativeMode)
                     {
                         ItemStack itemstack3 = PotionUtils.addPotionToItemStack(new ItemStack(Items.POTIONITEM), PotionTypes.WATER);
@@ -185,7 +198,7 @@
                     }
 
                     worldIn.playSound(null, pos, SoundEvents.ITEM_BOTTLE_FILL, SoundCategory.BLOCKS, 1.0F, 1.0F);
-                    this.setWaterLevel(worldIn, pos, state, i - 1);
+//                    this.setWaterLevel(worldIn, pos, state, i - 1);
                 }
 
                 return true;
@@ -194,6 +207,9 @@
             {
                 if (i < 3 && !worldIn.isRemote)
                 {
+                    if (!this.changeLevel(worldIn, pos, state, i + 1, playerIn, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                        return true;
+                    }
                     if (!playerIn.capabilities.isCreativeMode)
                     {
                         ItemStack itemstack2 = new ItemStack(Items.GLASS_BOTTLE);
@@ -207,7 +223,7 @@
                     }
 
                     worldIn.playSound(null, pos, SoundEvents.ITEM_BOTTLE_EMPTY, SoundCategory.BLOCKS, 1.0F, 1.0F);
-                    this.setWaterLevel(worldIn, pos, state, i + 1);
+//                    this.setWaterLevel(worldIn, pos, state, i + 1);
                 }
 
                 return true;
@@ -220,8 +236,11 @@
 
                     if (itemarmor.getArmorMaterial() == ItemArmor.ArmorMaterial.LEATHER && itemarmor.hasColor(itemstack) && !worldIn.isRemote)
                     {
+                        if (!this.changeLevel(worldIn, pos, state, i - 1, playerIn, CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                            return true;
+                        }
                         itemarmor.removeColor(itemstack);
-                        this.setWaterLevel(worldIn, pos, state, i - 1);
+//                        this.setWaterLevel(worldIn, pos, state, i - 1);
                         playerIn.addStat(StatList.ARMOR_CLEANED);
                         return true;
                     }
@@ -231,6 +250,9 @@
                 {
                     if (TileEntityBanner.getPatterns(itemstack) > 0 && !worldIn.isRemote)
                     {
+                        if (!this.changeLevel(worldIn, pos, state, i - 1, playerIn, CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                            return true;
+                        }
                         ItemStack itemstack1 = itemstack.copy();
                         itemstack1.setCount(1);
                         TileEntityBanner.removeBannerData(itemstack1);
@@ -239,7 +261,7 @@
                         if (!playerIn.capabilities.isCreativeMode)
                         {
                             itemstack.shrink(1);
-                            this.setWaterLevel(worldIn, pos, state, i - 1);
+//                            this.setWaterLevel(worldIn, pos, state, i - 1);
                         }
 
                         if (itemstack.isEmpty())
@@ -268,8 +290,27 @@
 
     public void setWaterLevel(World worldIn, BlockPos pos, IBlockState state, int level)
     {
-        worldIn.setBlockState(pos, state.withProperty(LEVEL, MathHelper.clamp(level, 0, 3)), 2);
-        worldIn.updateComparatorOutputLevel(pos, this);
+        this.changeLevel(worldIn, pos, state, level, null, CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
+    }
+
+    public void changeLevel(World world, BlockPos blockposition, IBlockState iblockdata, int i) {
+        this.changeLevel(world, blockposition, iblockdata, i, null, CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
+    }
+
+    private boolean changeLevel(World world, BlockPos blockposition, IBlockState iblockdata, int i, Entity entity, CauldronLevelChangeEvent.ChangeReason reason) {
+        int newLevel = MathHelper.clamp(i, 0, 3);
+        CauldronLevelChangeEvent event = new CauldronLevelChangeEvent(
+                world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()),
+                (entity == null) ? null : entity.getBukkitEntity(), reason, iblockdata.getValue(BlockCauldron.LEVEL), newLevel
+        );
+        world.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return false;
+        }
+        world.setBlockState(blockposition, iblockdata.withProperty(BlockCauldron.LEVEL, event.getNewLevel()), 2);
+        world.updateComparatorOutputLevel(blockposition, this);
+        world.updateComparatorOutputLevel(blockposition, this);
+        return true;
     }
 
     @Override
@@ -279,13 +320,13 @@
         {
             float f = worldIn.getBiome(pos).getTemperature(pos);
 
-            if (!(worldIn.getBiomeProvider().getTemperatureAtHeight(f, pos.getY()) < 0.15F))
+            if (!(worldIn.getBiomeProvider().getTemperatureAtHeight(f, pos.getY()) >= 0.15F))
             {
                 IBlockState iblockstate = worldIn.getBlockState(pos);
 
                 if (iblockstate.getValue(LEVEL) < 3)
                 {
-                    worldIn.setBlockState(pos, iblockstate.cycleProperty(LEVEL), 2);
+                    this.changeLevel(worldIn, pos, iblockstate.cycleProperty(LEVEL), 2);
                 }
             }
         }
