--- before/net/minecraft/world/storage/MapData.java
+++ after/net/minecraft/world/storage/MapData.java
@@ -20,7 +20,7 @@
 {
     public int xCenter;
     public int zCenter;
-    public byte dimension;
+    public int dimension; //FML byte -> int
     public boolean trackingPosition;
     public boolean unlimitedTracking;
     public byte scale;
@@ -29,9 +29,15 @@
     private final Map<EntityPlayer, MapData.MapInfo> playersHashMap = Maps.newHashMap();
     public Map<String, MapDecoration> mapDecorations = Maps.newLinkedHashMap();
 
+    public final org.bukkit.craftbukkit.v1_12_R1.map.CraftMapView mapView;
+    private org.bukkit.craftbukkit.v1_12_R1.CraftServer server;
+    private java.util.UUID uniqueId = null;
+
     public MapData(String mapname)
     {
         super(mapname);
+        mapView = new org.bukkit.craftbukkit.v1_12_R1.map.CraftMapView(this);
+        server = (org.bukkit.craftbukkit.v1_12_R1.CraftServer) org.bukkit.Bukkit.getServer();
     }
 
     public void calculateMapCenter(double x, double z, int mapScale)
@@ -46,7 +52,31 @@
     @Override
     public void readFromNBT(NBTTagCompound nbt)
     {
-        this.dimension = nbt.getByte("dimension");
+        // this.dimension = nbt.getInteger("dimension");
+        // CraftBukkit start
+        int dimension = nbt.getInteger("dimension");
+
+        if (dimension >= 10) {
+            long least = nbt.getLong("UUIDLeast");
+            long most = nbt.getLong("UUIDMost");
+
+            if (least != 0L && most != 0L) {
+                this.uniqueId = new java.util.UUID(most, least);
+
+                org.bukkit.craftbukkit.v1_12_R1.CraftWorld world = (org.bukkit.craftbukkit.v1_12_R1.CraftWorld) server.getWorld(this.uniqueId);
+                // Check if the stored world details are correct.
+                if (world == null) {
+                    /* All Maps which do not have their valid world loaded are set to a dimension which hopefully won't be reached.
+                       This is to prevent them being corrupted with the wrong map data. */
+                    dimension = 127;
+                } else {
+                    dimension = (byte) world.getHandle().dimension;
+                }
+            }
+        }
+
+        this.dimension = dimension;
+        // CraftBukkit end
         this.xCenter = nbt.getInteger("xCenter");
         this.zCenter = nbt.getInteger("zCenter");
         this.scale = nbt.getByte("scale");
@@ -99,7 +129,26 @@
     @Override
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
     {
-        compound.setByte("dimension", this.dimension);
+        // CraftBukkit start
+        if (this.dimension >= 10) {
+            if (this.uniqueId == null) {
+                for (org.bukkit.World world : server.getWorlds()) {
+                    org.bukkit.craftbukkit.v1_12_R1.CraftWorld cWorld = (org.bukkit.craftbukkit.v1_12_R1.CraftWorld) world;
+                    if (cWorld.getHandle().dimension == this.dimension) {
+                        this.uniqueId = cWorld.getUID();
+                        break;
+                    }
+                }
+            }
+            /* Perform a second check to see if a matching world was found, this is a necessary
+               change incase Maps are forcefully unlinked from a World and lack a UID.*/
+            if (this.uniqueId != null) {
+                compound.setLong("UUIDLeast", this.uniqueId.getLeastSignificantBits());
+                compound.setLong("UUIDMost", this.uniqueId.getMostSignificantBits());
+            }
+        }
+        // CraftBukkit end
+        compound.setInteger("dimension", this.dimension);
         compound.setInteger("xCenter", this.xCenter);
         compound.setInteger("zCenter", this.zCenter);
         compound.setByte("scale", this.scale);
@@ -232,9 +281,9 @@
             rotationIn += rotationIn < 0.0 ? -8.0 : 8.0;
             b2 = (byte)(rotationIn * 16.0 / 360.0);
 
-            if (this.dimension < 0)
+            if (worldIn.provider.shouldMapSpin(decorationName, worldX, worldZ, rotationIn))
             {
-                int l = (int)(worldIn.getWorldInfo().getWorldTime() / 10L);
+                int l = (int)(worldIn.getWorldTime() / 10L);
                 b2 = (byte)(l * l * 34187121 + l * 121 >> 15 & 15);
             }
         }
@@ -339,6 +388,17 @@
         @Nullable
         public Packet<?> getPacket(ItemStack stack)
         {
+            org.bukkit.craftbukkit.v1_12_R1.map.RenderData render = MapData.this.mapView.render((org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer) this.player.getBukkitEntity());
+
+            java.util.Collection<MapDecoration> icons = new java.util.ArrayList<>();
+
+            for ( org.bukkit.map.MapCursor cursor : render.cursors) {
+
+                if (cursor.isVisible()) {
+                    icons.add(new MapDecoration(MapDecoration.Type.byIcon(cursor.getRawType()), cursor.getX(), cursor.getY(), cursor.getDirection()));
+                }
+            }
+
             if (this.isDirty)
             {
                 this.isDirty = false;
@@ -346,8 +406,8 @@
                            stack.getMetadata(),
                            MapData.this.scale,
                            MapData.this.trackingPosition,
-                           MapData.this.mapDecorations.values(),
-                           MapData.this.colors,
+                           icons,
+                           render.buffer,
                            this.minX,
                            this.minY,
                            this.maxX + 1 - this.minX,
@@ -361,8 +421,8 @@
                            stack.getMetadata(),
                            MapData.this.scale,
                            MapData.this.trackingPosition,
-                           MapData.this.mapDecorations.values(),
-                           MapData.this.colors,
+                           icons,
+                           render.buffer,
                            0,
                            0,
                            0,
