--- before/net/minecraft/world/storage/ThreadedFileIOBase.java
+++ after/net/minecraft/world/storage/ThreadedFileIOBase.java
@@ -8,8 +8,8 @@
 {
     private static final ThreadedFileIOBase INSTANCE = new ThreadedFileIOBase();
     private final List<IThreadedFileIO> threadedIOQueue = Collections.synchronizedList(Lists.newArrayList());
-    private volatile long writeQueuedCounter;
-    private volatile long savedIOCounter;
+    private final java.util.concurrent.atomic.AtomicLong writeQueuedCounter = new java.util.concurrent.atomic.AtomicLong();
+    private final java.util.concurrent.atomic.AtomicLong savedIOCounter = new java.util.concurrent.atomic.AtomicLong();
     private volatile boolean isThreadWaiting;
 
     private ThreadedFileIOBase()
@@ -35,7 +35,7 @@
 
     private void processQueue()
     {
-        for (int i = 0; i < this.threadedIOQueue.size(); i++)
+        for (int i = 0; i < this.threadedIOQueue.size(); ++i)
         {
             IThreadedFileIO ithreadedfileio = this.threadedIOQueue.get(i);
             boolean flag = ithreadedfileio.writeNextIO();
@@ -43,9 +43,10 @@
             if (!flag)
             {
                 this.threadedIOQueue.remove(i--);
-                this.savedIOCounter++;
+                this.savedIOCounter.incrementAndGet();
             }
 
+            /*
             try
             {
                 Thread.sleep(this.isThreadWaiting ? 0L : 10L);
@@ -54,6 +55,7 @@
             {
                 interruptedexception1.printStackTrace();
             }
+            */
         }
 
         if (this.threadedIOQueue.isEmpty())
@@ -73,7 +75,7 @@
     {
         if (!this.threadedIOQueue.contains(fileIo))
         {
-            this.writeQueuedCounter++;
+            this.writeQueuedCounter.incrementAndGet();
             this.threadedIOQueue.add(fileIo);
         }
     }
@@ -82,7 +84,7 @@
     {
         this.isThreadWaiting = true;
 
-        while (this.writeQueuedCounter != this.savedIOCounter)
+        while (this.writeQueuedCounter.get() == this.savedIOCounter.get())
         {
             Thread.sleep(10L);
         }
