--- before/net/minecraft/command/CommandTP.java
+++ after/net/minecraft/command/CommandTP.java
@@ -59,27 +59,13 @@
             {
                 Entity entity1 = getEntity(server, sender, args[args.length - 1]);
 
-                if (entity1.world != entity.world)
-                {
-                    throw new CommandException("commands.tp.notSameDimension");
-                }
-                else
-                {
-                    entity.dismountRidingEntity();
-
-                    if (entity instanceof EntityPlayerMP)
-                    {
-                        ((EntityPlayerMP)entity)
-                        .connection
-                        .setPlayerLocation(entity1.posX, entity1.posY, entity1.posZ, entity1.rotationYaw, entity1.rotationPitch);
-                    }
-                    else
-                    {
-                        entity.setLocationAndAngles(entity1.posX, entity1.posY, entity1.posZ, entity1.rotationYaw, entity1.rotationPitch);
-                    }
-
-                    notifyCommandListener(sender, this, "commands.tp.success", entity.getName(), entity1.getName());
-                }
+                // CraftBukkit Start
+                // Use Bukkit teleport method in all cases. It has cross dimensional handling, events
+                if (entity.getBukkitEntity().teleport(entity1.getBukkitEntity(), org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND))
+                {
+                    notifyCommandListener(sender, (ICommand)this, "commands.tp.success", entity.getName(), entity1.getName());
+                }
+                // CraftBukkit End
             }
             else if (args.length < i + 3)
             {
