--- before/net/minecraft/command/CommandBase.java
+++ after/net/minecraft/command/CommandBase.java
@@ -40,6 +40,7 @@
     private static ICommandListener commandListener;
     private static final Splitter COMMA_SPLITTER = Splitter.on(',');
     private static final Splitter EQUAL_SPLITTER = Splitter.on('=').limit(2);
+    public String permissionNode = "minecraft";
 
     protected static SyntaxErrorException toSyntaxException(JsonParseException e)
     {
@@ -398,7 +399,7 @@
                 itextcomponent.appendText(" ");
             }
 
-            ITextComponent itextcomponent1 = new TextComponentString(args[i]);
+            ITextComponent itextcomponent1 = net.minecraftforge.common.ForgeHooks.newChatWithLinks(args[i]); // Forge: links for messages
 
             if (p_147176_3_)
             {
