--- before/net/minecraft/command/CommandBase.java
+++ after/net/minecraft/command/CommandBase.java
@@ -40,6 +40,7 @@
     private static ICommandListener commandListener;
     private static final Splitter COMMA_SPLITTER = Splitter.on(',');
     private static final Splitter EQUAL_SPLITTER = Splitter.on('=').limit(2);
+    public String permissionNode = "minecraft";
 
     protected static SyntaxErrorException toSyntaxException(JsonParseException e)
     {
@@ -56,7 +57,7 @@
             }
         }
 
-        return new SyntaxErrorException("commands.tellraw.jsonException", s);
+        return new SyntaxErrorException("commands.tellraw.jsonException", new Object[] {s});
     }
 
     public static NBTTagCompound entityToNBT(Entity theEntity)
@@ -81,22 +82,19 @@
         return 4;
     }
 
-    @Override
     public List<String> getAliases()
     {
-        return Collections.emptyList();
+        return Collections.<String>emptyList();
     }
 
-    @Override
     public boolean checkPermission(MinecraftServer server, ICommandSender sender)
     {
         return sender.canUseCommand(this.getRequiredPermissionLevel(), this.getName());
     }
 
-    @Override
     public List<String> getTabCompletions(MinecraftServer server, ICommandSender sender, String[] args, @Nullable BlockPos targetPos)
     {
-        return Collections.emptyList();
+        return Collections.<String>emptyList();
     }
 
     public static int parseInt(String input) throws NumberInvalidException
@@ -105,9 +103,9 @@
         {
             return Integer.parseInt(input);
         }
-        catch (NumberFormatException numberformatexception)
+        catch (NumberFormatException var2)
         {
-            throw new NumberInvalidException("commands.generic.num.invalid", input);
+            throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {input});
         }
     }
 
@@ -122,11 +120,11 @@
 
         if (i < min)
         {
-            throw new NumberInvalidException("commands.generic.num.tooSmall", i, min);
+            throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {i, min});
         }
         else if (i > max)
         {
-            throw new NumberInvalidException("commands.generic.num.tooBig", i, max);
+            throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {i, max});
         }
         else
         {
@@ -140,9 +138,9 @@
         {
             return Long.parseLong(input);
         }
-        catch (NumberFormatException numberformatexception)
+        catch (NumberFormatException var2)
         {
-            throw new NumberInvalidException("commands.generic.num.invalid", input);
+            throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {input});
         }
     }
 
@@ -152,11 +150,11 @@
 
         if (i < min)
         {
-            throw new NumberInvalidException("commands.generic.num.tooSmall", i, min);
+            throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {i, min});
         }
         else if (i > max)
         {
-            throw new NumberInvalidException("commands.generic.num.tooBig", i, max);
+            throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {i, max});
         }
         else
         {
@@ -167,11 +165,7 @@
     public static BlockPos parseBlockPos(ICommandSender sender, String[] args, int startIndex, boolean centerBlock) throws NumberInvalidException
     {
         BlockPos blockpos = sender.getPosition();
-        return new BlockPos(
-                   parseDouble((double)blockpos.getX(), args[startIndex], -30000000, 30000000, centerBlock),
-                   parseDouble((double)blockpos.getY(), args[startIndex + 1], 0, 256, false),
-                   parseDouble((double)blockpos.getZ(), args[startIndex + 2], -30000000, 30000000, centerBlock)
-               );
+        return new BlockPos(parseDouble((double)blockpos.getX(), args[startIndex], -30000000, 30000000, centerBlock), parseDouble((double)blockpos.getY(), args[startIndex + 1], 0, 256, false), parseDouble((double)blockpos.getZ(), args[startIndex + 2], -30000000, 30000000, centerBlock));
     }
 
     public static double parseDouble(String input) throws NumberInvalidException
@@ -182,16 +176,16 @@
 
             if (!Doubles.isFinite(d0))
             {
-                throw new NumberInvalidException("commands.generic.num.invalid", input);
+                throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {input});
             }
             else
             {
                 return d0;
             }
         }
-        catch (NumberFormatException numberformatexception)
+        catch (NumberFormatException var3)
         {
-            throw new NumberInvalidException("commands.generic.num.invalid", input);
+            throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {input});
         }
     }
 
@@ -206,11 +200,11 @@
 
         if (d0 < min)
         {
-            throw new NumberInvalidException("commands.generic.num.tooSmall", String.format("%.2f", d0), String.format("%.2f", min));
+            throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {String.format("%.2f", d0), String.format("%.2f", min)});
         }
         else if (d0 > max)
         {
-            throw new NumberInvalidException("commands.generic.num.tooBig", String.format("%.2f", d0), String.format("%.2f", max));
+            throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {String.format("%.2f", d0), String.format("%.2f", max)});
         }
         else
         {
@@ -220,18 +214,21 @@
 
     public static boolean parseBoolean(String input) throws CommandException
     {
-        if ("true".equals(input) || "1".equals(input))
+        if (!"true".equals(input) && !"1".equals(input))
+        {
+            if (!"false".equals(input) && !"0".equals(input))
+            {
+                throw new CommandException("commands.generic.boolean.invalid", new Object[] {input});
+            }
+            else
+            {
+                return false;
+            }
+        }
+        else
         {
             return true;
         }
-        else if (!"false".equals(input) && !"0".equals(input))
-        {
-            throw new CommandException("commands.generic.boolean.invalid", input);
-        }
-        else
-        {
-            return false;
-        }
     }
 
     public static EntityPlayerMP getCommandSenderAsPlayer(ICommandSender sender) throws PlayerNotFoundException
@@ -249,7 +246,7 @@
     public static List<EntityPlayerMP> getPlayers(MinecraftServer p_193513_0_, ICommandSender p_193513_1_, String p_193513_2_) throws CommandException
     {
         List<EntityPlayerMP> list = EntitySelector.getPlayers(p_193513_1_, p_193513_2_);
-        return (List<EntityPlayerMP>)(list.isEmpty() ? Lists.newArrayList(getPlayer(p_193513_0_, null, p_193513_2_)) : list);
+        return (List<EntityPlayerMP>)(list.isEmpty() ? Lists.newArrayList(getPlayer(p_193513_0_, (EntityPlayerMP)null, p_193513_2_)) : list);
     }
 
     public static EntityPlayerMP getPlayer(MinecraftServer server, ICommandSender sender, String target) throws PlayerNotFoundException, CommandException
@@ -265,8 +262,9 @@
             {
                 p_193512_1_ = p_193512_0_.getPlayerList().getPlayerByUUID(UUID.fromString(p_193512_2_));
             }
-            catch (IllegalArgumentException illegalargumentexception)
+            catch (IllegalArgumentException var4)
             {
+                ;
             }
         }
 
@@ -277,7 +275,7 @@
 
         if (p_193512_1_ == null)
         {
-            throw new PlayerNotFoundException("commands.generic.player.notFound", p_193512_2_);
+            throw new PlayerNotFoundException("commands.generic.player.notFound", new Object[] {p_193512_2_});
         }
         else
         {
@@ -311,11 +309,11 @@
                     entity = server.getPlayerList().getPlayerByUUID(uuid);
                 }
             }
-            catch (IllegalArgumentException illegalargumentexception)
+            catch (IllegalArgumentException var6)
             {
                 if (target.split("-").length == 5)
                 {
-                    throw new EntityNotFoundException("commands.generic.entity.invalidUuid", target);
+                    throw new EntityNotFoundException("commands.generic.entity.invalidUuid", new Object[] {target});
                 }
             }
         }
@@ -332,9 +330,7 @@
 
     public static List<Entity> getEntityList(MinecraftServer server, ICommandSender sender, String target) throws EntityNotFoundException, CommandException
     {
-        return (List<Entity>)(EntitySelector.isSelector(target)
-                              ? EntitySelector.matchEntities(sender, target, Entity.class)
-                              : Lists.newArrayList(getEntity(server, sender, target)));
+        return (List<Entity>)(EntitySelector.isSelector(target) ? EntitySelector.matchEntities(sender, target, Entity.class) : Lists.newArrayList(getEntity(server, sender, target)));
     }
 
     public static String getPlayerName(MinecraftServer server, ICommandSender sender, String target) throws PlayerNotFoundException, CommandException
@@ -362,7 +358,7 @@
         {
             return getPlayer(server, sender, target).getName();
         }
-        catch (PlayerNotFoundException playernotfoundexception)
+        catch (PlayerNotFoundException var6)
         {
             try
             {
@@ -391,14 +387,14 @@
     {
         ITextComponent itextcomponent = new TextComponentString("");
 
-        for (int i = index; i < args.length; i++)
+        for (int i = index; i < args.length; ++i)
         {
             if (i > index)
             {
                 itextcomponent.appendText(" ");
             }
 
-            ITextComponent itextcomponent1 = new TextComponentString(args[i]);
+            ITextComponent itextcomponent1 = net.minecraftforge.common.ForgeHooks.newChatWithLinks(args[i]); // Forge: links for messages
 
             if (p_147176_3_)
             {
@@ -408,7 +404,7 @@
                 {
                     if (EntitySelector.isSelector(args[i]))
                     {
-                        throw new PlayerNotFoundException("commands.generic.selector.notFound", args[i]);
+                        throw new PlayerNotFoundException("commands.generic.selector.notFound", new Object[] {args[i]});
                     }
                 }
                 else
@@ -427,7 +423,7 @@
     {
         StringBuilder stringbuilder = new StringBuilder();
 
-        for (int i = startPos; i < args.length; i++)
+        for (int i = startPos; i < args.length; ++i)
         {
             if (i > startPos)
             {
@@ -452,11 +448,11 @@
 
         if (flag && Double.isNaN(base))
         {
-            throw new NumberInvalidException("commands.generic.num.invalid", base);
+            throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {base});
         }
         else
         {
-            double d0 = 0.0;
+            double d0 = 0.0D;
 
             if (!flag || selectorArg.length() > 1)
             {
@@ -471,22 +467,22 @@
 
                 if (!flag1 && !flag && centerBlock)
                 {
-                    d0 += 0.5;
+                    d0 += 0.5D;
                 }
             }
 
-            double d1 = d0 + (flag ? base : 0.0);
+            double d1 = d0 + (flag ? base : 0.0D);
 
             if (min != 0 || max != 0)
             {
                 if (d1 < (double)min)
                 {
-                    throw new NumberInvalidException("commands.generic.num.tooSmall", String.format("%.2f", d1), min);
+                    throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {String.format("%.2f", d1), min});
                 }
 
                 if (d1 > (double)max)
                 {
-                    throw new NumberInvalidException("commands.generic.num.tooBig", String.format("%.2f", d1), max);
+                    throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {String.format("%.2f", d1), max});
                 }
             }
 
@@ -505,11 +501,11 @@
 
         if (flag && Double.isNaN(base))
         {
-            throw new NumberInvalidException("commands.generic.num.invalid", base);
+            throw new NumberInvalidException("commands.generic.num.invalid", new Object[] {base});
         }
         else
         {
-            double d0 = flag ? base : 0.0;
+            double d0 = flag ? base : 0.0D;
 
             if (!flag || input.length() > 1)
             {
@@ -524,7 +520,7 @@
 
                 if (!flag1 && !flag && centerBlock)
                 {
-                    d0 += 0.5;
+                    d0 += 0.5D;
                 }
             }
 
@@ -532,12 +528,12 @@
             {
                 if (d0 < (double)min)
                 {
-                    throw new NumberInvalidException("commands.generic.num.tooSmall", String.format("%.2f", d0), min);
+                    throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {String.format("%.2f", d0), min});
                 }
 
                 if (d0 > (double)max)
                 {
-                    throw new NumberInvalidException("commands.generic.num.tooBig", String.format("%.2f", d0), max);
+                    throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {String.format("%.2f", d0), max});
                 }
             }
 
@@ -552,7 +548,7 @@
 
         if (item == null)
         {
-            throw new NumberInvalidException("commands.give.item.notFound", resourcelocation);
+            throw new NumberInvalidException("commands.give.item.notFound", new Object[] {resourcelocation});
         }
         else
         {
@@ -566,7 +562,7 @@
 
         if (!Block.REGISTRY.containsKey(resourcelocation))
         {
-            throw new NumberInvalidException("commands.give.block.notFound", resourcelocation);
+            throw new NumberInvalidException("commands.give.block.notFound", new Object[] {resourcelocation});
         }
         else
         {
@@ -582,18 +578,18 @@
 
             if (i < 0)
             {
-                throw new NumberInvalidException("commands.generic.num.tooSmall", i, 0);
+                throw new NumberInvalidException("commands.generic.num.tooSmall", new Object[] {i, Integer.valueOf(0)});
             }
             else if (i > 15)
             {
-                throw new NumberInvalidException("commands.generic.num.tooBig", i, 15);
+                throw new NumberInvalidException("commands.generic.num.tooBig", new Object[] {i, Integer.valueOf(15)});
             }
             else
             {
                 return p_190794_0_.getStateFromMeta(Integer.parseInt(p_190794_1_));
             }
         }
-        catch (RuntimeException runtimeexception1)
+        catch (RuntimeException var7)
         {
             try
             {
@@ -607,16 +603,16 @@
 
                 return iblockstate;
             }
-            catch (RuntimeException runtimeexception)
+            catch (RuntimeException var6)
             {
-                throw new InvalidBlockStateException("commands.generic.blockstate.invalid", p_190794_1_, Block.REGISTRY.getNameForObject(p_190794_0_));
+                throw new InvalidBlockStateException("commands.generic.blockstate.invalid", new Object[] {p_190794_1_, Block.REGISTRY.getNameForObject(p_190794_0_)});
             }
         }
     }
 
     private static <T extends Comparable<T>> IBlockState getBlockState(IBlockState p_190793_0_, IProperty<T> p_190793_1_, Comparable<?> p_190793_2_)
     {
-        return p_190793_0_.withProperty(p_190793_1_, (T) p_190793_2_);
+        return p_190793_0_.withProperty(p_190793_1_, (T)p_190793_2_);
     }
 
     public static Predicate<IBlockState> convertArgToBlockStatePredicate(final Block p_190791_0_, String p_190791_1_) throws InvalidBlockStateException
@@ -634,7 +630,7 @@
                     }
                 };
             }
-            catch (RuntimeException runtimeexception)
+            catch (RuntimeException var3)
             {
                 final Map < IProperty<?>, Comparable<? >> map = getBlockStatePropertyValueMap(p_190791_0_, p_190791_1_);
                 return new Predicate<IBlockState>()
@@ -669,7 +665,7 @@
 
     private static Map < IProperty<?>, Comparable<? >> getBlockStatePropertyValueMap(Block p_190795_0_, String p_190795_1_) throws InvalidBlockStateException
     {
-        Map < IProperty<?>, Comparable<? >> map = Maps.newHashMap();
+        Map < IProperty<?>, Comparable<? >> map = Maps. < IProperty<?>, Comparable<? >> newHashMap();
 
         if ("default".equals(p_190795_1_))
         {
@@ -678,7 +674,7 @@
         else
         {
             BlockStateContainer blockstatecontainer = p_190795_0_.getBlockState();
-            Iterator<String> iterator = COMMA_SPLITTER.split(p_190795_1_).iterator();
+            Iterator iterator = COMMA_SPLITTER.split(p_190795_1_).iterator();
 
             while (true)
             {
@@ -702,7 +698,7 @@
                     break;
                 }
 
-                Comparable<?> comparable = getValueHelper((IProperty<?>)iproperty, iterator1.next());
+                Comparable<?> comparable = getValueHelper(iproperty, iterator1.next());
 
                 if (comparable == null)
                 {
@@ -712,21 +708,21 @@
                 map.put(iproperty, comparable);
             }
 
-            throw new InvalidBlockStateException("commands.generic.blockstate.invalid", p_190795_1_, Block.REGISTRY.getNameForObject(p_190795_0_));
+            throw new InvalidBlockStateException("commands.generic.blockstate.invalid", new Object[] {p_190795_1_, Block.REGISTRY.getNameForObject(p_190795_0_)});
         }
     }
 
     @Nullable
     private static <T extends Comparable<T>> T getValueHelper(IProperty<T> p_190792_0_, String p_190792_1_)
     {
-        return p_190792_0_.parseValue(p_190792_1_).orNull();
+        return (T)(p_190792_0_.parseValue(p_190792_1_).orNull());
     }
 
     public static String joinNiceString(Object[] elements)
     {
         StringBuilder stringbuilder = new StringBuilder();
 
-        for (int i = 0; i < elements.length; i++)
+        for (int i = 0; i < elements.length; ++i)
         {
             String s = elements[i].toString();
 
@@ -752,7 +748,7 @@
     {
         ITextComponent itextcomponent = new TextComponentString("");
 
-        for (int i = 0; i < components.size(); i++)
+        for (int i = 0; i < components.size(); ++i)
         {
             if (i > 0)
             {
@@ -800,7 +796,7 @@
             {
                 if (i != index + 2)
                 {
-                    return Collections.emptyList();
+                    return Collections.<String>emptyList();
                 }
 
                 s = Integer.toString(pos.getZ());
@@ -829,7 +825,7 @@
             {
                 if (i != index + 1)
                 {
-                    return Collections.emptyList();
+                    return Collections.<String>emptyList();
                 }
 
                 s = Integer.toString(lookedPos.getZ());
@@ -852,11 +848,11 @@
     public static List<String> getListOfStringsMatchingLastWord(String[] inputArgs, Collection<?> possibleCompletions)
     {
         String s = inputArgs[inputArgs.length - 1];
-        List<String> list = Lists.newArrayList();
+        List<String> list = Lists.<String>newArrayList();
 
         if (!possibleCompletions.isEmpty())
         {
-            for (String s1 : Iterables.transform((Iterable<Object>)possibleCompletions, Functions.toStringFunction()))
+            for (String s1 : Iterables.transform(possibleCompletions, Functions.toStringFunction()))
             {
                 if (doesStringStartWith(s, s1))
                 {
@@ -879,7 +875,6 @@
         return list;
     }
 
-    @Override
     public boolean isUsernameIndex(String[] args, int index)
     {
         return false;
