--- before/net/minecraft/util/datafix/walkers/BlockEntityTag.java
+++ after/net/minecraft/util/datafix/walkers/BlockEntityTag.java
@@ -45,7 +45,7 @@
 
                 if (s1 == null)
                 {
-                    LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", s);
+                   // LOGGER.warn("Unable to resolve BlockEntity for ItemInstance: {}", (Object)s); // CraftBukkit - Remove unnecessary warning (occurs when deserializing a Shulker Box item)
                     flag = false;
                 }
                 else
