--- before/net/minecraft/tileentity/TileEntityBrewingStand.java
+++ after/net/minecraft/tileentity/TileEntityBrewingStand.java
@@ -35,6 +35,32 @@
     private String customName;
     private int fuel;
 
+    private int lastTick = catserver.server.CatServer.getCurrentTick(); // CatServer - implement realtime
+    // CraftBukkit start - add fields and methods
+    public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+    private int maxStack = MAX_STACK;
+    
+    public java.util.List<ItemStack> getContents() {
+        return this.brewingItemStacks;
+    }
+        
+    public void onOpen(org.bukkit.craftbukkit.v1_12_R1.entity.CraftHumanEntity who) {
+        transaction.add(who);
+    }
+    
+    public void onClose(org.bukkit.craftbukkit.v1_12_R1.entity.CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+        
+    public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return transaction;
+    }
+        
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+    
     @Override
     public String getName()
     {
@@ -79,8 +105,19 @@
 
         if (this.fuel <= 0 && itemstack.getItem() == Items.BLAZE_POWDER)
         {
-            this.fuel = 20;
-            itemstack.shrink(1);
+            // this.fuel = 20;
+            // itemstack.shrink(1);
+            org.bukkit.event.inventory.BrewingStandFuelEvent event = new org.bukkit.event.inventory.BrewingStandFuelEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack), 20);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
+
+            this.fuel = event.getFuelPower();
+            if (this.fuel > 0 && event.isConsuming()) {
+                itemstack.shrink(1);
+            }
             this.markDirty();
         }
 
@@ -88,10 +125,14 @@
         boolean flag1 = this.brewTime > 0;
         ItemStack itemstack1 = this.brewingItemStacks.get(3);
 
+        // CraftBukkit start - Use wall time instead of ticks for brewing
+        int elapsedTicks = catserver.server.CatServer.getCurrentTick() - this.lastTick; // CatServer - implement realtime
+        this.lastTick = catserver.server.CatServer.getCurrentTick(); // CatServer - implement realtime
+
         if (flag1)
         {
-            this.brewTime--;
-            boolean flag2 = this.brewTime == 0;
+            this.brewTime -= elapsedTicks;
+            boolean flag2 = this.brewTime <= 0; // == -> <=
 
             if (flag2 && flag)
             {
@@ -158,6 +199,7 @@
 
     private boolean canBrew()
     {
+        if (1 == 1) return net.minecraftforge.common.brewing.BrewingRecipeRegistry.canBrew(brewingItemStacks, brewingItemStacks.get(3), OUTPUT_SLOTS); // divert to VanillaBrewingRegistry
         ItemStack itemstack = this.brewingItemStacks.get(3);
 
         if (itemstack.isEmpty())
@@ -186,19 +228,26 @@
 
     private void brewPotions()
     {
+        if (net.minecraftforge.event.ForgeEventFactory.onPotionAttemptBrew(brewingItemStacks)) return;
         ItemStack itemstack = this.brewingItemStacks.get(3);
-
-        for (int i = 0; i < 3; i++)
-        {
-            this.brewingItemStacks.set(i, PotionHelper.doReaction(itemstack, this.brewingItemStacks.get(i)));
+        
+        org.bukkit.inventory.InventoryHolder owner = this.getOwner();
+        if (owner != null) {
+            org.bukkit.event.inventory.BrewEvent event = new org.bukkit.event.inventory.BrewEvent(world.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ()), (org.bukkit.inventory.BrewerInventory) owner.getInventory(), this.fuel);
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
         }
+
+        net.minecraftforge.common.brewing.BrewingRecipeRegistry.brewPotions(brewingItemStacks, brewingItemStacks.get(3), OUTPUT_SLOTS);
 
         itemstack.shrink(1);
         BlockPos blockpos = this.getPos();
 
-        if (itemstack.getItem().hasContainerItem())
+        if (itemstack.getItem().hasContainerItem(itemstack))
         {
-            ItemStack itemstack1 = new ItemStack(itemstack.getItem().getContainerItem());
+            ItemStack itemstack1 = itemstack.getItem().getContainerItem(itemstack);
 
             if (itemstack.isEmpty())
             {
@@ -212,6 +261,7 @@
 
         this.brewingItemStacks.set(3, itemstack);
         this.world.playEvent(1035, blockpos, 0);
+        net.minecraftforge.event.ForgeEventFactory.onPotionBrewed(brewingItemStacks);
     }
 
     public static void registerFixesBrewingStand(DataFixer fixer)
@@ -281,7 +331,7 @@
     @Override
     public int getInventoryStackLimit()
     {
-        return 64;
+        return this.maxStack;
     }
 
     @Override
@@ -312,14 +362,14 @@
     {
         if (index == 3)
         {
-            return PotionHelper.isReagent(stack);
+            return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(stack);
         }
         else
         {
             Item item = stack.getItem();
             return index == 4
                    ? item == Items.BLAZE_POWDER
-                   : (item == Items.POTIONITEM || item == Items.SPLASH_POTION || item == Items.LINGERING_POTION || item == Items.GLASS_BOTTLE)
+                   : net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(stack)
                    && this.getStackInSlot(index).isEmpty();
         }
     }
@@ -386,6 +436,27 @@
             case 1:
                 this.fuel = value;
         }
+    }
+
+    net.minecraftforge.items.IItemHandler handlerInput = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.UP);
+    net.minecraftforge.items.IItemHandler handlerOutput = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.DOWN);
+    net.minecraftforge.items.IItemHandler handlerSides = new net.minecraftforge.items.wrapper.SidedInvWrapper(this, net.minecraft.util.EnumFacing.NORTH);
+
+    @SuppressWarnings("unchecked")
+    @Override
+    @javax.annotation.Nullable
+    public <T> T getCapability(net.minecraftforge.common.capabilities.Capability<T> capability, @javax.annotation.Nullable net.minecraft.util.EnumFacing facing)
+    {
+        if (facing != null && capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY)
+        {
+            if (facing == EnumFacing.UP)
+                return (T) handlerInput;
+            else if (facing == EnumFacing.DOWN)
+                return (T) handlerOutput;
+            else
+                return (T) handlerSides;
+        }
+        return super.getCapability(capability, facing);
     }
 
     @Override
