--- before/net/minecraft/tileentity/TileEntitySign.java
+++ after/net/minecraft/tileentity/TileEntitySign.java
@@ -27,7 +27,7 @@
         new TextComponentString(""), new TextComponentString(""), new TextComponentString(""), new TextComponentString("")
     };
     public int lineBeingEdited = -1;
-    private boolean isEditable = true;
+    public boolean isEditable = true;
     private EntityPlayer player;
     private final CommandResultStats stats = new CommandResultStats();
 
@@ -42,6 +42,10 @@
             compound.setString("Text" + (i + 1), s);
         }
 
+        if (Boolean.getBoolean("convertLegacySigns")) {
+            compound.setBoolean("Bukkit.isConverted", true);
+        }
+
         this.stats.writeStatsToNBT(compound);
         return compound;
     }
@@ -67,7 +71,7 @@
             @Override
             public boolean canUseCommand(int permLevel, String commandName)
             {
-                return true;
+                return permLevel <= 2; //Forge: Fixes  MC-75630 - Exploit with signs and command blocks
             }
             @Override
             public BlockPos getPosition()
@@ -95,18 +99,38 @@
             }
         };
 
+        // CraftBukkit start - Add an option to convert signs correctly
+        // This is done with a flag instead of all the time because
+        // we have no way to tell whether a sign is from 1.7.10 or 1.8
+
+        boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !compound.getBoolean("Bukkit.isConverted");
+
         for (int i = 0; i < 4; i++)
         {
             String s = compound.getString("Text" + (i + 1));
-            ITextComponent itextcomponent = ITextComponent.Serializer.jsonToComponent(s);
+            // ITextComponent itextcomponent = ITextComponent.Serializer.jsonToComponent(s);
+            if (s != null && s.length() > 2048) {
+                s = "\"\"";
+            }
 
             try
             {
-                this.signText[i] = TextComponentUtils.processComponent(icommandsender, itextcomponent, null);
+                ITextComponent ichatbasecomponent = ITextComponent.Serializer.jsonToComponent(s);
+
+                if (oldSign) {
+                    signText[i] = org.bukkit.craftbukkit.v1_12_R1.util.CraftChatMessage.fromString(s)[0];
+                    continue;
+                }
+                // CraftBukkit end
+
+                try {
+                    this.signText[i] = TextComponentUtils.processComponent(icommandsender, ichatbasecomponent, null);
+                } catch (CommandException commandexception) {
+                    this.signText[i] = ichatbasecomponent;
+                }
             }
-            catch (CommandException commandexception)
-            {
-                this.signText[i] = itextcomponent;
+            catch (com.google.gson.JsonParseException jsonparseexception) {
+                this.signText[i] = new TextComponentString(s);
             }
         }
 
@@ -235,7 +259,12 @@
 
                 if (clickevent.getAction() == ClickEvent.Action.RUN_COMMAND)
                 {
-                    playerIn.getServer().getCommandManager().executeCommand(icommandsender, clickevent.getValue());
+                    // playerIn.getServer().getCommandManager().executeCommand(icommandsender, clickevent.getValue());
+                    CommandBlockBaseLogic.executeSafely(icommandsender, new org.bukkit.craftbukkit.v1_12_R1.command.ProxiedNativeCommandSender(
+                            icommandsender,
+                            new org.bukkit.craftbukkit.v1_12_R1.command.CraftBlockCommandSender(icommandsender),
+                            playerIn.getBukkitEntity()
+                    ), clickevent.getValue());
                 }
             }
         }
