--- before/net/minecraft/tileentity/TileEntityEndGateway.java
+++ after/net/minecraft/tileentity/TileEntityEndGateway.java
@@ -5,6 +5,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.Blocks;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.nbt.NBTUtil;
@@ -24,14 +25,18 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_12_R1.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class TileEntityEndGateway extends TileEntityEndPortal implements ITickable
 {
     private static final Logger LOGGER = LogManager.getLogger();
     private long age;
     private int teleportCooldown;
-    private BlockPos exitPortal;
-    private boolean exactTeleport;
+    public BlockPos exitPortal;
+    public boolean exactTeleport;
 
     @Override
     public NBTTagCompound writeToNBT(NBTTagCompound compound)
@@ -78,11 +83,11 @@
     {
         boolean flag = this.isSpawning();
         boolean flag1 = this.isCoolingDown();
-        this.age++;
+        ++this.age;
 
         if (flag1)
         {
-            this.teleportCooldown--;
+            --this.teleportCooldown;
         }
         else if (!this.world.isRemote)
         {
@@ -178,7 +183,23 @@
             if (this.exitPortal != null)
             {
                 BlockPos blockpos = this.exactTeleport ? this.exitPortal : this.findExitPosition();
-                entityIn.setPositionAndUpdate(blockpos.getX() + 0.5, blockpos.getY() + 0.5, blockpos.getZ() + 0.5);
+                if (entityIn instanceof EntityPlayerMP)
+                {
+                    CraftPlayer player = (CraftPlayer)entityIn.getBukkitEntity();
+                    Location location = new Location(world.getWorld(), blockpos.getX() + 0.5D, blockpos.getY() + 0.5D, blockpos.getZ() + 0.5D);
+                    location.setPitch(player.getLocation().getPitch());
+                    location.setYaw(player.getLocation().getYaw());
+                    PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+                    Bukkit.getPluginManager().callEvent(teleEvent);
+                    if (teleEvent.isCancelled())
+                    {
+                        return;
+                    }
+                    ((EntityPlayerMP)entityIn).connection.teleport(teleEvent.getTo());
+                    this.triggerCooldown();
+                    return;
+                }
+                entityIn.setPositionAndUpdate(blockpos.getX() + 0.5D, blockpos.getY() + 0.5D, blockpos.getZ() + 0.5D);
             }
 
             this.triggerCooldown();
@@ -233,13 +254,13 @@
     {
         BlockPos blockpos = null;
 
-        for (int i = -p_184308_2_; i <= p_184308_2_; i++)
+        for (int i = -p_184308_2_; i <= p_184308_2_; ++i)
         {
-            for (int j = -p_184308_2_; j <= p_184308_2_; j++)
+            for (int j = -p_184308_2_; j <= p_184308_2_; ++j)
             {
                 if (i != 0 || j != 0 || p_184308_3_)
                 {
-                    for (int k = 255; k > (blockpos == null ? 0 : blockpos.getY()); k--)
+                    for (int k = 255; k > (blockpos == null ? 0 : blockpos.getY()); --k)
                     {
                         BlockPos blockpos1 = new BlockPos(p_184308_1_.getX() + i, k, p_184308_1_.getZ() + j);
                         IBlockState iblockstate = p_184308_0_.getBlockState(blockpos1);
