--- before/net/minecraft/tileentity/TileEntityHopper.java
+++ after/net/minecraft/tileentity/TileEntityHopper.java
@@ -33,6 +33,32 @@
     private NonNullList<ItemStack> inventory = NonNullList.withSize(5, ItemStack.EMPTY);
     private int transferCooldown = -1;
     private long tickedGameTime;
+    // CraftBukkit start - add fields and methods
+    public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+    private int maxStack = MAX_STACK;
+    
+    public List<ItemStack> getContents() {
+        return this.inventory;
+    }
+        
+    public void onOpen(org.bukkit.craftbukkit.v1_12_R1.entity.CraftHumanEntity who) {
+        transaction.add(who);
+    }
+    
+    public void onClose(org.bukkit.craftbukkit.v1_12_R1.entity.CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+        
+    public List<org.bukkit.entity.HumanEntity> getViewers() {
+        return transaction;
+    }
+        
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+
+    public static boolean skipHopperEvents = true; // CatServer
 
     public static void registerFixesHopper(DataFixer fixer)
     {
@@ -88,7 +114,8 @@
     public ItemStack decrStackSize(int index, int count)
     {
         this.fillWithLoot(null);
-        return ItemStackHelper.getAndSplit(this.getItems(), index, count);
+        ItemStack itemstack = ItemStackHelper.getAndSplit(this.getItems(), index, count);
+        return itemstack;
     }
 
     @Override
@@ -112,7 +139,7 @@
     @Override
     public int getInventoryStackLimit()
     {
-        return 64;
+        return this.maxStack;
     }
 
     @Override
@@ -151,7 +178,7 @@
 
                 if (flag)
                 {
-                    this.setTransferCooldown(8);
+                    this.setTransferCooldown(world.spigotConfig.hopperTransfer); // Spigot
                     this.markDirty();
                     return true;
                 }
@@ -199,6 +226,7 @@
 
     private boolean transferItemsOut()
     {
+        if (net.minecraftforge.items.VanillaInventoryCodeHooks.insertHook(this)) { return true; }
         IInventory iinventory = this.getInventoryForHopperTransfer();
 
         if (iinventory == null)
@@ -305,6 +333,8 @@
 
     public static boolean pullItems(IHopper hopper)
     {
+        Boolean ret = net.minecraftforge.items.VanillaInventoryCodeHooks.extractHook(hopper);
+        if (ret != null) return ret;
         IInventory iinventory = getSourceInventory(hopper);
 
         if (iinventory != null)
@@ -358,6 +388,7 @@
         return false;
     }
 
+    // CatServer - TODO: move into Forge, support handle Forge container
     private static boolean pullItemFromSlot(IHopper hopper, IInventory inventoryIn, int index, EnumFacing direction)
     {
         ItemStack itemstack = inventoryIn.getStackInSlot(index);
@@ -365,14 +396,50 @@
         if (!itemstack.isEmpty() && canExtractItemFromSlot(inventoryIn, itemstack, index, direction))
         {
             ItemStack itemstack1 = itemstack.copy();
-            ItemStack itemstack2 = putStackInInventoryAllSlots(inventoryIn, hopper, inventoryIn.decrStackSize(index, 1), null);
+            // CatServer start - Optimized of call event on collection of items from inventories into the hopper
+            ItemStack originNMSStack = inventoryIn.decrStackSize(index, hopper.getWorld().spigotConfig.hopperAmount);
+            ItemStack resultNMSStack = originNMSStack;
+
+            org.bukkit.event.inventory.InventoryMoveItemEvent event = null;
+            if (!TileEntityHopper.skipHopperEvents) {
+                org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(originNMSStack);
+
+                org.bukkit.inventory.Inventory sourceInventory;
+                // Have to special case large chests as they work oddly
+                if (inventoryIn instanceof net.minecraft.inventory.InventoryLargeChest) {
+                    sourceInventory = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryDoubleChest((net.minecraft.inventory.InventoryLargeChest) inventoryIn);
+                } else {
+                    sourceInventory = catserver.server.inventory.CatInventoryUtils.getBukkitInventory(inventoryIn);
+                }
+
+                if (sourceInventory != null) {
+                    event = new org.bukkit.event.inventory.InventoryMoveItemEvent(sourceInventory, oitemstack, catserver.server.inventory.CatInventoryUtils.getBukkitInventory(hopper), false);
+                    hopper.getWorld().getServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        inventoryIn.setInventorySlotContents(index, itemstack1);
+                        return true;
+                    }
+
+                    if (event.isCallSetItem) resultNMSStack = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asNMSCopy(event.getRawItem());
+                }
+            }
+
+            int origCount = resultNMSStack.getCount();
+            ItemStack itemstack2 = putStackInInventoryAllSlots(inventoryIn, hopper, resultNMSStack, null);
 
             if (itemstack2.isEmpty())
             {
-                inventoryIn.markDirty();
+                // inventoryIn.markDirty();
+                if (event == null || !event.isCallSetItem || ItemStack.areItemStacksEqual(resultNMSStack, originNMSStack)) {
+                    inventoryIn.markDirty();
+                } else {
+                    inventoryIn.setInventorySlotContents(index, itemstack1);
+                }
                 return true;
             }
-
+            itemstack1.shrink(origCount - itemstack2.getCount());
+            // CatServer end
             inventoryIn.setInventorySlotContents(index, itemstack1);
         }
 
@@ -389,6 +456,11 @@
         }
         else
         {
+            org.bukkit.event.inventory.InventoryPickupItemEvent event = new org.bukkit.event.inventory.InventoryPickupItemEvent(catserver.server.inventory.CatInventoryUtils.getBukkitInventory(destination) , (org.bukkit.entity.Item) entity.getBukkitEntity());
+            entity.world.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return false;
+            }
             ItemStack itemstack = entity.getItem().copy();
             ItemStack itemstack1 = putStackInInventoryAllSlots(source, destination, itemstack, null);
 
@@ -406,6 +478,12 @@
         }
     }
 
+
+    protected net.minecraftforge.items.IItemHandler createUnSidedHandler()
+    {
+        return new net.minecraftforge.items.VanillaHopperItemHandler(this);
+    }
+
     public static ItemStack putStackInInventoryAllSlots(IInventory source, IInventory destination, ItemStack stack, @Nullable EnumFacing direction)
     {
         if (destination instanceof ISidedInventory && direction != null)
@@ -487,7 +565,7 @@
                             }
                         }
 
-                        tileentityhopper1.setTransferCooldown(8 - k);
+                        tileentityhopper1.setTransferCooldown(tileentityhopper1.world.spigotConfig.hopperTransfer - k); // Spigot
                     }
                 }
 
@@ -530,9 +608,11 @@
         int j = MathHelper.floor(y);
         int k = MathHelper.floor(z);
         BlockPos blockpos = new BlockPos(i, j, k);
-        Block block = worldIn.getBlockState(blockpos).getBlock();
+        if (!worldIn.isBlockLoaded(blockpos)) return null; // Spigot
+        net.minecraft.block.state.IBlockState state = worldIn.getBlockState(blockpos);
+        Block block = state.getBlock();
 
-        if (block.hasTileEntity())
+        if (block.hasTileEntity(state))
         {
             TileEntity tileentity = worldIn.getTileEntity(blockpos);
 
@@ -631,4 +711,6 @@
     {
         return this.inventory;
     }
+
+    public long getLastUpdateTime() { return tickedGameTime; } // Forge
 }
