--- before/net/minecraft/entity/monster/EntityIronGolem.java
+++ after/net/minecraft/entity/monster/EntityIronGolem.java
@@ -125,7 +125,7 @@
     {
         if (entityIn instanceof IMob && !(entityIn instanceof EntityCreeper) && this.getRNG().nextInt(20) == 0)
         {
-            this.setAttackTarget((EntityLivingBase)entityIn);
+            this.setGoalTarget((EntityLivingBase)entityIn, org.bukkit.event.entity.EntityTargetLivingEntityEvent.TargetReason.COLLISION, true);
         }
 
         super.collideWithEntity(entityIn);
@@ -321,7 +321,7 @@
     {
         if (!this.isPlayerCreated() && this.attackingPlayer != null && this.village != null)
         {
-            this.village.modifyPlayerReputation(this.attackingPlayer.getName(), -5);
+            this.village.modifyPlayerReputation(this.attackingPlayer.getUniqueID(), -5);
         }
 
         super.onDeath(cause);
