--- before/net/minecraft/entity/monster/EntityZombie.java
+++ after/net/minecraft/entity/monster/EntityZombie.java
@@ -89,7 +89,7 @@
         this.tasks.addTask(6, new EntityAIMoveThroughVillage(this, 1.0, false));
         this.targetTasks.addTask(1, new EntityAIHurtByTarget(this, true, EntityPigZombie.class));
         this.targetTasks.addTask(2, new EntityAINearestAttackableTarget<>(this, EntityPlayer.class, true));
-        this.targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(this, EntityVillager.class, false));
+        if (world.spigotConfig.zombieAggressiveTowardsVillager) this.targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(this, EntityVillager.class, false)); // Spigot
         this.targetTasks.addTask(3, new EntityAINearestAttackableTarget<>(this, EntityIronGolem.class, true));
     }
 
@@ -101,7 +101,7 @@
         this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.23F);
         this.getEntityAttribute(SharedMonsterAttributes.ATTACK_DAMAGE).setBaseValue(3.0);
         this.getEntityAttribute(SharedMonsterAttributes.ARMOR).setBaseValue(2.0);
-        this.getAttributeMap().registerAttribute(SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.rand.nextDouble() * 0.1F);
+        this.getAttributeMap().registerAttribute(SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.rand.nextDouble() * net.minecraftforge.common.ForgeModContainer.zombieSummonBaseChance);
     }
 
     @Override
@@ -225,7 +225,13 @@
 
                 if (flag)
                 {
-                    this.setFire(8);
+                    // this.setFire(8);
+                    org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustEvent(this.getBukkitEntity(), 8);
+                    this.world.getServer().getPluginManager().callEvent(event);
+
+                    if (!event.isCancelled()) {
+                        this.setFire(event.getDuration());
+                    }
                 }
             }
         }
@@ -250,15 +256,31 @@
                 entitylivingbase = (EntityLivingBase)source.getTrueSource();
             }
 
-            if (entitylivingbase != null
+            int i = MathHelper.floor(this.posX);
+            int j = MathHelper.floor(this.posY);
+            int k = MathHelper.floor(this.posZ);
+            net.minecraftforge.event.entity.living.ZombieEvent.SummonAidEvent summonAid
+                = net.minecraftforge.event.ForgeEventFactory.fireZombieSummonAid(
+                    this, world, i, j, k, entitylivingbase, this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).getAttributeValue()
+            );
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return true;
+
+            if (summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW
+                    || entitylivingbase != null
                     && this.world.getDifficulty() == EnumDifficulty.HARD
                     && this.rand.nextFloat() < this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).getAttributeValue()
                     && this.world.getGameRules().getBoolean("doMobSpawning"))
             {
-                int i = MathHelper.floor(this.posX);
-                int j = MathHelper.floor(this.posY);
-                int k = MathHelper.floor(this.posZ);
-                EntityZombie entityzombie = new EntityZombie(this.world);
+                EntityZombie entityzombie;
+                if (summonAid.getCustomSummonedAid() != null
+                        && summonAid.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW)
+                {
+                    entityzombie = summonAid.getCustomSummonedAid();
+                }
+                else
+                {
+                    entityzombie = new EntityZombie(this.world);
+                }
 
                 for (int l = 0; l < 50; l++)
                 {
@@ -266,7 +288,7 @@
                     int j1 = j + MathHelper.getInt(this.rand, 7, 40) * MathHelper.getInt(this.rand, -1, 1);
                     int k1 = k + MathHelper.getInt(this.rand, 7, 40) * MathHelper.getInt(this.rand, -1, 1);
 
-                    if (this.world.getBlockState(new BlockPos(i1, j1 - 1, k1)).isTopSolid()
+                    if (this.world.getBlockState(new BlockPos(i1, j1 - 1, k1)).isSideSolid(this.world, new BlockPos(i1, j1 - 1, k1), net.minecraft.util.EnumFacing.UP)
                             && this.world.getLightFromNeighbors(new BlockPos(i1, j1, k1)) < 10)
                     {
                         entityzombie.setPosition(i1, j1, k1);
@@ -276,8 +298,8 @@
                                 && this.world.getCollisionBoxes(entityzombie, entityzombie.getEntityBoundingBox()).isEmpty()
                                 && !this.world.containsAnyLiquid(entityzombie.getEntityBoundingBox()))
                         {
-                            this.world.spawnEntity(entityzombie);
-                            entityzombie.setAttackTarget(entitylivingbase);
+                            this.world.addEntity(entityzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.REINFORCEMENTS);
+                            if (entitylivingbase != null) entityzombie.setGoalTarget(entitylivingbase, org.bukkit.event.entity.EntityTargetEvent.TargetReason.REINFORCEMENT_TARGET, true);
                             entityzombie.onInitialSpawn(this.world.getDifficultyForLocation(new BlockPos(entityzombie)), null);
                             this.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement caller charge", -0.05F, 0));
                             entityzombie.getEntityAttribute(SPAWN_REINFORCEMENTS_CHANCE).applyModifier(new AttributeModifier("Zombie reinforcement callee charge", -0.05F, 0));
@@ -306,7 +328,13 @@
 
             if (this.getHeldItemMainhand().isEmpty() && this.isBurning() && this.rand.nextFloat() < f * 0.3F)
             {
-                entityIn.setFire(2 * (int)f);
+                // entityIn.setFire(2 * (int)f);
+                org.bukkit.event.entity.EntityCombustByEntityEvent event = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entityIn.getBukkitEntity(), 2 * (int) f); // PAIL: fixme
+                this.world.getServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    entityIn.setFire(event.getDuration());
+                }
             }
         }
 
@@ -434,7 +462,7 @@
                 entityzombievillager.setAlwaysRenderNameTag(entityvillager.getAlwaysRenderNameTag());
             }
 
-            this.world.spawnEntity(entityzombievillager);
+            this.world.addEntity(entityzombievillager, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.INFECTION);
             this.world.playEvent(null, 1026, new BlockPos(this), 0);
         }
     }
@@ -468,7 +496,7 @@
 
         if (livingdata == null)
         {
-            livingdata = new EntityZombie.GroupData(this.world.rand.nextFloat() < 0.05F);
+            livingdata = new EntityZombie.GroupData(this.world.rand.nextFloat() < net.minecraftforge.common.ForgeModContainer.zombieBabyChance);
         }
 
         if (livingdata instanceof EntityZombie.GroupData)
@@ -497,7 +525,7 @@
                     entitychicken1.setLocationAndAngles(this.posX, this.posY, this.posZ, this.rotationYaw, 0.0F);
                     entitychicken1.onInitialSpawn(difficulty, null);
                     entitychicken1.setChickenJockey(true);
-                    this.world.spawnEntity(entitychicken1);
+                    this.world.addEntity(entitychicken1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.MOUNT);
                     this.startRiding(entitychicken1);
                 }
             }
@@ -570,7 +598,7 @@
     @Override
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+        // super.onDeath(cause); // CraftBukkit - moved down
 
         if (cause.getTrueSource() instanceof EntityCreeper)
         {
@@ -587,6 +615,7 @@
                 }
             }
         }
+        super.onDeath(cause);
     }
 
     protected ItemStack getSkullDrop()
