--- before/net/minecraft/entity/monster/EntityShulker.java
+++ after/net/minecraft/entity/monster/EntityShulker.java
@@ -47,6 +47,7 @@
 import net.minecraft.world.storage.loot.LootTableList;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.Location;
 
 public class EntityShulker extends EntityGolem implements IMob
 {
@@ -57,7 +58,7 @@
                     EntityShulker.class, DataSerializers.OPTIONAL_BLOCK_POS
                 );
     protected static final DataParameter<Byte> PEEK_TICK = EntityDataManager.createKey(EntityShulker.class, DataSerializers.BYTE);
-    protected static final DataParameter<Byte> COLOR = EntityDataManager.createKey(EntityShulker.class, DataSerializers.BYTE);
+    public static final DataParameter<Byte> COLOR = EntityDataManager.createKey(EntityShulker.class, DataSerializers.BYTE);
     public static final EnumDyeColor DEFAULT_COLOR = EnumDyeColor.PURPLE;
     private float prevPeekAmount;
     private float peekAmount;
@@ -327,6 +328,7 @@
             this.posX = blockpos.getX() + 0.5;
             this.posY = blockpos.getY();
             this.posZ = blockpos.getZ() + 0.5;
+            if (this.isAddedToWorld() && !this.world.isRemote) this.world.updateEntityWithOptionalForce(this, false); // Forge - Process chunk registration after moving.
             this.prevPosX = this.posX;
             this.prevPosY = this.posY;
             this.prevPosZ = this.posZ;
@@ -495,10 +497,23 @@
                     {
                         if (this.world.isBlockNormalCube(blockpos1.offset(enumfacing), false))
                         {
-                            this.dataManager.set(ATTACHED_FACE, enumfacing);
-                            flag = true;
-                            break;
+                            org.bukkit.event.entity.EntityTeleportEvent teleport = new org.bukkit.event.entity.EntityTeleportEvent(this.getBukkitEntity(), this.getBukkitEntity().getLocation(), new Location(this.world.getWorld(), blockpos1.getX(), blockpos1.getY(), blockpos1.getZ()));
+                            this.world.getServer().getPluginManager().callEvent(teleport);
+                            if (!teleport.isCancelled()) {
+                                org.bukkit.Location to = teleport.getTo();
+                                blockpos1 = new BlockPos(to.getX(), to.getY(), to.getZ());
+                                this.dataManager.set(ATTACHED_FACE, enumfacing);
+                                flag = true;
+                                break;
+                            }
                         }
+                    }
+
+                    if (flag)
+                    {
+                        net.minecraftforge.event.entity.living.EnderTeleportEvent event = new net.minecraftforge.event.entity.living.EnderTeleportEvent(this, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ(), 0);
+                        if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event)) flag = false;
+                        blockpos1 = new BlockPos(event.getTargetX(), event.getTargetY(), event.getTargetZ());
                     }
 
                     if (flag)
@@ -553,6 +568,7 @@
                 this.posX = blockpos.getX() + 0.5;
                 this.posY = blockpos.getY();
                 this.posZ = blockpos.getZ() + 0.5;
+                if (this.isAddedToWorld() && !this.world.isRemote) this.world.updateEntityWithOptionalForce(this, false); // Forge - Process chunk registration after moving.
                 this.prevPosX = this.posX;
                 this.prevPosY = this.posY;
                 this.prevPosZ = this.posZ;
