--- before/net/minecraft/entity/monster/EntitySpellcasterIllager.java
+++ after/net/minecraft/entity/monster/EntitySpellcasterIllager.java
@@ -63,7 +63,7 @@
         this.dataManager.set(SPELL, (byte)spellType.id);
     }
 
-    protected EntitySpellcasterIllager.SpellType getSpellType()
+    public EntitySpellcasterIllager.SpellType getSpellType()
     {
         return !this.world.isRemote
                ? this.activeSpell
