--- before/net/minecraft/entity/monster/EntityMagmaCube.java
+++ after/net/minecraft/entity/monster/EntityMagmaCube.java
@@ -50,7 +50,7 @@
     }
 
     @Override
-    protected void setSlimeSize(int size, boolean resetHealth)
+    public void setSlimeSize(int size, boolean resetHealth)
     {
         super.setSlimeSize(size, resetHealth);
         this.getEntityAttribute(SharedMonsterAttributes.ARMOR).setBaseValue(size * 3);
@@ -111,6 +111,7 @@
     {
         this.motionY = 0.42F + this.getSlimeSize() * 0.1F;
         this.isAirBorne = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     @Override
