--- before/net/minecraft/entity/monster/EntityCreeper.java
+++ after/net/minecraft/entity/monster/EntityCreeper.java
@@ -44,8 +44,8 @@
     private static final DataParameter<Boolean> IGNITED = EntityDataManager.createKey(EntityCreeper.class, DataSerializers.BOOLEAN);
     private int lastActiveTime;
     private int timeSinceIgnited;
-    private int fuseTime = 30;
-    private int explosionRadius = 3;
+    public int fuseTime = 30;
+    public int explosionRadius = 3;
     private int droppedSkulls;
 
     public EntityCreeper(World worldIn)
@@ -195,7 +195,7 @@
     @Override
     public void onDeath(DamageSource cause)
     {
-        super.onDeath(cause);
+//        super.onDeath(cause); // CraftBukkit - Moved to end
 
         if (this.world.getGameRules().getBoolean("doMobLoot"))
         {
@@ -215,6 +215,7 @@
                 this.entityDropItem(new ItemStack(Items.SKULL, 1, 4), 0.0F);
             }
         }
+        super.onDeath(cause); // CraftBukkit - Moved from above
     }
 
     @Override
@@ -254,8 +255,18 @@
     @Override
     public void onStruckByLightning(EntityLightningBolt lightningBolt)
     {
+        if (lightningBolt == null) lightningBolt = new EntityLightningBolt(this.world, this.posX, this.posY, this.posZ, true); // CatServer - if null, create EntityLightningBolt for Bukkit
         super.onStruckByLightning(lightningBolt);
-        this.dataManager.set(POWERED, true);
+
+        if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callCreeperPowerEvent(this, lightningBolt, org.bukkit.event.entity.CreeperPowerEvent.PowerCause.LIGHTNING).isCancelled()) {
+            return;
+        }
+
+        this.setPowered(true);
+    }
+
+    public void setPowered(boolean powered) {
+        this.dataManager.set(POWERED, powered);
     }
 
     @Override
@@ -293,12 +304,22 @@
     {
         if (!this.world.isRemote)
         {
-            boolean flag = this.world.getGameRules().getBoolean("mobGriefing");
+            boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.world, this);
             float f = this.getPowered() ? 2.0F : 1.0F;
-            this.dead = true;
-            this.world.createExplosion(this, this.posX, this.posY, this.posZ, this.explosionRadius * f, flag);
-            this.setDead();
-            this.spawnLingeringCloud();
+//            this.dead = true;
+//            this.world.createExplosion(this, this.posX, this.posY, this.posZ, this.explosionRadius * f, flag);
+//            this.setDead();
+//            this.spawnLingeringCloud();
+            org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+            this.world.getServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                this.dead = true;
+                this.world.newExplosion(this, this.posX, this.posY, this.posZ, event.getRadius(), event.getFire(), flag);
+                this.setDead();
+                this.spawnLingeringCloud();
+            } else {
+                timeSinceIgnited = 0;
+            }
         }
     }
 
@@ -309,6 +330,7 @@
         if (!collection.isEmpty())
         {
             EntityAreaEffectCloud entityareaeffectcloud = new EntityAreaEffectCloud(this.world, this.posX, this.posY, this.posZ);
+            entityareaeffectcloud.setOwner(this);
             entityareaeffectcloud.setRadius(2.5F);
             entityareaeffectcloud.setRadiusOnUse(-0.5F);
             entityareaeffectcloud.setWaitTime(10);
