--- before/net/minecraft/entity/monster/EntitySlime.java
+++ after/net/minecraft/entity/monster/EntitySlime.java
@@ -68,7 +68,7 @@
         this.dataManager.register(SLIME_SIZE, 1);
     }
 
-    protected void setSlimeSize(int size, boolean resetHealth)
+    public void setSlimeSize(int size, boolean resetHealth)
     {
         this.dataManager.set(SLIME_SIZE, size);
         this.setSize(0.51000005F * size, 0.51000005F * size);
@@ -142,7 +142,7 @@
         if (this.onGround && !this.wasOnGround)
         {
             int i = this.getSlimeSize();
-
+            if (spawnCustomParticles()) { i = 0; } // don't spawn particles if it's handled by the implementation itself
             for (int j = 0; j < i * 8; j++)
             {
                 float f = this.rand.nextFloat() * (float)(Math.PI * 2);
@@ -212,6 +212,16 @@
         if (!this.world.isRemote && i > 1 && this.getHealth() <= 0.0F)
         {
             int j = 2 + this.rand.nextInt(3);
+            
+            org.bukkit.event.entity.SlimeSplitEvent event = new org.bukkit.event.entity.SlimeSplitEvent((org.bukkit.entity.Slime) this.getBukkitEntity(), j);
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled() && event.getCount() > 0) {
+                j = event.getCount();
+            } else {
+                super.setDead();
+                return;
+            }
 
             for (int k = 0; k < j; k++)
             {
@@ -231,7 +241,7 @@
 
                 entityslime.setSlimeSize(i / 2, true);
                 entityslime.setLocationAndAngles(this.posX + f, this.posY + 0.5, this.posZ + f1, this.rand.nextFloat() * 360.0F, 0.0F);
-                this.world.spawnEntity(entityslime);
+                this.world.addEntity(entityslime, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT);
             }
         }
 
@@ -323,7 +333,7 @@
         BlockPos blockpos = new BlockPos(MathHelper.floor(this.posX), 0, MathHelper.floor(this.posZ));
         Chunk chunk = this.world.getChunk(blockpos);
 
-        if (this.world.getWorldInfo().getTerrainType() == WorldType.FLAT && this.rand.nextInt(4) != 1)
+        if (this.world.getWorldInfo().getTerrainType().handleSlimeSpawnReduction(rand, world))
         {
             return false;
         }
@@ -343,7 +353,7 @@
                     return super.getCanSpawnHere();
                 }
 
-                if (this.rand.nextInt(10) == 0 && chunk.getRandomWithSeed(987234911L).nextInt(10) == 0 && this.posY < 40.0)
+                if (this.rand.nextInt(10) == 0 && chunk.getRandomWithSeed(world.spigotConfig.slimeSeed).nextInt(10) == 0 && this.posY < 40.0)
                 {
                     return super.getCanSpawnHere();
                 }
@@ -375,6 +385,7 @@
     {
         this.motionY = 0.42F;
         this.isAirBorne = true;
+        net.minecraftforge.common.ForgeHooks.onLivingJump(this);
     }
 
     @Nullable
@@ -397,6 +408,14 @@
     {
         return this.isSmallSlime() ? SoundEvents.ENTITY_SMALL_SLIME_JUMP : SoundEvents.ENTITY_SLIME_JUMP;
     }
+
+    /* ======================================== FORGE START =====================================*/
+    /**
+     * Called when the slime spawns particles on landing, see onUpdate.
+     * Return true to prevent the spawning of the default particles.
+     */
+    protected boolean spawnCustomParticles() { return false; }
+    /* ======================================== FORGE END   =====================================*/
 
     static class AISlimeAttack extends EntityAIBase
         {
