--- before/net/minecraft/entity/monster/EntityGhast.java
+++ after/net/minecraft/entity/monster/EntityGhast.java
@@ -229,7 +229,7 @@
                 if (entitylivingbase.getDistanceSq(this.parentEntity) < 4096.0 && this.parentEntity.canEntityBeSeen(entitylivingbase))
                 {
                     World world = this.parentEntity.world;
-                    this.attackTimer++;
+                    ++this.attackTimer;
 
                     if (this.attackTimer == 10)
                     {
@@ -247,7 +247,8 @@
                         double d4 = entitylivingbase.posZ - (this.parentEntity.posZ + vec3d.z * 4.0);
                         world.playEvent(null, 1016, new BlockPos(this.parentEntity), 0);
                         EntityLargeFireball entitylargefireball = new EntityLargeFireball(world, this.parentEntity, d2, d3, d4);
-                        entitylargefireball.explosionPower = this.parentEntity.getFireballStrength();
+                        // CraftBukkit - set bukkitYield when setting explosionpower
+                        entitylargefireball.bukkitYield = entitylargefireball.explosionPower = this.parentEntity.getFireballStrength();
                         entitylargefireball.posX = this.parentEntity.posX + vec3d.x * 4.0;
                         entitylargefireball.posY = this.parentEntity.posY + this.parentEntity.height / 2.0F + 0.5;
                         entitylargefireball.posZ = this.parentEntity.posZ + vec3d.z * 4.0;
@@ -257,7 +258,7 @@
                 }
                 else if (this.attackTimer > 0)
                 {
-                    this.attackTimer--;
+                    --this.attackTimer;
                 }
 
                 this.parentEntity.setAttacking(this.attackTimer > 10);
@@ -374,7 +375,7 @@
 
                     if (this.courseChangeCooldown-- <= 0)
                     {
-                        this.courseChangeCooldown = this.courseChangeCooldown + this.parentEntity.getRNG().nextInt(5) + 2;
+                        this.courseChangeCooldown += this.parentEntity.getRNG().nextInt(5) + 2;
                         d3 = MathHelper.sqrt(d3);
 
                         if (this.isNotColliding(this.posX, this.posY, this.posZ, d3))
