--- before/net/minecraft/entity/monster/EntityMob.java
+++ after/net/minecraft/entity/monster/EntityMob.java
@@ -73,7 +73,8 @@
     @Override
     public boolean attackEntityFrom(DamageSource source, float amount)
     {
-        return this.isEntityInvulnerable(source) ? false : super.attackEntityFrom(source, amount);
+//        return this.isEntityInvulnerable(source) ? false : super.attackEntityFrom(source, amount);
+        return super.attackEntityFrom(source, amount); // CREF - we're calling this method in super.attackEntityFrom
     }
 
     @Override
@@ -127,7 +128,15 @@
 
             if (j > 0)
             {
-                entityIn.setFire(j * 4);
+                // entityIn.setFire(j * 4);
+                // CraftBukkit start - Call a combust event when somebody hits with a fire enchanted item
+                org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entityIn.getBukkitEntity(), j * 4);
+                org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+
+                if (!combustEvent.isCancelled()) {
+                    entityIn.setFire(combustEvent.getDuration());
+                }
+                // CraftBukkit end
             }
 
             if (entityIn instanceof EntityPlayer)
@@ -138,14 +147,14 @@
 
                 if (!itemstack.isEmpty()
                         && !itemstack1.isEmpty()
-                        && itemstack.getItem() instanceof ItemAxe
-                        && itemstack1.getItem() == Items.SHIELD)
+                        && itemstack.getItem().canDisableShield(itemstack, itemstack1, entityplayer, this)
+                        && itemstack1.getItem().isShield(itemstack1, entityplayer))
                 {
                     float f1 = 0.25F + EnchantmentHelper.getEfficiencyModifier(this) * 0.05F;
 
                     if (this.rand.nextFloat() < f1)
                     {
-                        entityplayer.getCooldownTracker().setCooldown(Items.SHIELD, 100);
+                        entityplayer.getCooldownTracker().setCooldown(itemstack1.getItem(), 100);
                         this.world.setEntityState(entityplayer, (byte)30);
                     }
                 }
