--- before/net/minecraft/client/resources/Locale.java
+++ after/net/minecraft/client/resources/Locale.java
@@ -20,7 +20,7 @@
 {
     private static final Splitter SPLITTER = Splitter.on('=').limit(2);
     private static final Pattern PATTERN = Pattern.compile("%(\\d+\\$)?[\\d\\.]*[df]");
-    Map<String, String> properties = Maps.newHashMap();
+    Map<String, String> properties = Maps.<String, String>newHashMap();
     private boolean unicode;
 
     public synchronized void loadLocaleDataFiles(IResourceManager resourceManager, List<String> languageList)
@@ -37,8 +37,9 @@
                 {
                     this.loadLocaleData(resourceManager.getAllResources(new ResourceLocation(s2, s1)));
                 }
-                catch (IOException ioexception)
+                catch (IOException var9)
                 {
+                    ;
                 }
             }
         }
@@ -53,26 +54,6 @@
 
     private void checkUnicode()
     {
-        this.unicode = false;
-        int i = 0;
-        int j = 0;
-
-        for (String s : this.properties.values())
-        {
-            int k = s.length();
-            j += k;
-
-            for (int l = 0; l < k; l++)
-            {
-                if (s.charAt(l) >= 256)
-                {
-                    i++;
-                }
-            }
-        }
-
-        float f = (float)i / j;
-        this.unicode = f > 0.1;
     }
 
     private void loadLocaleData(List<IResource> resourcesList) throws IOException
@@ -94,11 +75,13 @@
 
     private void loadLocaleData(InputStream inputStreamIn) throws IOException
     {
+        inputStreamIn = net.minecraftforge.fml.common.FMLCommonHandler.instance().loadLanguage(properties, inputStreamIn);
+        if (inputStreamIn == null) return;
         for (String s : IOUtils.readLines(inputStreamIn, StandardCharsets.UTF_8))
         {
             if (!s.isEmpty() && s.charAt(0) != '#')
             {
-                String[] astring = Iterables.toArray(SPLITTER.split(s), String.class);
+                String[] astring = (String[])Iterables.toArray(SPLITTER.split(s), String.class);
 
                 if (astring != null && astring.length == 2)
                 {
@@ -124,7 +107,7 @@
         {
             return String.format(s, parameters);
         }
-        catch (IllegalFormatException illegalformatexception)
+        catch (IllegalFormatException var5)
         {
             return "Format error: " + s;
         }
