--- before/net/minecraft/advancements/AdvancementRewards.java
+++ after/net/minecraft/advancements/AdvancementRewards.java
@@ -11,9 +11,11 @@
 import java.util.Arrays;
 import net.minecraft.command.CommandResultStats;
 import net.minecraft.command.FunctionObject;
+import net.minecraft.command.ICommandListener;
 import net.minecraft.command.ICommandSender;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.item.EntityItem;
+import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.SoundEvents;
 import net.minecraft.item.ItemStack;
@@ -50,15 +52,15 @@
     public void apply(final EntityPlayerMP player)
     {
         player.addExperience(this.experience);
-        LootContext lootcontext = new LootContext.Builder(player.getServerWorld()).withLootedEntity(player).build();
+        LootContext lootcontext = new LootContext.Builder(player.getServerWorld()).withLootedEntity(player).withPlayer(player).withLuck(player.getLuck()).build(); // Forge: add player & luck to LootContext
         boolean flag = false;
 
         for (ResourceLocation resourcelocation : this.loot)
         {
             for (ItemStack itemstack : player.world
-                    .getLootTableManager()
-                    .getLootTableFromLocation(resourcelocation)
-                    .generateLootForPools(player.getRNG(), lootcontext))
+                .getLootTableManager()
+                .getLootTableFromLocation(resourcelocation)
+                .generateLootForPools(player.getRNG(), lootcontext))
             {
                 if (player.addItemStackToInventory(itemstack))
                 {
@@ -103,64 +105,74 @@
 
         if (functionobject != null)
         {
-            ICommandSender icommandsender = new ICommandSender()
-            {
-                @Override
-                public String getName()
-                {
-                    return player.getName();
-                }
-                @Override
-                public ITextComponent getDisplayName()
-                {
-                    return player.getDisplayName();
-                }
-                @Override
-                public void sendMessage(ITextComponent component)
-                {
-                }
-                @Override
-                public boolean canUseCommand(int permLevel, String commandName)
-                {
-                    return permLevel <= 2;
-                }
-                @Override
-                public BlockPos getPosition()
-                {
-                    return player.getPosition();
-                }
-                @Override
-                public Vec3d getPositionVector()
-                {
-                    return player.getPositionVector();
-                }
-                @Override
-                public World getEntityWorld()
-                {
-                    return player.world;
-                }
-                @Override
-                public Entity getCommandSenderEntity()
-                {
-                    return player;
-                }
-                @Override
-                public boolean sendCommandFeedback()
-                {
-                    return minecraftserver.worlds[0].getGameRules().getBoolean("commandBlockOutput");
-                }
-                @Override
-                public void setCommandStat(CommandResultStats.Type type, int amount)
-                {
-                    player.setCommandStat(type, amount);
-                }
-                @Override
-                public MinecraftServer getServer()
-                {
-                    return player.getServer();
-                }
-            };
+            ICommandSender icommandsender = new AdvancementCommandListener(player, minecraftserver);
+
             minecraftserver.getFunctionManager().execute(functionobject, icommandsender);
+        }
+
+    }
+
+    public static class AdvancementCommandListener implements ICommandSender {
+        private final EntityPlayerMP player;
+        private final MinecraftServer minecraftserver;
+
+        public AdvancementCommandListener(EntityPlayerMP entityplayer, MinecraftServer minecraftserver) {
+            this.player = entityplayer;
+            this.minecraftserver = minecraftserver;
+        }
+
+        @Override
+        public String getName() {
+            return player.getName();
+        }
+
+        @Override
+        public ITextComponent getDisplayName() {
+            return player.getDisplayName();
+        }
+
+        @Override
+        public void sendMessage(ITextComponent component) {
+        }
+
+        @Override
+        public boolean canUseCommand(int permLevel, String commandName) {
+            return permLevel <= 2;
+        }
+
+        @Override
+        public BlockPos getPosition() {
+            return player.getPosition();
+        }
+
+        @Override
+        public Vec3d getPositionVector() {
+            return player.getPositionVector();
+        }
+
+        @Override
+        public World getEntityWorld() {
+            return player.world;
+        }
+
+        @Override
+        public Entity getCommandSenderEntity() {
+            return player;
+        }
+
+        @Override
+        public boolean sendCommandFeedback() {
+            return minecraftserver.worlds[0].getGameRules().getBoolean("commandBlockOutput");
+        }
+
+        @Override
+        public void setCommandStat(CommandResultStats.Type type, int amount) {
+            player.setCommandStat(type, amount);
+        }
+
+        @Override
+        public MinecraftServer getServer() {
+            return player.getServer();
         }
     }
 
