--- before/net/minecraft/advancements/FunctionManager.java
+++ after/net/minecraft/advancements/FunctionManager.java
@@ -29,29 +29,27 @@
     private FunctionObject gameLoopFunction;
     private final ArrayDeque<FunctionManager.QueuedCommand> commandQueue = new ArrayDeque<>();
     private boolean isExecuting = false;
-    private final ICommandSender gameLoopFunctionSender = new ICommandSender()
-    {
-        @Override
+    private final ICommandSender gameLoopFunctionSender = new CustomFunctionListener();
+
+    public class CustomFunctionListener implements ICommandSender {
+        public org.bukkit.command.CommandSender sender = new org.bukkit.craftbukkit.v1_12_R1.command.CraftFunctionCommandSender(this);
         public String getName()
         {
             return FunctionManager.this.currentGameLoopFunctionId;
         }
-        @Override
         public boolean canUseCommand(int permLevel, String commandName)
         {
             return permLevel <= 2;
         }
-        @Override
         public World getEntityWorld()
         {
             return FunctionManager.this.server.worlds[0];
         }
-        @Override
         public MinecraftServer getServer()
         {
             return FunctionManager.this.server;
         }
-    };
+    }
 
     public FunctionManager(@Nullable File functionDirIn, MinecraftServer serverIn)
     {
@@ -121,7 +119,7 @@
                 int j = 0;
                 FunctionObject.Entry[] afunctionobject$entry = function.getEntries();
 
-                for (int k = afunctionobject$entry.length - 1; k >= 0; k--)
+                for (int k = afunctionobject$entry.length - 1; k >= 0; --k)
                 {
                     this.commandQueue.push(new FunctionManager.QueuedCommand(this, sender, afunctionobject$entry[k]));
                 }
