--- before/net/minecraft/server/management/PlayerList.java
+++ after/net/minecraft/server/management/PlayerList.java
@@ -76,7 +76,7 @@
     private static final Logger LOGGER = LogManager.getLogger();
     private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z");
     private final MinecraftServer server;
-    private final List<EntityPlayerMP> playerEntityList = Lists.newArrayList();
+    public final List<EntityPlayerMP> playerEntityList = new java.util.concurrent.CopyOnWriteArrayList<>(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
     private final Map<UUID, EntityPlayerMP> uuidToPlayerMap = Maps.newHashMap();
     private final UserListBans bannedPlayers = new UserListBans(FILE_PLAYERBANS);
     private final UserListIPBans bannedIPs = new UserListIPBans(FILE_IPBANS);
@@ -84,7 +84,7 @@
     private final UserListWhitelist whiteListedPlayers = new UserListWhitelist(FILE_WHITELIST);
     private final Map<UUID, StatisticsManagerServer> playerStatFiles = Maps.newHashMap();
     private final Map<UUID, PlayerAdvancements> advancements = Maps.newHashMap();
-    private IPlayerFileData playerDataManager;
+    public IPlayerFileData playerDataManager;
     private boolean whiteListEnforced;
     protected int maxPlayers;
     private int viewDistance;
@@ -92,6 +92,8 @@
     private boolean commandsAllowedForAll;
     private int playerPingIndex;
 
+    private org.bukkit.craftbukkit.v1_12_R1.CraftServer cserver;
+
     public PlayerList(MinecraftServer server)
     {
         this.server = server;
@@ -100,8 +102,22 @@
         this.maxPlayers = 8;
     }
 
-    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn)
+    // CatServer start
+    public PlayerList initCraftServer() {
+        this.cserver = server.server = new org.bukkit.craftbukkit.v1_12_R1.CraftServer(server, this);
+        server.console = org.bukkit.craftbukkit.v1_12_R1.command.ColouredConsoleSender.getInstance();
+        return this;
+    }
+    // CatServer end
+
+    public void initializeConnectionToPlayer(NetworkManager netManager, EntityPlayerMP playerIn, NetHandlerPlayServer nethandlerplayserver)
     {
+        // CatServer start
+        if (this.playerEntityList.stream().anyMatch(p -> p.getName().toLowerCase(java.util.Locale.ROOT).equals(playerIn.getName().toLowerCase(java.util.Locale.ROOT)) || p.getUniqueID().equals(playerIn.getUniqueID()))) {
+            nethandlerplayserver.disconnect(new TextComponentTranslation("multiplayer.disconnect.duplicate_login", new Object[0]));
+            return;
+        }
+        // CatServer end
         GameProfile gameprofile = playerIn.getGameProfile();
         PlayerProfileCache playerprofilecache = this.server.getPlayerProfileCache();
         GameProfile gameprofile1 = playerprofilecache.getProfileByUUID(gameprofile.getId());
@@ -109,6 +125,24 @@
         playerprofilecache.addEntry(gameprofile);
         NBTTagCompound nbttagcompound = this.readPlayerDataFromFile(playerIn);
         playerIn.setWorld(this.server.getWorld(playerIn.dimension));
+
+        World playerWorld = this.server.getWorld(playerIn.dimension);
+        if (playerWorld == null)
+        {
+            playerIn.dimension = 0;
+            playerWorld = this.server.getWorld(0);
+            BlockPos spawnPoint = playerWorld.provider.getRandomizedSpawnPoint();
+            playerIn.setPosition(spawnPoint.getX(), spawnPoint.getY(), spawnPoint.getZ());
+        }
+
+        // CraftBukkit start - Better rename detection
+        if (nbttagcompound != null && nbttagcompound.hasKey("bukkit")) {
+            NBTTagCompound bukkit = nbttagcompound.getCompoundTag("bukkit");
+            s = bukkit.hasKey("lastKnownName", 8) ? bukkit.getString("lastKnownName") : s;
+        }
+        // CraftBukkit end
+
+        playerIn.setWorld(playerWorld);
         playerIn.interactionManager.setWorld((WorldServer)playerIn.world);
         String s1 = "local";
 
@@ -117,31 +151,44 @@
             s1 = netManager.getRemoteAddress().toString();
         }
 
-        LOGGER.info(
-            "{}[{}] logged in with entity id {} at ({}, {}, {})",
-            playerIn.getName(),
-            s1,
-            playerIn.getEntityId(),
-            playerIn.posX,
-            playerIn.posY,
-            playerIn.posZ
-        );
+        catserver.server.PlayerDataFixer.checkLocation(playerIn); // CatServer - fix invalid location
+        catserver.server.PlayerDataFixer.checkHealth(playerIn); // CatServer - fix invalid health
+
+        // Spigot start - spawn location event
+        org.bukkit.entity.Player bukkitPlayer = playerIn.getBukkitEntity();
+        org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new org.spigotmc.event.player.PlayerSpawnLocationEvent(bukkitPlayer, bukkitPlayer.getLocation());
+        org.bukkit.Bukkit.getPluginManager().callEvent(ev);
+
+        org.bukkit.Location loc = ev.getSpawnLocation();
+        WorldServer world = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) loc.getWorld()).getHandle();
+
+        playerIn.setWorld(world);
+        playerIn.setPosition(loc.getX(), loc.getY(), loc.getZ());
+        playerIn.setRotation(loc.getYaw(), loc.getPitch()); 
+        // Spigot end
+
+
+        // CraftBukkit - Moved message to after join
+        // LOGGER.info("{}[{}] logged in with entity id {} at ({}, {}, {})", playerIn.getName(), s1, Integer.valueOf(playerIn.getEntityId()), Double.valueOf(playerIn.posX), Double.valueOf(playerIn.posY), Double.valueOf(playerIn.posZ));
         WorldServer worldserver = this.server.getWorld(playerIn.dimension);
         WorldInfo worldinfo = worldserver.getWorldInfo();
         this.setPlayerGameTypeBasedOnOther(playerIn, null, worldserver);
-        NetHandlerPlayServer nethandlerplayserver = new NetHandlerPlayServer(this.server, netManager, playerIn);
+        playerIn.connection = nethandlerplayserver;
+        catserver.server.CatServerNetwork.registerBukkitWorldToClient(playerIn, playerIn.dimension); // send DimensionRegisterMessage to client before attempting to login to a Bukkit dimension
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().fireServerConnectionEvent(netManager);
         nethandlerplayserver.sendPacket(
             new SPacketJoinGame(
                 playerIn.getEntityId(),
                 playerIn.interactionManager.getGameType(),
                 worldinfo.isHardcoreModeEnabled(),
-                worldserver.provider.getDimensionType().getId(),
+                worldserver.provider.getDimension(),
                 worldserver.getDifficulty(),
                 this.getMaxPlayers(),
                 worldinfo.getTerrainType(),
                 worldserver.getGameRules().getBoolean("reducedDebugInfo")
             )
         );
+        playerIn.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
         nethandlerplayserver.sendPacket(
             new SPacketCustomPayload("MC|Brand", new PacketBuffer(Unpooled.buffer()).writeString(this.getServerInstance().getServerModName()))
         );
@@ -153,20 +200,23 @@
         playerIn.getRecipeBook().init(playerIn);
         this.sendScoreboard((ServerScoreboard)worldserver.getScoreboard(), playerIn);
         this.server.refreshStatusNextTick();
-        TextComponentTranslation textcomponenttranslation;
-
+        // CraftBukkit start - login message is handled in the event
+        // TextComponentTranslation textcomponenttranslation;
+        String joinMessage;
         if (playerIn.getName().equalsIgnoreCase(s))
         {
-            textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined", playerIn.getDisplayName());
+            // textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined", new Object[] {playerIn.getDisplayName()});
+            joinMessage = "\u00A7e" + net.minecraft.util.text.translation.I18n.translateToLocalFormatted("multiplayer.player.joined", playerIn.getName());
         }
         else
         {
-            textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined.renamed", playerIn.getDisplayName(), s);
+            // textcomponenttranslation = new TextComponentTranslation("multiplayer.player.joined.renamed", new Object[] {playerIn.getDisplayName(), s});
+            joinMessage = "\u00A7e" + net.minecraft.util.text.translation.I18n.translateToLocalFormatted("multiplayer.player.joined.renamed", playerIn.getName(), s);
         }
 
-        textcomponenttranslation.getStyle().setColor(TextFormatting.YELLOW);
-        this.sendMessage(textcomponenttranslation);
-        this.playerLoggedIn(playerIn);
+        // textcomponenttranslation.getStyle().setColor(TextFormatting.YELLOW);
+        // this.sendMessage(textcomponenttranslation);
+        this.playerLoggedIn(playerIn, joinMessage);
         nethandlerplayserver.setPlayerLocation(
             playerIn.posX, playerIn.posY, playerIn.posZ, playerIn.rotationYaw, playerIn.rotationPitch
         );
@@ -221,9 +271,12 @@
         }
 
         playerIn.addSelfToInternalCraftingInventory();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedIn(playerIn);
+        // CraftBukkit - Moved from above, added world
+        PlayerList.LOGGER.info(playerIn.getName() + "[" + s1 + "] logged in with entity id " + playerIn.getEntityId() + " at ([" + playerIn.world.worldInfo.getWorldName() + "]" + playerIn.posX + ", " + playerIn.posY + ", " + playerIn.posZ + ")");
     }
 
-    protected void sendScoreboard(ServerScoreboard scoreboardIn, EntityPlayerMP playerIn)
+    public void sendScoreboard(ServerScoreboard scoreboardIn, EntityPlayerMP playerIn)
     {
         Set<ScoreObjective> set = Sets.newHashSet();
 
@@ -250,6 +303,7 @@
 
     public void setPlayerManager(WorldServer[] worldServers)
     {
+        if (playerDataManager != null) return;
         this.playerDataManager = worldServers[0].getSaveHandler().getPlayerNBTManager();
         worldServers[0].getWorldBorder().addListener(new IBorderListener()
         {
@@ -322,7 +376,7 @@
     @Nullable
     public NBTTagCompound readPlayerDataFromFile(EntityPlayerMP playerIn)
     {
-        NBTTagCompound nbttagcompound = this.server.worlds[0].getWorldInfo().getPlayerNBTTagCompound();
+        NBTTagCompound nbttagcompound = this.server.worldServerList.get(0).getWorldInfo().getPlayerNBTTagCompound();
         NBTTagCompound nbttagcompound1;
 
         if (playerIn.getName().equals(this.server.getServerOwner()) && nbttagcompound != null)
@@ -330,6 +384,7 @@
             nbttagcompound1 = nbttagcompound;
             playerIn.readFromNBT(nbttagcompound);
             LOGGER.debug("loading single player");
+            net.minecraftforge.event.ForgeEventFactory.firePlayerLoadingEvent(playerIn, this.playerDataManager, playerIn.getUniqueID().toString());
         }
         else
         {
@@ -339,8 +394,24 @@
         return nbttagcompound1;
     }
 
+    public NBTTagCompound getPlayerNBT(EntityPlayerMP player)
+    {
+        // Hacky method to allow loading the NBT for a player prior to login
+        NBTTagCompound nbttagcompound = this.server.worldServerList.get(0).getWorldInfo().getPlayerNBTTagCompound();
+        if (player.getName().equals(this.server.getServerOwner()) && nbttagcompound != null)
+        {
+            return nbttagcompound;
+        }
+        else
+        {
+            return ((net.minecraft.world.storage.SaveHandler)this.playerDataManager).getPlayerNBT(player);
+        }
+    }
+
     protected void writePlayerData(EntityPlayerMP playerIn)
     {
+        if (playerIn.connection == null) return;
+
         this.playerDataManager.writePlayerData(playerIn);
         StatisticsManagerServer statisticsmanagerserver = this.playerStatFiles.get(playerIn.getUniqueID());
 
@@ -369,19 +440,90 @@
             playerIn.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, this.playerEntityList.get(i)));
         }
 
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
         worldserver.spawnEntity(playerIn);
         this.preparePlayer(playerIn, null);
     }
 
+    public void playerLoggedIn(EntityPlayerMP playerIn, String joinMessage)
+    {
+        this.playerEntityList.add(playerIn);
+        this.uuidToPlayerMap.put(playerIn.getUniqueID(), playerIn);
+        // this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {playerIn})); // CraftBukkit - replaced with loop below
+        WorldServer worldserver = this.server.getWorld(playerIn.dimension);
+        
+        org.bukkit.event.player.PlayerJoinEvent playerJoinEvent = new org.bukkit.event.player.PlayerJoinEvent(cserver.getPlayer(playerIn), joinMessage);
+        cserver.getPluginManager().callEvent(playerJoinEvent);
+
+        if (!playerIn.connection.netManager.isChannelOpen()) {
+            return;
+        }
+
+        joinMessage = playerJoinEvent.getJoinMessage();
+
+        if (joinMessage != null && joinMessage.length() > 0) {
+            for (ITextComponent line : org.bukkit.craftbukkit.v1_12_R1.util.CraftChatMessage.fromString(joinMessage)) {
+                server.getPlayerList().sendPacketToAllPlayers(new SPacketChat(line));
+            }
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.chunkio.ChunkIOExecutor.adjustPoolSize(getCurrentPlayerCount());
+
+        // CraftBukkit start - sendAll above replaced with this loop
+        SPacketPlayerListItem packet = new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, playerIn);
+        for (int i = 0; i < this.playerEntityList.size(); i++)
+        {
+            // playerIn.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] {this.playerEntityList.get(i)}));
+            EntityPlayerMP entityplayer1 = this.playerEntityList.get(i);
+
+            if (entityplayer1.getBukkitEntity().canSee(playerIn.getBukkitEntity())) {
+                entityplayer1.connection.sendPacket(packet);
+            }
+
+            if (!playerIn.getBukkitEntity().canSee(entityplayer1.getBukkitEntity())) {
+                continue;
+            }
+
+            playerIn.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.ADD_PLAYER, new EntityPlayerMP[] { entityplayer1}));
+        }
+        playerIn.sentListPacket = true;
+        // CraftBukkit end
+
+        playerIn.connection.sendPacket(new net.minecraft.network.play.server.SPacketEntityMetadata(playerIn.getEntityId(), playerIn.getDataManager(), true)); // CraftBukkit - BungeeCord#2321, send complete data to self on spawn
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
+        // CraftBukkit start - Only add if the player wasn't moved in the event
+        if (playerIn.world == worldserver && !worldserver.playerEntities.contains(playerIn)) {
+            worldserver.spawnEntity(playerIn);
+            this.preparePlayer(playerIn, null);
+        }
+        // CraftBukkit end
+    }
+
     public void serverUpdateMovingPlayer(EntityPlayerMP playerIn)
     {
         playerIn.getServerWorld().getPlayerChunkMap().updateMovingPlayer(playerIn);
     }
 
-    public void playerLoggedOut(EntityPlayerMP playerIn)
+    public String playerLoggedOut(EntityPlayerMP playerIn)
     {
+        if (catserver.server.CatAsyncCatcher.checkAsync("kick player")) return catserver.server.CatAsyncCatcher.ensureExecuteOnPrimaryThread(() -> playerLoggedOut(playerIn)); // CatServer
+        String quitMessage = null;
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerLoggedOut(playerIn);
         WorldServer worldserver = playerIn.getServerWorld();
         playerIn.addStat(StatList.LEAVE_GAME);
+
+        // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
+        org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.handleInventoryCloseEvent(playerIn);
+        if(playerIn.connection != null) { // CatServer - fix crash when player is Forge Handshake
+            org.bukkit.event.player.PlayerQuitEvent playerQuitEvent = new org.bukkit.event.player.PlayerQuitEvent(cserver.getPlayer(playerIn), "\u00A7e" + playerIn.getName() + " left the game");
+            cserver.getPluginManager().callEvent(playerQuitEvent);
+            playerIn.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
+    
+            playerIn.onUpdateEntity();// SPIGOT-924
+            quitMessage = playerQuitEvent.getQuitMessage();
+        }
+        // CraftBukkit end
+
         this.writePlayerData(playerIn);
 
         if (playerIn.isRiding())
@@ -402,6 +544,7 @@
                 worldserver.getChunk(playerIn.chunkCoordX, playerIn.chunkCoordZ).markDirty();
             }
         }
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
 
         worldserver.removeEntity(playerIn);
         worldserver.getPlayerChunkMap().removePlayer(playerIn);
@@ -417,7 +560,23 @@
             this.advancements.remove(uuid);
         }
 
-        this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, playerIn));
+        // this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, new EntityPlayerMP[] {playerIn}));
+        SPacketPlayerListItem packet = new SPacketPlayerListItem(SPacketPlayerListItem.Action.REMOVE_PLAYER, playerIn);
+        for (int i = 0; i < playerEntityList.size(); i++) {
+            EntityPlayerMP entityplayer2 = (EntityPlayerMP) this.playerEntityList.get(i);
+
+            if (entityplayer2.getBukkitEntity().canSee(playerIn.getBukkitEntity())) {
+                entityplayer2.connection.sendPacket(packet);
+            } else {
+                entityplayer2.getBukkitEntity().removeDisconnectingPlayer(playerIn.getBukkitEntity());
+            }
+        }
+        // This removes the scoreboard (and player reference) for the specific player in the manager
+        cserver.getScoreboardManager().removePlayer(playerIn.getBukkitEntity());
+
+        org.bukkit.craftbukkit.v1_12_R1.chunkio.ChunkIOExecutor.adjustPoolSize(this.getCurrentPlayerCount());
+
+        return quitMessage;
     }
 
     public String allowUserToConnect(SocketAddress address, GameProfile profile)
@@ -456,6 +615,87 @@
         }
     }
 
+    @Nullable
+    public EntityPlayerMP allowUserToConnect(net.minecraft.server.network.NetHandlerLoginServer loginServer, GameProfile profile, String hostname)
+    {
+        // Moved from processLogin
+        UUID uuid = EntityPlayer.getUUID(profile);
+        java.util.ArrayList<EntityPlayerMP> arraylist = Lists.newArrayList();
+
+        EntityPlayerMP entityplayer;
+
+        for (int i = 0; i < this.playerEntityList.size(); i++) {
+            entityplayer = (EntityPlayerMP) this.playerEntityList.get(i);
+            if (entityplayer.getUniqueID().equals(uuid)) {
+                arraylist.add(entityplayer);
+            }
+        }
+        
+        java.util.Iterator<EntityPlayerMP> iterator = arraylist.iterator();
+
+        while (iterator.hasNext()) {
+            entityplayer = (EntityPlayerMP) iterator.next();
+            writePlayerData(entityplayer); // CraftBukkit - Force the player's inventory to be saved
+            entityplayer.connection.disconnect(new TextComponentTranslation("multiplayer.disconnect.duplicate_login", new Object[0]));
+        }
+
+        // Instead of kicking then returning, we need to store the kick reason
+        // in the event, check with plugins to see if it's ok, and THEN kick
+        // depending on the outcome.
+        SocketAddress socketaddress = loginServer.networkManager.getRemoteAddress();
+
+        EntityPlayerMP entity = new EntityPlayerMP(server, server.getWorldServer(0), profile, new PlayerInteractionManager(server.getWorldServer(0)));
+        org.bukkit.entity.Player player = entity.getBukkitEntity();
+        //Spigot Start
+        org.bukkit.event.player.PlayerLoginEvent event = new org.bukkit.event.player.PlayerLoginEvent(player, hostname, ((java.net.InetSocketAddress) socketaddress).getAddress(), ((java.net.InetSocketAddress) loginServer.networkManager.getRawAddress()).getAddress());
+        //Spigot End
+        if (getBannedPlayers().isBanned(profile) && !getBannedPlayers().getEntry(profile).hasBanExpired())
+        {
+            UserListBansEntry userlistbansentry = (UserListBansEntry)this.bannedPlayers.getEntry(profile);
+            String s1 = "You are banned from this server!\nReason: " + userlistbansentry.getBanReason();
+
+            if (userlistbansentry.getBanEndDate() != null)
+            {
+                s1 = s1 + "\nYour ban will be removed on " + DATE_FORMAT.format(userlistbansentry.getBanEndDate());
+            }
+
+            // return s1;
+            event.disallow(org.bukkit.event.player.PlayerLoginEvent.Result.KICK_BANNED, s1);
+        }
+        else if (!this.canJoin(profile))
+        {
+            // return "You are not white-listed on this server!";
+            event.disallow(org.bukkit.event.player.PlayerLoginEvent.Result.KICK_WHITELIST, org.spigotmc.SpigotConfig.whitelistMessage); // Spigot
+        }
+        else if (getBannedIPs().isBanned(socketaddress) && !getBannedIPs().getBanEntry(socketaddress).hasBanExpired())
+        {
+            UserListIPBansEntry userlistipbansentry = this.bannedIPs.getBanEntry(socketaddress);
+            String s = "Your IP address is banned from this server!\nReason: " + userlistipbansentry.getBanReason();
+
+            if (userlistipbansentry.getBanEndDate() != null)
+            {
+                s = s + "\nYour ban will be removed on " + DATE_FORMAT.format(userlistipbansentry.getBanEndDate());
+            }
+
+            // return s;
+            if (!userlistipbansentry.hasBanExpired()) event.disallow(org.bukkit.event.player.PlayerLoginEvent.Result.KICK_BANNED, s);
+        }
+        else
+        {
+            // return this.playerEntityList.size() >= this.maxPlayers && !this.bypassesPlayerLimit(profile) ? "The server is full!" : null;
+            if (this.playerEntityList.size() >= this.maxPlayers && !this.bypassesPlayerLimit(profile)) {
+                event.disallow(org.bukkit.event.player.PlayerLoginEvent.Result.KICK_FULL, org.spigotmc.SpigotConfig.serverFullMessage); // Spigot
+            }
+        }
+
+        cserver.getPluginManager().callEvent(event);
+        if (event.getResult() != org.bukkit.event.player.PlayerLoginEvent.Result.ALLOWED) {
+            loginServer.disconnect(event.getKickMessage());
+            return null;
+        }
+        return entity;
+    }
+
     public EntityPlayerMP createPlayerForUser(GameProfile profile)
     {
         UUID uuid = EntityPlayer.getUUID(profile);
@@ -497,15 +737,99 @@
         return new EntityPlayerMP(this.server, this.server.getWorld(0), profile, playerinteractionmanager);
     }
 
+    public EntityPlayerMP createPlayerForUser(GameProfile profile, EntityPlayerMP entityPlayerMP)
+    {
+        /* CraftBukkit - Moved up
+        UUID uuid = EntityPlayer.getUUID(profile);
+        List<EntityPlayerMP> list = Lists.<EntityPlayerMP>newArrayList();
+
+        for (int i = 0; i < this.playerEntityList.size(); i++)
+        {
+            EntityPlayerMP entityplayermp = this.playerEntityList.get(i);
+
+            if (entityplayermp.getUniqueID().equals(uuid))
+            {
+                list.add(entityplayermp);
+            }
+        }
+
+        EntityPlayerMP entityplayermp2 = this.uuidToPlayerMap.get(profile.getId());
+
+        if (entityplayermp2 != null && !list.contains(entityplayermp2))
+        {
+            list.add(entityplayermp2);
+        }
+
+        for (EntityPlayerMP entityplayermp1 : list)
+        {
+            entityplayermp1.connection.disconnect(new TextComponentTranslation("multiplayer.disconnect.duplicate_login", new Object[0]));
+        }
+
+        PlayerInteractionManager playerinteractionmanager;
+
+        if (this.mcServer.isDemo())
+        {
+            playerinteractionmanager = new DemoPlayerInteractionManager(this.mcServer.getWorld(0));
+        }
+        else
+        {
+            playerinteractionmanager = new PlayerInteractionManager(this.mcServer.getWorld(0));
+        }
+
+        return new EntityPlayerMP(this.mcServer, this.mcServer.getWorld(0), profile, playerinteractionmanager);
+        */
+        return entityPlayerMP;
+    }
+
     public EntityPlayerMP recreatePlayerEntity(EntityPlayerMP playerIn, int dimension, boolean conqueredEnd)
     {
+        return this.moveToWorld(playerIn, dimension, conqueredEnd, null, true);
+    }
+
+    public EntityPlayerMP moveToWorld(EntityPlayerMP playerIn, int dimension, boolean conqueredEnd, org.bukkit.Location location, boolean avoidSuffocation) {
+        playerIn.dismountRidingEntity(); // CraftBukkit
+        World world = server.getWorld(dimension);
+        if (world == null)
+        {
+            dimension = playerIn.getSpawnDimension();
+        }
+        else if (location == null && !world.provider.canRespawnHere())
+        {
+            dimension = world.provider.getRespawnDimension(playerIn);
+        }
+        if (server.getWorld(dimension) == null) dimension = 0;
+
+        // handle return from End
+        if (conqueredEnd && location == null)
+        {
+            WorldServer exitWorld = this.server.getWorld(dimension);
+            org.bukkit.Location enter = playerIn.getBukkitEntity().getLocation();
+            org.bukkit.Location exit = null;
+            // THE_END -> NORMAL; use bed if available, otherwise default spawn
+            exit = playerIn.getBukkitEntity().getBedSpawnLocation();
+
+            if (exit == null || ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) exit.getWorld()).getHandle().dimension != 0)
+            {
+                exit = exitWorld.getWorld().getSpawnLocation();
+            }
+            org.bukkit.event.player.PlayerPortalEvent event = new org.bukkit.event.player.PlayerPortalEvent(playerIn.getBukkitEntity(), enter, exit, org.bukkit.craftbukkit.v1_12_R1.CraftTravelAgent.DEFAULT, org.bukkit.event.player.PlayerPortalEvent.TeleportCause.END_PORTAL);
+            event.useTravelAgent(false);
+            org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+            if (event.isCancelled() || event.getTo() == null)
+            {
+                return null;
+            }
+            location = event.getTo();
+        }
+
         playerIn.getServerWorld().getEntityTracker().removePlayerFromTrackers(playerIn);
-        playerIn.getServerWorld().getEntityTracker().untrack(playerIn);
+        // playerIn.getServerWorld().getEntityTracker().untrack(playerIn);
         playerIn.getServerWorld().getPlayerChunkMap().removePlayer(playerIn);
         this.playerEntityList.remove(playerIn);
         this.server.getWorld(playerIn.dimension).removeEntityDangerously(playerIn);
-        BlockPos blockpos = playerIn.getBedLocation();
-        boolean flag = playerIn.isSpawnForced();
+        BlockPos blockpos = playerIn.getBedLocation(dimension);
+        boolean flag = playerIn.isSpawnForced(dimension);
+        /* CraftBukkit start
         playerIn.dimension = dimension;
         PlayerInteractionManager playerinteractionmanager;
 
@@ -521,8 +845,14 @@
         EntityPlayerMP entityplayermp = new EntityPlayerMP(
             this.server, this.server.getWorld(playerIn.dimension), playerIn.getGameProfile(), playerinteractionmanager
         );
+        */
+        EntityPlayerMP entityplayermp = playerIn;
+        org.bukkit.World fromWorld = playerIn.getBukkitEntity().getWorld();
+        playerIn.queuedEndExit = false;
+
         entityplayermp.connection = playerIn.connection;
         entityplayermp.copyFrom(playerIn, conqueredEnd);
+        entityplayermp.dimension = dimension;
         entityplayermp.setEntityId(playerIn.getEntityId());
         entityplayermp.setCommandStats(playerIn);
         entityplayermp.setPrimaryHand(playerIn.getPrimaryHand());
@@ -532,60 +862,112 @@
             entityplayermp.addTag(s);
         }
 
-        WorldServer worldserver = this.server.getWorld(playerIn.dimension);
-        this.setPlayerGameTypeBasedOnOther(entityplayermp, playerIn, worldserver);
-
-        if (blockpos != null)
-        {
-            BlockPos blockpos1 = EntityPlayer.getBedSpawnLocation(this.server.getWorld(playerIn.dimension), blockpos, flag);
-
-            if (blockpos1 != null)
-            {
-                entityplayermp.setLocationAndAngles(blockpos1.getX() + 0.5F, blockpos1.getY() + 0.1F, blockpos1.getZ() + 0.5F, 0.0F, 0.0F);
-                entityplayermp.setSpawnPoint(blockpos, flag);
-            }
-            else
-            {
-                entityplayermp.connection.sendPacket(new SPacketChangeGameState(0, 0.0F));
-            }
+        // WorldServer worldserver = this.mcServer.getWorld(playerIn.dimension); // CraftBukkit - handled later
+        // this.setPlayerGameTypeBasedOnOther(entityplayermp, playerIn, worldserver);
+        BlockPos blockposition1;
+        // CraftBukkit start - fire PlayerRespawnEvent
+        if (location == null) {
+            boolean isBedSpawn = false;
+            org.bukkit.craftbukkit.v1_12_R1.CraftWorld cworld = (org.bukkit.craftbukkit.v1_12_R1.CraftWorld) this.server.server.getWorld(playerIn.spawnWorld);
+            if (cworld != null && blockpos != null) {
+                blockposition1 = EntityPlayer.getBedSpawnLocation(cworld.getHandle(), blockpos, flag);
+                if (blockposition1 != null){
+                    isBedSpawn = true;
+                    location = new org.bukkit.Location(cworld, (float) blockposition1.getX() + 0.5F, (float) blockposition1.getY() + 0.1F, (float) blockposition1.getZ() + 0.5F);
+                } else {
+                    entityplayermp.setSpawnPoint(null, true);
+                    entityplayermp.connection.sendPacket(new SPacketChangeGameState(0, 0.0F));
+                }
+            }
+
+            if (location == null) {
+                cworld = (org.bukkit.craftbukkit.v1_12_R1.CraftWorld) this.server.server.getWorlds().get(0);
+                blockpos = entityplayermp.getSpawnPoint(this.server, cworld.getHandle());
+                location = new org.bukkit.Location(cworld, (float) blockpos.getX() + 0.5F, (float) blockpos.getY() + 0.1F, (float) blockpos.getZ() + 0.5F);
+            }
+            org.bukkit.entity.Player respawnPlayer = cserver.getPlayer(entityplayermp);
+            org.bukkit.event.player.PlayerRespawnEvent respawnEvent = new org.bukkit.event.player.PlayerRespawnEvent(respawnPlayer, location, isBedSpawn);
+            cserver.getPluginManager().callEvent(respawnEvent);
+            // Spigot Start
+            if (playerIn.connection.isDisconnected()) {
+                return playerIn;
+            }
+            // Spigot End
+
+
+            location = respawnEvent.getRespawnLocation();
+            playerIn.reset();
+        } else {
+            location.setWorld(server.getWorld(dimension).getWorld());
         }
+
+        WorldServer worldserver = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) location.getWorld()).getHandle();
+        entityplayermp.forceSetPositionRotation(location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch());
 
         worldserver.getChunkProvider().provideChunk((int)entityplayermp.posX >> 4, (int)entityplayermp.posZ >> 4);
 
-        while (!worldserver.getCollisionBoxes(entityplayermp, entityplayermp.getEntityBoundingBox()).isEmpty() && entityplayermp.posY < 256.0)
+        while (avoidSuffocation && !worldserver.getCollisionBoxes(entityplayermp, entityplayermp.getEntityBoundingBox()).isEmpty() && entityplayermp.posY < 256.0)
         {
             entityplayermp.setPosition(entityplayermp.posX, entityplayermp.posY + 1.0, entityplayermp.posZ);
         }
 
+        int actualDimension = worldserver.provider.getDimension();
+        catserver.server.CatServerNetwork.registerBukkitWorldToClient(entityplayermp, actualDimension); // CatServer - change dim for bukkit added dimensions
+        // entityplayermp.connection.sendPacket(new SPacketRespawn(entityplayermp.dimension, entityplayermp.world.getDifficulty(), entityplayermp.world.getWorldInfo().getTerrainType(), entityplayermp.interactionManager.getGameType()));
         entityplayermp.connection
         .sendPacket(
             new SPacketRespawn(
-                entityplayermp.dimension,
-                entityplayermp.world.getDifficulty(),
-                entityplayermp.world.getWorldInfo().getTerrainType(),
+                actualDimension,
+                worldserver.getDifficulty(),
+                worldserver.getWorldInfo().getTerrainType(),
                 entityplayermp.interactionManager.getGameType()
             )
         );
+        entityplayermp.setWorld(worldserver);
+        entityplayermp.isDead = false;
+        entityplayermp.connection.teleport(new org.bukkit.Location(worldserver.getWorld(), entityplayermp.posX, entityplayermp.posY, entityplayermp.posZ, entityplayermp.rotationYaw, entityplayermp.rotationPitch));
+        entityplayermp.setSneaking(false);
         BlockPos blockpos2 = worldserver.getSpawnPoint();
-        entityplayermp.connection
-        .setPlayerLocation(
-            entityplayermp.posX,
-            entityplayermp.posY,
-            entityplayermp.posZ,
-            entityplayermp.rotationYaw,
-            entityplayermp.rotationPitch
-        );
+        // entityplayermp.connection.setPlayerLocation(entityplayermp.posX, entityplayermp.posY, entityplayermp.posZ, entityplayermp.rotationYaw, entityplayermp.rotationPitch);
         entityplayermp.connection.sendPacket(new SPacketSpawnPosition(blockpos2));
         entityplayermp.connection
         .sendPacket(new SPacketSetExperience(entityplayermp.experience, entityplayermp.experienceTotal, entityplayermp.experienceLevel));
         this.updateTimeAndWeatherForPlayer(entityplayermp, worldserver);
         this.updatePermissionLevel(entityplayermp);
-        worldserver.getPlayerChunkMap().addPlayer(entityplayermp);
-        worldserver.spawnEntity(entityplayermp);
-        this.playerEntityList.add(entityplayermp);
-        this.uuidToPlayerMap.put(entityplayermp.getUniqueID(), entityplayermp);
-        entityplayermp.addSelfToInternalCraftingInventory();
+        if (!playerIn.connection.isDisconnected()) {
+            worldserver.getPlayerChunkMap().addPlayer(entityplayermp);
+            worldserver.spawnEntity(entityplayermp);
+            this.playerEntityList.add(entityplayermp);
+            this.uuidToPlayerMap.put(entityplayermp.getUniqueID(), entityplayermp);
+        }
+        // entityplayermp.addSelfToInternalCraftingInventory();
         entityplayermp.setHealth(entityplayermp.getHealth());
+        // Added from changeDimension
+        syncPlayerInventory(playerIn); // Update health, etc...
+        playerIn.sendPlayerAbilities();
+        for (Object o1 : playerIn.getActivePotionEffects()) {
+            PotionEffect mobEffect = (PotionEffect) o1;
+            playerIn.connection.sendPacket(new SPacketEntityEffect(playerIn.getEntityId(), mobEffect));
+        }
+
+        // Fire advancement trigger
+        CriteriaTriggers.CHANGED_DIMENSION.trigger(playerIn, ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) fromWorld).getHandle().provider.getDimensionType(), worldserver.provider.getDimensionType());
+        if (((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) fromWorld).getHandle().provider.getDimensionType() == DimensionType.NETHER && worldserver.provider.getDimensionType() == DimensionType.OVERWORLD && playerIn.getEnteredNetherPosition() != null) {
+            CriteriaTriggers.NETHER_TRAVEL.trigger(playerIn, playerIn.getEnteredNetherPosition());
+        }
+
+        // Don't fire on respawn
+        if (fromWorld != location.getWorld()) {
+            org.bukkit.event.player.PlayerChangedWorldEvent event = new org.bukkit.event.player.PlayerChangedWorldEvent(playerIn.getBukkitEntity(), fromWorld);
+            server.server.getPluginManager().callEvent(event);
+        }
+
+        // Save player file again if they were disconnected
+        if (playerIn.connection.isDisconnected()) {
+            this.writePlayerData(playerIn);
+        }
+        // CraftBukkit end
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerRespawnEvent(entityplayermp, conqueredEnd);
         return entityplayermp;
     }
 
@@ -593,53 +975,125 @@
     {
         GameProfile gameprofile = player.getGameProfile();
         int i = this.canSendCommands(gameprofile) ? this.ops.getPermissionLevel(gameprofile) : 0;
-        i = this.server.isSinglePlayer() && this.server.worlds[0].getWorldInfo().areCommandsAllowed() ? 4 : i;
+        i = this.server.isSinglePlayer() && this.server.worldServerList.get(0).getWorldInfo().areCommandsAllowed() ? 4 : i;
         i = this.commandsAllowedForAll ? 4 : i;
         this.sendPlayerPermissionLevel(player, i);
     }
 
     public void changePlayerDimension(EntityPlayerMP player, int dimensionIn)
     {
-        int i = player.dimension;
-        WorldServer worldserver = this.server.getWorld(player.dimension);
+        transferPlayerToDimension(player, dimensionIn, server.getWorld(dimensionIn).getDefaultTeleporter());
+    }
+
+    // TODO: Remove (1.13)
+    public void transferPlayerToDimension(EntityPlayerMP player, int dimensionIn, net.minecraft.world.Teleporter teleporter)
+    {
+        transferPlayerToDimension(player, dimensionIn, (net.minecraftforge.common.util.ITeleporter) teleporter);
+    }
+
+    // CatServer start - Replaced the standard handling of portals with a more customised method.
+    public void transferPlayerToDimension(EntityPlayerMP player, int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter)
+    {
+        transferPlayerToDimension(player, dimensionIn, teleporter, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.MOD);
+    }
+
+    public void transferPlayerToDimension(EntityPlayerMP player, int dimensionIn, net.minecraftforge.common.util.ITeleporter teleporter, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause)
+    {
+        WorldServer fromWorld = this.server.getWorld(player.dimension);
+        WorldServer exitWorld = this.server.getWorld(dimensionIn);
+
+        org.bukkit.Location enter = player.getBukkitEntity().getLocation();
+        org.bukkit.Location exit = null;
+        boolean useTravelAgent = false;
+        if (exitWorld != null) {
+            exit = this.calculateTarget(enter, exitWorld);
+            if (cause != cause.MOD) // don't use travel agent for custom dimensions
+            {
+                useTravelAgent = true;
+            }
+        }
+
+        org.bukkit.TravelAgent agent = teleporter instanceof org.bukkit.TravelAgent ? (org.bukkit.TravelAgent)teleporter : org.bukkit.craftbukkit.v1_12_R1.CraftTravelAgent.DEFAULT; // return arbitrary TA to compensate for implementation dependent plugins
+        
+        org.bukkit.event.player.PlayerPortalEvent event = new org.bukkit.event.player.PlayerPortalEvent(player.getBukkitEntity(), enter, exit, agent, cause);
+        event.useTravelAgent(useTravelAgent);
+        org.bukkit.Bukkit.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled() || event.getTo() == null) {
+            return;
+        }
+
+        exit = event.useTravelAgent() && cause != cause.MOD ? event.getPortalTravelAgent().findOrCreate(event.getTo()) : event.getTo();
+        if (exit == null) {
+            return;
+        }
+        exitWorld = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) exit.getWorld()).getHandle();
+
+        org.bukkit.event.player.PlayerTeleportEvent tpEvent = new org.bukkit.event.player.PlayerTeleportEvent(player.getBukkitEntity(), enter, exit, cause);
+        org.bukkit.Bukkit.getServer().getPluginManager().callEvent(tpEvent);
+        if (tpEvent.isCancelled() || tpEvent.getTo() == null) {
+            return;
+        }
+        
+        org.bukkit.util.Vector velocity = player.getBukkitEntity().getVelocity();
+        exitWorld.getDefaultTeleporter().adjustExit(player, exit, velocity);
+
+        player.invulnerableDimensionChange = true; // CraftBukkit - Set teleport invulnerability only if player changing worlds
+
+        int fromDimension = player.dimension;
         player.dimension = dimensionIn;
-        WorldServer worldserver1 = this.server.getWorld(player.dimension);
-        player.connection
-        .sendPacket(
-            new SPacketRespawn(
-                player.dimension,
-                player.world.getDifficulty(),
-                player.world.getWorldInfo().getTerrainType(),
-                player.interactionManager.getGameType()
-            )
-        );
+
+        catserver.server.CatServerNetwork.registerBukkitWorldToClient(player, player.dimension); // CatServer - change dim for bukkit added dimensions
+        player.connection.sendPacket(new SPacketRespawn(player.dimension, exitWorld.getDifficulty(), exitWorld.getWorldInfo().getTerrainType(), player.interactionManager.getGameType()));
         this.updatePermissionLevel(player);
-        worldserver.removeEntityDangerously(player);
+        fromWorld.removeEntityDangerously(player);
         player.isDead = false;
-        this.transferEntityToWorld(player, i, worldserver, worldserver1);
-        this.preparePlayer(player, worldserver);
-        player.connection
-        .setPlayerLocation(player.posX, player.posY, player.posZ, player.rotationYaw, player.rotationPitch);
-        player.interactionManager.setWorld(worldserver1);
+        this.transferEntityToWorld(player, fromDimension, fromWorld, exitWorld, teleporter);
+        this.preparePlayer(player, fromWorld);
+        player.connection.setPlayerLocation(player.posX, player.posY, player.posZ, player.rotationYaw, player.rotationPitch);
+        player.interactionManager.setWorld(exitWorld);
         player.connection.sendPacket(new SPacketPlayerAbilities(player.capabilities));
-        this.updateTimeAndWeatherForPlayer(player, worldserver1);
+        this.updateTimeAndWeatherForPlayer(player, exitWorld);
         this.syncPlayerInventory(player);
 
         for (PotionEffect potioneffect : player.getActivePotionEffects())
         {
             player.connection.sendPacket(new SPacketEntityEffect(player.getEntityId(), potioneffect));
         }
-    }
+        // Fix MC-88179: on non-death SPacketRespawn, also resend attributes
+        net.minecraft.entity.ai.attributes.AttributeMap attributemap = (net.minecraft.entity.ai.attributes.AttributeMap) player.getAttributeMap();
+        java.util.Collection<net.minecraft.entity.ai.attributes.IAttributeInstance> watchedAttribs = attributemap.getWatchedAttributes();
+        if (!watchedAttribs.isEmpty()) player.connection.sendPacket(new net.minecraft.network.play.server.SPacketEntityProperties(player.getEntityId(), watchedAttribs));
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerChangedDimensionEvent(player, fromDimension, dimensionIn);
+    }
+
+    public void changeDimension(EntityPlayerMP player, int dimensionIn, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause) {
+        transferPlayerToDimension(player, dimensionIn, server.getWorld(dimensionIn).getDefaultTeleporter(), cause);
+    }
+    // CatServer end
 
     public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn)
     {
-        double d0 = entityIn.posX;
-        double d1 = entityIn.posZ;
+        // transferEntityToWorld(entityIn, lastDimension, oldWorldIn, toWorldIn, toWorldIn.getDefaultTeleporter());
+        org.bukkit.Location exit = this.calculateTarget(entityIn.getBukkitEntity().getLocation(), toWorldIn);
+        this.repositionEntity(entityIn, exit, true);
+    }
+
+    // TODO: Remove (1.13)
+    public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn, net.minecraft.world.Teleporter teleporter)
+    {
+        transferEntityToWorld(entityIn, lastDimension, oldWorldIn, toWorldIn, (net.minecraftforge.common.util.ITeleporter) teleporter);
+    }
+
+    public void transferEntityToWorld(Entity entityIn, int lastDimension, WorldServer oldWorldIn, WorldServer toWorldIn, net.minecraftforge.common.util.ITeleporter teleporter)
+    {
+        double moveFactor = oldWorldIn.provider.getMovementFactor() / toWorldIn.provider.getMovementFactor();
+        double d0 = MathHelper.clamp(entityIn.posX * moveFactor, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
+        double d1 = MathHelper.clamp(entityIn.posZ * moveFactor, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
         double d2 = 8.0;
         float f = entityIn.rotationYaw;
         oldWorldIn.profiler.startSection("moving");
 
-        if (entityIn.dimension == -1)
+        if (false && entityIn.dimension == -1)
         {
             d0 = MathHelper.clamp(d0 / 8.0, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
             d1 = MathHelper.clamp(d1 / 8.0, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
@@ -650,7 +1104,7 @@
                 oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else if (entityIn.dimension == 0)
+        else if (false && entityIn.dimension == 0)
         {
             d0 = MathHelper.clamp(d0 * 8.0, toWorldIn.getWorldBorder().minX() + 16.0, toWorldIn.getWorldBorder().maxX() - 16.0);
             d1 = MathHelper.clamp(d1 * 8.0, toWorldIn.getWorldBorder().minZ() + 16.0, toWorldIn.getWorldBorder().maxZ() - 16.0);
@@ -661,7 +1115,7 @@
                 oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
         }
-        else
+        if (entityIn.dimension == 1 && teleporter.isVanilla())
         {
             BlockPos blockpos;
 
@@ -687,7 +1141,7 @@
 
         oldWorldIn.profiler.endSection();
 
-        if (lastDimension != 1)
+        if (lastDimension != 1 || !teleporter.isVanilla())
         {
             oldWorldIn.profiler.startSection("placing");
             d0 = MathHelper.clamp((int)d0, -29999872, 29999872);
@@ -696,7 +1150,8 @@
             if (entityIn.isEntityAlive())
             {
                 entityIn.setLocationAndAngles(d0, entityIn.posY, d1, entityIn.rotationYaw, entityIn.rotationPitch);
-                toWorldIn.getDefaultTeleporter().placeInPortal(entityIn, f);
+                oldWorldIn.updateEntityWithOptionalForce(entityIn, false);
+                teleporter.placeEntity(toWorldIn, entityIn, f);
                 toWorldIn.spawnEntity(entityIn);
                 toWorldIn.updateEntityWithOptionalForce(entityIn, false);
             }
@@ -707,11 +1162,155 @@
         entityIn.setWorld(toWorldIn);
     }
 
+    // Copy of original transferEntityToWorld(Entity, int, WorldServer, WorldServer) method with only location calculation logic
+    public org.bukkit.Location calculateTarget(org.bukkit.Location enter, World target) {
+        WorldServer worldserver = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) enter.getWorld()).getHandle();
+        WorldServer worldserver1 = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) target.getWorld()).getHandle();
+        int i = worldserver.dimension;
+
+        double y = enter.getY();
+        float yaw = enter.getYaw();
+        float pitch = enter.getPitch();
+        double d0 = enter.getX();
+        double d1 = enter.getZ();
+        double d2 = 8.0D;
+
+        worldserver.profiler.startSection("moving");
+        if (worldserver1.dimension == -1) {
+            d0 = MathHelper.clamp(d0 / d2, worldserver1.getWorldBorder().minX()+ 16.0D, worldserver1.getWorldBorder().maxX() - 16.0D);
+            d1 = MathHelper.clamp(d1 / d2, worldserver1.getWorldBorder().minZ() + 16.0D, worldserver1.getWorldBorder().maxZ() - 16.0D);
+        } else if (worldserver1.dimension == 0) {
+            d0 = MathHelper.clamp(d0 * d2, worldserver1.getWorldBorder().minX() + 16.0D, worldserver1.getWorldBorder().maxX() - 16.0D);
+            d1 = MathHelper.clamp(d1 * d2, worldserver1.getWorldBorder().minZ() + 16.0D, worldserver1.getWorldBorder().maxZ() - 16.0D);
+        } else {
+            BlockPos blockposition;
+
+            if (i == 1) {
+                // use default NORMAL world spawn instead of target
+                worldserver1 = this.server.worldServerList.get(0);
+                blockposition = worldserver1.getSpawnPoint();
+            } else {
+                blockposition = worldserver1.getSpawnCoordinate();
+            }
+
+            // CatServer start - check null
+            if (blockposition != null)
+            {
+                d0 = (double) blockposition.getX();
+                y = (double) blockposition.getY();
+                d1 = (double) blockposition.getZ();
+            }
+            // CatServer end
+        }
+
+        worldserver.profiler.endSection();
+        if (i != 1) {
+            worldserver.profiler.startSection("placing");
+            d0 = (double) MathHelper.clamp((int) d0, -29999872, 29999872);
+            d1 = (double) MathHelper.clamp((int) d1, -29999872, 29999872);
+            worldserver.profiler.endSection();
+        }
+
+        return new org.bukkit.Location(worldserver1.getWorld(), d0, y, d1, yaw, pitch);
+    }
+
+    // copy of original transferEntityToWorld(Entity, int, WorldServer, WorldServer) method with only entity repositioning logic
+    public void repositionEntity(Entity entity, org.bukkit.Location exit, boolean portal) {
+        WorldServer worldserver = (WorldServer) entity.world;
+        WorldServer worldserver1 = ((org.bukkit.craftbukkit.v1_12_R1.CraftWorld) exit.getWorld()).getHandle();
+        int i = worldserver.dimension;
+
+        /*
+        double d0 = entity.locX;
+        double d1 = entity.locZ;
+        double d2 = 8.0D;
+        float f = entity.yaw;
+        */
+
+        worldserver.profiler.startSection("moving");
+        entity.setLocationAndAngles(exit.getX(), exit.getY(), exit.getZ(), exit.getYaw(), exit.getPitch());
+        if (entity.isEntityAlive()) {
+            worldserver.updateEntityWithOptionalForce(entity, false);
+        }
+        /*
+        if (entity.dimension == -1) {
+            d0 = MathHelper.a(d0 / 8.0D, worldserver1.getWorldBorder().b() + 16.0D, worldserver1.getWorldBorder().d() - 16.0D);
+            d1 = MathHelper.a(d1 / 8.0D, worldserver1.getWorldBorder().c() + 16.0D, worldserver1.getWorldBorder().e() - 16.0D);
+            entity.setPositionRotation(d0, entity.locY, d1, entity.yaw, entity.pitch);
+            if (entity.isAlive()) {
+                worldserver.entityJoinedWorld(entity, false);
+            }
+        } else if (entity.dimension == 0) {
+            d0 = MathHelper.a(d0 * 8.0D, worldserver1.getWorldBorder().b() + 16.0D, worldserver1.getWorldBorder().d() - 16.0D);
+            d1 = MathHelper.a(d1 * 8.0D, worldserver1.getWorldBorder().c() + 16.0D, worldserver1.getWorldBorder().e() - 16.0D);
+            entity.setPositionRotation(d0, entity.locY, d1, entity.yaw, entity.pitch);
+            if (entity.isAlive()) {
+                worldserver.entityJoinedWorld(entity, false);
+            }
+        } else {
+            BlockPosition blockposition;
+
+            if (i == 1) {
+                // use default NORMAL world spawn instead of target
+                worldserver1 = this.server.worlds.get(0);
+                blockposition = worldserver1.getSpawn();
+            } else {
+                blockposition = worldserver1.getDimensionSpawn();
+            }
+
+            d0 = (double) blockposition.getX();
+            entity.locY = (double) blockposition.getY();
+            d1 = (double) blockposition.getZ();
+            entity.setPositionRotation(d0, entity.locY, d1, 90.0F, 0.0F);
+            if (entity.isAlive()) {
+                worldserver.entityJoinedWorld(entity, false);
+            }
+        }
+        */
+
+        worldserver.profiler.endSection();
+        if (i != 1) {
+            worldserver.profiler.startSection("placing");
+            /*
+            d0 = (double) MathHelper.clamp((int) d0, -29999872, 29999872);
+            d1 = (double) MathHelper.clamp((int) d1, -29999872, 29999872);
+            */
+            if (entity.isEntityAlive()) {
+                // entity.setPositionRotation(d0, entity.locY, d1, entity.yaw, entity.pitch);
+                // worldserver1.getTravelAgent().a(entity, f);
+                if (portal) {
+                    org.bukkit.util.Vector velocity = entity.getBukkitEntity().getVelocity();
+                    worldserver1.getDefaultTeleporter().adjustExit(entity, exit, velocity);
+                    entity.setLocationAndAngles(exit.getX(), exit.getY(), exit.getZ(), exit.getYaw(), exit.getPitch());
+                    if (entity.motionX != velocity.getX() || entity.motionY != velocity.getY() || entity.motionZ != velocity.getZ()) {
+                        entity.getBukkitEntity().setVelocity(velocity);
+                    }
+                }
+                // worldserver1.addEntity(entity);
+                worldserver1.updateEntityWithOptionalForce(entity, false);
+            }
+
+            worldserver.profiler.endSection();
+        }
+
+        entity.setWorld(worldserver1);
+    }
+
     public void onTick()
     {
         if (++this.playerPingIndex > 600)
         {
-            this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.UPDATE_LATENCY, this.playerEntityList));
+            // this.sendPacketToAllPlayers(new SPacketPlayerListItem(SPacketPlayerListItem.Action.UPDATE_LATENCY, this.playerEntityList));
+            for (int i = 0; i < this.playerEntityList.size(); i++) {
+                final EntityPlayerMP target = this.playerEntityList.get(i);
+
+                target.connection.sendPacket(new SPacketPlayerListItem(SPacketPlayerListItem.Action.UPDATE_LATENCY, com.google.common.collect.Iterables.filter(this.playerEntityList, new com.google.common.base.Predicate<EntityPlayerMP>() {
+                    @Override
+                    public boolean apply(EntityPlayerMP input) {
+                        return target.getBukkitEntity().canSee(input.getBukkitEntity());
+                    }
+                })));
+            }
             this.playerPingIndex = 0;
         }
     }
@@ -724,6 +1323,24 @@
         }
     }
 
+    // CraftBukkit start - add a world/entity limited version
+    public void sendAll(Packet packet, EntityPlayer entityhuman) {
+        for (int i = 0; i < this.playerEntityList.size(); i++) {
+            EntityPlayerMP entityplayer =  this.playerEntityList.get(i);
+            if (entityhuman != null && entityhuman instanceof EntityPlayerMP && !entityplayer.getBukkitEntity().canSee(((EntityPlayerMP) entityhuman).getBukkitEntity())) {
+                continue;
+            }
+            ((EntityPlayerMP) this.playerEntityList.get(i)).connection.sendPacket(packet);
+        }
+    }
+
+    public void sendAll(Packet packet, World world) {
+        for (int i = 0; i < world.playerEntities.size(); i++) {
+            ((EntityPlayerMP) world.playerEntities.get(i)).connection.sendPacket(packet);
+        }
+    }
+    // CraftBukkit end
+
     public void sendPacketToAllPlayersInDimension(Packet<?> packetIn, int dimension)
     {
         for (int i = 0; i < this.playerEntityList.size(); i++)
@@ -839,12 +1456,20 @@
         int i = this.server.getOpPermissionLevel();
         this.ops.addEntry(new UserListOpsEntry(profile, this.server.getOpPermissionLevel(), this.ops.bypassesPlayerLimit(profile)));
         this.sendPlayerPermissionLevel(this.getPlayerByUUID(profile.getId()), i);
+        org.bukkit.entity.Player player = server.server.getPlayer(profile.getId());
+        if (player != null) {
+            player.recalculatePermissions();
+        }
     }
 
     public void removeOp(GameProfile profile)
     {
         this.ops.removeEntry(profile);
         this.sendPlayerPermissionLevel(this.getPlayerByUUID(profile.getId()), 0);
+        org.bukkit.entity.Player player = server.server.getPlayer(profile.getId());
+        if (player != null) {
+            player.recalculatePermissions();
+        }
     }
 
     private void sendPlayerPermissionLevel(EntityPlayerMP player, int permLevel)
@@ -879,7 +1504,7 @@
     {
         return this.ops.hasEntry(profile)
                || this.server.isSinglePlayer()
-               && this.server.worlds[0].getWorldInfo().areCommandsAllowed()
+               && this.server.worldServerList.get(0).getWorldInfo().areCommandsAllowed()
                && this.server.getServerOwner().equalsIgnoreCase(profile.getName())
                || this.commandsAllowedForAll;
     }
@@ -912,6 +1537,12 @@
         {
             EntityPlayerMP entityplayermp = this.playerEntityList.get(i);
 
+            // CraftBukkit start - Test if player receiving packet can see the source of the packet
+            if (except != null && except instanceof EntityPlayerMP && !entityplayermp.getBukkitEntity().canSee(((EntityPlayerMP) except).getBukkitEntity())) {
+                continue;
+            }
+            // CraftBukkit end
+
             if (entityplayermp != except && entityplayermp.dimension == dimension)
             {
                 double d0 = x - entityplayermp.posX;
@@ -970,7 +1601,7 @@
 
     public void updateTimeAndWeatherForPlayer(EntityPlayerMP playerIn, WorldServer worldIn)
     {
-        WorldBorder worldborder = this.server.worlds[0].getWorldBorder();
+        WorldBorder worldborder = this.server.worldServerList.get(0).getWorldBorder();
         playerIn.connection.sendPacket(new SPacketWorldBorder(worldborder, SPacketWorldBorder.Action.INITIALIZE));
         playerIn.connection
         .sendPacket(new SPacketTimeUpdate(worldIn.getTotalWorldTime(), worldIn.getWorldTime(), worldIn.getGameRules().getBoolean("doDaylightCycle")));
@@ -979,16 +1610,21 @@
 
         if (worldIn.isRaining())
         {
-            playerIn.connection.sendPacket(new SPacketChangeGameState(1, 0.0F));
-            playerIn.connection.sendPacket(new SPacketChangeGameState(7, worldIn.getRainStrength(1.0F)));
-            playerIn.connection.sendPacket(new SPacketChangeGameState(8, worldIn.getThunderStrength(1.0F)));
+            // CraftBukkit start - handle player weather
+            // playerIn.connection.sendPacket(new SPacketChangeGameState(1, 0.0F));
+            // playerIn.connection.sendPacket(new SPacketChangeGameState(7, worldIn.getRainStrength(1.0F)));
+            // playerIn.connection.sendPacket(new SPacketChangeGameState(8, worldIn.getThunderStrength(1.0F)));
+            playerIn.setPlayerWeather(org.bukkit.WeatherType.DOWNFALL, false);
+            playerIn.updateWeather(-worldIn.rainingStrength, worldIn.rainingStrength, -worldIn.thunderingStrength, worldIn.thunderingStrength);
+            // CraftBukkit end
         }
     }
 
     public void syncPlayerInventory(EntityPlayerMP playerIn)
     {
         playerIn.sendContainerToPlayer(playerIn.inventoryContainer);
-        playerIn.setPlayerHealthUpdated();
+        // playerIn.setPlayerHealthUpdated();
+        playerIn.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
         playerIn.connection.sendPacket(new SPacketHeldItemChange(playerIn.inventory.currentItem));
     }
 
@@ -1004,7 +1640,7 @@
 
     public String[] getAvailablePlayerDat()
     {
-        return this.server.worlds[0].getSaveHandler().getPlayerNBTManager().getAvailablePlayerDat();
+        return this.server.worldServerList.get(0).getSaveHandler().getPlayerNBTManager().getAvailablePlayerDat();
     }
 
     @SideOnly(Side.SERVER)
@@ -1076,9 +1712,15 @@
 
     public void removeAllPlayers()
     {
-        for (int i = 0; i < this.playerEntityList.size(); i++)
-        {
-            this.playerEntityList.get(i).connection.disconnect(new TextComponentTranslation("multiplayer.disconnect.server_shutdown"));
+        // CraftBukkit start - disconnect safely
+        for (EntityPlayerMP player : this.playerEntityList) {
+            player.connection.disconnect(this.server.server.getShutdownMessage()); // CraftBukkit - add custom shutdown message
+        }
+    }
+
+    public void sendMessage(ITextComponent[] iChatBaseComponents) {
+        for (ITextComponent component : iChatBaseComponents) {
+            sendMessage(component, true);
         }
     }
 
@@ -1086,7 +1728,10 @@
     {
         this.server.sendMessage(component);
         ChatType chattype = isSystem ? ChatType.SYSTEM : ChatType.CHAT;
-        this.sendPacketToAllPlayers(new SPacketChat(component, chattype));
+        // this.sendPacketToAllPlayers(new SPacketChat(component, chattype));
+        // CraftBukkit start - we run this through our processor first so we can get web links etc
+        this.sendPacketToAllPlayers(new SPacketChat(org.bukkit.craftbukkit.v1_12_R1.util.CraftChatMessage.fixComponent(component), chattype));
+        // CraftBukkit end
     }
 
     public void sendMessage(ITextComponent component)
@@ -1122,6 +1767,8 @@
         return statisticsmanagerserver;
     }
 
+    public StatisticsManagerServer getStatisticManager(EntityPlayerMP playerIn) { return getPlayerStatsFile(playerIn); } // CatServer - Add CraftBukkit method
+
     public PlayerAdvancements getPlayerAdvancements(EntityPlayerMP p_192054_1_)
     {
         UUID uuid = p_192054_1_.getUniqueID();
@@ -1145,7 +1792,7 @@
 
         if (this.server.worlds != null)
         {
-            for (WorldServer worldserver : this.server.worlds)
+            for (WorldServer worldserver : this.server.worldServerList)
             {
                 if (worldserver != null)
                 {
