--- before/net/minecraft/entity/EntityCreature.java
+++ after/net/minecraft/entity/EntityCreature.java
@@ -7,6 +7,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.Vec3d;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityUnleashEvent;
 
 public abstract class EntityCreature extends EntityLiving
 {
@@ -88,6 +89,7 @@
             {
                 if (f > 10.0F)
                 {
+                    this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE));
                     this.clearLeashed(true, true);
                 }
 
@@ -98,6 +100,7 @@
 
             if (f > 10.0F)
             {
+                this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE));
                 this.clearLeashed(true, true);
                 this.tasks.disableControlFlag(1);
             }
@@ -106,9 +109,9 @@
                 double d0 = (entity.posX - this.posX) / f;
                 double d1 = (entity.posY - this.posY) / f;
                 double d2 = (entity.posZ - this.posZ) / f;
-                this.motionX = this.motionX + d0 * Math.abs(d0) * 0.4;
-                this.motionY = this.motionY + d1 * Math.abs(d1) * 0.4;
-                this.motionZ = this.motionZ + d2 * Math.abs(d2) * 0.4;
+                this.motionX += d0 * Math.abs(d0) * 0.4;
+                this.motionY += d1 * Math.abs(d1) * 0.4;
+                this.motionZ += d2 * Math.abs(d2) * 0.4;
             }
             else
             {
