--- before/net/minecraft/entity/EntityLeashKnot.java
+++ after/net/minecraft/entity/EntityLeashKnot.java
@@ -54,6 +54,7 @@
         this.posX = this.hangingPosition.getX() + 0.5;
         this.posY = this.hangingPosition.getY() + 0.5;
         this.posZ = this.hangingPosition.getZ() + 0.5;
+        if (this.isAddedToWorld() && !this.world.isRemote) this.world.updateEntityWithOptionalForce(this, false); // Forge - Process chunk registration after moving.
     }
 
     @Override
@@ -136,6 +137,10 @@
             {
                 if (entityliving.getLeashed() && entityliving.getLeashHolder() == player)
                 {
+                    if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerLeashEntityEvent(entityliving, this, player).isCancelled()) {
+                        ((net.minecraft.entity.player.EntityPlayerMP) player).connection.sendPacket(new net.minecraft.network.play.server.SPacketEntityAttach(entityliving, entityliving.getLeashHolder()));
+                        continue;
+                    }
                     entityliving.setLeashHolder(this, true);
                     flag = true;
                 }
@@ -143,16 +148,24 @@
 
             if (!flag)
             {
-                this.setDead();
-
-                if (player.capabilities.isCreativeMode)
-                {
+                // CraftBukkit start - Move below
+                // this.setDead();
+                if (true || player.capabilities.isCreativeMode) { // CraftBukkit - Process for non-creative as well
+                boolean die = true;
                     for (EntityLiving entityliving1 : list)
                     {
                         if (entityliving1.getLeashed() && entityliving1.getLeashHolder() == this)
                         {
-                            entityliving1.clearLeashed(true, false);
+                            // entityliving1.clearLeashed(true, false);
+                            if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPlayerUnleashEntityEvent(entityliving1, player).isCancelled()) {
+                                die = false;
+                                continue;
+                            }
+                            entityliving1.clearLeashed(true, !player.capabilities.isCreativeMode); // false -> survival mode boolean
                         }
+                    }
+                    if (die) {
+                        this.setDead();
                     }
                 }
             }
