--- before/net/minecraft/command/server/CommandWhitelist.java
+++ after/net/minecraft/command/server/CommandWhitelist.java
@@ -69,7 +69,7 @@
                     throw new WrongUsageException("commands.whitelist.add.usage");
                 }
 
-                GameProfile gameprofile = server.getPlayerProfileCache().getGameProfileForUsername(args[1]);
+                GameProfile gameprofile = server.isServerInOnlineMode() ? server.getPlayerProfileCache().getGameProfileForUsername(args[1]) : new GameProfile(net.minecraft.entity.player.EntityPlayer.getOfflineUUID(args[1]), args[1]); // CatServer - offline mode support
 
                 if (gameprofile == null)
                 {
