--- before/net/minecraft/server/MinecraftServer.java
+++ after/net/minecraft/server/MinecraftServer.java
@@ -88,22 +88,22 @@
 
 public abstract class MinecraftServer implements ICommandSender, Runnable, IThreadListener, ISnooperInfo
 {
-    private static final Logger LOGGER = LogManager.getLogger();
+    public static final Logger LOGGER = LogManager.getLogger();
     public static final File USER_CACHE_FILE = new File("usercache.json");
-    private final ISaveFormat anvilConverterForAnvilFile;
+    public ISaveFormat anvilConverterForAnvilFile;
     private final Snooper usageSnooper = new Snooper("server", this, getCurrentTimeMillis());
-    private final File anvilFile;
+    public File anvilFile;
     private final List<ITickable> tickables = Lists.newArrayList();
     public final ICommandManager commandManager;
     public final Profiler profiler = new Profiler();
     private final NetworkSystem networkSystem;
     private final ServerStatusResponse statusResponse = new ServerStatusResponse();
     private final Random random = new Random();
-    private final DataFixer dataFixer;
+    public final DataFixer dataFixer;
     @SideOnly(Side.SERVER)
     private String hostname;
     private int serverPort = -1;
-    public WorldServer[] worlds;
+    public WorldServer[] worlds = new WorldServer[0];
     private PlayerList playerList;
     private boolean serverRunning = true;
     private boolean serverStopped;
@@ -121,7 +121,8 @@
     private int buildLimit;
     private int maxPlayerIdleMinutes;
     public final long[] tickTimeArray = new long[100];
-    public long[][] timeOfLastDimensionTick;
+    //public long[][] timeOfLastDimensionTick;
+    public it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<long[]> worldTickTimes = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>();
     private KeyPair serverKeyPair;
     private String serverOwner;
     private String folderName;
@@ -141,14 +142,34 @@
     private final GameProfileRepository profileRepo;
     private final PlayerProfileCache profileCache;
     private long nanoTimeSinceStatusRefresh;
-    public final Queue < FutureTask<? >> futureTaskQueue = Queues.newArrayDeque();
+    public final Queue < FutureTask<? >> futureTaskQueue = new catserver.server.utils.CachedSizeConcurrentLinkedQueue<>(); // Paper - Make size() constant-time
     private Thread serverThread;
     protected long currentTime = getCurrentTimeMillis();
     @SideOnly(Side.CLIENT)
     private boolean worldIconSet;
+    private boolean hasStopped = false;
+    private final Object stopLock = new Object();
+    // CraftBukkit start
+    public List<WorldServer> worldServerList = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+    public org.bukkit.craftbukkit.v1_12_R1.CraftServer server;
+    public joptsimple.OptionSet options;
+    public org.bukkit.command.ConsoleCommandSender console;
+    public org.bukkit.command.RemoteConsoleCommandSender remoteConsole;
+    // public ConsoleReader reader; // Paper
+    public static int currentTick = 0; // Paper - Further improve tick loop
+    public Thread primaryThread;
+    public Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<>(); // TODO: CREF - look for jctools
+    public int autosavePeriod;
+    // CraftBukkit end
+    // Spigot start
+    public final org.spigotmc.SlackActivityAccountant slackActivityAccountant = new org.spigotmc.SlackActivityAccountant();
+    // Spigot end
+    //
+    private static long lastTickNanos = System.nanoTime();
+    public static long realTimeTicks = 1;
 
     public MinecraftServer(
-        File anvilFileIn,
+        joptsimple.OptionSet options,
         Proxy proxyIn,
         DataFixer dataFixerIn,
         YggdrasilAuthenticationService authServiceIn,
@@ -162,12 +183,45 @@
         this.sessionService = sessionServiceIn;
         this.profileRepo = profileRepoIn;
         this.profileCache = profileCacheIn;
-        this.anvilFile = anvilFileIn;
+        // this.anvilFile = anvilFileIn;
         this.networkSystem = new NetworkSystem(this);
         this.commandManager = this.createCommandManager();
-        this.anvilConverterForAnvilFile = new AnvilSaveConverter(anvilFileIn, dataFixerIn);
+        // this.anvilConverterForAnvilFile = new AnvilSaveConverter(anvilFileIn, dataFixerIn); // CraftBukkit - moved to DedicatedServer.init
         this.dataFixer = dataFixerIn;
+        this.options = options;
+        // Paper start - Handled by TerminalConsoleAppender
+        /*
+        // Try to see if we're actually running in a terminal, disable jline if not
+        if (System.console() == null && System.getProperty("jline.terminal") == null) {
+            System.setProperty("jline.terminal", "jline.UnsupportedTerminal");
+            Main.useJline = false;
+        }
+
+        try {
+            reader = new ConsoleReader(System.in, System.out);
+            reader.setExpandEvents(false); // Avoid parsing exceptions for uncommonly used event designators
+        } catch (Throwable e) {
+            try {
+                // Try again with jline disabled for Windows users without C++ 2008 Redistributable
+                System.setProperty("jline.terminal", "jline.UnsupportedTerminal");
+                System.setProperty("user.language", "en");
+                Main.useJline = false;
+                reader = new ConsoleReader(System.in, System.out);
+                reader.setExpandEvents(false);
+            } catch (IOException ex) {
+                LOGGER.warn((String) null, ex);
+            }
+        }
+        */
+        // Paper end
+        Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.v1_12_R1.util.ServerShutdownThread(this));
+        this.serverThread = this.primaryThread = new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, this, "Server thread"); // Moved from main
+        // CatRoom Start
+        this.serverThread.setPriority(Thread.MAX_PRIORITY);
+        // CatRoom End
     }
+
+    public abstract net.minecraft.server.dedicated.PropertyManager getPropertyManager();
 
     public ServerCommandManager createCommandManager()
     {
@@ -184,7 +238,7 @@
             this.setUserMessage("menu.convertingLevel");
             this.getActiveAnvilConverter().convertMapFormat(worldNameIn, new IProgressUpdate()
             {
-                private long startTime = MinecraftServer.getCurrentTimeMillis();
+                private long startTime = System.currentTimeMillis();
                 @Override
                 public void displaySavingString(String message)
                 {
@@ -197,9 +251,9 @@
                 @Override
                 public void setLoadingProgress(int progress)
                 {
-                    if (MinecraftServer.getCurrentTimeMillis() - this.startTime >= 1000L)
+                    if (System.currentTimeMillis() - this.startTime >= 1000L)
                     {
-                        this.startTime = MinecraftServer.getCurrentTimeMillis();
+                        this.startTime = System.currentTimeMillis();
                         MinecraftServer.LOGGER.info("Converting... {}%", progress);
                     }
                 }
@@ -231,83 +285,101 @@
 
     public void loadAllWorlds(String saveName, String worldNameIn, long seed, WorldType type, String generatorOptions)
     {
+        ServerCommandManager vanillaCommandManager = (ServerCommandManager) this.getCommandManager();
+        vanillaCommandManager.registerVanillaCommands();
         this.convertMapIfNeeded(saveName);
         this.setUserMessage("menu.loadingLevel");
-        this.worlds = new WorldServer[3];
-        this.timeOfLastDimensionTick = new long[this.worlds.length][100];
-        ISaveHandler isavehandler = this.anvilConverterForAnvilFile.getSaveLoader(saveName, true);
-        this.setResourcePackFromWorld(this.getFolderName(), isavehandler);
-        WorldInfo worldinfo = isavehandler.loadWorldInfo();
-        WorldSettings worldsettings;
-
-        if (worldinfo == null)
-        {
-            if (this.isDemo())
-            {
-                worldsettings = WorldServerDemo.DEMO_WORLD_SETTINGS;
-            }
-            else
-            {
-                worldsettings = new WorldSettings(seed, this.getGameType(), this.canStructuresSpawn(), this.isHardcore(), type);
-                worldsettings.setGeneratorOptions(generatorOptions);
-
-                if (this.enableBonusChest)
-                {
-                    worldsettings.enableBonusChest();
-                }
-            }
-
-            worldinfo = new WorldInfo(worldsettings, worldNameIn);
-        }
-        else
-        {
-            worldinfo.setWorldName(worldNameIn);
-            worldsettings = new WorldSettings(worldinfo);
-        }
-
-        for (int i = 0; i < this.worlds.length; i++)
-        {
-            int j = 0;
-
-            if (i == 1)
-            {
-                j = -1;
-            }
-
-            if (i == 2)
-            {
-                j = 1;
-            }
-
-            if (i == 0)
-            {
-                if (this.isDemo())
-                {
-                    this.worlds[i] = (WorldServer)new WorldServerDemo(this, isavehandler, worldinfo, j, this.profiler).init();
-                }
-                else
-                {
-                    this.worlds[i] = (WorldServer)new WorldServer(this, isavehandler, worldinfo, j, this.profiler).init();
-                }
-
-                this.worlds[i].initialize(worldsettings);
-            }
-            else
-            {
-                this.worlds[i] = (WorldServer)new WorldServerMulti(this, isavehandler, j, this.worlds[0], this.profiler).init();
-            }
-
-            this.worlds[i].addEventListener(new ServerWorldEventHandler(this, this.worlds[i]));
+
+        WorldSettings worldsettings = new WorldSettings(seed, this.getGameType(), this.canStructuresSpawn(), this.isHardcore(), type);
+        worldsettings.setGeneratorOptions(generatorOptions);
+        WorldServer world;
+
+        // WorldServer overWorld = (WorldServer)(isDemo() ? new WorldServerDemo(this, new AnvilSaveHandler(server.getWorldContainer(), worldNameIn , true, this.dataFixer), worldinfo, 0, profiler).init() : new WorldServer(this, new AnvilSaveHandler(server.getWorldContainer(), worldNameIn , true, this.dataFixer), worldinfo, 0, profiler).init());
+        // TODO: Reimplement this!
+        Integer[] dimIds = net.minecraftforge.common.DimensionManager.getStaticDimensionIDs();
+        Arrays.sort(dimIds, (o1, o2) -> {
+            // Zero-dimension must always be the first in array!
+            if (o1 == 0) {
+                return -1;
+            } else {
+                return Math.max(o1, o2);
+            }
+        });
+        for (int dim : dimIds)
+        {
+            // World validation
+            if (dim != 0) {
+                if ((dim == -1 && !this.getAllowNether()) || (dim == 1 && !server.getAllowEnd())) {
+                    continue;
+                }
+            }
+
+            org.bukkit.World.Environment worldEnvironment = org.bukkit.World.Environment.getEnvironment(dim);
+            if (worldEnvironment == null)
+                worldEnvironment = org.bukkit.World.Environment.getEnvironment(net.minecraftforge.common.DimensionManager.getProviderType(dim).getId());
+            String name = (dim == 0) ? saveName : "DIM" + dim;
+            org.bukkit.generator.ChunkGenerator gen = null;
+
+            if (dim == 0)
+            {
+                ISaveHandler idatamanager = new net.minecraft.world.chunk.storage.AnvilSaveHandler(server.getWorldContainer(), worldNameIn, true, this.dataFixer);
+                WorldInfo worlddata = idatamanager.loadWorldInfo();
+                // CatServer start
+                if (!catserver.server.BukkitInjector.initializedBukkit) { // CatServer - inject bukkit materials before plugins load
+                    catserver.server.BukkitInjector.injectBlockBukkitMaterials();
+                    catserver.server.BukkitInjector.injectItemBukkitMaterials();
+                    catserver.server.BukkitInjector.injectBiomes();
+                    catserver.server.BukkitInjector.injectEntityType();
+                    catserver.server.BukkitInjector.registerEnchantments();
+                    catserver.server.BukkitInjector.registerPotions();
+                    catserver.server.BukkitInjector.registerBannerPatterns();
+                    catserver.server.BukkitInjector.initializedBukkit = true;
+                }
+                server.loadPlugins();
+                server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+                // CatServer end
+                if (worlddata == null) {
+                    worlddata = new WorldInfo(worldsettings, worldNameIn);
+                }
+                worlddata.checkName(worldNameIn); // CraftBukkit - Migration did not rewrite the level.dat; This forces 1.8 to take the last loaded world as respawn (in this case the end)
+                if (this.isDemo()) {
+                    world = (WorldServer) (new WorldServerDemo(this, idatamanager, worlddata, dim, this.profiler)).init();
+                } else {
+                    world = (WorldServer) (new WorldServer(this, idatamanager, worlddata, dim, this.profiler, worldEnvironment, gen)).init();
+                }
+
+                world.initialize(worldsettings);
+                this.server.scoreboardManager = new org.bukkit.craftbukkit.v1_12_R1.scoreboard.CraftScoreboardManager(this, world.getScoreboard());
+            } else {
+                gen = this.server.getGenerator(name);
+
+                ISaveHandler idatamanager = new net.minecraft.world.chunk.storage.AnvilSaveHandler(server.getWorldContainer(), name, true, this.dataFixer);
+                // world =, b0 to dimension, s1 to name, added Environment and gen
+                WorldInfo worlddata = idatamanager.loadWorldInfo();
+                if (worlddata == null) {
+                    worlddata = new WorldInfo(worldsettings, name);
+                }
+                worlddata.checkName(name); // CraftBukkit - Migration did not rewrite the level.dat; This forces 1.8 to take the last loaded world as respawn (in this case the end)
+                world = (WorldServer) new WorldServerMulti(this, idatamanager, dim, this.worldServerList.get(0), this.profiler, worlddata, worldEnvironment, gen).init();
+                worlddata.setServerInitialized(true);
+            }
+            this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(world.getWorld()));
+            world.addEventListener(new ServerWorldEventHandler(this, world));
 
             if (!this.isSinglePlayer())
             {
-                this.worlds[i].getWorldInfo().setGameType(this.getGameType());
+                world.getWorldInfo().setGameType(this.getGameType());
             }
+            getPlayerList().setPlayerManager(worldServerList.toArray(new WorldServer[worldServerList.size()]));
+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(world));
         }
 
         this.playerList.setPlayerManager(this.worlds);
         this.setDifficultyForAllWorlds(this.getDifficulty());
         this.initialWorldChunkLoad();
+
+        this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        catserver.server.CatServer.onServerStart();
     }
 
     public void initialWorldChunkLoad()
@@ -320,7 +392,7 @@
         this.setUserMessage("menu.generatingTerrain");
         int j1 = 0;
         LOGGER.info("Preparing start region for level 0");
-        WorldServer worldserver = this.worlds[0];
+        WorldServer worldserver = net.minecraftforge.common.DimensionManager.getWorld(j1);
         BlockPos blockpos = worldserver.getSpawnPoint();
         long k1 = getCurrentTimeMillis();
 
@@ -341,6 +413,7 @@
             }
         }
 
+        catroom.CatRoom.callBukkitWorldLoadEvent(this.server, this.worldServerList); // fix Nether-API
         this.clearCurrentTask();
     }
 
@@ -390,7 +463,7 @@
 
     public void saveAllWorlds(boolean isSilent)
     {
-        for (WorldServer worldserver : this.worlds)
+        for (WorldServer worldserver : this.worldServerList)
         {
             if (worldserver != null)
             {
@@ -415,8 +488,19 @@
 
     public void stopServer()
     {
+        org.spigotmc.AsyncCatcher.enabled = false; // Spigot
+        // CraftBukkit start - prevent double stopping on multiple threads
+        synchronized(stopLock) {
+            if (hasStopped) return;
+            hasStopped = true;
+        }
+        // CraftBukkit end
         LOGGER.info("Stopping server");
 
+        if (this.server != null) {
+            this.server.disablePlugins();
+        }
+
         if (this.getNetworkSystem() != null)
         {
             this.getNetworkSystem().terminateEndpoints();
@@ -427,6 +511,7 @@
             LOGGER.info("Saving players");
             this.playerList.saveAllPlayerData();
             this.playerList.removeAllPlayers();
+            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
         }
 
         if (this.worlds != null)
@@ -443,19 +528,38 @@
 
             this.saveAllWorlds(false);
 
-            for (WorldServer worldserver1 : this.worlds)
+            // CraftBukkit start - Handled in saveChunks
+            for (WorldServer worldserver1 : this.worldServerList)
             {
                 if (worldserver1 != null)
                 {
+                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Unload(worldserver1));
                     worldserver1.flush();
                 }
             }
+            // CraftBukkit end
+
+            WorldServer[] tmp = worlds;
+            for (WorldServer world : tmp)
+            {
+                net.minecraftforge.common.DimensionManager.setWorld(world.provider.getDimension(), null, this);
+            }
         }
 
         if (this.usageSnooper.isSnooperRunning())
         {
             this.usageSnooper.stopSnooper();
         }
+
+        CommandBase.setCommandListener(null); // Forge: fix MC-128561
+
+        // Spigot start
+        if (org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) {
+            LOGGER.info("Saving usercache.json");
+            this.profileCache.save(false); // Paper
+        }
+        // Spigot end
+        catserver.server.CatServer.onServerStop();
     }
 
     @SideOnly(Side.SERVER)
@@ -470,6 +574,8 @@
         this.hostname = host;
     }
 
+    public boolean isServerStopping() { return this.hasStopped; } // CatServer
+
     public boolean isServerRunning()
     {
         return this.serverRunning;
@@ -480,65 +586,136 @@
         this.serverRunning = false;
     }
 
+    // Paper start - Further improve server tick loop
+    private static final int TPS = 20;
+    private static final long SEC_IN_NANO = 1000000000;
+    public static final long TICK_TIME = SEC_IN_NANO / TPS;
+    private static final long MAX_CATCHUP_BUFFER = TICK_TIME * TPS * 60L;
+    private static final int SAMPLE_INTERVAL = 20;
+    public final RollingAverage tps1 = new RollingAverage(60);
+    public final RollingAverage tps5 = new RollingAverage(60 * 5);
+    public final RollingAverage tps15 = new RollingAverage(60 * 15);
+    public double[] recentTps = new double[3]; // Paper - Fine have your darn compat with bad plugins
+
+    public static class RollingAverage {
+        private final int size;
+        private long time;
+        private double total;
+        private int index = 0;
+        private final double[] samples;
+        private final long[] times;
+
+        RollingAverage(int size) {
+            this.size = size;
+            this.time = size * SEC_IN_NANO;
+            this.total = TPS * SEC_IN_NANO * size;
+            this.samples = new double[size];
+            this.times = new long[size];
+            for (int i = 0; i < size; i++) {
+                this.samples[i] = TPS;
+                this.times[i] = SEC_IN_NANO;
+            }
+        }
+
+        public synchronized void add(double x, long t) {
+            time -= times[index];
+            total -= samples[index] * times[index];
+            samples[index] = x;
+            times[index] = t;
+            time += t;
+            total += x * t;
+            if (++index == size) {
+                index = 0;
+            }
+        }
+
+        public synchronized double getAverage() {
+            return total / time;
+        }
+    }
+    // Paper End
+
     @Override
     public void run()
     {
         try
         {
+            // CatRoom Start
+            catserver.server.CatServerConfig config = catserver.server.CatServer.getConfig();
+            config.loadConfig();
+            Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
+            // CatRoom End
             if (this.init())
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStarted();
                 this.currentTime = getCurrentTimeMillis();
-                long i = 0L;
                 this.statusResponse.setServerDescription(new TextComponentString(this.motd));
                 this.statusResponse.setVersion(new ServerStatusResponse.Version("1.12.2", 340));
                 this.applyServerIconToResponse(this.statusResponse);
 
+                // Spigot start
+                Arrays.fill( recentTps, 20 );
+                long start = System.nanoTime(), lastTick = start - TICK_TIME, catchupTime = 0, curTime, wait, tickSection = start; // Paper - Further improve server tick loop
                 while (this.serverRunning)
                 {
-                    long k = getCurrentTimeMillis();
-                    long j = k - this.currentTime;
-
-                    if (j > 2000L && this.currentTime - this.timeOfLastWarning >= 15000L)
-                    {
-                        LOGGER.warn(
-                            "Can't keep up! Did the system time change, or is the server overloaded? Running {}ms behind, skipping {} tick(s)", j, j / 50L
-                        );
-                        j = 2000L;
-                        this.timeOfLastWarning = this.currentTime;
-                    }
-
-                    if (j < 0L)
-                    {
-                        LOGGER.warn("Time ran backwards! Did the system time change?");
-                        j = 0L;
-                    }
-
-                    i += j;
-                    this.currentTime = k;
-
-                    if (this.worlds[0].areAllPlayersAsleep())
-                    {
-                        this.tick();
-                        i = 0L;
-                    }
-                    else
-                    {
-                        while (i > 50L)
-                        {
-                            i -= 50L;
-                            this.tick();
+                    curTime = System.nanoTime();
+                    // Paper start - Further improve server tick loop
+                    wait = TICK_TIME - (curTime - lastTick);
+                    if (wait > 0) {
+                        if (catchupTime < 2E6) {
+                            wait += Math.abs(catchupTime);
+                        } else if (wait < catchupTime) {
+                            catchupTime -= wait;
+                            wait = 0;
+                        } else {
+                            wait -= catchupTime;
+                            catchupTime = 0;
                         }
                     }
-
-                    Thread.sleep(Math.max(1L, 50L - i));
+                    if (wait > 0) {
+                        Thread.sleep(wait / 1000000);
+                        curTime = System.nanoTime();
+                        wait = TICK_TIME - (curTime - lastTick);
+                    }
+
+                    catchupTime = Math.min(MAX_CATCHUP_BUFFER, catchupTime - wait);
+                    if ( ++MinecraftServer.currentTick % SAMPLE_INTERVAL == 0 )
+                    {
+                        final long diff = curTime - tickSection;
+                        double currentTps = 1E9 / diff * SAMPLE_INTERVAL;
+                        tps1.add(currentTps, diff);
+                        tps5.add(currentTps, diff);
+                        tps15.add(currentTps, diff);
+                        // Backwards compat with bad plugins
+                        recentTps[0] = tps1.getAverage();
+                        recentTps[1] = tps5.getAverage();
+                        recentTps[2] = tps15.getAverage();
+                        // Paper end
+                        tickSection = curTime;
+                    }
+                    lastTick = curTime;
+
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callServerTickStartEvent(this.tickCounter + 1); // Paper - Server Tick Events
+                    this.tick();
+                    org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callServerTickEndEvent(this.tickCounter, lastTick, catchupTime); // Paper - Server Tick Events
+
                     this.serverIsRunning = true;
                 }
+                // Spigot end
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStopping();
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
             }
             else
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
                 this.finalTick(null);
             }
         }
+        catch (net.minecraftforge.fml.common.StartupQuery.AbortedException e)
+        {
+            // ignore silently
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
+        }
         catch (Throwable throwable1)
         {
             LOGGER.error("Encountered an unexpected exception", throwable1);
@@ -566,13 +743,13 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.minecraftforge.fml.common.FMLCommonHandler.instance().expectServerStopped(); // has to come before finalTick to avoid race conditions
             this.finalTick(crashreport);
         }
         finally
         {
             try
             {
-                this.serverStopped = true;
                 this.stopServer();
             }
             catch (Throwable throwable)
@@ -581,6 +758,8 @@
             }
             finally
             {
+                net.minecraftforge.fml.common.FMLCommonHandler.instance().handleServerStopped();
+                this.serverStopped = true;
                 this.systemExitNow();
             }
         }
@@ -607,6 +786,7 @@
                 ImageIO.write(bufferedimage, "PNG", new ByteBufOutputStream(bytebuf));
                 ByteBuf bytebuf1 = Base64.encode(bytebuf);
                 response.setFavicon("data:image/png;base64," + bytebuf1.toString(StandardCharsets.UTF_8));
+                bytebuf1.release(); // Forge: fix MC-122085
             }
             catch (Exception exception)
             {
@@ -647,7 +827,18 @@
 
     public void tick()
     {
+        if (catserver.server.CatServer.getConfig().enableRealtime) {
+            long currentNanos = System.nanoTime();
+            realTimeTicks = (currentNanos - lastTickNanos) / 50000000;
+            if (realTimeTicks < 1) {
+                realTimeTicks = 1;
+            }
+            lastTickNanos = currentNanos;
+        }
+
+        this.slackActivityAccountant.tickStarted(); // Spigot
         long i = System.nanoTime();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onPreServerTick();
         this.tickCounter++;
 
         if (this.startProfiling)
@@ -674,15 +865,24 @@
 
             Collections.shuffle(Arrays.asList(agameprofile));
             this.statusResponse.getPlayers().setPlayers(agameprofile);
+            this.statusResponse.invalidateJson();
         }
 
-        if (this.tickCounter % 900 == 0)
+        if (autosavePeriod > 0 && this.tickCounter % autosavePeriod == 0) // CraftBukkit
         {
             this.profiler.startSection("save");
             this.playerList.saveAllPlayerData();
-            this.saveAllWorlds(true);
-            this.profiler.endSection();
-        }
+            catserver.server.WeakWorldSaveManager.saveAllWorlds(); // this.saveAllWorlds(true); // CatServer - Don't save all worlds at once
+            this.profiler.endSection();
+        }
+
+        // CatServer start
+        if (catserver.server.WeakWorldSaveManager.isNeedTick()) {
+            this.profiler.startSection("save");
+            catserver.server.WeakWorldSaveManager.onTick(); // CatServer - Save a world per tick
+            this.profiler.endSection();
+        }
+        // CatServer end
 
         this.profiler.startSection("tallying");
         this.tickTimeArray[this.tickCounter % 100] = System.nanoTime() - i;
@@ -701,21 +901,43 @@
 
         this.profiler.endSection();
         this.profiler.endSection();
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().onPostServerTick();
+        org.spigotmc.WatchdogThread.tick(); // Spigot
+        this.slackActivityAccountant.tickEnded(System.nanoTime() - i); // Spigot
     }
 
     public void updateTimeLightAndEntities()
     {
+        this.server.getScheduler().mainThreadHeartbeat(this.tickCounter); // CraftBukkit
         this.profiler.startSection("jobs");
 
-        synchronized (this.futureTaskQueue)
-        {
-            while (!this.futureTaskQueue.isEmpty())
-            {
-                Util.runTask(this.futureTaskQueue.poll(), LOGGER);
-            }
-        }
+        // Spigot start
+        FutureTask<?> entry;
+        int count = this.futureTaskQueue.size();
+        while (count-- > 0 && (entry = this.futureTaskQueue.poll()) != null) {
+            Util.runTask(entry, MinecraftServer.LOGGER);
+         }
+        // Spigot end
 
         this.profiler.endStartSection("levels");
+        // CraftBukkit start
+        // Run tasks that are waiting on processing
+        while (!processQueue.isEmpty()) {
+            processQueue.remove().run();
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.chunkio.ChunkIOExecutor.tick();
+
+        // Send time updates to everyone, it will get the right time from the world the player is in.
+        if (this.tickCounter % 20 == 0) {
+            for (int i = 0; i < this.getPlayerList().getPlayers().size(); ++i) {
+                net.minecraft.entity.player.EntityPlayerMP entityplayer = this.getPlayerList().getPlayers().get(i);
+                entityplayer.connection.sendPacket(new SPacketTimeUpdate(entityplayer.world.getTotalWorldTime(), entityplayer.getPlayerTime(), entityplayer.world.getGameRules().getBoolean("doDaylightCycle"))); // Add support for per player time
+            }
+        }
+        net.minecraftforge.common.chunkio.ChunkIOExecutor.tick();
+        catserver.server.command.internal.CommandChunkStats.onServerTick();
+        /*
 
         for (int j = 0; j < this.worlds.length; j++)
         {
@@ -770,7 +992,11 @@
 
             this.timeOfLastDimensionTick[j][this.tickCounter % 100] = System.nanoTime() - i;
         }
-
+        */
+
+        catroom.voidstudio.VoidStudio.worldTick(this.tickCounter, this.worldTickTimes);
+
+        this.profiler.endStartSection("dim_unloading");
         this.profiler.endStartSection("connection");
         this.getNetworkSystem().networkTick();
         this.profiler.endStartSection("players");
@@ -801,6 +1027,20 @@
     @SideOnly(Side.SERVER)
     public static void main(String[] p_main_0_)
     {
+        joptsimple.OptionSet options = org.bukkit.craftbukkit.Main.main(p_main_0_);
+        if (options == null)
+            return;
+        catserver.server.utils.ServerUtils.acceptEula(); // CatServer
+        //Forge: Copied from DedicatedServer.init as to run as early as possible, Old code left in place intentionally.
+        //Done in good faith with permission: https://github.com/MinecraftForge/MinecraftForge/issues/3659#issuecomment-390467028
+        ServerEula eula = new ServerEula(new File("eula.txt"));
+        if (!eula.hasAcceptedEULA())
+        {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            eula.createEULAFile();
+            return;
+        }
+
         Bootstrap.register();
 
         try
@@ -808,110 +1048,31 @@
             boolean flag = true;
             String s = null;
             String s1 = ".";
-            String s2 = null;
-            boolean flag1 = false;
-            boolean flag2 = false;
-            int i = -1;
-
-            for (int j = 0; j < p_main_0_.length; j++)
-            {
-                String s3 = p_main_0_[j];
-                String s4 = j == p_main_0_.length - 1 ? null : p_main_0_[j + 1];
-                boolean flag3 = false;
-
-                if ("nogui".equals(s3) || "--nogui".equals(s3))
-                {
-                    flag = false;
-                }
-                else if ("--port".equals(s3) && s4 != null)
-                {
-                    flag3 = true;
-
-                    try
-                    {
-                        i = Integer.parseInt(s4);
-                    }
-                    catch (NumberFormatException numberformatexception)
-                    {
-                    }
-                }
-                else if ("--singleplayer".equals(s3) && s4 != null)
-                {
-                    flag3 = true;
-                    s = s4;
-                }
-                else if ("--universe".equals(s3) && s4 != null)
-                {
-                    flag3 = true;
-                    s1 = s4;
-                }
-                else if ("--world".equals(s3) && s4 != null)
-                {
-                    flag3 = true;
-                    s2 = s4;
-                }
-                else if ("--demo".equals(s3))
-                {
-                    flag1 = true;
-                }
-                else if ("--bonusChest".equals(s3))
-                {
-                    flag2 = true;
-                }
-
-                if (flag3)
-                {
-                    j++;
-                }
-            }
-
-            YggdrasilAuthenticationService yggdrasilauthenticationservice = new YggdrasilAuthenticationService(Proxy.NO_PROXY, UUID.randomUUID().toString());
+            YggdrasilAuthenticationService yggdrasilauthenticationservice = catserver.server.CatServer.getConfig().disableUpdateGameProfile ? new catserver.server.mcauth.CatProxyAuthenticationService(Proxy.NO_PROXY, UUID.randomUUID().toString()) : new YggdrasilAuthenticationService(Proxy.NO_PROXY, UUID.randomUUID().toString());;
             MinecraftSessionService minecraftsessionservice = yggdrasilauthenticationservice.createMinecraftSessionService();
             GameProfileRepository gameprofilerepository = yggdrasilauthenticationservice.createProfileRepository();
             PlayerProfileCache playerprofilecache = new PlayerProfileCache(gameprofilerepository, new File(s1, USER_CACHE_FILE.getName()));
             final DedicatedServer dedicatedserver = new DedicatedServer(
-                new File(s1), DataFixesManager.createFixer(), yggdrasilauthenticationservice, minecraftsessionservice, gameprofilerepository, playerprofilecache
+                options, DataFixesManager.createFixer(), yggdrasilauthenticationservice, minecraftsessionservice, gameprofilerepository, playerprofilecache
             );
 
-            if (s != null)
-            {
-                dedicatedserver.setServerOwner(s);
-            }
-
-            if (s2 != null)
-            {
-                dedicatedserver.setFolderName(s2);
-            }
-
-            if (i >= 0)
-            {
-                dedicatedserver.setServerPort(i);
-            }
-
-            if (flag1)
-            {
-                dedicatedserver.setDemo(true);
-            }
-
-            if (flag2)
-            {
-                dedicatedserver.canCreateBonusChest(true);
-            }
-
-            if (flag && !GraphicsEnvironment.isHeadless())
-            {
-                dedicatedserver.setGuiEnabled();
-            }
-
-            dedicatedserver.startServerThread();
-            Runtime.getRuntime().addShutdownHook(new Thread("Server Shutdown Thread")
-            {
-                @Override
-                public void run()
-                {
-                    dedicatedserver.stopServer();
+            if (options.has("port"))
+            {
+                int port = (Integer) options.valueOf("port");
+                if (port > 0) {
+                    dedicatedserver.setServerPort(port);
                 }
-            });
+            }
+
+            if (options.has("universe")) {
+                dedicatedserver.anvilFile = (File) options.valueOf("universe");
+            }
+
+            if (options.has("world")) {
+                dedicatedserver.setWorldName((String) options.valueOf("world"));
+            }
+
+            dedicatedserver.primaryThread.start();
         }
         catch (Exception exception)
         {
@@ -921,8 +1082,12 @@
 
     public void startServerThread()
     {
-        this.serverThread = new Thread(this, "Server thread");
+        if (this instanceof MinecraftServer) return; // CatServer - Don't remove the code, or it will break asm of the fucking coremod, so just return always
+        // /* CraftBukkit start - prevent abuse
+        net.minecraftforge.fml.common.StartupQuery.reset();
+        this.serverThread = new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, this, "Server thread");
         this.serverThread.start();
+        // // CraftBukkit end */
     }
 
     public File getFile(String fileName)
@@ -943,14 +1108,18 @@
 
     public WorldServer getWorld(int dimension)
     {
-        if (dimension == -1)
-        {
-            return this.worlds[1];
-        }
-        else
-        {
-            return dimension == 1 ? this.worlds[2] : this.worlds[0];
-        }
+        WorldServer ret = net.minecraftforge.common.DimensionManager.getWorld(dimension, true);
+        if (ret == null)
+        {
+            net.minecraftforge.common.DimensionManager.initDimension(dimension);
+            ret = net.minecraftforge.common.DimensionManager.getWorld(dimension);
+        }
+        return ret;
+    }
+
+    public WorldServer getWorldServer(int i) {
+        WorldServer world = getWorld(i);
+        return world != null ? world : worldServerList.get(0);
     }
 
     public String getMinecraftVersion()
@@ -981,7 +1150,8 @@
     @SideOnly(Side.SERVER)
     public boolean isDebuggingEnabled()
     {
-        return false;
+        // return false;
+        return this.getPropertyManager().getBooleanProperty("debug", false); // CraftBukkit - don't hardcode
     }
 
     @SideOnly(Side.SERVER)
@@ -1001,7 +1171,8 @@
 
     public String getServerModName()
     {
-        return "vanilla";
+        // TODO: Should we change this for CraftBukkit's server name?
+        return net.minecraftforge.fml.common.FMLCommonHandler.instance().getModName();
     }
 
     public CrashReport addServerInfoToCrashReport(CrashReport report)
@@ -1038,7 +1209,7 @@
 
     public List<String> getTabCompletions(ICommandSender sender, String input, @Nullable BlockPos pos, boolean hasTargetBlock)
     {
-        List<String> list = Lists.newArrayList();
+        it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<String> completionsSet = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(server.tabComplete(sender, input, pos, hasTargetBlock));
         boolean flag = input.startsWith("/");
 
         if (flag)
@@ -1055,11 +1226,9 @@
             {
                 if (CommandBase.doesStringStartWith(s2, s1))
                 {
-                    list.add(s1);
+                    completionsSet.add(s1);
                 }
             }
-
-            return list;
         }
         else
         {
@@ -1072,22 +1241,24 @@
                 {
                     if (flag1 && !hasTargetBlock)
                     {
-                        list.add("/" + s);
+                        completionsSet.add("/" + s);
                     }
                     else
                     {
-                        list.add(s);
+                        completionsSet.add(s);
                     }
                 }
             }
-
-            return list;
         }
+        List<String> finalCompletionsList = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(completionsSet);
+        finalCompletionsList.sort(null);
+        return finalCompletionsList;
     }
 
     public boolean isAnvilFileSet()
     {
-        return this.anvilFile != null;
+        // return this.anvilFile != null;
+        return true; // CraftBukkit
     }
 
     @Override
@@ -1099,7 +1270,8 @@
     @Override
     public void sendMessage(ITextComponent component)
     {
-        LOGGER.info(component.getUnformattedText());
+        // Paper - Log message with colors
+        MinecraftServer.LOGGER.info(org.bukkit.craftbukkit.v1_12_R1.util.CraftChatMessage.fromComponent(component, net.minecraft.util.text.TextFormatting.WHITE));
     }
 
     @Override
@@ -1174,7 +1346,7 @@
 
     public void setDifficultyForAllWorlds(EnumDifficulty difficulty)
     {
-        for (WorldServer worldserver : this.worlds)
+        for (WorldServer worldserver : this.worldServerList)
         {
             if (worldserver != null)
             {
@@ -1257,9 +1429,9 @@
         playerSnooper.addClientStat("avg_tick_ms", (int)(MathHelper.average(this.tickTimeArray) * 1.0E-6));
         int i = 0;
 
-        if (this.worlds != null)
+        if (this.worldServerList != null)
         {
-            for (WorldServer worldserver : this.worlds)
+            for (WorldServer worldserver : this.worldServerList)
             {
                 if (worldserver != null)
                 {
@@ -1299,7 +1471,8 @@
 
     public boolean isServerInOnlineMode()
     {
-        return this.onlineMode;
+        // return this.onlineMode;
+        return server.getOnlineMode(); // CraftBukkit
     }
 
     public void setOnlineMode(boolean online)
@@ -1394,12 +1567,12 @@
 
     public void setPlayerList(PlayerList list)
     {
-        this.playerList = list;
+        if (this.playerList == null) this.playerList = list.initCraftServer(); else new UnsupportedOperationException("CatServer is not support the use of setPlayerList. If you installed FastWorkbench, Please replace with this fork: https://github.com/Luohuayu/FastWorkbench/").printStackTrace(); // CatServer
     }
 
     public void setGameType(GameType gameMode)
     {
-        for (WorldServer worldserver : this.worlds)
+        for (WorldServer worldserver : this.worldServerList)
         {
             worldserver.getWorldInfo().setGameType(gameMode);
         }
@@ -1442,7 +1615,7 @@
     @Override
     public World getEntityWorld()
     {
-        return this.worlds[0];
+        return this.worldServerList.get(0); // CraftBukkit
     }
 
     @SideOnly(Side.SERVER)
@@ -1515,7 +1688,7 @@
     @Nullable
     public Entity getEntityFromUuid(UUID uuid)
     {
-        for (WorldServer worldserver : this.worlds)
+        for (WorldServer worldserver : this.worldServerList)
         {
             if (worldserver != null)
             {
@@ -1534,7 +1707,7 @@
     @Override
     public boolean sendCommandFeedback()
     {
-        return this.worlds[0].getGameRules().getBoolean("sendCommandFeedback");
+        return this.worldServerList.get(0).getGameRules().getBoolean("sendCommandFeedback");
     }
 
     @Override
@@ -1543,6 +1716,11 @@
         return this;
     }
 
+    public static MinecraftServer getServerInst()
+    {
+        return (org.bukkit.Bukkit.getServer() instanceof org.bukkit.craftbukkit.v1_12_R1.CraftServer) ? ((org.bukkit.craftbukkit.v1_12_R1.CraftServer) org.bukkit.Bukkit.getServer()).getServer() : null;
+    }
+
     public int getMaxWorldSize()
     {
         return 29999984;
@@ -1552,15 +1730,14 @@
     {
         Validate.notNull(callable);
 
-        if (!this.isCallingFromMinecraftThread() && !this.isServerStopped())
+        if (!this.isCallingFromMinecraftThread()/* && !this.isServerStopped()*/) // CraftBukkit
         {
             ListenableFutureTask<V> listenablefuturetask = ListenableFutureTask.create(callable);
 
-            synchronized (this.futureTaskQueue)
-            {
-                this.futureTaskQueue.add(listenablefuturetask);
-                return listenablefuturetask;
-            }
+            // Spigot start
+            this.futureTaskQueue.add(listenablefuturetask);
+            return listenablefuturetask;
+            // Spigot end
         }
         else
         {
@@ -1570,7 +1747,7 @@
             }
             catch (Exception exception)
             {
-                return Futures.immediateFailedCheckedFuture(exception);
+                return Futures.immediateFailedFuture(exception);
             }
         }
     }
@@ -1612,12 +1789,12 @@
 
     public AdvancementManager getAdvancementManager()
     {
-        return this.worlds[0].getAdvancementManager();
+        return this.worldServerList.get(0).getAdvancementManager();
     }
 
     public FunctionManager getFunctionManager()
     {
-        return this.worlds[0].getFunctionManager();
+        return this.worldServerList.get(0).getFunctionManager();
     }
 
     public void reload()
@@ -1625,7 +1802,7 @@
         if (this.isCallingFromMinecraftThread())
         {
             this.getPlayerList().saveAllPlayerData();
-            this.worlds[0].getLootTableManager().reloadLootTables();
+            this.worldServerList.get(0).getLootTableManager().reloadLootTables();
             this.getAdvancementManager().reload();
             this.getFunctionManager().reload();
             this.getPlayerList().reloadResources();
@@ -1634,5 +1811,10 @@
         {
             this.addScheduledTask(this::reload);
         }
+    }
+
+    public DataFixer getDataFixer()
+    {
+        return this.dataFixer;
     }
 }
