--- before/net/minecraft/entity/passive/EntityLlama.java
+++ after/net/minecraft/entity/passive/EntityLlama.java
@@ -65,7 +65,7 @@
         this.setSize(0.9F, 1.87F);
     }
 
-    private void setStrength(int strengthIn)
+    public void setStrength(int strengthIn)
     {
         this.dataManager.set(DATA_STRENGTH_ID, Math.max(1, Math.min(5, strengthIn)));
     }
@@ -486,6 +486,9 @@
     @Override
     public void fall(float distance, float damageMultiplier)
     {
+        float[] ret = net.minecraftforge.common.ForgeHooks.onLivingFall(this, distance, damageMultiplier);
+        if (ret == null) return;
+        distance = ret[0]; damageMultiplier = ret[1];
         int i = MathHelper.ceil((distance * 0.5F - 3.0F) * damageMultiplier);
 
         if (i > 0)
