--- before/net/minecraft/entity/passive/EntityRabbit.java
+++ after/net/minecraft/entity/passive/EntityRabbit.java
@@ -68,6 +68,11 @@
         this.setSize(0.4F, 0.5F);
         this.jumpHelper = new EntityRabbit.RabbitJumpHelper(this);
         this.moveHelper = new EntityRabbit.RabbitMoveHelper(this);
+        this.initializePathFinderGoals(); // CraftBukkit - moved code
+    }
+
+    // CraftBukkit start - code from constructor
+    public void initializePathFinderGoals(){
         this.setMovementSpeed(0.0);
     }
 
@@ -592,7 +597,7 @@
             {
                 if (this.runDelay <= 0)
                 {
-                    if (!this.rabbit.world.getGameRules().getBoolean("mobGriefing"))
+                    if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.rabbit.world, this.rabbit))
                     {
                         return false;
                     }
@@ -638,11 +643,21 @@
 
                         if (integer == 0)
                         {
+                            if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, blockpos, Blocks.AIR, 0).isCancelled()) {
+                                return;
+                            }
                             world.setBlockState(blockpos, Blocks.AIR.getDefaultState(), 2);
                             world.destroyBlock(blockpos, true);
                         }
                         else
                         {
+                            if (org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityChangeBlockEvent(
+                                    this.rabbit,
+                                    blockpos,
+                                    block, block.getMetaFromState(iblockstate.withProperty(BlockCarrot.AGE, integer - 1))
+                            ).isCancelled()) {
+                                return;
+                            }
                             world.setBlockState(blockpos, iblockstate.withProperty(BlockCarrot.AGE, integer - 1), 2);
                             world.playEvent(2001, blockpos, Block.getStateId(iblockstate));
                         }
