--- before/net/minecraft/entity/passive/EntityVillager.java
+++ after/net/minecraft/entity/passive/EntityVillager.java
@@ -104,13 +104,13 @@
     @Nullable
     private EntityPlayer buyingPlayer;
     @Nullable
-    private MerchantRecipeList buyingList;
+    public MerchantRecipeList buyingList;
     private int timeUntilReset;
     private boolean needsInitilization;
     private boolean isWillingToMate;
-    private int wealth;
-    private String lastBuyingPlayer;
-    private int careerId;
+    public int wealth;
+    private java.util.UUID lastBuyingPlayer;
+    public int careerId;
     private int careerLevel;
     private boolean isLookingForHome;
     private boolean areAdditionalTasksSet;
@@ -413,7 +413,22 @@
         this.getEntityAttribute(SharedMonsterAttributes.MOVEMENT_SPEED).setBaseValue(0.5);
     }
 
+    // Spigot Start
     @Override
+    public void inactiveTick() {
+        // SPIGOT-3874
+        if (world.spigotConfig.tickInactiveVillagers) {
+            // SPIGOT-3894
+            net.minecraft.world.chunk.Chunk startingChunk = this.world.getChunkIfLoaded(MathHelper.floor(this.posX) >> 4, MathHelper.floor(this.posZ) >> 4);
+            if (!(startingChunk != null && startingChunk.areNeighborsLoaded(1))) {
+                return;
+            }
+            this.updateAITasks(); // SPIGOT-3846
+        }
+        super.inactiveTick();
+    }
+    // Spigot End
+
     protected void updateAITasks()
     {
         if (--this.randomTickDivider <= 0)
@@ -452,7 +467,13 @@
                     {
                         if (merchantrecipe.isRecipeDisabled())
                         {
-                            merchantrecipe.increaseMaxTradeUses(this.rand.nextInt(6) + this.rand.nextInt(6) + 2);
+                            // merchantrecipe.increaseMaxTradeUses(this.rand.nextInt(6) + this.rand.nextInt(6) + 2);
+                            int bonus = this.rand.nextInt(6) + this.rand.nextInt(6) + 2;
+                            org.bukkit.event.entity.VillagerReplenishTradeEvent event = new org.bukkit.event.entity.VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), merchantrecipe.asBukkit(), bonus);
+                            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                            if (!event.isCancelled()) {
+                                merchantrecipe.increaseMaxTradeUses(event.getBonus());
+                            }
                         }
                     }
 
@@ -487,7 +508,8 @@
         else if (!this.holdingSpawnEggOfClass(itemstack, (Class <? extends Entity >)this.getClass())
                  && this.isEntityAlive()
                  && !this.isTrading()
-                 && !this.isChild())
+                 && !this.isChild()
+                 && !player.isSneaking())
         {
             if (this.buyingList == null)
             {
@@ -566,6 +588,7 @@
     {
         super.writeEntityToNBT(compound);
         compound.setInteger("Profession", this.getProfession());
+        compound.setString("ProfessionName", this.getProfessionForge().getRegistryName().toString());
         compound.setInteger("Riches", this.wealth);
         compound.setInteger("Career", this.careerId);
         compound.setInteger("CareerLevel", this.careerLevel);
@@ -596,6 +619,14 @@
     {
         super.readEntityFromNBT(compound);
         this.setProfession(compound.getInteger("Profession"));
+        if (compound.hasKey("ProfessionName"))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p =
+                net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation(compound.getString("ProfessionName")));
+            if (p == null)
+                p = net.minecraftforge.fml.common.registry.ForgeRegistries.VILLAGER_PROFESSIONS.getValue(new net.minecraft.util.ResourceLocation("minecraft:farmer"));
+            this.setProfession(p);
+        }
         this.wealth = compound.getInteger("Riches");
         this.careerId = compound.getInteger("Career");
         this.careerLevel = compound.getInteger("CareerLevel");
@@ -624,7 +655,7 @@
     }
 
     @Override
-    protected boolean canDespawn()
+    public boolean canDespawn()
     {
         return false;
     }
@@ -656,12 +687,42 @@
 
     public void setProfession(int professionId)
     {
-        this.dataManager.set(PROFESSION, professionId);
+        this.dataManager.set(PROFESSION, Integer.valueOf(professionId));
+        net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, professionId);
     }
 
+    @Deprecated //Use Forge Variant below
     public int getProfession()
     {
-        return Math.max(this.dataManager.get(PROFESSION) % 6, 0);
+        return Math.max(((Integer)this.dataManager.get(PROFESSION)).intValue(), 0);
+    }
+
+    private net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof;
+    public void setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession prof)
+    {
+        this.prof = prof;
+        this.setProfession(net.minecraftforge.fml.common.registry.VillagerRegistry.getId(prof));
+    }
+
+    public net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession getProfessionForge()
+    {
+        if (this.prof == null)
+        {
+            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.getProfession());
+            if (this.prof == null)
+                return net.minecraftforge.fml.common.registry.VillagerRegistry.getById(0); //Farmer
+        }
+        return this.prof;
+    }
+
+    @Override
+    public void notifyDataManagerChange(DataParameter<?> key)
+    {
+        super.notifyDataManagerChange(key);
+        if (key.equals(PROFESSION))
+        {
+            net.minecraftforge.fml.common.registry.VillagerRegistry.onSetProfession(this, this.dataManager.get(PROFESSION));
+        }
     }
 
     public boolean isMating()
@@ -702,7 +763,7 @@
                     i = -3;
                 }
 
-                this.village.modifyPlayerReputation(livingBase.getName(), i);
+                this.village.modifyPlayerReputation(livingBase.getUniqueID(), i);
 
                 if (this.isEntityAlive())
                 {
@@ -723,7 +784,7 @@
             {
                 if (entity instanceof EntityPlayer)
                 {
-                    this.village.modifyPlayerReputation(entity.getName(), -2);
+                    this.village.modifyPlayerReputation(entity.getUniqueID(), -2);
                 }
                 else if (entity instanceof IMob)
                 {
@@ -820,7 +881,7 @@
 
             if (this.buyingPlayer != null)
             {
-                this.lastBuyingPlayer = this.buyingPlayer.getName();
+                this.lastBuyingPlayer = this.buyingPlayer.getUniqueID();
             }
             else
             {
@@ -865,20 +926,18 @@
             this.populateBuyingList();
         }
 
-        return this.buyingList;
+        return net.minecraftforge.event.ForgeEventFactory.listTradeOffers(this, player, buyingList);
     }
 
-    private void populateBuyingList()
+    public void populateBuyingList()
     {
-        EntityVillager.ITradeList[][][] aentityvillager$itradelist = DEFAULT_TRADE_LIST_MAP[this.getProfession()];
-
         if (this.careerId != 0 && this.careerLevel != 0)
         {
             this.careerLevel++;
         }
         else
         {
-            this.careerId = this.rand.nextInt(aentityvillager$itradelist.length) + 1;
+            this.careerId = this.getProfessionForge().getRandomCareer(this.rand) + 1;
             this.careerLevel = 1;
         }
 
@@ -889,19 +948,26 @@
 
         int i = this.careerId - 1;
         int j = this.careerLevel - 1;
+        java.util.List<EntityVillager.ITradeList> trades = this.getProfessionForge().getCareer(i).getTrades(j);
 
-        if (i >= 0 && i < aentityvillager$itradelist.length)
+        if (trades != null)
         {
-            EntityVillager.ITradeList[][] aentityvillager$itradelist1 = aentityvillager$itradelist[i];
-
-            if (j >= 0 && j < aentityvillager$itradelist1.length)
+            for (EntityVillager.ITradeList entityvillager$itradelist : trades)
             {
-                EntityVillager.ITradeList[] aentityvillager$itradelist2 = aentityvillager$itradelist1[j];
-
-                for (EntityVillager.ITradeList entityvillager$itradelist : aentityvillager$itradelist2)
-                {
-                    entityvillager$itradelist.addMerchantRecipe(this, this.buyingList, this.rand);
+                // CraftBukkit start
+                // this is a hack. this must be done because otherwise, if
+                // mojang adds a new type of villager merchant option, it will need to
+                // have event handling added manually. this is better than having to do that.
+                MerchantRecipeList list = new MerchantRecipeList();
+                entityvillager$itradelist.addMerchantRecipe(this, list /*this.buyingList*/, this.rand);
+                for (MerchantRecipe recipe : list) {
+                    org.bukkit.event.entity.VillagerAcquireTradeEvent event = new org.bukkit.event.entity.VillagerAcquireTradeEvent((org.bukkit.entity.Villager) getBukkitEntity(), recipe.asBukkit());
+                    org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                    if (!event.isCancelled()) {
+                        this.buyingList.add(org.bukkit.craftbukkit.v1_12_R1.inventory.CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+                    }
                 }
+                // CraftBukkit end
             }
         }
     }
@@ -1015,7 +1081,7 @@
                     s1 = "nitwit";
             }
 
-            if (s1 != null)
+            s1 = this.getProfessionForge().getCareer(this.careerId-1).getName();
             {
                 ITextComponent itextcomponent = new TextComponentTranslation("entity.Villager." + s1);
                 itextcomponent.getStyle().setHoverEvent(this.getHoverEvent());
@@ -1028,10 +1094,6 @@
 
                 return itextcomponent;
             }
-            else
-            {
-                return super.getDisplayName();
-            }
         }
     }
 
@@ -1097,7 +1159,7 @@
 
         if (p_190672_3_)
         {
-            this.setProfession(this.world.rand.nextInt(6));
+            net.minecraftforge.fml.common.registry.VillagerRegistry.setRandomProfession(this, this.world.rand);
         }
 
         this.setAdditionalAItasks();
@@ -1194,7 +1256,15 @@
     public boolean wantsMoreFood()
     {
         boolean flag = this.getProfession() == 0;
-        return flag ? !this.hasEnoughItems(5) : !this.hasEnoughItems(1);
+
+        if (flag)
+        {
+            return !this.hasEnoughItems(5);
+        }
+        else
+        {
+            return !this.hasEnoughItems(1);
+        }
     }
 
     private boolean hasEnoughItems(int multiplier)
@@ -1381,6 +1451,9 @@
                 recipeList.add(new MerchantRecipe(itemstack, itemstack1));
             }
         }
+    //MODDERS DO NOT USE OR EDIT THIS IN ANY WAY IT WILL HAVE NO EFFECT, THIS IS JUST IN HERE TO ALLOW FORGE TO ACCESS IT
+    @Deprecated
+    public static ITradeList[][][][] GET_TRADES_DONT_USE(){ return DEFAULT_TRADE_LIST_MAP; }
 
     public static class ListItemForEmeralds implements EntityVillager.ITradeList
         {
