--- before/net/minecraft/entity/passive/EntitySquid.java
+++ after/net/minecraft/entity/passive/EntitySquid.java
@@ -111,7 +111,7 @@
         this.prevSquidYaw = this.squidYaw;
         this.prevSquidRotation = this.squidRotation;
         this.lastTentacleAngle = this.tentacleAngle;
-        this.squidRotation = this.squidRotation + this.rotationVelocity;
+        this.squidRotation += this.rotationVelocity;
 
         if (this.squidRotation > Math.PI * 2)
         {
@@ -205,7 +205,7 @@
     @Override
     public boolean getCanSpawnHere()
     {
-        return this.posY > 45.0 && this.posY < this.world.getSeaLevel() && super.getCanSpawnHere();
+        return this.posY > this.world.spigotConfig.squidSpawnRangeMin && this.posY < this.world.getSeaLevel() && super.getCanSpawnHere();
     }
 
     @SideOnly(Side.CLIENT)
