--- before/net/minecraft/inventory/ContainerEnchantment.java
+++ after/net/minecraft/inventory/ContainerEnchantment.java
@@ -24,21 +24,8 @@
 
 public class ContainerEnchantment extends Container
 {
-    public IInventory tableInventory = new InventoryBasic("Enchant", true, 2)
-    {
-        @Override
-        public int getInventoryStackLimit()
-        {
-            return 64;
-        }
-        @Override
-        public void markDirty()
-        {
-            super.markDirty();
-            ContainerEnchantment.this.onCraftMatrixChanged(this);
-        }
-    };
-    private final World world;
+    public IInventory tableInventory;
+    private World world;
     private final BlockPos position;
     private final Random rand = new Random();
     public int xpSeed;
@@ -46,6 +33,9 @@
     public int[] enchantClue = new int[] { -1, -1, -1};
     public int[] worldClue = new int[] { -1, -1, -1};
 
+    private org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
     @SideOnly(Side.CLIENT)
     public ContainerEnchantment(InventoryPlayer playerInv, World worldIn)
     {
@@ -54,6 +44,23 @@
 
     public ContainerEnchantment(InventoryPlayer playerInv, World worldIn, BlockPos pos)
     {
+        this.tableInventory = new InventoryBasic("Enchant", true, 2)
+        {
+            public int getInventoryStackLimit()
+            {
+                return 64;
+            }
+            public void markDirty()
+            {
+                super.markDirty();
+                ContainerEnchantment.this.onCraftMatrixChanged(this);
+            }
+
+            @Override
+            public org.bukkit.Location getLocation() {
+                return new org.bukkit.Location(worldIn.getWorld(), position.getX(), position.getY(), position.getZ());
+            }
+        };
         this.world = worldIn;
         this.position = pos;
         this.xpSeed = playerInv.player.getXPSeed();
@@ -72,10 +79,13 @@
         });
         this.addSlotToContainer(new Slot(this.tableInventory, 1, 35, 47)
         {
+            java.util.List<ItemStack> ores = net.minecraftforge.oredict.OreDictionary.getOres("gemLapis");
             @Override
             public boolean isItemValid(ItemStack stack)
             {
-                return stack.getItem() == Items.DYE && EnumDyeColor.byDyeDamage(stack.getMetadata()) == EnumDyeColor.BLUE;
+                for (ItemStack ore : ores)
+                    if (net.minecraftforge.oredict.OreDictionary.itemMatches(ore, stack, false)) return true;
+                return false;
             }
         });
 
@@ -91,6 +101,8 @@
         {
             this.addSlotToContainer(new Slot(playerInv, k, 8 + k * 18, 142));
         }
+
+        player = (org.bukkit.entity.Player) playerInv.player.getBukkitEntity();
     }
 
     protected void broadcastData(IContainerListener crafting)
@@ -159,11 +171,12 @@
         {
             ItemStack itemstack = inventoryIn.getStackInSlot(0);
 
-            if (!itemstack.isEmpty() && itemstack.isItemEnchantable())
+            if (!itemstack.isEmpty()/* && itemstack.isItemEnchantable()*/) // CraftBukkit - relax condition
             {
                 if (!this.world.isRemote)
                 {
                     int l = 0;
+                    float power = 0;
 
                     for (int j = -1; j <= 1; j++)
                     {
@@ -173,37 +186,14 @@
                                     && this.world.isAirBlock(this.position.add(k, 0, j))
                                     && this.world.isAirBlock(this.position.add(k, 1, j)))
                             {
-                                if (this.world.getBlockState(this.position.add(k * 2, 0, j * 2)).getBlock() == Blocks.BOOKSHELF)
-                                {
-                                    l++;
-                                }
-
-                                if (this.world.getBlockState(this.position.add(k * 2, 1, j * 2)).getBlock() == Blocks.BOOKSHELF)
-                                {
-                                    l++;
-                                }
-
+                                power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k * 2, 0, j * 2));
+                                power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k * 2, 1, j * 2));
                                 if (k != 0 && j != 0)
                                 {
-                                    if (this.world.getBlockState(this.position.add(k * 2, 0, j)).getBlock() == Blocks.BOOKSHELF)
-                                    {
-                                        l++;
-                                    }
-
-                                    if (this.world.getBlockState(this.position.add(k * 2, 1, j)).getBlock() == Blocks.BOOKSHELF)
-                                    {
-                                        l++;
-                                    }
-
-                                    if (this.world.getBlockState(this.position.add(k, 0, j * 2)).getBlock() == Blocks.BOOKSHELF)
-                                    {
-                                        l++;
-                                    }
-
-                                    if (this.world.getBlockState(this.position.add(k, 1, j * 2)).getBlock() == Blocks.BOOKSHELF)
-                                    {
-                                        l++;
-                                    }
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k * 2, 0, j));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k * 2, 1, j));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k, 0, j * 2));
+                                    power += net.minecraftforge.common.ForgeHooks.getEnchantPower(world, position.add(k, 1, j * 2));
                                 }
                             }
                         }
@@ -213,7 +203,7 @@
 
                     for (int i1 = 0; i1 < 3; i1++)
                     {
-                        this.enchantLevels[i1] = EnchantmentHelper.calcItemStackEnchantability(this.rand, i1, l, itemstack);
+                        this.enchantLevels[i1] = EnchantmentHelper.calcItemStackEnchantability(this.rand, i1, (int)power, itemstack);
                         this.enchantClue[i1] = -1;
                         this.worldClue[i1] = -1;
 
@@ -221,6 +211,7 @@
                         {
                             this.enchantLevels[i1] = 0;
                         }
+                        this.enchantLevels[i1] = net.minecraftforge.event.ForgeEventFactory.onEnchantmentLevelSet(world, position, i1, (int)power, itemstack, enchantLevels[i1]);
                     }
 
                     for (int j1 = 0; j1 < 3; j1++)
@@ -238,6 +229,41 @@
                         }
                     }
 
+                    // CraftBukkit start
+                    org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack item = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+                    org.bukkit.enchantments.EnchantmentOffer[] offers = new org.bukkit.enchantments.EnchantmentOffer[3];
+                    for (int j = 0; j < 3; ++j) {
+                        org.bukkit.enchantments.Enchantment enchantment = (this.enchantClue[j] >= 0) ? org.bukkit.enchantments.Enchantment.getById(this.enchantClue[j]) : null;
+                        offers[j] = (enchantment != null) ? new org.bukkit.enchantments.EnchantmentOffer(enchantment, this.worldClue[j], this.enchantLevels[j]) : null;
+                    }
+
+                    org.bukkit.event.enchantment.PrepareItemEnchantEvent event = new org.bukkit.event.enchantment.PrepareItemEnchantEvent(player, this.getBukkitView(), this.world.getWorld().getBlockAt(position.getX(), position.getY(), position.getZ()), item, offers, (int) power);
+                    event.setCancelled(!itemstack.isItemEnchantable());
+                    if (this.getBukkitView() != null) this.world.getServer().getPluginManager().callEvent(event); // CatServer - mods bypass
+
+                    if (event.isCancelled()) {
+                        for (int j = 0; j < 3; ++j) {
+                            this.enchantLevels[j] = 0;
+                            this.enchantClue[j] = -1;
+                            this.worldClue[j] = -1;
+                        }
+                        return;
+                    }
+
+                    for (int j = 0; j < 3; j++) {
+                        org.bukkit.enchantments.EnchantmentOffer offer = event.getOffers()[j];
+                        if (offer != null) {
+                            this.enchantLevels[j] = offer.getCost();
+                            this.enchantClue[j] = offer.getEnchantment().getId();
+                            this.worldClue[j] = offer.getEnchantmentLevel();
+                        } else {
+                            this.enchantLevels[j] = 0;
+                            this.enchantClue[j] = -1;
+                            this.worldClue[j] = -1;
+                        }
+                    }
+                    // CraftBukkit end
+
                     this.detectAndSendChanges();
                 }
             }
@@ -272,31 +298,63 @@
             {
                 List<EnchantmentData> list = this.getEnchantmentList(itemstack, id, this.enchantLevels[id]);
 
-                if (!list.isEmpty())
+                if (true || !list.isEmpty())
                 {
-                    playerIn.onEnchant(itemstack, i);
+//                    playerIn.onEnchant(itemstack, i); // Moved down
                     boolean flag = itemstack.getItem() == Items.BOOK;
 
+                    java.util.Map<org.bukkit.enchantments.Enchantment, Integer> enchants = new java.util.HashMap<org.bukkit.enchantments.Enchantment, Integer>();
+                    for (Object obj : list) {
+                        EnchantmentData instance = (EnchantmentData) obj;
+                        enchants.put(org.bukkit.enchantments.Enchantment.getById(Enchantment.getEnchantmentID(instance.enchantment)), instance.enchantmentLevel);
+                    }
+                    org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack item = org.bukkit.craftbukkit.v1_12_R1.inventory.CraftItemStack.asCraftMirror(itemstack);
+
+                    org.bukkit.event.enchantment.EnchantItemEvent event = new org.bukkit.event.enchantment.EnchantItemEvent((org.bukkit.entity.Player) playerIn.getBukkitEntity(), this.getBukkitView(), this.world.getWorld().getBlockAt(position.getX(), position.getY(), position.getZ()), item, this.enchantLevels[id], enchants, id);
+                    if (this.getBukkitView() != null) this.world.getServer().getPluginManager().callEvent(event); // CatServer - mods bypass
+
+                    int level = event.getExpLevelCost();
+                    if (event.isCancelled() || (level > playerIn.experienceLevel && !playerIn.capabilities.isCreativeMode) || event.getEnchantsToAdd().isEmpty()) {
+                        return false;
+                    }
+
                     if (flag)
                     {
                         itemstack = new ItemStack(Items.ENCHANTED_BOOK);
                         this.tableInventory.setInventorySlotContents(0, itemstack);
                     }
 
-                    for (int j = 0; j < list.size(); j++)
-                    {
-                        EnchantmentData enchantmentdata = list.get(j);
-
-                        if (flag)
-                        {
-                            ItemEnchantedBook.addEnchantment(itemstack, enchantmentdata);
-                        }
-                        else
-                        {
-                            itemstack.addEnchantment(enchantmentdata.enchantment, enchantmentdata.enchantmentLevel);
+//                    for (int j = 0; j < list.size(); ++j)
+//                    {
+//                        EnchantmentData enchantmentdata = list.get(j);
+                    for (java.util.Map.Entry<org.bukkit.enchantments.Enchantment, Integer> entry : event.getEnchantsToAdd().entrySet()) {
+                        try {
+                            if (flag) {
+                                int enchantId = entry.getKey().getId();
+                                if (Enchantment.getEnchantmentByID(enchantId) == null) {
+                                    continue;
+                                }
+//                        if (flag)
+//                        {
+//                            ItemEnchantedBook.addEnchantment(itemstack, enchantmentdata);
+//                        }
+//                        else
+//                        {
+//                            itemstack.addEnchantment(enchantmentdata.enchantment, enchantmentdata.enchantmentLevel);
+                                EnchantmentData weightedrandomenchant = new EnchantmentData(Enchantment.getEnchantmentByID(enchantId), entry.getValue());
+                                ItemEnchantedBook.addEnchantment(itemstack, weightedrandomenchant);
+                            } else {
+                                item.addUnsafeEnchantment(entry.getKey(), entry.getValue());
+                            }
+                        } catch (IllegalArgumentException e) {
+                            /* Just swallow invalid enchantments */
                         }
                     }
 
+                    playerIn.onEnchant(itemstack, i);
+
+
+                    // CraftBukkit - TODO: let plugins change this
                     if (!playerIn.capabilities.isCreativeMode)
                     {
                         itemstack1.shrink(i);
@@ -361,6 +419,11 @@
     public void onContainerClosed(EntityPlayer playerIn)
     {
         super.onContainerClosed(playerIn);
+        // CraftBukkit Start - If an enchantable was opened from a null location, set the world to the player's world, preventing a crash
+        if (this.world == null) {
+            this.world = playerIn.getEntityWorld();
+        }
+        // CraftBukkit end
 
         if (!this.world.isRemote)
         {
@@ -371,6 +434,7 @@
     @Override
     public boolean canInteractWith(EntityPlayer playerIn)
     {
+        if (!this.checkReachable) return true;
         return this.world.getBlockState(this.position).getBlock() != Blocks.ENCHANTING_TABLE
                ? false
                : !(
@@ -420,10 +484,9 @@
                     return ItemStack.EMPTY;
                 }
 
-                if (itemstack1.hasTagCompound() && itemstack1.getCount() == 1)
+                if (itemstack1.hasTagCompound())// Forge: Fix MC-17431
                 {
-                    this.inventorySlots.get(0).putStack(itemstack1.copy());
-                    itemstack1.setCount(0);
+                    this.inventorySlots.get(0).putStack(itemstack1.splitStack(1));
                 }
                 else if (!itemstack1.isEmpty())
                 {
@@ -450,5 +513,16 @@
         }
 
         return itemstack;
+    }
+
+    @Override
+    public org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryEnchanting inventory = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryEnchanting(this.tableInventory);
+        bukkitEntity = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView(this.player, inventory, this);
+        return bukkitEntity;
     }
 }
