--- before/net/minecraft/inventory/ContainerBrewingStand.java
+++ after/net/minecraft/inventory/ContainerBrewingStand.java
@@ -20,8 +20,12 @@
     private int prevBrewTime;
     private int prevFuel;
 
+    private org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView bukkitEntity = null;
+    private InventoryPlayer player;
+
     public ContainerBrewingStand(InventoryPlayer playerInventory, IInventory tileBrewingStandIn)
     {
+        this.player = playerInventory;
         this.tileBrewingStand = tileBrewingStandIn;
         this.addSlotToContainer(new ContainerBrewingStand.Potion(tileBrewingStandIn, 0, 56, 51));
         this.addSlotToContainer(new ContainerBrewingStand.Potion(tileBrewingStandIn, 1, 79, 58));
@@ -84,6 +88,7 @@
     @Override
     public boolean canInteractWith(EntityPlayer playerIn)
     {
+        if (!this.checkReachable) return true;
         return this.tileBrewingStand.isUsableByPlayer(playerIn);
     }
 
@@ -205,7 +210,7 @@
             @Override
             public boolean isItemValid(ItemStack stack)
             {
-                return PotionHelper.isReagent(stack);
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidIngredient(stack);
             }
 
             @Override
@@ -241,6 +246,7 @@
 
                 if (thePlayer instanceof EntityPlayerMP)
                 {
+                    net.minecraftforge.event.ForgeEventFactory.onPlayerBrewedPotion(thePlayer, stack);
                     CriteriaTriggers.BREWED_POTION.trigger((EntityPlayerMP)thePlayer, potiontype);
                 }
 
@@ -250,8 +256,18 @@
 
             public static boolean canHoldPotion(ItemStack stack)
             {
-                Item item = stack.getItem();
-                return item == Items.POTIONITEM || item == Items.SPLASH_POTION || item == Items.LINGERING_POTION || item == Items.GLASS_BOTTLE;
+                return net.minecraftforge.common.brewing.BrewingRecipeRegistry.isValidInput(stack);
             }
         }
+
+    @Override
+    public org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryBrewer inventory = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryBrewer(this.tileBrewingStand);
+        bukkitEntity = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
 }
