--- before/net/minecraft/inventory/ContainerRepair.java
+++ after/net/minecraft/inventory/ContainerRepair.java
@@ -36,9 +36,13 @@
     private final BlockPos pos;
     public int maximumCost;
     public int materialCost;
-    private String repairedItemName;
+    public String repairedItemName;
     private final EntityPlayer player;
 
+    private int lastLevelCost;
+    private org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView bukkitEntity;
+    private InventoryPlayer playerInventory;
+
     @SideOnly(Side.CLIENT)
     public ContainerRepair(InventoryPlayer playerInventory, World worldIn, EntityPlayer player)
     {
@@ -47,6 +51,7 @@
 
     public ContainerRepair(InventoryPlayer playerInventory, final World worldIn, final BlockPos blockPosIn, EntityPlayer player)
     {
+        this.playerInventory = playerInventory;
         this.pos = blockPosIn;
         this.world = worldIn;
         this.player = player;
@@ -75,6 +80,8 @@
                     thePlayer.addExperienceLevel(-ContainerRepair.this.maximumCost);
                 }
 
+                float breakChance = net.minecraftforge.common.ForgeHooks.onAnvilRepair(thePlayer, stack, ContainerRepair.this.inputSlots.getStackInSlot(0), ContainerRepair.this.inputSlots.getStackInSlot(1));
+
                 ContainerRepair.this.inputSlots.setInventorySlotContents(0, ItemStack.EMPTY);
 
                 if (ContainerRepair.this.materialCost > 0)
@@ -102,7 +109,7 @@
                 if (!thePlayer.capabilities.isCreativeMode
                         && !worldIn.isRemote
                         && iblockstate.getBlock() == Blocks.ANVIL
-                        && thePlayer.getRNG().nextFloat() < 0.12F)
+                        && thePlayer.getRNG().nextFloat() < breakChance)
                 {
                     int l = iblockstate.getValue(BlockAnvil.DAMAGE);
 
@@ -162,7 +169,8 @@
 
         if (itemstack.isEmpty())
         {
-            this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+            // this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+            if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.EMPTY); // CatServer - mods bypass
             this.maximumCost = 0;
         }
         else
@@ -172,10 +180,17 @@
             Map<Enchantment, Integer> map = EnchantmentHelper.getEnchantments(itemstack1);
             j += itemstack.getRepairCost() + (itemstack2.isEmpty() ? 0 : itemstack2.getRepairCost());
             this.materialCost = 0;
+            boolean flag = false;
 
             if (!itemstack2.isEmpty())
             {
-                boolean flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && !ItemEnchantedBook.getEnchantments(itemstack2).isEmpty();
+                // CatServer start - call bukkit event when forge event cancel
+                if (!net.minecraftforge.common.ForgeHooks.onAnvilChange(this, itemstack, itemstack2, outputSlot, repairedItemName, j, player)) {
+                    if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), outputSlot.getStackInSlot(0));
+                    return;
+                }
+                // CatServer end
+                flag = itemstack2.getItem() == Items.ENCHANTED_BOOK && !ItemEnchantedBook.getEnchantments(itemstack2).isEmpty();
 
                 if (itemstack1.isItemStackDamageable() && itemstack1.getItem().getIsRepairable(itemstack, itemstack2))
                 {
@@ -183,7 +198,8 @@
 
                     if (l2 <= 0)
                     {
-                        this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+//                        this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.EMPTY); // CatServer - mods bypass
                         this.maximumCost = 0;
                         return;
                     }
@@ -204,7 +220,8 @@
                 {
                     if (!flag && (itemstack1.getItem() != itemstack2.getItem() || !itemstack1.isItemStackDamageable()))
                     {
-                        this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        // this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.EMPTY); // CatServer - mods bypass
                         this.maximumCost = 0;
                         return;
                     }
@@ -222,7 +239,7 @@
                             l1 = 0;
                         }
 
-                        if (l1 < itemstack1.getMetadata())
+                        if (l1 < itemstack1.getItemDamage()) // vanilla uses metadata here instead of damage.
                         {
                             itemstack1.setItemDamage(l1);
                             i += 2;
@@ -304,7 +321,8 @@
 
                     if (flag3 && !flag2)
                     {
-                        this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+//                        this.outputSlot.setInventorySlotContents(0, ItemStack.EMPTY);
+                        if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), ItemStack.EMPTY); // CatServer - mods bypass
                         this.maximumCost = 0;
                         return;
                     }
@@ -326,6 +344,7 @@
                 i += k;
                 itemstack1.setStackDisplayName(this.repairedItemName);
             }
+            if (flag && !itemstack1.getItem().isBookEnchantable(itemstack1, itemstack2)) itemstack1 = ItemStack.EMPTY;
 
             this.maximumCost = j + i;
 
@@ -362,7 +381,8 @@
                 EnchantmentHelper.setEnchantments(map, itemstack1);
             }
 
-            this.outputSlot.setInventorySlotContents(0, itemstack1);
+//            this.outputSlot.setInventorySlotContents(0, itemstack1);
+            if (getBukkitView() != null) org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callPrepareAnvilEvent(getBukkitView(), itemstack1); // CatServer - mods bypass
             this.detectAndSendChanges();
         }
     }
@@ -398,6 +418,7 @@
     @Override
     public boolean canInteractWith(EntityPlayer playerIn)
     {
+        if (!this.checkReachable) return true;
         return this.world.getBlockState(this.pos).getBlock() != Blocks.ANVIL
                ? false
                : playerIn.getDistanceSq(
@@ -477,5 +498,31 @@
         }
 
         this.updateRepairOutput();
+    }
+
+    @Override
+    public void detectAndSendChanges() {
+        super.detectAndSendChanges();
+        for (int i = 0; i < this.listeners.size(); ++i) {
+            IContainerListener icrafting = this.listeners.get(i);
+
+            if (this.lastLevelCost != this.maximumCost) {
+                icrafting.sendWindowProperty(this, 0, this.maximumCost);
+            }
+        }
+
+        this.lastLevelCost = this.maximumCost;
+    }
+
+    @Override
+    public org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryAnvil(
+                new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()), this.inputSlots, this.outputSlot, this);
+        bukkitEntity = new org.bukkit.craftbukkit.v1_12_R1.inventory.CraftInventoryView(this.playerInventory.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
     }
 }
