--- before/net/minecraft/inventory/SlotFurnaceOutput.java
+++ after/net/minecraft/inventory/SlotFurnaceOutput.java
@@ -75,6 +75,19 @@
                 i = j;
             }
 
+            if (this.inventory instanceof net.minecraft.tileentity.TileEntityFurnace) { // CatServer - mods bypass
+                // CraftBukkit - FurnaceExtractEvent
+                org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.player.getBukkitEntity();
+                net.minecraft.tileentity.TileEntityFurnace furnace = ((net.minecraft.tileentity.TileEntityFurnace) this.inventory);
+                org.bukkit.block.Block block = this.player.world.getWorld().getBlockAt(furnace.getPos().getX(), furnace.getPos().getY(), furnace.getPos().getZ());
+    
+                if (removeCount != 0) {
+                    org.bukkit.event.inventory.FurnaceExtractEvent event = new org.bukkit.event.inventory.FurnaceExtractEvent(player, block, org.bukkit.craftbukkit.v1_12_R1.util.CraftMagicNumbers.getMaterial(stack.getItem()), removeCount, i);
+                    this.player.world.getServer().getPluginManager().callEvent(event);
+                    i = event.getExpToDrop();
+                }
+            }
+
             while (i > 0)
             {
                 int k = EntityXPOrb.getXPSplit(i);
@@ -94,5 +107,6 @@
         }
 
         this.removeCount = 0;
+        net.minecraftforge.fml.common.FMLCommonHandler.instance().firePlayerSmeltedEvent(player, stack);
     }
 }
