--- before/net/minecraft/entity/ai/EntityAIMate.java
+++ after/net/minecraft/entity/ai/EntityAIMate.java
@@ -96,8 +96,23 @@
     {
         EntityAgeable entityageable = this.animal.createChild(this.targetMate);
 
+        final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, targetMate, entityageable);
+        final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+        entityageable = event.getChild();
+        if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.animal.setGrowingAge(6000);
+            this.targetMate.setGrowingAge(6000);
+            this.animal.resetInLove();
+            this.targetMate.resetInLove();
+            return;
+        }
+        
         if (entityageable != null)
         {
+            if (entityageable instanceof net.minecraft.entity.passive.EntityTameable && ((net.minecraft.entity.passive.EntityTameable) entityageable).isTamed()) {
+                entityageable.persistenceRequired = true;
+            }
             EntityPlayerMP entityplayermp = this.animal.getLoveCause();
 
             if (entityplayermp == null && this.targetMate.getLoveCause() != null)
@@ -105,6 +120,13 @@
                 entityplayermp = this.targetMate.getLoveCause();
             }
 
+            int experience = this.animal.getRNG().nextInt(catserver.server.utils.ModFixUtils.fixNetherAPI()) + 1; // CatRoom - Fix NetherAPI
+            org.bukkit.event.entity.EntityBreedEvent entityBreedEvent = org.bukkit.craftbukkit.v1_12_R1.event.CraftEventFactory.callEntityBreedEvent(entityageable, animal, targetMate, entityplayermp, this.animal.breedItem, experience);
+            if (entityBreedEvent.isCancelled()) {
+                return;
+            }
+            experience = entityBreedEvent.getExperience();
+
             if (entityplayermp != null)
             {
                 entityplayermp.addStat(StatList.ANIMALS_BRED);
@@ -117,7 +139,8 @@
             this.targetMate.resetInLove();
             entityageable.setGrowingAge(-24000);
             entityageable.setLocationAndAngles(this.animal.posX, this.animal.posY, this.animal.posZ, 0.0F, 0.0F);
-            this.world.spawnEntity(entityageable);
+//            this.world.spawnEntity(entityageable);
+            this.world.addEntity(entityageable, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BREEDING);
             Random random = this.animal.getRNG();
 
             for (int i = 0; i < 7; i++)
@@ -142,6 +165,7 @@
 
             if (this.world.getGameRules().getBoolean("doMobLoot"))
             {
+                if (experience > 0)
                 this.world
                 .spawnEntity(
                     new EntityXPOrb(
@@ -149,7 +173,7 @@
                         this.animal.posX,
                         this.animal.posY,
                         this.animal.posZ,
-                        random.nextInt(7) + 1
+                        experience
                     )
                 );
             }
